<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 5 Compositional analyses of biological sequences | Nested genes code</title>
  <meta name="description" content="This is the description of the code employed in the research." />
  <meta name="generator" content="bookdown 0.34 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 5 Compositional analyses of biological sequences | Nested genes code" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="This is the description of the code employed in the research." />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 5 Compositional analyses of biological sequences | Nested genes code" />
  
  <meta name="twitter:description" content="This is the description of the code employed in the research." />
  

<meta name="author" content="Esmeralda G. Legarda" />


<meta name="date" content="2023-11-08" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="distances-with-chimerax.html"/>
<link rel="next" href="comparative-analyses-of-selection.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A Minimal Book Example</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> About</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#abstract"><i class="fa fa-check"></i><b>1.1</b> Abstract</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#abstract-1"><i class="fa fa-check"></i><b>1.2</b> Abstract</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="blast-and-e-utilities.html"><a href="blast-and-e-utilities.html"><i class="fa fa-check"></i><b>2</b> Blast and E-utilities</a>
<ul>
<li class="chapter" data-level="2.1" data-path="blast-and-e-utilities.html"><a href="blast-and-e-utilities.html#psi-blast"><i class="fa fa-check"></i><b>2.1</b> PSI-BLAST</a></li>
<li class="chapter" data-level="2.2" data-path="blast-and-e-utilities.html"><a href="blast-and-e-utilities.html#protein-id-to-nucleotides"><i class="fa fa-check"></i><b>2.2</b> Protein ID to nucleotides</a></li>
<li class="chapter" data-level="2.3" data-path="blast-and-e-utilities.html"><a href="blast-and-e-utilities.html#species-to-family"><i class="fa fa-check"></i><b>2.3</b> Species to family</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html"><i class="fa fa-check"></i><b>3</b> Distances with clustering in CLANS</a>
<ul>
<li class="chapter" data-level="3.1" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html#read-clans-file"><i class="fa fa-check"></i><b>3.1</b> Read CLANS file</a></li>
<li class="chapter" data-level="3.2" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html#create-a-dataframe"><i class="fa fa-check"></i><b>3.2</b> Create a DataFrame</a></li>
<li class="chapter" data-level="3.3" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html#calculate-centroids-and-distances"><i class="fa fa-check"></i><b>3.3</b> Calculate centroids and distances</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html#plot-the-clustermap"><i class="fa fa-check"></i><b>3.3.1</b> Plot the clustermap</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html#run"><i class="fa fa-check"></i><b>3.4</b> Run</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html"><i class="fa fa-check"></i><b>4</b> Distances with ChimeraX</a>
<ul>
<li class="chapter" data-level="4.1" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#obtaining-rmsd-with-chimerax"><i class="fa fa-check"></i><b>4.1</b> Obtaining RMSD with ChimeraX</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#import-libraries-and-data"><i class="fa fa-check"></i><b>4.1.1</b> Import libraries and data</a></li>
<li class="chapter" data-level="4.1.2" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#matchmaker"><i class="fa fa-check"></i><b>4.1.2</b> MatchMaker</a></li>
<li class="chapter" data-level="4.1.3" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#save-csv-file"><i class="fa fa-check"></i><b>4.1.3</b> Save csv file</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#hiarachical-clustering"><i class="fa fa-check"></i><b>4.2</b> Hiarachical clustering</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#libraries-and-read-csv"><i class="fa fa-check"></i><b>4.2.1</b> Libraries and read csv</a></li>
<li class="chapter" data-level="4.2.2" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#create-matrix"><i class="fa fa-check"></i><b>4.2.2</b> Create matrix</a></li>
<li class="chapter" data-level="4.2.3" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#plot-the-clustermap-1"><i class="fa fa-check"></i><b>4.2.3</b> Plot the clustermap</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="compositional-analyses-of-biological-sequences.html"><a href="compositional-analyses-of-biological-sequences.html"><i class="fa fa-check"></i><b>5</b> Compositional analyses of biological sequences</a>
<ul>
<li class="chapter" data-level="5.1" data-path="compositional-analyses-of-biological-sequences.html"><a href="compositional-analyses-of-biological-sequences.html#amino-acids"><i class="fa fa-check"></i><b>5.1</b> Amino acids</a></li>
<li class="chapter" data-level="5.2" data-path="compositional-analyses-of-biological-sequences.html"><a href="compositional-analyses-of-biological-sequences.html#nucleotides"><i class="fa fa-check"></i><b>5.2</b> Nucleotides</a></li>
<li class="chapter" data-level="5.3" data-path="compositional-analyses-of-biological-sequences.html"><a href="compositional-analyses-of-biological-sequences.html#dinuclotides"><i class="fa fa-check"></i><b>5.3</b> Dinuclotides</a></li>
<li class="chapter" data-level="5.4" data-path="compositional-analyses-of-biological-sequences.html"><a href="compositional-analyses-of-biological-sequences.html#codons"><i class="fa fa-check"></i><b>5.4</b> Codons</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html"><i class="fa fa-check"></i><b>6</b> Comparative analyses of selection</a>
<ul>
<li class="chapter" data-level="6.1" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#evolution-speed-with-absrel"><i class="fa fa-check"></i><b>6.1</b> Evolution speed with aBSREL</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#libraries-and-json-files"><i class="fa fa-check"></i><b>6.1.1</b> Libraries and json files</a></li>
<li class="chapter" data-level="6.1.2" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#omega-and-significance"><i class="fa fa-check"></i><b>6.1.2</b> Omega and significance</a></li>
<li class="chapter" data-level="6.1.3" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#relative-asinh-transformation"><i class="fa fa-check"></i><b>6.1.3</b> Relative asinh transformation</a></li>
<li class="chapter" data-level="6.1.4" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#utilities"><i class="fa fa-check"></i><b>6.1.4</b> Utilities</a></li>
<li class="chapter" data-level="6.1.5" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#boxplots-with-significance"><i class="fa fa-check"></i><b>6.1.5</b> Boxplots with significance</a></li>
<li class="chapter" data-level="6.1.6" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#test-selection"><i class="fa fa-check"></i><b>6.1.6</b> Test selection</a></li>
<li class="chapter" data-level="6.1.7" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#run-1"><i class="fa fa-check"></i><b>6.1.7</b> Run</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#fel-for-correlations"><i class="fa fa-check"></i><b>6.2</b> FEL for correlations</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#significance-and-difference"><i class="fa fa-check"></i><b>6.2.1</b> Significance and difference</a></li>
<li class="chapter" data-level="6.2.2" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#plot-correlation"><i class="fa fa-check"></i><b>6.2.2</b> Plot correlation</a></li>
<li class="chapter" data-level="6.2.3" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#test-correlation-of-selection"><i class="fa fa-check"></i><b>6.2.3</b> Test correlation of selection</a></li>
<li class="chapter" data-level="6.2.4" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#correlation-2d"><i class="fa fa-check"></i><b>6.2.4</b> Correlation 2D</a></li>
<li class="chapter" data-level="6.2.5" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#selection-in-2d"><i class="fa fa-check"></i><b>6.2.5</b> Selection in 2D</a></li>
<li class="chapter" data-level="6.2.6" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#run-2"><i class="fa fa-check"></i><b>6.2.6</b> Run</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="miscellania.html"><a href="miscellania.html"><i class="fa fa-check"></i><b>7</b> Miscellania</a>
<ul>
<li class="chapter" data-level="7.1" data-path="miscellania.html"><a href="miscellania.html#representation-of-predicted-disordered-amino-acid-data"><i class="fa fa-check"></i><b>7.1</b> Representation of predicted disordered amino acid data</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Nested genes code</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="compositional-analyses-of-biological-sequences" class="section level1 hasAnchor" number="5">
<h1><span class="header-section-number">Chapter 5</span> Compositional analyses of biological sequences<a href="compositional-analyses-of-biological-sequences.html#compositional-analyses-of-biological-sequences" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>Compositional data might be useful to be compared between two or more groups and could reveal some overrepresented characteristics. We hereby show some code for the comparison between movement proteins, and also their corresponding encoding genes, with and without overlapping frames.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb22-1"><a href="compositional-analyses-of-biological-sequences.html#cb22-1" tabindex="-1"></a><span class="im">from</span> Bio <span class="im">import</span> SeqIO</span>
<span id="cb22-2"><a href="compositional-analyses-of-biological-sequences.html#cb22-2" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb22-3"><a href="compositional-analyses-of-biological-sequences.html#cb22-3" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb22-4"><a href="compositional-analyses-of-biological-sequences.html#cb22-4" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb22-5"><a href="compositional-analyses-of-biological-sequences.html#cb22-5" tabindex="-1"></a><span class="im">import</span> scipy.stats</span>
<span id="cb22-6"><a href="compositional-analyses-of-biological-sequences.html#cb22-6" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb22-7"><a href="compositional-analyses-of-biological-sequences.html#cb22-7" tabindex="-1"></a><span class="im">from</span> statannotations.Annotator <span class="im">import</span> Annotator</span></code></pre></div>
<p>We define the table of codons and amino acids as well as a classification of amino acids based on their physicochemical properties. Notice that the table of codons is ordered by the degeneracy of the genetic code.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb23-1"><a href="compositional-analyses-of-biological-sequences.html#cb23-1" tabindex="-1"></a>table_DNA_to_codon <span class="op">=</span> {<span class="st">&quot;M&quot;</span>: [<span class="st">&quot;ATG&quot;</span>],</span>
<span id="cb23-2"><a href="compositional-analyses-of-biological-sequences.html#cb23-2" tabindex="-1"></a>                      <span class="st">&quot;W&quot;</span>: [<span class="st">&quot;TGG&quot;</span>],</span>
<span id="cb23-3"><a href="compositional-analyses-of-biological-sequences.html#cb23-3" tabindex="-1"></a>                      <span class="st">&quot;C&quot;</span>: [<span class="st">&quot;TGT&quot;</span>, <span class="st">&quot;TGC&quot;</span>],</span>
<span id="cb23-4"><a href="compositional-analyses-of-biological-sequences.html#cb23-4" tabindex="-1"></a>                      <span class="st">&quot;D&quot;</span>: [<span class="st">&quot;GAT&quot;</span>, <span class="st">&quot;GAC&quot;</span>],</span>
<span id="cb23-5"><a href="compositional-analyses-of-biological-sequences.html#cb23-5" tabindex="-1"></a>                      <span class="st">&quot;E&quot;</span>: [<span class="st">&quot;GAA&quot;</span>, <span class="st">&quot;GAG&quot;</span>],</span>
<span id="cb23-6"><a href="compositional-analyses-of-biological-sequences.html#cb23-6" tabindex="-1"></a>                      <span class="st">&quot;F&quot;</span>: [<span class="st">&quot;TTT&quot;</span>, <span class="st">&quot;TTC&quot;</span>],</span>
<span id="cb23-7"><a href="compositional-analyses-of-biological-sequences.html#cb23-7" tabindex="-1"></a>                      <span class="st">&quot;H&quot;</span>: [<span class="st">&quot;CAT&quot;</span>, <span class="st">&quot;CAC&quot;</span>],</span>
<span id="cb23-8"><a href="compositional-analyses-of-biological-sequences.html#cb23-8" tabindex="-1"></a>                      <span class="st">&quot;K&quot;</span>: [<span class="st">&quot;AAA&quot;</span>, <span class="st">&quot;AAG&quot;</span>],</span>
<span id="cb23-9"><a href="compositional-analyses-of-biological-sequences.html#cb23-9" tabindex="-1"></a>                      <span class="st">&quot;N&quot;</span>: [<span class="st">&quot;AAT&quot;</span>, <span class="st">&quot;AAC&quot;</span>], </span>
<span id="cb23-10"><a href="compositional-analyses-of-biological-sequences.html#cb23-10" tabindex="-1"></a>                      <span class="st">&quot;Q&quot;</span>: [<span class="st">&quot;CAA&quot;</span>, <span class="st">&quot;CAG&quot;</span>],</span>
<span id="cb23-11"><a href="compositional-analyses-of-biological-sequences.html#cb23-11" tabindex="-1"></a>                      <span class="st">&quot;Y&quot;</span>: [<span class="st">&quot;TAT&quot;</span>, <span class="st">&quot;TAC&quot;</span>],</span>
<span id="cb23-12"><a href="compositional-analyses-of-biological-sequences.html#cb23-12" tabindex="-1"></a>                      <span class="st">&quot;I&quot;</span>: [<span class="st">&quot;ATT&quot;</span>, <span class="st">&quot;ATC&quot;</span>, <span class="st">&quot;ATA&quot;</span>],</span>
<span id="cb23-13"><a href="compositional-analyses-of-biological-sequences.html#cb23-13" tabindex="-1"></a>                      <span class="st">&quot;A&quot;</span>: [<span class="st">&quot;GCT&quot;</span>, <span class="st">&quot;GCC&quot;</span>, <span class="st">&quot;GCA&quot;</span>, <span class="st">&quot;GCG&quot;</span>], </span>
<span id="cb23-14"><a href="compositional-analyses-of-biological-sequences.html#cb23-14" tabindex="-1"></a>                      <span class="st">&quot;G&quot;</span>: [<span class="st">&quot;GGT&quot;</span>, <span class="st">&quot;GGC&quot;</span>, <span class="st">&quot;GGA&quot;</span>, <span class="st">&quot;GGG&quot;</span>],</span>
<span id="cb23-15"><a href="compositional-analyses-of-biological-sequences.html#cb23-15" tabindex="-1"></a>                      <span class="st">&quot;P&quot;</span>: [<span class="st">&quot;CCT&quot;</span>, <span class="st">&quot;CCC&quot;</span>, <span class="st">&quot;CCA&quot;</span>, <span class="st">&quot;CCG&quot;</span>],</span>
<span id="cb23-16"><a href="compositional-analyses-of-biological-sequences.html#cb23-16" tabindex="-1"></a>                      <span class="st">&quot;T&quot;</span>: [<span class="st">&quot;ACT&quot;</span>, <span class="st">&quot;ACC&quot;</span>, <span class="st">&quot;ACA&quot;</span>, <span class="st">&quot;ACG&quot;</span>],</span>
<span id="cb23-17"><a href="compositional-analyses-of-biological-sequences.html#cb23-17" tabindex="-1"></a>                      <span class="st">&quot;V&quot;</span>: [<span class="st">&quot;GTT&quot;</span>, <span class="st">&quot;GTC&quot;</span>, <span class="st">&quot;GTA&quot;</span>, <span class="st">&quot;GTG&quot;</span>],</span>
<span id="cb23-18"><a href="compositional-analyses-of-biological-sequences.html#cb23-18" tabindex="-1"></a>                      <span class="st">&quot;L&quot;</span>: [<span class="st">&quot;CTT&quot;</span>, <span class="st">&quot;CTC&quot;</span>, <span class="st">&quot;CTA&quot;</span>, <span class="st">&quot;CTG&quot;</span>, <span class="st">&quot;TTA&quot;</span>, <span class="st">&quot;TTG&quot;</span>],</span>
<span id="cb23-19"><a href="compositional-analyses-of-biological-sequences.html#cb23-19" tabindex="-1"></a>                      <span class="st">&quot;R&quot;</span>: [<span class="st">&quot;CGT&quot;</span>, <span class="st">&quot;CGC&quot;</span>, <span class="st">&quot;CGA&quot;</span>, <span class="st">&quot;CGG&quot;</span>, <span class="st">&quot;AGA&quot;</span>, <span class="st">&quot;AGG&quot;</span>],</span>
<span id="cb23-20"><a href="compositional-analyses-of-biological-sequences.html#cb23-20" tabindex="-1"></a>                      <span class="st">&quot;S&quot;</span>: [<span class="st">&quot;TCT&quot;</span>, <span class="st">&quot;TCC&quot;</span>, <span class="st">&quot;TCA&quot;</span>, <span class="st">&quot;TCG&quot;</span>, <span class="st">&quot;AGT&quot;</span>, <span class="st">&quot;AGC&quot;</span>],</span>
<span id="cb23-21"><a href="compositional-analyses-of-biological-sequences.html#cb23-21" tabindex="-1"></a>                      <span class="st">&quot;STOP&quot;</span>: [<span class="st">&quot;TAA&quot;</span>, <span class="st">&quot;TGA&quot;</span>, <span class="st">&quot;TAG&quot;</span>]</span>
<span id="cb23-22"><a href="compositional-analyses-of-biological-sequences.html#cb23-22" tabindex="-1"></a>                      }</span>
<span id="cb23-23"><a href="compositional-analyses-of-biological-sequences.html#cb23-23" tabindex="-1"></a></span>
<span id="cb23-24"><a href="compositional-analyses-of-biological-sequences.html#cb23-24" tabindex="-1"></a>features_aa <span class="op">=</span> {<span class="st">&quot;nonpolarbig&quot;</span>: [<span class="st">&quot;F&quot;</span>, <span class="st">&quot;W&quot;</span>, <span class="st">&quot;L&quot;</span>, <span class="st">&quot;I&quot;</span>, <span class="st">&quot;M&quot;</span>],</span>
<span id="cb23-25"><a href="compositional-analyses-of-biological-sequences.html#cb23-25" tabindex="-1"></a>               <span class="st">&quot;nonpolar&quot;</span>: [<span class="st">&quot;V&quot;</span>, <span class="st">&quot;P&quot;</span>, <span class="st">&quot;A&quot;</span>, <span class="st">&quot;G&quot;</span>],</span>
<span id="cb23-26"><a href="compositional-analyses-of-biological-sequences.html#cb23-26" tabindex="-1"></a>               <span class="st">&quot;polar&quot;</span>: [<span class="st">&quot;S&quot;</span>, <span class="st">&quot;C&quot;</span>, <span class="st">&quot;N&quot;</span>, <span class="st">&quot;Q&quot;</span>, <span class="st">&quot;T&quot;</span>, <span class="st">&quot;Y&quot;</span>],</span>
<span id="cb23-27"><a href="compositional-analyses-of-biological-sequences.html#cb23-27" tabindex="-1"></a>               <span class="st">&quot;basic&quot;</span>: [<span class="st">&quot;K&quot;</span>, <span class="st">&quot;R&quot;</span>, <span class="st">&quot;H&quot;</span>],</span>
<span id="cb23-28"><a href="compositional-analyses-of-biological-sequences.html#cb23-28" tabindex="-1"></a>               <span class="st">&quot;acidic&quot;</span>: [<span class="st">&quot;D&quot;</span>, <span class="st">&quot;E&quot;</span>]}</span>
<span id="cb23-29"><a href="compositional-analyses-of-biological-sequences.html#cb23-29" tabindex="-1"></a></span>
<span id="cb23-30"><a href="compositional-analyses-of-biological-sequences.html#cb23-30" tabindex="-1"></a><span class="co"># Bind the codon correspondence of amino acids and amino acid features</span></span>
<span id="cb23-31"><a href="compositional-analyses-of-biological-sequences.html#cb23-31" tabindex="-1"></a>codons_byfeatures <span class="op">=</span> {}</span>
<span id="cb23-32"><a href="compositional-analyses-of-biological-sequences.html#cb23-32" tabindex="-1"></a><span class="cf">for</span> feat <span class="kw">in</span> features_aa:</span>
<span id="cb23-33"><a href="compositional-analyses-of-biological-sequences.html#cb23-33" tabindex="-1"></a>    codons_byfeatures[feat] <span class="op">=</span> []</span>
<span id="cb23-34"><a href="compositional-analyses-of-biological-sequences.html#cb23-34" tabindex="-1"></a>    <span class="cf">for</span> aminoacid <span class="kw">in</span> features_aa[feat]:</span>
<span id="cb23-35"><a href="compositional-analyses-of-biological-sequences.html#cb23-35" tabindex="-1"></a>        <span class="cf">for</span> codon <span class="kw">in</span> table_DNA_to_codon[aminoacid]:</span>
<span id="cb23-36"><a href="compositional-analyses-of-biological-sequences.html#cb23-36" tabindex="-1"></a>            codons_byfeatures[feat].append(codon)</span>
<span id="cb23-37"><a href="compositional-analyses-of-biological-sequences.html#cb23-37" tabindex="-1"></a></span>
<span id="cb23-38"><a href="compositional-analyses-of-biological-sequences.html#cb23-38" tabindex="-1"></a><span class="co"># Add STOP codons</span></span>
<span id="cb23-39"><a href="compositional-analyses-of-biological-sequences.html#cb23-39" tabindex="-1"></a>codons_byfeatures[<span class="st">&quot;STOP&quot;</span>] <span class="op">=</span> table_DNA_to_codon[<span class="st">&quot;STOP&quot;</span>]</span></code></pre></div>
<div id="amino-acids" class="section level2 hasAnchor" number="5.1">
<h2><span class="header-section-number">5.1</span> Amino acids<a href="compositional-analyses-of-biological-sequences.html#amino-acids" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>We counted the number of amino acids within each group of sequences, including the 20 essential amino acids.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb24-1"><a href="compositional-analyses-of-biological-sequences.html#cb24-1" tabindex="-1"></a></span>
<span id="cb24-2"><a href="compositional-analyses-of-biological-sequences.html#cb24-2" tabindex="-1"></a><span class="kw">def</span> getAA(seq, letter):</span>
<span id="cb24-3"><a href="compositional-analyses-of-biological-sequences.html#cb24-3" tabindex="-1"></a>    <span class="cf">return</span>(<span class="bu">round</span>((<span class="bu">sum</span>([<span class="fl">1.0</span> <span class="cf">for</span> nucl <span class="kw">in</span> seq <span class="cf">if</span> nucl <span class="kw">in</span> [letter]]) <span class="op">/</span> (<span class="bu">len</span>(seq)<span class="op">-</span>seq.count(<span class="st">&quot;-&quot;</span>))), ndigits<span class="op">=</span><span class="dv">2</span>))</span>
<span id="cb24-4"><a href="compositional-analyses-of-biological-sequences.html#cb24-4" tabindex="-1"></a></span>
<span id="cb24-5"><a href="compositional-analyses-of-biological-sequences.html#cb24-5" tabindex="-1"></a>paths <span class="op">=</span> [<span class="st">&quot;p19_aligned.fa&quot;</span>, <span class="st">&quot;p22_aligned.fa&quot;</span>, <span class="st">&quot;p22 core_aligned.fa&quot;</span>, <span class="st">&quot;other cores_allMP.fa&quot;</span>, <span class="st">&quot;RdRp_aligned.fa&quot;</span>, <span class="st">&quot;CP_aligned.fa&quot;</span>] </span>
<span id="cb24-6"><a href="compositional-analyses-of-biological-sequences.html#cb24-6" tabindex="-1"></a>letters <span class="op">=</span> [<span class="st">&quot;M&quot;</span>, <span class="st">&quot;W&quot;</span>, <span class="st">&quot;C&quot;</span>, <span class="st">&quot;D&quot;</span>, <span class="st">&quot;E&quot;</span>, <span class="st">&quot;F&quot;</span>, <span class="st">&quot;H&quot;</span>, <span class="st">&quot;K&quot;</span>, <span class="st">&quot;N&quot;</span>, <span class="st">&quot;Q&quot;</span>, </span>
<span id="cb24-7"><a href="compositional-analyses-of-biological-sequences.html#cb24-7" tabindex="-1"></a>            <span class="st">&quot;Y&quot;</span>, <span class="st">&quot;I&quot;</span>, <span class="st">&quot;A&quot;</span>, <span class="st">&quot;G&quot;</span>, <span class="st">&quot;P&quot;</span>, <span class="st">&quot;T&quot;</span>, <span class="st">&quot;V&quot;</span>, <span class="st">&quot;L&quot;</span>, <span class="st">&quot;R&quot;</span>, <span class="st">&quot;S&quot;</span>]</span>
<span id="cb24-8"><a href="compositional-analyses-of-biological-sequences.html#cb24-8" tabindex="-1"></a>df_aa <span class="op">=</span> []</span>
<span id="cb24-9"><a href="compositional-analyses-of-biological-sequences.html#cb24-9" tabindex="-1"></a>aa_means <span class="op">=</span> []</span>
<span id="cb24-10"><a href="compositional-analyses-of-biological-sequences.html#cb24-10" tabindex="-1"></a>aa_medians <span class="op">=</span> []</span>
<span id="cb24-11"><a href="compositional-analyses-of-biological-sequences.html#cb24-11" tabindex="-1"></a>aa_std <span class="op">=</span> []</span>
<span id="cb24-12"><a href="compositional-analyses-of-biological-sequences.html#cb24-12" tabindex="-1"></a></span>
<span id="cb24-13"><a href="compositional-analyses-of-biological-sequences.html#cb24-13" tabindex="-1"></a><span class="cf">for</span> i, aminoacid <span class="kw">in</span> <span class="bu">enumerate</span>(letters): </span>
<span id="cb24-14"><a href="compositional-analyses-of-biological-sequences.html#cb24-14" tabindex="-1"></a>    AA_content <span class="op">=</span> []</span>
<span id="cb24-15"><a href="compositional-analyses-of-biological-sequences.html#cb24-15" tabindex="-1"></a>    <span class="cf">for</span> alignment <span class="kw">in</span> paths:</span>
<span id="cb24-16"><a href="compositional-analyses-of-biological-sequences.html#cb24-16" tabindex="-1"></a>        name <span class="op">=</span> alignment[<span class="dv">0</span>:alignment.find(<span class="st">&quot;_&quot;</span>)]</span>
<span id="cb24-17"><a href="compositional-analyses-of-biological-sequences.html#cb24-17" tabindex="-1"></a>        <span class="cf">for</span> record <span class="kw">in</span> SeqIO.parse(alignment, <span class="st">&quot;fasta&quot;</span>):</span>
<span id="cb24-18"><a href="compositional-analyses-of-biological-sequences.html#cb24-18" tabindex="-1"></a>            count <span class="op">=</span> getAA(<span class="bu">str</span>(record.seq), aminoacid)</span>
<span id="cb24-19"><a href="compositional-analyses-of-biological-sequences.html#cb24-19" tabindex="-1"></a>            AA_content.append([count, name])</span>
<span id="cb24-20"><a href="compositional-analyses-of-biological-sequences.html#cb24-20" tabindex="-1"></a>    name <span class="op">=</span> aminoacid</span>
<span id="cb24-21"><a href="compositional-analyses-of-biological-sequences.html#cb24-21" tabindex="-1"></a>    df_aa.append(pd.DataFrame(AA_content, columns<span class="op">=</span>[name, <span class="st">&quot;Sequence&quot;</span>]))</span>
<span id="cb24-22"><a href="compositional-analyses-of-biological-sequences.html#cb24-22" tabindex="-1"></a></span>
<span id="cb24-23"><a href="compositional-analyses-of-biological-sequences.html#cb24-23" tabindex="-1"></a>    aa_means.append(df_aa[i].groupby(<span class="st">&quot;Sequence&quot;</span>).mean())</span>
<span id="cb24-24"><a href="compositional-analyses-of-biological-sequences.html#cb24-24" tabindex="-1"></a>    aa_medians.append(df_aa[i].groupby(<span class="st">&quot;Sequence&quot;</span>).median())</span>
<span id="cb24-25"><a href="compositional-analyses-of-biological-sequences.html#cb24-25" tabindex="-1"></a>    aa_std.append(df_aa[i].groupby(<span class="st">&quot;Sequence&quot;</span>).std())</span></code></pre></div>
<p>In this case, we decided to plot and compare the medians of the amino acid proportions ( of total of amino acids in the sequence).</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb25-1"><a href="compositional-analyses-of-biological-sequences.html#cb25-1" tabindex="-1"></a>hm_medians <span class="op">=</span> pd.concat(aa_medians, axis <span class="op">=</span> <span class="dv">1</span>)</span>
<span id="cb25-2"><a href="compositional-analyses-of-biological-sequences.html#cb25-2" tabindex="-1"></a>hm_medians <span class="op">=</span> hm_medians.reindex(index <span class="op">=</span> [<span class="st">&#39;RdRp&#39;</span>,<span class="st">&#39;CP&#39;</span>,<span class="st">&#39;p19&#39;</span>, <span class="st">&#39;p22&#39;</span>, <span class="st">&#39;p22 core&#39;</span>, <span class="st">&#39;other cores&#39;</span>])</span>
<span id="cb25-3"><a href="compositional-analyses-of-biological-sequences.html#cb25-3" tabindex="-1"></a></span>
<span id="cb25-4"><a href="compositional-analyses-of-biological-sequences.html#cb25-4" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots()</span>
<span id="cb25-5"><a href="compositional-analyses-of-biological-sequences.html#cb25-5" tabindex="-1"></a>plt.rcParams[<span class="st">&quot;figure.figsize&quot;</span>] <span class="op">=</span> [<span class="dv">12</span>,<span class="fl">1.5</span>]</span>
<span id="cb25-6"><a href="compositional-analyses-of-biological-sequences.html#cb25-6" tabindex="-1"></a>sns.heatmap(hm_medians, ax<span class="op">=</span>ax, cmap<span class="op">=</span><span class="st">&quot;viridis&quot;</span>)</span>
<span id="cb25-7"><a href="compositional-analyses-of-biological-sequences.html#cb25-7" tabindex="-1"></a>plt.xlabel(<span class="st">&quot;Amino acid&quot;</span>)</span>
<span id="cb25-8"><a href="compositional-analyses-of-biological-sequences.html#cb25-8" tabindex="-1"></a></span>
<span id="cb25-9"><a href="compositional-analyses-of-biological-sequences.html#cb25-9" tabindex="-1"></a>plt.tight_layout()</span></code></pre></div>
</div>
<div id="nucleotides" class="section level2 hasAnchor" number="5.2">
<h2><span class="header-section-number">5.2</span> Nucleotides<a href="compositional-analyses-of-biological-sequences.html#nucleotides" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The first approach we tried was counting the proportion of nucleotides in different genes and also within the overlapping region of p19 and p22 genes.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb26-1"><a href="compositional-analyses-of-biological-sequences.html#cb26-1" tabindex="-1"></a><span class="kw">def</span> getNT(seq, letter):</span>
<span id="cb26-2"><a href="compositional-analyses-of-biological-sequences.html#cb26-2" tabindex="-1"></a>    <span class="cf">return</span>(<span class="bu">round</span>((<span class="bu">sum</span>([<span class="fl">1.0</span> <span class="cf">for</span> nucl <span class="kw">in</span> seq <span class="cf">if</span> nucl <span class="kw">in</span> [letter]]) <span class="op">/</span> (<span class="bu">len</span>(seq)<span class="op">-</span> seq.count(<span class="st">&quot;-&quot;</span>))), ndigits<span class="op">=</span><span class="dv">2</span>))</span>
<span id="cb26-3"><a href="compositional-analyses-of-biological-sequences.html#cb26-3" tabindex="-1"></a></span>
<span id="cb26-4"><a href="compositional-analyses-of-biological-sequences.html#cb26-4" tabindex="-1"></a>paths <span class="op">=</span> [<span class="st">&quot;RdRp_nt_aligned.fa&quot;</span>, <span class="st">&quot;CP_nt_aligned.fa&quot;</span>, <span class="st">&quot;p19_nt_aligned.fa&quot;</span>, <span class="st">&quot;p22_nt_aligned.fa&quot;</span>, <span class="st">&quot;/Users/esmeralda/Documents/TFM/RevTrans/final/nt/alignments/concatenate/p22 core_nt_alignment.fa&quot;</span>, <span class="st">&quot;other cores_nt.fa&quot;</span>] </span>
<span id="cb26-5"><a href="compositional-analyses-of-biological-sequences.html#cb26-5" tabindex="-1"></a>letters <span class="op">=</span> [<span class="st">&quot;A&quot;</span>, <span class="st">&quot;T&quot;</span>, <span class="st">&quot;C&quot;</span>, <span class="st">&quot;G&quot;</span>]</span>
<span id="cb26-6"><a href="compositional-analyses-of-biological-sequences.html#cb26-6" tabindex="-1"></a>df_nt <span class="op">=</span> []</span>
<span id="cb26-7"><a href="compositional-analyses-of-biological-sequences.html#cb26-7" tabindex="-1"></a>nt_only_means <span class="op">=</span> []</span>
<span id="cb26-8"><a href="compositional-analyses-of-biological-sequences.html#cb26-8" tabindex="-1"></a>nt_only_medians <span class="op">=</span> []</span>
<span id="cb26-9"><a href="compositional-analyses-of-biological-sequences.html#cb26-9" tabindex="-1"></a>nt_only_std <span class="op">=</span> []</span>
<span id="cb26-10"><a href="compositional-analyses-of-biological-sequences.html#cb26-10" tabindex="-1"></a></span>
<span id="cb26-11"><a href="compositional-analyses-of-biological-sequences.html#cb26-11" tabindex="-1"></a><span class="cf">for</span> i, nucleot <span class="kw">in</span> <span class="bu">enumerate</span>(letters):</span>
<span id="cb26-12"><a href="compositional-analyses-of-biological-sequences.html#cb26-12" tabindex="-1"></a>    NT_content <span class="op">=</span> []</span>
<span id="cb26-13"><a href="compositional-analyses-of-biological-sequences.html#cb26-13" tabindex="-1"></a>    <span class="cf">for</span> alignment <span class="kw">in</span> paths:</span>
<span id="cb26-14"><a href="compositional-analyses-of-biological-sequences.html#cb26-14" tabindex="-1"></a>        name <span class="op">=</span> alignment[<span class="dv">0</span>:alignment.find(<span class="st">&quot;_&quot;</span>)]</span>
<span id="cb26-15"><a href="compositional-analyses-of-biological-sequences.html#cb26-15" tabindex="-1"></a>        <span class="cf">if</span> <span class="bu">len</span>(name) <span class="op">&gt;</span> <span class="dv">50</span>:</span>
<span id="cb26-16"><a href="compositional-analyses-of-biological-sequences.html#cb26-16" tabindex="-1"></a>             name <span class="op">=</span> <span class="st">&quot;p22 core&quot;</span></span>
<span id="cb26-17"><a href="compositional-analyses-of-biological-sequences.html#cb26-17" tabindex="-1"></a>        <span class="cf">for</span> record <span class="kw">in</span> SeqIO.parse(alignment, <span class="st">&quot;fasta&quot;</span>):</span>
<span id="cb26-18"><a href="compositional-analyses-of-biological-sequences.html#cb26-18" tabindex="-1"></a>            count <span class="op">=</span> getNT(<span class="bu">str</span>(record.seq), nucleot)</span>
<span id="cb26-19"><a href="compositional-analyses-of-biological-sequences.html#cb26-19" tabindex="-1"></a>            NT_content.append([count, name])</span>
<span id="cb26-20"><a href="compositional-analyses-of-biological-sequences.html#cb26-20" tabindex="-1"></a>    <span class="cf">if</span> nucleot <span class="op">==</span> <span class="st">&quot;T&quot;</span>:</span>
<span id="cb26-21"><a href="compositional-analyses-of-biological-sequences.html#cb26-21" tabindex="-1"></a>        nucleot <span class="op">=</span> <span class="st">&quot;U&quot;</span></span>
<span id="cb26-22"><a href="compositional-analyses-of-biological-sequences.html#cb26-22" tabindex="-1"></a>    </span>
<span id="cb26-23"><a href="compositional-analyses-of-biological-sequences.html#cb26-23" tabindex="-1"></a>    name <span class="op">=</span> nucleot</span>
<span id="cb26-24"><a href="compositional-analyses-of-biological-sequences.html#cb26-24" tabindex="-1"></a>    df_nt.append(pd.DataFrame(NT_content, columns<span class="op">=</span>[name, <span class="st">&quot;Sequence&quot;</span>]))</span>
<span id="cb26-25"><a href="compositional-analyses-of-biological-sequences.html#cb26-25" tabindex="-1"></a></span>
<span id="cb26-26"><a href="compositional-analyses-of-biological-sequences.html#cb26-26" tabindex="-1"></a>    nt_only_means.append(df_nt[i].groupby(<span class="st">&quot;Sequence&quot;</span>).mean())</span>
<span id="cb26-27"><a href="compositional-analyses-of-biological-sequences.html#cb26-27" tabindex="-1"></a>    nt_only_medians.append(df_nt[i].groupby(<span class="st">&quot;Sequence&quot;</span>).median())</span>
<span id="cb26-28"><a href="compositional-analyses-of-biological-sequences.html#cb26-28" tabindex="-1"></a>    nt_only_std.append(df_nt[i].groupby(<span class="st">&quot;Sequence&quot;</span>).std())</span></code></pre></div>
<p>Besides, we were interested in the positions of codons (1st, 2nd and 3rd nucleotides) and its nucleotidic abundances.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb27-1"><a href="compositional-analyses-of-biological-sequences.html#cb27-1" tabindex="-1"></a><span class="kw">def</span> getPosCodon(seq, letter, start):</span>
<span id="cb27-2"><a href="compositional-analyses-of-biological-sequences.html#cb27-2" tabindex="-1"></a>    count <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb27-3"><a href="compositional-analyses-of-biological-sequences.html#cb27-3" tabindex="-1"></a>    <span class="cf">for</span> nucl <span class="kw">in</span> <span class="bu">range</span>(start, <span class="bu">len</span>(seq), <span class="dv">3</span>):</span>
<span id="cb27-4"><a href="compositional-analyses-of-biological-sequences.html#cb27-4" tabindex="-1"></a>        <span class="cf">if</span> seq[nucl] <span class="op">==</span> letter:</span>
<span id="cb27-5"><a href="compositional-analyses-of-biological-sequences.html#cb27-5" tabindex="-1"></a>            count <span class="op">+=</span> <span class="fl">1.0</span></span>
<span id="cb27-6"><a href="compositional-analyses-of-biological-sequences.html#cb27-6" tabindex="-1"></a>    </span>
<span id="cb27-7"><a href="compositional-analyses-of-biological-sequences.html#cb27-7" tabindex="-1"></a>    <span class="cf">return</span> <span class="bu">round</span>(count<span class="op">/</span>((<span class="bu">len</span>(seq)<span class="op">-</span> seq.count(<span class="st">&quot;-&quot;</span>))<span class="op">/</span><span class="dv">3</span>), ndigits<span class="op">=</span><span class="dv">2</span>)</span>
<span id="cb27-8"><a href="compositional-analyses-of-biological-sequences.html#cb27-8" tabindex="-1"></a></span>
<span id="cb27-9"><a href="compositional-analyses-of-biological-sequences.html#cb27-9" tabindex="-1"></a>general <span class="op">=</span> []</span>
<span id="cb27-10"><a href="compositional-analyses-of-biological-sequences.html#cb27-10" tabindex="-1"></a></span>
<span id="cb27-11"><a href="compositional-analyses-of-biological-sequences.html#cb27-11" tabindex="-1"></a><span class="cf">for</span> start <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">3</span>):</span>
<span id="cb27-12"><a href="compositional-analyses-of-biological-sequences.html#cb27-12" tabindex="-1"></a>    df_nt <span class="op">=</span> []</span>
<span id="cb27-13"><a href="compositional-analyses-of-biological-sequences.html#cb27-13" tabindex="-1"></a>    nt_pos_means <span class="op">=</span> []</span>
<span id="cb27-14"><a href="compositional-analyses-of-biological-sequences.html#cb27-14" tabindex="-1"></a>    nt_pos_medians <span class="op">=</span> []</span>
<span id="cb27-15"><a href="compositional-analyses-of-biological-sequences.html#cb27-15" tabindex="-1"></a>    nt_pos_std <span class="op">=</span> []</span>
<span id="cb27-16"><a href="compositional-analyses-of-biological-sequences.html#cb27-16" tabindex="-1"></a></span>
<span id="cb27-17"><a href="compositional-analyses-of-biological-sequences.html#cb27-17" tabindex="-1"></a>    <span class="cf">for</span> i, nucleot <span class="kw">in</span> <span class="bu">enumerate</span>(letters):</span>
<span id="cb27-18"><a href="compositional-analyses-of-biological-sequences.html#cb27-18" tabindex="-1"></a>        NT_content <span class="op">=</span> []</span>
<span id="cb27-19"><a href="compositional-analyses-of-biological-sequences.html#cb27-19" tabindex="-1"></a>        <span class="cf">for</span> alignment <span class="kw">in</span> paths:</span>
<span id="cb27-20"><a href="compositional-analyses-of-biological-sequences.html#cb27-20" tabindex="-1"></a>            name <span class="op">=</span> alignment[<span class="dv">0</span>:alignment.find(<span class="st">&quot;_&quot;</span>)]</span>
<span id="cb27-21"><a href="compositional-analyses-of-biological-sequences.html#cb27-21" tabindex="-1"></a>            <span class="cf">if</span> <span class="bu">len</span>(name) <span class="op">&gt;</span> <span class="dv">50</span>:</span>
<span id="cb27-22"><a href="compositional-analyses-of-biological-sequences.html#cb27-22" tabindex="-1"></a>                name <span class="op">=</span> <span class="st">&quot;p22 core&quot;</span></span>
<span id="cb27-23"><a href="compositional-analyses-of-biological-sequences.html#cb27-23" tabindex="-1"></a>            <span class="cf">for</span> record <span class="kw">in</span> SeqIO.parse(alignment, <span class="st">&quot;fasta&quot;</span>):</span>
<span id="cb27-24"><a href="compositional-analyses-of-biological-sequences.html#cb27-24" tabindex="-1"></a>                count <span class="op">=</span> getPosCodon(<span class="bu">str</span>(record.seq), nucleot, start)</span>
<span id="cb27-25"><a href="compositional-analyses-of-biological-sequences.html#cb27-25" tabindex="-1"></a>                NT_content.append([count, name])</span>
<span id="cb27-26"><a href="compositional-analyses-of-biological-sequences.html#cb27-26" tabindex="-1"></a>        </span>
<span id="cb27-27"><a href="compositional-analyses-of-biological-sequences.html#cb27-27" tabindex="-1"></a>        <span class="cf">if</span> nucleot <span class="op">==</span> <span class="st">&quot;T&quot;</span>:</span>
<span id="cb27-28"><a href="compositional-analyses-of-biological-sequences.html#cb27-28" tabindex="-1"></a>            name <span class="op">=</span> <span class="st">&quot;U&quot;</span> <span class="op">+</span> <span class="bu">str</span>(start <span class="op">+</span> <span class="dv">1</span>)</span>
<span id="cb27-29"><a href="compositional-analyses-of-biological-sequences.html#cb27-29" tabindex="-1"></a>        <span class="cf">else</span>:</span>
<span id="cb27-30"><a href="compositional-analyses-of-biological-sequences.html#cb27-30" tabindex="-1"></a>            name <span class="op">=</span> nucleot <span class="op">+</span> <span class="bu">str</span>(start <span class="op">+</span> <span class="dv">1</span>)</span>
<span id="cb27-31"><a href="compositional-analyses-of-biological-sequences.html#cb27-31" tabindex="-1"></a>            </span>
<span id="cb27-32"><a href="compositional-analyses-of-biological-sequences.html#cb27-32" tabindex="-1"></a>        df_nt.append(pd.DataFrame(NT_content, columns<span class="op">=</span>[name, <span class="st">&quot;Sequence&quot;</span>]))</span>
<span id="cb27-33"><a href="compositional-analyses-of-biological-sequences.html#cb27-33" tabindex="-1"></a></span>
<span id="cb27-34"><a href="compositional-analyses-of-biological-sequences.html#cb27-34" tabindex="-1"></a>        nt_pos_means.append(df_nt[i].groupby(<span class="st">&quot;Sequence&quot;</span>).mean())</span>
<span id="cb27-35"><a href="compositional-analyses-of-biological-sequences.html#cb27-35" tabindex="-1"></a>        nt_pos_medians.append(df_nt[i].groupby(<span class="st">&quot;Sequence&quot;</span>).median())</span>
<span id="cb27-36"><a href="compositional-analyses-of-biological-sequences.html#cb27-36" tabindex="-1"></a>        nt_pos_std.append(df_nt[i].groupby(<span class="st">&quot;Sequence&quot;</span>).std())</span>
<span id="cb27-37"><a href="compositional-analyses-of-biological-sequences.html#cb27-37" tabindex="-1"></a></span>
<span id="cb27-38"><a href="compositional-analyses-of-biological-sequences.html#cb27-38" tabindex="-1"></a>    general.append(nt_pos_medians)</span></code></pre></div>
<p>Now we join all data of nucleotides and plot a heatmap.</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb28-1"><a href="compositional-analyses-of-biological-sequences.html#cb28-1" tabindex="-1"></a>nucleotides_matrix <span class="op">=</span> pd.concat([nt_only_medians[<span class="dv">0</span>],nt_only_medians[<span class="dv">1</span>],nt_only_medians[<span class="dv">2</span>],nt_only_medians[<span class="dv">3</span>],</span>
<span id="cb28-2"><a href="compositional-analyses-of-biological-sequences.html#cb28-2" tabindex="-1"></a>                                                general[<span class="dv">0</span>][<span class="dv">0</span>], general[<span class="dv">0</span>][<span class="dv">1</span>], general[<span class="dv">0</span>][<span class="dv">2</span>], general[<span class="dv">0</span>][<span class="dv">3</span>],</span>
<span id="cb28-3"><a href="compositional-analyses-of-biological-sequences.html#cb28-3" tabindex="-1"></a>                                                general[<span class="dv">1</span>][<span class="dv">0</span>], general[<span class="dv">1</span>][<span class="dv">1</span>], general[<span class="dv">1</span>][<span class="dv">2</span>], general[<span class="dv">1</span>][<span class="dv">3</span>],</span>
<span id="cb28-4"><a href="compositional-analyses-of-biological-sequences.html#cb28-4" tabindex="-1"></a>                                                general[<span class="dv">2</span>][<span class="dv">0</span>],general[<span class="dv">2</span>][<span class="dv">1</span>], general[<span class="dv">2</span>][<span class="dv">2</span>], general[<span class="dv">2</span>][<span class="dv">3</span>]],</span>
<span id="cb28-5"><a href="compositional-analyses-of-biological-sequences.html#cb28-5" tabindex="-1"></a>                                                axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb28-6"><a href="compositional-analyses-of-biological-sequences.html#cb28-6" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots()</span>
<span id="cb28-7"><a href="compositional-analyses-of-biological-sequences.html#cb28-7" tabindex="-1"></a>plt.rcParams[<span class="st">&quot;figure.figsize&quot;</span>] <span class="op">=</span> [<span class="dv">20</span>,<span class="fl">1.5</span>]</span>
<span id="cb28-8"><a href="compositional-analyses-of-biological-sequences.html#cb28-8" tabindex="-1"></a>nucleotides_matrix <span class="op">=</span> nucleotides_matrix.reindex(index <span class="op">=</span> [<span class="st">&#39;RdRp&#39;</span>,<span class="st">&#39;CP&#39;</span>,<span class="st">&#39;p19&#39;</span>, <span class="st">&#39;p22&#39;</span>, <span class="st">&#39;p22 core&#39;</span>, <span class="st">&#39;other cores&#39;</span>])</span>
<span id="cb28-9"><a href="compositional-analyses-of-biological-sequences.html#cb28-9" tabindex="-1"></a>sns.heatmap(nucleotides_matrix, ax<span class="op">=</span>ax, cmap<span class="op">=</span><span class="st">&quot;viridis&quot;</span>)</span>
<span id="cb28-10"><a href="compositional-analyses-of-biological-sequences.html#cb28-10" tabindex="-1"></a>plt.xlabel(<span class="st">&quot;Nucleotide&quot;</span>)</span>
<span id="cb28-11"><a href="compositional-analyses-of-biological-sequences.html#cb28-11" tabindex="-1"></a></span>
<span id="cb28-12"><a href="compositional-analyses-of-biological-sequences.html#cb28-12" tabindex="-1"></a>plt.tight_layout()</span></code></pre></div>
</div>
<div id="dinuclotides" class="section level2 hasAnchor" number="5.3">
<h2><span class="header-section-number">5.3</span> Dinuclotides<a href="compositional-analyses-of-biological-sequences.html#dinuclotides" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>We read dinucleotides, calculate a relativized value and plot the medians.</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb29-1"><a href="compositional-analyses-of-biological-sequences.html#cb29-1" tabindex="-1"></a></span>
<span id="cb29-2"><a href="compositional-analyses-of-biological-sequences.html#cb29-2" tabindex="-1"></a><span class="kw">def</span> read_diNT(path):</span>
<span id="cb29-3"><a href="compositional-analyses-of-biological-sequences.html#cb29-3" tabindex="-1"></a>    chains_dint <span class="op">=</span> SeqIO.parse(path, <span class="st">&quot;fasta&quot;</span>)</span>
<span id="cb29-4"><a href="compositional-analyses-of-biological-sequences.html#cb29-4" tabindex="-1"></a></span>
<span id="cb29-5"><a href="compositional-analyses-of-biological-sequences.html#cb29-5" tabindex="-1"></a>    chain_dint_p22_core <span class="op">=</span> []</span>
<span id="cb29-6"><a href="compositional-analyses-of-biological-sequences.html#cb29-6" tabindex="-1"></a>    <span class="cf">for</span> chain <span class="kw">in</span> chains_dint:</span>
<span id="cb29-7"><a href="compositional-analyses-of-biological-sequences.html#cb29-7" tabindex="-1"></a>        dint <span class="op">=</span> {}</span>
<span id="cb29-8"><a href="compositional-analyses-of-biological-sequences.html#cb29-8" tabindex="-1"></a>        seq <span class="op">=</span> <span class="bu">str</span>(chain.seq)</span>
<span id="cb29-9"><a href="compositional-analyses-of-biological-sequences.html#cb29-9" tabindex="-1"></a>        <span class="cf">for</span> ch <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(seq)<span class="op">-</span><span class="dv">1</span>):</span>
<span id="cb29-10"><a href="compositional-analyses-of-biological-sequences.html#cb29-10" tabindex="-1"></a>            dinuc <span class="op">=</span> seq[ch:ch<span class="op">+</span><span class="dv">2</span>].replace(<span class="st">&quot;T&quot;</span>, <span class="st">&quot;U&quot;</span>)</span>
<span id="cb29-11"><a href="compositional-analyses-of-biological-sequences.html#cb29-11" tabindex="-1"></a>            <span class="cf">if</span> dinuc <span class="kw">in</span> dint:</span>
<span id="cb29-12"><a href="compositional-analyses-of-biological-sequences.html#cb29-12" tabindex="-1"></a>                dint[dinuc] <span class="op">+=</span> <span class="dv">1</span></span>
<span id="cb29-13"><a href="compositional-analyses-of-biological-sequences.html#cb29-13" tabindex="-1"></a>            <span class="cf">else</span>:</span>
<span id="cb29-14"><a href="compositional-analyses-of-biological-sequences.html#cb29-14" tabindex="-1"></a>                <span class="cf">if</span> dinuc.find(<span class="st">&quot;-&quot;</span>)<span class="op">&lt;</span><span class="dv">0</span> <span class="kw">and</span> dinuc.find(<span class="st">&quot;N&quot;</span>)<span class="op">&lt;</span><span class="dv">0</span>:</span>
<span id="cb29-15"><a href="compositional-analyses-of-biological-sequences.html#cb29-15" tabindex="-1"></a>                    dint[dinuc] <span class="op">=</span> <span class="dv">1</span></span>
<span id="cb29-16"><a href="compositional-analyses-of-biological-sequences.html#cb29-16" tabindex="-1"></a>        chain_dint_p22_core.append(dint)</span>
<span id="cb29-17"><a href="compositional-analyses-of-biological-sequences.html#cb29-17" tabindex="-1"></a></span>
<span id="cb29-18"><a href="compositional-analyses-of-biological-sequences.html#cb29-18" tabindex="-1"></a>    <span class="cf">return</span> pd.DataFrame.from_dict(chain_dint_p22_core)</span>
<span id="cb29-19"><a href="compositional-analyses-of-biological-sequences.html#cb29-19" tabindex="-1"></a></span>
<span id="cb29-20"><a href="compositional-analyses-of-biological-sequences.html#cb29-20" tabindex="-1"></a><span class="kw">def</span> rel_diNT(df_dint):</span>
<span id="cb29-21"><a href="compositional-analyses-of-biological-sequences.html#cb29-21" tabindex="-1"></a>    <span class="cf">return</span> df_dint.divide(df_dint.<span class="bu">sum</span>(axis<span class="op">=</span><span class="dv">1</span>), axis<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb29-22"><a href="compositional-analyses-of-biological-sequences.html#cb29-22" tabindex="-1"></a></span>
<span id="cb29-23"><a href="compositional-analyses-of-biological-sequences.html#cb29-23" tabindex="-1"></a><span class="kw">def</span> stats_composition(df_composition, stat<span class="op">=</span><span class="st">&quot;median&quot;</span>):</span>
<span id="cb29-24"><a href="compositional-analyses-of-biological-sequences.html#cb29-24" tabindex="-1"></a>    <span class="cf">if</span> stat <span class="op">==</span> <span class="st">&quot;median&quot;</span>:</span>
<span id="cb29-25"><a href="compositional-analyses-of-biological-sequences.html#cb29-25" tabindex="-1"></a>        <span class="cf">return</span> pd.DataFrame.from_dict(df_composition.mean())</span>
<span id="cb29-26"><a href="compositional-analyses-of-biological-sequences.html#cb29-26" tabindex="-1"></a>    <span class="cf">elif</span> stat <span class="op">==</span> <span class="st">&quot;mean&quot;</span>:</span>
<span id="cb29-27"><a href="compositional-analyses-of-biological-sequences.html#cb29-27" tabindex="-1"></a>        <span class="cf">return</span> pd.DataFrame.from_dict(df_composition.median())</span>
<span id="cb29-28"><a href="compositional-analyses-of-biological-sequences.html#cb29-28" tabindex="-1"></a>    <span class="cf">elif</span> stat <span class="op">==</span> <span class="st">&quot;std&quot;</span>:</span>
<span id="cb29-29"><a href="compositional-analyses-of-biological-sequences.html#cb29-29" tabindex="-1"></a>        <span class="cf">return</span> pd.DataFrame.from_dict(df_composition.std())</span>
<span id="cb29-30"><a href="compositional-analyses-of-biological-sequences.html#cb29-30" tabindex="-1"></a>    </span>
<span id="cb29-31"><a href="compositional-analyses-of-biological-sequences.html#cb29-31" tabindex="-1"></a><span class="kw">def</span> join_medians(median_df):</span>
<span id="cb29-32"><a href="compositional-analyses-of-biological-sequences.html#cb29-32" tabindex="-1"></a>    dint_comparison <span class="op">=</span> pd.concat([median_df[<span class="dv">0</span>],</span>
<span id="cb29-33"><a href="compositional-analyses-of-biological-sequences.html#cb29-33" tabindex="-1"></a>                                median_df[<span class="dv">1</span>],</span>
<span id="cb29-34"><a href="compositional-analyses-of-biological-sequences.html#cb29-34" tabindex="-1"></a>                                median_df[<span class="dv">2</span>],</span>
<span id="cb29-35"><a href="compositional-analyses-of-biological-sequences.html#cb29-35" tabindex="-1"></a>                                median_df[<span class="dv">3</span>],</span>
<span id="cb29-36"><a href="compositional-analyses-of-biological-sequences.html#cb29-36" tabindex="-1"></a>                                median_df[<span class="dv">4</span>],</span>
<span id="cb29-37"><a href="compositional-analyses-of-biological-sequences.html#cb29-37" tabindex="-1"></a>                                median_df[<span class="dv">5</span>][<span class="dv">0</span>][<span class="bu">list</span>(median_df[<span class="dv">0</span>].index)]], axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb29-38"><a href="compositional-analyses-of-biological-sequences.html#cb29-38" tabindex="-1"></a></span>
<span id="cb29-39"><a href="compositional-analyses-of-biological-sequences.html#cb29-39" tabindex="-1"></a>    dint_comparison.columns <span class="op">=</span> [<span class="st">&quot;RdRp&quot;</span>, <span class="st">&quot;CP&quot;</span>, <span class="st">&quot;p19&quot;</span>, <span class="st">&quot;p22&quot;</span>, <span class="st">&quot;p22_core&quot;</span>, <span class="st">&quot;other cores&quot;</span>]</span>
<span id="cb29-40"><a href="compositional-analyses-of-biological-sequences.html#cb29-40" tabindex="-1"></a></span>
<span id="cb29-41"><a href="compositional-analyses-of-biological-sequences.html#cb29-41" tabindex="-1"></a>    <span class="cf">return</span> dint_comparison.T</span></code></pre></div>
<p>Then, we plot a heatmap with the shares of both nucleotides and nucleotides per codon position.</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb30-1"><a href="compositional-analyses-of-biological-sequences.html#cb30-1" tabindex="-1"></a><span class="co"># Paths of fasta files with nucleotide sequences</span></span>
<span id="cb30-2"><a href="compositional-analyses-of-biological-sequences.html#cb30-2" tabindex="-1"></a>paths <span class="op">=</span> [<span class="st">&quot;RdRp_nt_aligned.fa&quot;</span>,</span>
<span id="cb30-3"><a href="compositional-analyses-of-biological-sequences.html#cb30-3" tabindex="-1"></a>         <span class="st">&quot;cp_nt_aligned.fa&quot;</span>,</span>
<span id="cb30-4"><a href="compositional-analyses-of-biological-sequences.html#cb30-4" tabindex="-1"></a>         <span class="st">&quot;p19_nt_aligned.fa&quot;</span>,</span>
<span id="cb30-5"><a href="compositional-analyses-of-biological-sequences.html#cb30-5" tabindex="-1"></a>         <span class="st">&quot;p22_nt_aligned.fa&quot;</span>,</span>
<span id="cb30-6"><a href="compositional-analyses-of-biological-sequences.html#cb30-6" tabindex="-1"></a>         <span class="st">&quot;/Users/esmeralda/Documents/TFM/RevTrans/final/nt/alignments/concatenate/p22 core_nt_alignment.fa&quot;</span>,</span>
<span id="cb30-7"><a href="compositional-analyses-of-biological-sequences.html#cb30-7" tabindex="-1"></a>         <span class="st">&quot;other cores_nt.fa&quot;</span>]</span>
<span id="cb30-8"><a href="compositional-analyses-of-biological-sequences.html#cb30-8" tabindex="-1"></a></span>
<span id="cb30-9"><a href="compositional-analyses-of-biological-sequences.html#cb30-9" tabindex="-1"></a>order_names <span class="op">=</span> []</span>
<span id="cb30-10"><a href="compositional-analyses-of-biological-sequences.html#cb30-10" tabindex="-1"></a>median_df <span class="op">=</span> []</span>
<span id="cb30-11"><a href="compositional-analyses-of-biological-sequences.html#cb30-11" tabindex="-1"></a></span>
<span id="cb30-12"><a href="compositional-analyses-of-biological-sequences.html#cb30-12" tabindex="-1"></a><span class="cf">for</span> path <span class="kw">in</span> paths:</span>
<span id="cb30-13"><a href="compositional-analyses-of-biological-sequences.html#cb30-13" tabindex="-1"></a>    df_dint <span class="op">=</span> read_diNT(path)</span>
<span id="cb30-14"><a href="compositional-analyses-of-biological-sequences.html#cb30-14" tabindex="-1"></a>    df_dint_rel <span class="op">=</span> rel_diNT(df_dint)</span>
<span id="cb30-15"><a href="compositional-analyses-of-biological-sequences.html#cb30-15" tabindex="-1"></a>    median_df.append(stats_composition(df_dint_rel))</span>
<span id="cb30-16"><a href="compositional-analyses-of-biological-sequences.html#cb30-16" tabindex="-1"></a></span>
<span id="cb30-17"><a href="compositional-analyses-of-biological-sequences.html#cb30-17" tabindex="-1"></a>summary_diNT <span class="op">=</span> join_medians(median_df)</span>
<span id="cb30-18"><a href="compositional-analyses-of-biological-sequences.html#cb30-18" tabindex="-1"></a></span>
<span id="cb30-19"><a href="compositional-analyses-of-biological-sequences.html#cb30-19" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots()</span>
<span id="cb30-20"><a href="compositional-analyses-of-biological-sequences.html#cb30-20" tabindex="-1"></a>plt.rcParams[<span class="st">&quot;figure.figsize&quot;</span>] <span class="op">=</span> [<span class="dv">12</span>,<span class="fl">1.5</span>]</span>
<span id="cb30-21"><a href="compositional-analyses-of-biological-sequences.html#cb30-21" tabindex="-1"></a>sns.heatmap(summary_diNT,ax<span class="op">=</span>ax, cmap<span class="op">=</span><span class="st">&quot;viridis&quot;</span>)</span>
<span id="cb30-22"><a href="compositional-analyses-of-biological-sequences.html#cb30-22" tabindex="-1"></a>plt.xlabel(<span class="st">&quot;Dinucleotide&quot;</span>)</span>
<span id="cb30-23"><a href="compositional-analyses-of-biological-sequences.html#cb30-23" tabindex="-1"></a></span>
<span id="cb30-24"><a href="compositional-analyses-of-biological-sequences.html#cb30-24" tabindex="-1"></a>plt.tight_layout()</span></code></pre></div>
</div>
<div id="codons" class="section level2 hasAnchor" number="5.4">
<h2><span class="header-section-number">5.4</span> Codons<a href="compositional-analyses-of-biological-sequences.html#codons" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>We write the paths of the fasta files.</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb31-1"><a href="compositional-analyses-of-biological-sequences.html#cb31-1" tabindex="-1"></a>concatenate_dir <span class="op">=</span> <span class="st">&quot;/Users/esmeralda/Documents/TFM/RevTrans/final/nt/alignments/concatenate/&quot;</span></span>
<span id="cb31-2"><a href="compositional-analyses-of-biological-sequences.html#cb31-2" tabindex="-1"></a>paths <span class="op">=</span> [ concatenate_dir <span class="op">+</span> <span class="st">&quot;RdRp_nt_aligned.fasta&quot;</span>,</span>
<span id="cb31-3"><a href="compositional-analyses-of-biological-sequences.html#cb31-3" tabindex="-1"></a>         concatenate_dir <span class="op">+</span> <span class="st">&quot;cp_nt_aligned_filtered.fasta&quot;</span>,</span>
<span id="cb31-4"><a href="compositional-analyses-of-biological-sequences.html#cb31-4" tabindex="-1"></a>         concatenate_dir <span class="op">+</span> <span class="st">&quot;p19_nt_alignment.fasta&quot;</span>,</span>
<span id="cb31-5"><a href="compositional-analyses-of-biological-sequences.html#cb31-5" tabindex="-1"></a>         concatenate_dir <span class="op">+</span> <span class="st">&quot;p22_nt_alignment.fasta&quot;</span>,</span>
<span id="cb31-6"><a href="compositional-analyses-of-biological-sequences.html#cb31-6" tabindex="-1"></a>         concatenate_dir <span class="op">+</span> <span class="st">&quot;p22 core_nt_alignment.fa&quot;</span>,</span>
<span id="cb31-7"><a href="compositional-analyses-of-biological-sequences.html#cb31-7" tabindex="-1"></a>         <span class="co"># concatenate_dir + &quot;union/sequence_data.fas&quot;,</span></span>
<span id="cb31-8"><a href="compositional-analyses-of-biological-sequences.html#cb31-8" tabindex="-1"></a>         <span class="st">&quot;/Users/esmeralda/Documents/TFM/overlap_comparisons/other cores_nt.fa&quot;</span>]</span>
<span id="cb31-9"><a href="compositional-analyses-of-biological-sequences.html#cb31-9" tabindex="-1"></a></span>
<span id="cb31-10"><a href="compositional-analyses-of-biological-sequences.html#cb31-10" tabindex="-1"></a>names <span class="op">=</span> [<span class="st">&quot;RdRp&quot;</span>, <span class="st">&quot;CP&quot;</span>, <span class="st">&quot;p19&quot;</span>, <span class="st">&quot;p22&quot;</span>, <span class="st">&quot;p22 core&quot;</span>, <span class="st">&quot;other cores&quot;</span>]</span></code></pre></div>
<p>The following steps consist of reading and counting codons. To emphasize the effect of codon usage we calculated the percentage of each codon within its group of codons that codify the same amino acid. For instance, if codons TGT and TGC are used 3 and 12 times, respectively, as both encode the amino acid C, the relative abundance is calculated as 0.2 for the former and 0.8 for the latter. Therefore, the comparison has to be made between different groups, but within the same amino acid.</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb32-1"><a href="compositional-analyses-of-biological-sequences.html#cb32-1" tabindex="-1"></a><span class="kw">def</span> read_codons(path):</span>
<span id="cb32-2"><a href="compositional-analyses-of-biological-sequences.html#cb32-2" tabindex="-1"></a>    n <span class="op">=</span> <span class="dv">3</span></span>
<span id="cb32-3"><a href="compositional-analyses-of-biological-sequences.html#cb32-3" tabindex="-1"></a>    record_dict <span class="op">=</span> SeqIO.to_dict(SeqIO.parse(path, <span class="st">&quot;fasta&quot;</span>))</span>
<span id="cb32-4"><a href="compositional-analyses-of-biological-sequences.html#cb32-4" tabindex="-1"></a>    codon_dict <span class="op">=</span> {}</span>
<span id="cb32-5"><a href="compositional-analyses-of-biological-sequences.html#cb32-5" tabindex="-1"></a>    <span class="cf">for</span> seq_name <span class="kw">in</span> record_dict.keys():</span>
<span id="cb32-6"><a href="compositional-analyses-of-biological-sequences.html#cb32-6" tabindex="-1"></a>        seq <span class="op">=</span> <span class="bu">str</span>(record_dict[seq_name].seq)</span>
<span id="cb32-7"><a href="compositional-analyses-of-biological-sequences.html#cb32-7" tabindex="-1"></a>        codon_dict[seq_name] <span class="op">=</span> [seq[i:i<span class="op">+</span>n] <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">0</span>, <span class="bu">len</span>(seq), n)]</span>
<span id="cb32-8"><a href="compositional-analyses-of-biological-sequences.html#cb32-8" tabindex="-1"></a>    </span>
<span id="cb32-9"><a href="compositional-analyses-of-biological-sequences.html#cb32-9" tabindex="-1"></a>    <span class="cf">return</span> pd.DataFrame.from_dict(codon_dict)</span>
<span id="cb32-10"><a href="compositional-analyses-of-biological-sequences.html#cb32-10" tabindex="-1"></a></span>
<span id="cb32-11"><a href="compositional-analyses-of-biological-sequences.html#cb32-11" tabindex="-1"></a><span class="kw">def</span> count_codons(df_codons):</span>
<span id="cb32-12"><a href="compositional-analyses-of-biological-sequences.html#cb32-12" tabindex="-1"></a>    counts <span class="op">=</span> []</span>
<span id="cb32-13"><a href="compositional-analyses-of-biological-sequences.html#cb32-13" tabindex="-1"></a></span>
<span id="cb32-14"><a href="compositional-analyses-of-biological-sequences.html#cb32-14" tabindex="-1"></a>    <span class="cf">for</span> column <span class="kw">in</span> df_codons.columns:</span>
<span id="cb32-15"><a href="compositional-analyses-of-biological-sequences.html#cb32-15" tabindex="-1"></a>        counts_num <span class="op">=</span> df_codons[column].value_counts()</span>
<span id="cb32-16"><a href="compositional-analyses-of-biological-sequences.html#cb32-16" tabindex="-1"></a>        counts.append(counts_num)</span>
<span id="cb32-17"><a href="compositional-analyses-of-biological-sequences.html#cb32-17" tabindex="-1"></a></span>
<span id="cb32-18"><a href="compositional-analyses-of-biological-sequences.html#cb32-18" tabindex="-1"></a>    codon_abs <span class="op">=</span> pd.DataFrame.from_dict(counts)</span>
<span id="cb32-19"><a href="compositional-analyses-of-biological-sequences.html#cb32-19" tabindex="-1"></a>    codon_abs.fillna(<span class="dv">0</span>, inplace<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb32-20"><a href="compositional-analyses-of-biological-sequences.html#cb32-20" tabindex="-1"></a></span>
<span id="cb32-21"><a href="compositional-analyses-of-biological-sequences.html#cb32-21" tabindex="-1"></a>    <span class="cf">return</span> codon_abs</span>
<span id="cb32-22"><a href="compositional-analyses-of-biological-sequences.html#cb32-22" tabindex="-1"></a></span>
<span id="cb32-23"><a href="compositional-analyses-of-biological-sequences.html#cb32-23" tabindex="-1"></a><span class="kw">def</span> relative_codons(codon_matrix, table_DNA_to_codon):</span>
<span id="cb32-24"><a href="compositional-analyses-of-biological-sequences.html#cb32-24" tabindex="-1"></a>    codon_rel <span class="op">=</span> pd.DataFrame()</span>
<span id="cb32-25"><a href="compositional-analyses-of-biological-sequences.html#cb32-25" tabindex="-1"></a>    </span>
<span id="cb32-26"><a href="compositional-analyses-of-biological-sequences.html#cb32-26" tabindex="-1"></a>    <span class="cf">for</span> amino <span class="kw">in</span> table_DNA_to_codon.keys():</span>
<span id="cb32-27"><a href="compositional-analyses-of-biological-sequences.html#cb32-27" tabindex="-1"></a>        <span class="cf">if</span> <span class="bu">len</span>(table_DNA_to_codon[amino]) <span class="op">&gt;</span> <span class="dv">1</span> <span class="kw">and</span> amino <span class="op">!=</span> <span class="st">&quot;STOP&quot;</span>:</span>
<span id="cb32-28"><a href="compositional-analyses-of-biological-sequences.html#cb32-28" tabindex="-1"></a>            total <span class="op">=</span> codon_matrix[table_DNA_to_codon[amino]].<span class="bu">sum</span>(axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb32-29"><a href="compositional-analyses-of-biological-sequences.html#cb32-29" tabindex="-1"></a>            res <span class="op">=</span> codon_matrix[table_DNA_to_codon[amino]].div(total, axis<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb32-30"><a href="compositional-analyses-of-biological-sequences.html#cb32-30" tabindex="-1"></a>            codon_rel <span class="op">=</span> pd.concat([codon_rel, res], axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb32-31"><a href="compositional-analyses-of-biological-sequences.html#cb32-31" tabindex="-1"></a>    </span>
<span id="cb32-32"><a href="compositional-analyses-of-biological-sequences.html#cb32-32" tabindex="-1"></a>    <span class="cf">return</span> codon_rel</span>
<span id="cb32-33"><a href="compositional-analyses-of-biological-sequences.html#cb32-33" tabindex="-1"></a></span>
<span id="cb32-34"><a href="compositional-analyses-of-biological-sequences.html#cb32-34" tabindex="-1"></a><span class="kw">def</span> join_codons(list_dfs, names):</span>
<span id="cb32-35"><a href="compositional-analyses-of-biological-sequences.html#cb32-35" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(names)):</span>
<span id="cb32-36"><a href="compositional-analyses-of-biological-sequences.html#cb32-36" tabindex="-1"></a>        list_dfs[i][<span class="st">&quot;Sequence&quot;</span>] <span class="op">=</span> names[i]</span>
<span id="cb32-37"><a href="compositional-analyses-of-biological-sequences.html#cb32-37" tabindex="-1"></a></span>
<span id="cb32-38"><a href="compositional-analyses-of-biological-sequences.html#cb32-38" tabindex="-1"></a>    <span class="cf">return</span> pd.concat([list_dfs[<span class="dv">0</span>],</span>
<span id="cb32-39"><a href="compositional-analyses-of-biological-sequences.html#cb32-39" tabindex="-1"></a>                     list_dfs[<span class="dv">1</span>], </span>
<span id="cb32-40"><a href="compositional-analyses-of-biological-sequences.html#cb32-40" tabindex="-1"></a>                     list_dfs[<span class="dv">2</span>], </span>
<span id="cb32-41"><a href="compositional-analyses-of-biological-sequences.html#cb32-41" tabindex="-1"></a>                     list_dfs[<span class="dv">3</span>], </span>
<span id="cb32-42"><a href="compositional-analyses-of-biological-sequences.html#cb32-42" tabindex="-1"></a>                     list_dfs[<span class="dv">4</span>],</span>
<span id="cb32-43"><a href="compositional-analyses-of-biological-sequences.html#cb32-43" tabindex="-1"></a>                     list_dfs[<span class="dv">5</span>]], axis<span class="op">=</span><span class="dv">0</span>).T.fillna(<span class="dv">0</span>, inplace<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb32-44"><a href="compositional-analyses-of-biological-sequences.html#cb32-44" tabindex="-1"></a></span>
<span id="cb32-45"><a href="compositional-analyses-of-biological-sequences.html#cb32-45" tabindex="-1"></a><span class="kw">def</span> join_codons_medians(list_dfs):</span>
<span id="cb32-46"><a href="compositional-analyses-of-biological-sequences.html#cb32-46" tabindex="-1"></a>    <span class="cf">return</span> pd.concat([list_dfs[<span class="dv">0</span>].median(axis<span class="op">=</span><span class="dv">0</span>), </span>
<span id="cb32-47"><a href="compositional-analyses-of-biological-sequences.html#cb32-47" tabindex="-1"></a>                          list_dfs[<span class="dv">1</span>].median(axis<span class="op">=</span><span class="dv">0</span>), </span>
<span id="cb32-48"><a href="compositional-analyses-of-biological-sequences.html#cb32-48" tabindex="-1"></a>                          list_dfs[<span class="dv">2</span>].median(axis<span class="op">=</span><span class="dv">0</span>), </span>
<span id="cb32-49"><a href="compositional-analyses-of-biological-sequences.html#cb32-49" tabindex="-1"></a>                          list_dfs[<span class="dv">3</span>].median(axis<span class="op">=</span><span class="dv">0</span>), </span>
<span id="cb32-50"><a href="compositional-analyses-of-biological-sequences.html#cb32-50" tabindex="-1"></a>                          list_dfs[<span class="dv">4</span>].median(axis<span class="op">=</span><span class="dv">0</span>),</span>
<span id="cb32-51"><a href="compositional-analyses-of-biological-sequences.html#cb32-51" tabindex="-1"></a>                          list_dfs[<span class="dv">5</span>].median(axis<span class="op">=</span><span class="dv">0</span>)], axis <span class="op">=</span> <span class="dv">1</span>).T</span></code></pre></div>
<p>Now we run the functions.</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb33-1"><a href="compositional-analyses-of-biological-sequences.html#cb33-1" tabindex="-1"></a>dfs_codons <span class="op">=</span> []</span>
<span id="cb33-2"><a href="compositional-analyses-of-biological-sequences.html#cb33-2" tabindex="-1"></a></span>
<span id="cb33-3"><a href="compositional-analyses-of-biological-sequences.html#cb33-3" tabindex="-1"></a><span class="cf">for</span> path <span class="kw">in</span> paths:</span>
<span id="cb33-4"><a href="compositional-analyses-of-biological-sequences.html#cb33-4" tabindex="-1"></a>    codons <span class="op">=</span> read_codons(path)</span>
<span id="cb33-5"><a href="compositional-analyses-of-biological-sequences.html#cb33-5" tabindex="-1"></a>    codons_abs <span class="op">=</span> count_codons(codons)</span>
<span id="cb33-6"><a href="compositional-analyses-of-biological-sequences.html#cb33-6" tabindex="-1"></a>    <span class="cf">if</span> path.find(<span class="st">&quot;p22 core&quot;</span>) <span class="op">&gt;</span> <span class="dv">0</span>:</span>
<span id="cb33-7"><a href="compositional-analyses-of-biological-sequences.html#cb33-7" tabindex="-1"></a>        codons_abs[<span class="st">&quot;CGT&quot;</span>] <span class="op">=</span> <span class="fl">0.0</span></span>
<span id="cb33-8"><a href="compositional-analyses-of-biological-sequences.html#cb33-8" tabindex="-1"></a>    dfs_codons.append(relative_codons(codons_abs, table_DNA_to_codon))</span>
<span id="cb33-9"><a href="compositional-analyses-of-biological-sequences.html#cb33-9" tabindex="-1"></a></span>
<span id="cb33-10"><a href="compositional-analyses-of-biological-sequences.html#cb33-10" tabindex="-1"></a>df_codon_all <span class="op">=</span> join_codons(dfs_codons, names<span class="op">=</span>names)</span>
<span id="cb33-11"><a href="compositional-analyses-of-biological-sequences.html#cb33-11" tabindex="-1"></a>df_codons_medians <span class="op">=</span> join_codons_medians(dfs_codons)</span></code></pre></div>
<p>Finally, we can get the heatmap with a seaborn function.</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb34-1"><a href="compositional-analyses-of-biological-sequences.html#cb34-1" tabindex="-1"></a>plt.rcParams[<span class="st">&quot;figure.figsize&quot;</span>] <span class="op">=</span> [<span class="dv">16</span>, <span class="dv">2</span>]</span>
<span id="cb34-2"><a href="compositional-analyses-of-biological-sequences.html#cb34-2" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots(nrows<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb34-3"><a href="compositional-analyses-of-biological-sequences.html#cb34-3" tabindex="-1"></a></span>
<span id="cb34-4"><a href="compositional-analyses-of-biological-sequences.html#cb34-4" tabindex="-1"></a>new_columns <span class="op">=</span> []</span>
<span id="cb34-5"><a href="compositional-analyses-of-biological-sequences.html#cb34-5" tabindex="-1"></a><span class="cf">for</span> codon_name <span class="kw">in</span> df_codons_medians.columns:</span>
<span id="cb34-6"><a href="compositional-analyses-of-biological-sequences.html#cb34-6" tabindex="-1"></a>    new <span class="op">=</span> codon_name.replace(<span class="st">&quot;T&quot;</span>, <span class="st">&quot;U&quot;</span>)</span>
<span id="cb34-7"><a href="compositional-analyses-of-biological-sequences.html#cb34-7" tabindex="-1"></a>    new_columns.append(new)</span>
<span id="cb34-8"><a href="compositional-analyses-of-biological-sequences.html#cb34-8" tabindex="-1"></a></span>
<span id="cb34-9"><a href="compositional-analyses-of-biological-sequences.html#cb34-9" tabindex="-1"></a>df_codons_medians.columns <span class="op">=</span> new_columns</span>
<span id="cb34-10"><a href="compositional-analyses-of-biological-sequences.html#cb34-10" tabindex="-1"></a>df_codons_medians.index <span class="op">=</span> [<span class="st">&quot;RdRp&quot;</span>, <span class="st">&quot;CP&quot;</span>, <span class="st">&quot;p19&quot;</span>, <span class="st">&quot;p22&quot;</span>, <span class="st">&quot;p22 core&quot;</span>, <span class="st">&quot;other cores&quot;</span>]</span>
<span id="cb34-11"><a href="compositional-analyses-of-biological-sequences.html#cb34-11" tabindex="-1"></a><span class="bu">print</span>(<span class="bu">len</span>(new_columns))</span>
<span id="cb34-12"><a href="compositional-analyses-of-biological-sequences.html#cb34-12" tabindex="-1"></a></span>
<span id="cb34-13"><a href="compositional-analyses-of-biological-sequences.html#cb34-13" tabindex="-1"></a>sns.heatmap(df_codons_medians, cmap<span class="op">=</span><span class="st">&quot;viridis&quot;</span>)</span>
<span id="cb34-14"><a href="compositional-analyses-of-biological-sequences.html#cb34-14" tabindex="-1"></a>plt.xlabel(<span class="st">&quot;Codon&quot;</span>)</span>
<span id="cb34-15"><a href="compositional-analyses-of-biological-sequences.html#cb34-15" tabindex="-1"></a>plt.xticks(rotation<span class="op">=</span><span class="dv">45</span>)</span>
<span id="cb34-16"><a href="compositional-analyses-of-biological-sequences.html#cb34-16" tabindex="-1"></a>plt.yticks(rotation<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb34-17"><a href="compositional-analyses-of-biological-sequences.html#cb34-17" tabindex="-1"></a>ax.yaxis.set_ticks_position(<span class="st">&#39;none&#39;</span>) </span>
<span id="cb34-18"><a href="compositional-analyses-of-biological-sequences.html#cb34-18" tabindex="-1"></a></span>
<span id="cb34-19"><a href="compositional-analyses-of-biological-sequences.html#cb34-19" tabindex="-1"></a>plt.tight_layout()</span></code></pre></div>
<p>This following code is useful to verify the significance between two groups. We use Mann-Whitney test to see if there is significance in the difference between p22 core and non-overlapping cores.</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb35-1"><a href="compositional-analyses-of-biological-sequences.html#cb35-1" tabindex="-1"></a>df_codon_all.fillna(<span class="dv">0</span>, inplace<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb35-2"><a href="compositional-analyses-of-biological-sequences.html#cb35-2" tabindex="-1"></a>codon <span class="op">=</span> <span class="st">&quot;CAT&quot;</span></span>
<span id="cb35-3"><a href="compositional-analyses-of-biological-sequences.html#cb35-3" tabindex="-1"></a>df_plot <span class="op">=</span> df_codon_all.T[[codon, <span class="st">&quot;Sequence&quot;</span>]]</span>
<span id="cb35-4"><a href="compositional-analyses-of-biological-sequences.html#cb35-4" tabindex="-1"></a>p22_core <span class="op">=</span> <span class="bu">list</span>(df_plot[df_plot[<span class="st">&quot;Sequence&quot;</span>] <span class="op">==</span> <span class="st">&quot;p22 core&quot;</span>][codon].values)</span>
<span id="cb35-5"><a href="compositional-analyses-of-biological-sequences.html#cb35-5" tabindex="-1"></a>other_cores <span class="op">=</span> <span class="bu">list</span>(df_plot[df_plot[<span class="st">&quot;Sequence&quot;</span>] <span class="op">==</span> <span class="st">&quot;other cores&quot;</span>][codon].values)</span>
<span id="cb35-6"><a href="compositional-analyses-of-biological-sequences.html#cb35-6" tabindex="-1"></a></span>
<span id="cb35-7"><a href="compositional-analyses-of-biological-sequences.html#cb35-7" tabindex="-1"></a>pvalues <span class="op">=</span> [</span>
<span id="cb35-8"><a href="compositional-analyses-of-biological-sequences.html#cb35-8" tabindex="-1"></a>    scipy.stats.mannwhitneyu(p22_core, other_cores, alternative<span class="op">=</span><span class="st">&quot;two-sided&quot;</span>).pvalue</span>
<span id="cb35-9"><a href="compositional-analyses-of-biological-sequences.html#cb35-9" tabindex="-1"></a>]</span>
<span id="cb35-10"><a href="compositional-analyses-of-biological-sequences.html#cb35-10" tabindex="-1"></a>formatted_pvalues <span class="op">=</span> [<span class="ss">f&#39;p=</span><span class="sc">{</span>pvalue<span class="sc">:.2e}</span><span class="ss">&#39;</span> <span class="cf">for</span> pvalue <span class="kw">in</span> pvalues]</span></code></pre></div>
<p>Then, we plot the data and the significance too.</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb36-1"><a href="compositional-analyses-of-biological-sequences.html#cb36-1" tabindex="-1"></a>plt.rcParams[<span class="st">&quot;figure.figsize&quot;</span>] <span class="op">=</span> [<span class="dv">16</span>,<span class="dv">9</span>]</span>
<span id="cb36-2"><a href="compositional-analyses-of-biological-sequences.html#cb36-2" tabindex="-1"></a>sns.set_style(<span class="st">&quot;white&quot;</span>)</span>
<span id="cb36-3"><a href="compositional-analyses-of-biological-sequences.html#cb36-3" tabindex="-1"></a>colors <span class="op">=</span> sns.color_palette(<span class="st">&quot;colorblind&quot;</span>, <span class="dv">5</span>)</span>
<span id="cb36-4"><a href="compositional-analyses-of-biological-sequences.html#cb36-4" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots()</span>
<span id="cb36-5"><a href="compositional-analyses-of-biological-sequences.html#cb36-5" tabindex="-1"></a></span>
<span id="cb36-6"><a href="compositional-analyses-of-biological-sequences.html#cb36-6" tabindex="-1"></a>pairs <span class="op">=</span> [(<span class="st">&#39;p22 core&#39;</span>, <span class="st">&#39;other cores&#39;</span>)]</span>
<span id="cb36-7"><a href="compositional-analyses-of-biological-sequences.html#cb36-7" tabindex="-1"></a></span>
<span id="cb36-8"><a href="compositional-analyses-of-biological-sequences.html#cb36-8" tabindex="-1"></a><span class="cf">with</span> sns.plotting_context(<span class="st">&#39;notebook&#39;</span>, font_scale <span class="op">=</span> <span class="fl">1.2</span>):</span>
<span id="cb36-9"><a href="compositional-analyses-of-biological-sequences.html#cb36-9" tabindex="-1"></a>        sns.swarmplot(data<span class="op">=</span>df_plot, x<span class="op">=</span><span class="st">&quot;Sequence&quot;</span>, y<span class="op">=</span>codon, ax<span class="op">=</span>ax, palette<span class="op">=</span>colors)</span>
<span id="cb36-10"><a href="compositional-analyses-of-biological-sequences.html#cb36-10" tabindex="-1"></a>        sns.boxplot(data<span class="op">=</span>df_plot, x<span class="op">=</span><span class="st">&quot;Sequence&quot;</span>, y<span class="op">=</span>codon, zorder<span class="op">=</span><span class="dv">0</span>, dodge<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>ax, boxprops<span class="op">=</span>{<span class="st">&#39;facecolor&#39;</span>:<span class="st">&#39;None&#39;</span>})</span>
<span id="cb36-11"><a href="compositional-analyses-of-biological-sequences.html#cb36-11" tabindex="-1"></a></span>
<span id="cb36-12"><a href="compositional-analyses-of-biological-sequences.html#cb36-12" tabindex="-1"></a>        <span class="co"># Add annotations</span></span>
<span id="cb36-13"><a href="compositional-analyses-of-biological-sequences.html#cb36-13" tabindex="-1"></a>        annotator <span class="op">=</span> Annotator(ax<span class="op">=</span>ax, pairs <span class="op">=</span> pairs, data<span class="op">=</span>df_plot, x<span class="op">=</span><span class="st">&quot;Sequence&quot;</span>, y<span class="op">=</span>codon)</span>
<span id="cb36-14"><a href="compositional-analyses-of-biological-sequences.html#cb36-14" tabindex="-1"></a>        annotator.set_custom_annotations(formatted_pvalues)</span>
<span id="cb36-15"><a href="compositional-analyses-of-biological-sequences.html#cb36-15" tabindex="-1"></a>        annotator.configure(test_short_name<span class="op">=</span><span class="st">&quot;MWW&quot;</span>)  <span class="co"># text_format is still simple</span></span>
<span id="cb36-16"><a href="compositional-analyses-of-biological-sequences.html#cb36-16" tabindex="-1"></a>        annotator.set_pvalues_and_annotate(pvalues)</span></code></pre></div>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="distances-with-chimerax.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="comparative-analyses-of-selection.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/USERNAME/REPO/edit/BRANCH/04-composition.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
