<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 7 Miscellania | Nested genes code</title>
  <meta name="description" content="This is the description of the code employed in the research." />
  <meta name="generator" content="bookdown 0.34 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 7 Miscellania | Nested genes code" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="This is the description of the code employed in the research." />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 7 Miscellania | Nested genes code" />
  
  <meta name="twitter:description" content="This is the description of the code employed in the research." />
  

<meta name="author" content="Esmeralda G. Legarda" />


<meta name="date" content="2023-11-08" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="comparative-analyses-of-selection.html"/>
<link rel="next" href="references.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A Minimal Book Example</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> About</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#abstract"><i class="fa fa-check"></i><b>1.1</b> Abstract</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#abstract-1"><i class="fa fa-check"></i><b>1.2</b> Abstract</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="blast-and-e-utilities.html"><a href="blast-and-e-utilities.html"><i class="fa fa-check"></i><b>2</b> Blast and E-utilities</a>
<ul>
<li class="chapter" data-level="2.1" data-path="blast-and-e-utilities.html"><a href="blast-and-e-utilities.html#psi-blast"><i class="fa fa-check"></i><b>2.1</b> PSI-BLAST</a></li>
<li class="chapter" data-level="2.2" data-path="blast-and-e-utilities.html"><a href="blast-and-e-utilities.html#protein-id-to-nucleotides"><i class="fa fa-check"></i><b>2.2</b> Protein ID to nucleotides</a></li>
<li class="chapter" data-level="2.3" data-path="blast-and-e-utilities.html"><a href="blast-and-e-utilities.html#species-to-family"><i class="fa fa-check"></i><b>2.3</b> Species to family</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html"><i class="fa fa-check"></i><b>3</b> Distances with clustering in CLANS</a>
<ul>
<li class="chapter" data-level="3.1" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html#read-clans-file"><i class="fa fa-check"></i><b>3.1</b> Read CLANS file</a></li>
<li class="chapter" data-level="3.2" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html#create-a-dataframe"><i class="fa fa-check"></i><b>3.2</b> Create a DataFrame</a></li>
<li class="chapter" data-level="3.3" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html#calculate-centroids-and-distances"><i class="fa fa-check"></i><b>3.3</b> Calculate centroids and distances</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html#plot-the-clustermap"><i class="fa fa-check"></i><b>3.3.1</b> Plot the clustermap</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html#run"><i class="fa fa-check"></i><b>3.4</b> Run</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html"><i class="fa fa-check"></i><b>4</b> Distances with ChimeraX</a>
<ul>
<li class="chapter" data-level="4.1" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#obtaining-rmsd-with-chimerax"><i class="fa fa-check"></i><b>4.1</b> Obtaining RMSD with ChimeraX</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#import-libraries-and-data"><i class="fa fa-check"></i><b>4.1.1</b> Import libraries and data</a></li>
<li class="chapter" data-level="4.1.2" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#matchmaker"><i class="fa fa-check"></i><b>4.1.2</b> MatchMaker</a></li>
<li class="chapter" data-level="4.1.3" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#save-csv-file"><i class="fa fa-check"></i><b>4.1.3</b> Save csv file</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#hiarachical-clustering"><i class="fa fa-check"></i><b>4.2</b> Hiarachical clustering</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#libraries-and-read-csv"><i class="fa fa-check"></i><b>4.2.1</b> Libraries and read csv</a></li>
<li class="chapter" data-level="4.2.2" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#create-matrix"><i class="fa fa-check"></i><b>4.2.2</b> Create matrix</a></li>
<li class="chapter" data-level="4.2.3" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#plot-the-clustermap-1"><i class="fa fa-check"></i><b>4.2.3</b> Plot the clustermap</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="compositional-analyses-of-biological-sequences.html"><a href="compositional-analyses-of-biological-sequences.html"><i class="fa fa-check"></i><b>5</b> Compositional analyses of biological sequences</a>
<ul>
<li class="chapter" data-level="5.1" data-path="compositional-analyses-of-biological-sequences.html"><a href="compositional-analyses-of-biological-sequences.html#amino-acids"><i class="fa fa-check"></i><b>5.1</b> Amino acids</a></li>
<li class="chapter" data-level="5.2" data-path="compositional-analyses-of-biological-sequences.html"><a href="compositional-analyses-of-biological-sequences.html#nucleotides"><i class="fa fa-check"></i><b>5.2</b> Nucleotides</a></li>
<li class="chapter" data-level="5.3" data-path="compositional-analyses-of-biological-sequences.html"><a href="compositional-analyses-of-biological-sequences.html#dinuclotides"><i class="fa fa-check"></i><b>5.3</b> Dinuclotides</a></li>
<li class="chapter" data-level="5.4" data-path="compositional-analyses-of-biological-sequences.html"><a href="compositional-analyses-of-biological-sequences.html#codons"><i class="fa fa-check"></i><b>5.4</b> Codons</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html"><i class="fa fa-check"></i><b>6</b> Comparative analyses of selection</a>
<ul>
<li class="chapter" data-level="6.1" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#evolution-speed-with-absrel"><i class="fa fa-check"></i><b>6.1</b> Evolution speed with aBSREL</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#libraries-and-json-files"><i class="fa fa-check"></i><b>6.1.1</b> Libraries and json files</a></li>
<li class="chapter" data-level="6.1.2" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#omega-and-significance"><i class="fa fa-check"></i><b>6.1.2</b> Omega and significance</a></li>
<li class="chapter" data-level="6.1.3" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#relative-asinh-transformation"><i class="fa fa-check"></i><b>6.1.3</b> Relative asinh transformation</a></li>
<li class="chapter" data-level="6.1.4" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#utilities"><i class="fa fa-check"></i><b>6.1.4</b> Utilities</a></li>
<li class="chapter" data-level="6.1.5" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#boxplots-with-significance"><i class="fa fa-check"></i><b>6.1.5</b> Boxplots with significance</a></li>
<li class="chapter" data-level="6.1.6" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#test-selection"><i class="fa fa-check"></i><b>6.1.6</b> Test selection</a></li>
<li class="chapter" data-level="6.1.7" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#run-1"><i class="fa fa-check"></i><b>6.1.7</b> Run</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#fel-for-correlations"><i class="fa fa-check"></i><b>6.2</b> FEL for correlations</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#significance-and-difference"><i class="fa fa-check"></i><b>6.2.1</b> Significance and difference</a></li>
<li class="chapter" data-level="6.2.2" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#plot-correlation"><i class="fa fa-check"></i><b>6.2.2</b> Plot correlation</a></li>
<li class="chapter" data-level="6.2.3" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#test-correlation-of-selection"><i class="fa fa-check"></i><b>6.2.3</b> Test correlation of selection</a></li>
<li class="chapter" data-level="6.2.4" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#correlation-2d"><i class="fa fa-check"></i><b>6.2.4</b> Correlation 2D</a></li>
<li class="chapter" data-level="6.2.5" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#selection-in-2d"><i class="fa fa-check"></i><b>6.2.5</b> Selection in 2D</a></li>
<li class="chapter" data-level="6.2.6" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#run-2"><i class="fa fa-check"></i><b>6.2.6</b> Run</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="miscellania.html"><a href="miscellania.html"><i class="fa fa-check"></i><b>7</b> Miscellania</a>
<ul>
<li class="chapter" data-level="7.1" data-path="miscellania.html"><a href="miscellania.html#representation-of-predicted-disordered-amino-acid-data"><i class="fa fa-check"></i><b>7.1</b> Representation of predicted disordered amino acid data</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Nested genes code</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="miscellania" class="section level1 hasAnchor" number="7">
<h1><span class="header-section-number">Chapter 7</span> Miscellania<a href="miscellania.html#miscellania" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>Here you can find out how we implement the visualization of different issues in Python.</p>
<div id="representation-of-predicted-disordered-amino-acid-data" class="section level2 hasAnchor" number="7.1">
<h2><span class="header-section-number">7.1</span> Representation of predicted disordered amino acid data<a href="miscellania.html#representation-of-predicted-disordered-amino-acid-data" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>We import some libraries.</p>
<div class="sourceCode" id="cb55"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb55-1"><a href="miscellania.html#cb55-1" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb55-2"><a href="miscellania.html#cb55-2" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb55-3"><a href="miscellania.html#cb55-3" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span></code></pre></div>
<p>Now we read a csv file with disordered data and few other characteristics of each protein.</p>
<div class="sourceCode" id="cb56"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb56-1"><a href="miscellania.html#cb56-1" tabindex="-1"></a>df_disorder <span class="op">=</span> pd.read_csv(<span class="st">&quot;/Users/esmeralda/Documents/TFM/article/disorder_data.csv&quot;</span>, sep<span class="op">=</span><span class="st">&quot;;&quot;</span>)</span>
<span id="cb56-2"><a href="miscellania.html#cb56-2" tabindex="-1"></a>df_disorder[<span class="st">&#39;Length&#39;</span>] <span class="op">=</span> pd.to_numeric(df_disorder[<span class="st">&#39;Length&#39;</span>], errors<span class="op">=</span><span class="st">&#39;coerce&#39;</span>)</span></code></pre></div>
<p>We define for the following plots in Seaborn:</p>
<ul>
<li>The style</li>
<li>The palette</li>
</ul>
<div class="sourceCode" id="cb57"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb57-1"><a href="miscellania.html#cb57-1" tabindex="-1"></a><span class="co"># Set Seaborn style to match a similar style to Plotly</span></span>
<span id="cb57-2"><a href="miscellania.html#cb57-2" tabindex="-1"></a>sns.<span class="bu">set</span>(rc<span class="op">=</span>{<span class="st">&#39;axes.facecolor&#39;</span>: <span class="st">&#39;white&#39;</span>,</span>
<span id="cb57-3"><a href="miscellania.html#cb57-3" tabindex="-1"></a> <span class="st">&#39;axes.edgecolor&#39;</span>: <span class="st">&#39;white&#39;</span>,</span>
<span id="cb57-4"><a href="miscellania.html#cb57-4" tabindex="-1"></a>            <span class="st">&quot;axes.grid&quot;</span>: <span class="va">True</span>, </span>
<span id="cb57-5"><a href="miscellania.html#cb57-5" tabindex="-1"></a>            <span class="st">&quot;grid.linestyle&quot;</span>: <span class="st">&quot;--&quot;</span>, </span>
<span id="cb57-6"><a href="miscellania.html#cb57-6" tabindex="-1"></a>            <span class="st">&quot;font.size&quot;</span>: <span class="dv">15</span>, </span>
<span id="cb57-7"><a href="miscellania.html#cb57-7" tabindex="-1"></a>            <span class="st">&quot;axes.labelsize&quot;</span>: <span class="dv">15</span>, </span>
<span id="cb57-8"><a href="miscellania.html#cb57-8" tabindex="-1"></a>            <span class="st">&quot;xtick.labelsize&quot;</span>: <span class="dv">12</span>, </span>
<span id="cb57-9"><a href="miscellania.html#cb57-9" tabindex="-1"></a>            <span class="st">&quot;ytick.labelsize&quot;</span>: <span class="dv">12</span>,</span>
<span id="cb57-10"><a href="miscellania.html#cb57-10" tabindex="-1"></a>            <span class="st">&#39;xtick.bottom&#39;</span>: <span class="va">True</span>,</span>
<span id="cb57-11"><a href="miscellania.html#cb57-11" tabindex="-1"></a>            <span class="st">&#39;figure.facecolor&#39;</span>: <span class="st">&#39;white&#39;</span>,</span>
<span id="cb57-12"><a href="miscellania.html#cb57-12" tabindex="-1"></a> <span class="st">&#39;xtick.top&#39;</span>: <span class="va">False</span>,</span>
<span id="cb57-13"><a href="miscellania.html#cb57-13" tabindex="-1"></a> <span class="st">&#39;xtick.bottom&#39;</span>: <span class="va">False</span>,</span>
<span id="cb57-14"><a href="miscellania.html#cb57-14" tabindex="-1"></a> <span class="st">&#39;ytick.left&#39;</span>: <span class="va">True</span>,</span>
<span id="cb57-15"><a href="miscellania.html#cb57-15" tabindex="-1"></a> <span class="st">&#39;ytick.right&#39;</span>: <span class="va">False</span>,</span>
<span id="cb57-16"><a href="miscellania.html#cb57-16" tabindex="-1"></a> <span class="st">&quot;grid.color&quot;</span>: <span class="st">&quot;#DFDFDF&quot;</span>,</span>
<span id="cb57-17"><a href="miscellania.html#cb57-17" tabindex="-1"></a> <span class="st">&#39;font.sans-serif&#39;</span>: <span class="st">&#39;Helvetica&#39;</span>})</span>
<span id="cb57-18"><a href="miscellania.html#cb57-18" tabindex="-1"></a></span>
<span id="cb57-19"><a href="miscellania.html#cb57-19" tabindex="-1"></a>safe_colorblind_palette <span class="op">=</span> [<span class="st">&quot;#E69F00&quot;</span>, <span class="st">&quot;#6D0000&quot;</span>, <span class="st">&quot;#56B4E9&quot;</span>, <span class="st">&quot;#97E4CF&quot;</span>, <span class="st">&quot;#F0E442&quot;</span>, <span class="st">&quot;#0F425F&quot;</span>, <span class="st">&quot;#D55E00&quot;</span>, <span class="st">&quot;#000877&quot;</span>, <span class="st">&quot;#CA4E93&quot;</span>, <span class="st">&quot;#ABB6D3&quot;</span>, <span class="st">&quot;#D9F221&quot;</span>]</span>
<span id="cb57-20"><a href="miscellania.html#cb57-20" tabindex="-1"></a></span>
<span id="cb57-21"><a href="miscellania.html#cb57-21" tabindex="-1"></a><span class="co"># Create a figure and axes</span></span>
<span id="cb57-22"><a href="miscellania.html#cb57-22" tabindex="-1"></a>plt.figure(figsize<span class="op">=</span>(<span class="dv">10</span>, <span class="dv">6</span>))</span>
<span id="cb57-23"><a href="miscellania.html#cb57-23" tabindex="-1"></a>ax <span class="op">=</span> sns.boxplot(data<span class="op">=</span>df_disorder[df_disorder[<span class="st">&quot;Movement protein&quot;</span>] <span class="op">==</span> <span class="st">&quot;Movement protein&quot;</span>], </span>
<span id="cb57-24"><a href="miscellania.html#cb57-24" tabindex="-1"></a>                 x<span class="op">=</span><span class="st">&quot;Taxonomy origin&quot;</span>, </span>
<span id="cb57-25"><a href="miscellania.html#cb57-25" tabindex="-1"></a>                 y<span class="op">=</span><span class="st">&quot;Percentage&quot;</span>, </span>
<span id="cb57-26"><a href="miscellania.html#cb57-26" tabindex="-1"></a>                 color<span class="op">=</span><span class="st">&quot;white&quot;</span>,</span>
<span id="cb57-27"><a href="miscellania.html#cb57-27" tabindex="-1"></a>                 boxprops<span class="op">=</span><span class="bu">dict</span>(edgecolor<span class="op">=</span><span class="st">&#39;black&#39;</span>),</span>
<span id="cb57-28"><a href="miscellania.html#cb57-28" tabindex="-1"></a>                 whiskerprops<span class="op">=</span><span class="bu">dict</span>(color<span class="op">=</span><span class="st">&#39;black&#39;</span>),</span>
<span id="cb57-29"><a href="miscellania.html#cb57-29" tabindex="-1"></a>                 medianprops<span class="op">=</span><span class="bu">dict</span>(color<span class="op">=</span><span class="st">&#39;black&#39;</span>),</span>
<span id="cb57-30"><a href="miscellania.html#cb57-30" tabindex="-1"></a>                 capprops<span class="op">=</span><span class="bu">dict</span>(linewidth<span class="op">=</span><span class="dv">2</span>),</span>
<span id="cb57-31"><a href="miscellania.html#cb57-31" tabindex="-1"></a>                 width<span class="op">=</span><span class="fl">0.7</span>, </span>
<span id="cb57-32"><a href="miscellania.html#cb57-32" tabindex="-1"></a>                 linewidth<span class="op">=</span><span class="fl">0.6</span>)</span>
<span id="cb57-33"><a href="miscellania.html#cb57-33" tabindex="-1"></a></span>
<span id="cb57-34"><a href="miscellania.html#cb57-34" tabindex="-1"></a>marker_style <span class="op">=</span> []</span>
<span id="cb57-35"><a href="miscellania.html#cb57-35" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> df_disorder[<span class="st">&quot;Movement protein&quot;</span>]:</span>
<span id="cb57-36"><a href="miscellania.html#cb57-36" tabindex="-1"></a>    <span class="cf">if</span> i <span class="op">==</span> <span class="st">&#39;o&#39;</span>:</span>
<span id="cb57-37"><a href="miscellania.html#cb57-37" tabindex="-1"></a>        marker_style.append(<span class="st">&#39;o&#39;</span>)</span>
<span id="cb57-38"><a href="miscellania.html#cb57-38" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb57-39"><a href="miscellania.html#cb57-39" tabindex="-1"></a>        marker_style.append(<span class="st">&#39;X&#39;</span>)</span></code></pre></div>
<p>We can plot disorder data classified by taxonomic families.</p>
<div class="sourceCode" id="cb58"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb58-1"><a href="miscellania.html#cb58-1" tabindex="-1"></a>ax <span class="op">=</span> sns.swarmplot(data<span class="op">=</span>df_disorder, </span>
<span id="cb58-2"><a href="miscellania.html#cb58-2" tabindex="-1"></a>                   x<span class="op">=</span><span class="st">&quot;Taxonomy origin&quot;</span>, </span>
<span id="cb58-3"><a href="miscellania.html#cb58-3" tabindex="-1"></a>                   y<span class="op">=</span><span class="st">&quot;Percentage&quot;</span>,</span>
<span id="cb58-4"><a href="miscellania.html#cb58-4" tabindex="-1"></a>                   color<span class="op">=</span><span class="st">&quot;black&quot;</span>, </span>
<span id="cb58-5"><a href="miscellania.html#cb58-5" tabindex="-1"></a>                   palette<span class="op">=</span>safe_colorblind_palette,</span>
<span id="cb58-6"><a href="miscellania.html#cb58-6" tabindex="-1"></a>                   size<span class="op">=</span><span class="dv">7</span>,</span>
<span id="cb58-7"><a href="miscellania.html#cb58-7" tabindex="-1"></a>                   marker<span class="op">=</span><span class="st">&quot;X&quot;</span>)</span>
<span id="cb58-8"><a href="miscellania.html#cb58-8" tabindex="-1"></a></span>
<span id="cb58-9"><a href="miscellania.html#cb58-9" tabindex="-1"></a>ax <span class="op">=</span> sns.swarmplot(data<span class="op">=</span>df_disorder[df_disorder[<span class="st">&quot;Movement protein&quot;</span>] <span class="op">==</span> <span class="st">&quot;Movement protein&quot;</span>], </span>
<span id="cb58-10"><a href="miscellania.html#cb58-10" tabindex="-1"></a>                   x<span class="op">=</span><span class="st">&quot;Taxonomy origin&quot;</span>, </span>
<span id="cb58-11"><a href="miscellania.html#cb58-11" tabindex="-1"></a>                   y<span class="op">=</span><span class="st">&quot;Percentage&quot;</span>,</span>
<span id="cb58-12"><a href="miscellania.html#cb58-12" tabindex="-1"></a>                   color<span class="op">=</span><span class="st">&quot;black&quot;</span>, </span>
<span id="cb58-13"><a href="miscellania.html#cb58-13" tabindex="-1"></a>                   palette<span class="op">=</span>safe_colorblind_palette,</span>
<span id="cb58-14"><a href="miscellania.html#cb58-14" tabindex="-1"></a>                   marker<span class="op">=</span><span class="st">&quot;o&quot;</span>,</span>
<span id="cb58-15"><a href="miscellania.html#cb58-15" tabindex="-1"></a>                   size<span class="op">=</span><span class="dv">7</span>)</span>
<span id="cb58-16"><a href="miscellania.html#cb58-16" tabindex="-1"></a></span>
<span id="cb58-17"><a href="miscellania.html#cb58-17" tabindex="-1"></a><span class="co"># Set plot labels and title</span></span>
<span id="cb58-18"><a href="miscellania.html#cb58-18" tabindex="-1"></a>plt.xlabel(<span class="st">&quot;Genus&quot;</span>)</span>
<span id="cb58-19"><a href="miscellania.html#cb58-19" tabindex="-1"></a>plt.ylabel(<span class="st">&quot;% predicted disodered residues&quot;</span>)</span>
<span id="cb58-20"><a href="miscellania.html#cb58-20" tabindex="-1"></a></span>
<span id="cb58-21"><a href="miscellania.html#cb58-21" tabindex="-1"></a><span class="co"># Adjust plot layout</span></span>
<span id="cb58-22"><a href="miscellania.html#cb58-22" tabindex="-1"></a>plt.xticks(rotation<span class="op">=</span><span class="dv">45</span>)</span>
<span id="cb58-23"><a href="miscellania.html#cb58-23" tabindex="-1"></a>plt.tight_layout()</span>
<span id="cb58-24"><a href="miscellania.html#cb58-24" tabindex="-1"></a></span>
<span id="cb58-25"><a href="miscellania.html#cb58-25" tabindex="-1"></a>plt.savefig(<span class="st">&quot;percentage_taxonomy.pdf&quot;</span>, <span class="bu">format</span><span class="op">=</span><span class="st">&quot;pdf&quot;</span>)</span>
<span id="cb58-26"><a href="miscellania.html#cb58-26" tabindex="-1"></a><span class="co"># Display the plot</span></span>
<span id="cb58-27"><a href="miscellania.html#cb58-27" tabindex="-1"></a>plt.show()</span></code></pre></div>
<p>In this case, we plot all the proteins only divided by if their corresponding open reading frames are overprinted.</p>
<div class="sourceCode" id="cb59"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb59-1"><a href="miscellania.html#cb59-1" tabindex="-1"></a>safe_colorblind_palette_2 <span class="op">=</span> [<span class="st">&quot;#054C6F&quot;</span>, <span class="st">&quot;#44AA99&quot;</span>]</span>
<span id="cb59-2"><a href="miscellania.html#cb59-2" tabindex="-1"></a></span>
<span id="cb59-3"><a href="miscellania.html#cb59-3" tabindex="-1"></a><span class="co"># Create a figure and axes</span></span>
<span id="cb59-4"><a href="miscellania.html#cb59-4" tabindex="-1"></a>plt.figure(figsize<span class="op">=</span>(<span class="dv">5</span>, <span class="dv">6</span>))</span>
<span id="cb59-5"><a href="miscellania.html#cb59-5" tabindex="-1"></a>ax <span class="op">=</span> sns.boxplot(data<span class="op">=</span>df_disorder[df_disorder[<span class="st">&quot;Movement protein&quot;</span>] <span class="op">==</span> <span class="st">&quot;Movement protein&quot;</span>], </span>
<span id="cb59-6"><a href="miscellania.html#cb59-6" tabindex="-1"></a>                 x<span class="op">=</span><span class="st">&quot;Overlap&quot;</span>, </span>
<span id="cb59-7"><a href="miscellania.html#cb59-7" tabindex="-1"></a>                 y<span class="op">=</span><span class="st">&quot;Percentage&quot;</span>, </span>
<span id="cb59-8"><a href="miscellania.html#cb59-8" tabindex="-1"></a>                 color<span class="op">=</span><span class="st">&quot;white&quot;</span>,</span>
<span id="cb59-9"><a href="miscellania.html#cb59-9" tabindex="-1"></a>                 boxprops<span class="op">=</span><span class="bu">dict</span>(edgecolor<span class="op">=</span><span class="st">&#39;black&#39;</span>),</span>
<span id="cb59-10"><a href="miscellania.html#cb59-10" tabindex="-1"></a>                 whiskerprops<span class="op">=</span><span class="bu">dict</span>(color<span class="op">=</span><span class="st">&#39;black&#39;</span>),</span>
<span id="cb59-11"><a href="miscellania.html#cb59-11" tabindex="-1"></a>                 medianprops<span class="op">=</span><span class="bu">dict</span>(color<span class="op">=</span><span class="st">&#39;black&#39;</span>),</span>
<span id="cb59-12"><a href="miscellania.html#cb59-12" tabindex="-1"></a>                 capprops<span class="op">=</span><span class="bu">dict</span>(linewidth<span class="op">=</span><span class="dv">2</span>),</span>
<span id="cb59-13"><a href="miscellania.html#cb59-13" tabindex="-1"></a>                 width<span class="op">=</span><span class="fl">0.7</span>, </span>
<span id="cb59-14"><a href="miscellania.html#cb59-14" tabindex="-1"></a>                 linewidth<span class="op">=</span><span class="fl">0.6</span>)</span>
<span id="cb59-15"><a href="miscellania.html#cb59-15" tabindex="-1"></a></span>
<span id="cb59-16"><a href="miscellania.html#cb59-16" tabindex="-1"></a>ax <span class="op">=</span> sns.swarmplot(data<span class="op">=</span>df_disorder,  </span>
<span id="cb59-17"><a href="miscellania.html#cb59-17" tabindex="-1"></a>                   x<span class="op">=</span><span class="st">&quot;Overlap&quot;</span>, </span>
<span id="cb59-18"><a href="miscellania.html#cb59-18" tabindex="-1"></a>                   y<span class="op">=</span><span class="st">&quot;Percentage&quot;</span>, </span>
<span id="cb59-19"><a href="miscellania.html#cb59-19" tabindex="-1"></a>                   color<span class="op">=</span><span class="st">&quot;black&quot;</span>, </span>
<span id="cb59-20"><a href="miscellania.html#cb59-20" tabindex="-1"></a>                   palette<span class="op">=</span>safe_colorblind_palette_2,</span>
<span id="cb59-21"><a href="miscellania.html#cb59-21" tabindex="-1"></a>                   marker<span class="op">=</span><span class="st">&quot;X&quot;</span>,      </span>
<span id="cb59-22"><a href="miscellania.html#cb59-22" tabindex="-1"></a>                   size<span class="op">=</span><span class="dv">7</span>)</span>
<span id="cb59-23"><a href="miscellania.html#cb59-23" tabindex="-1"></a></span>
<span id="cb59-24"><a href="miscellania.html#cb59-24" tabindex="-1"></a>ax <span class="op">=</span> sns.swarmplot(data<span class="op">=</span>df_disorder[df_disorder[<span class="st">&quot;Movement protein&quot;</span>] <span class="op">==</span> <span class="st">&quot;Movement protein&quot;</span>],  </span>
<span id="cb59-25"><a href="miscellania.html#cb59-25" tabindex="-1"></a>                   x<span class="op">=</span><span class="st">&quot;Overlap&quot;</span>, </span>
<span id="cb59-26"><a href="miscellania.html#cb59-26" tabindex="-1"></a>                   y<span class="op">=</span><span class="st">&quot;Percentage&quot;</span>, </span>
<span id="cb59-27"><a href="miscellania.html#cb59-27" tabindex="-1"></a>                   color<span class="op">=</span><span class="st">&quot;black&quot;</span>, </span>
<span id="cb59-28"><a href="miscellania.html#cb59-28" tabindex="-1"></a>                   palette<span class="op">=</span>safe_colorblind_palette_2, </span>
<span id="cb59-29"><a href="miscellania.html#cb59-29" tabindex="-1"></a>                   marker<span class="op">=</span><span class="st">&quot;o&quot;</span>,   </span>
<span id="cb59-30"><a href="miscellania.html#cb59-30" tabindex="-1"></a>                   size<span class="op">=</span><span class="dv">7</span>)</span>
<span id="cb59-31"><a href="miscellania.html#cb59-31" tabindex="-1"></a></span>
<span id="cb59-32"><a href="miscellania.html#cb59-32" tabindex="-1"></a><span class="co"># Set plot labels and title</span></span>
<span id="cb59-33"><a href="miscellania.html#cb59-33" tabindex="-1"></a>plt.xlabel(<span class="st">&quot;&quot;</span>)</span>
<span id="cb59-34"><a href="miscellania.html#cb59-34" tabindex="-1"></a>plt.ylabel(<span class="st">&quot;% predicted disodered residues&quot;</span>)</span>
<span id="cb59-35"><a href="miscellania.html#cb59-35" tabindex="-1"></a></span>
<span id="cb59-36"><a href="miscellania.html#cb59-36" tabindex="-1"></a><span class="co"># Adjust plot layout</span></span>
<span id="cb59-37"><a href="miscellania.html#cb59-37" tabindex="-1"></a>plt.xticks(rotation<span class="op">=</span><span class="dv">45</span>)</span>
<span id="cb59-38"><a href="miscellania.html#cb59-38" tabindex="-1"></a>plt.tight_layout()</span>
<span id="cb59-39"><a href="miscellania.html#cb59-39" tabindex="-1"></a></span>
<span id="cb59-40"><a href="miscellania.html#cb59-40" tabindex="-1"></a>plt.savefig(<span class="st">&quot;percentage_overprinted.pdf&quot;</span>, <span class="bu">format</span><span class="op">=</span><span class="st">&quot;pdf&quot;</span>)</span>
<span id="cb59-41"><a href="miscellania.html#cb59-41" tabindex="-1"></a></span>
<span id="cb59-42"><a href="miscellania.html#cb59-42" tabindex="-1"></a><span class="co"># Display the plot</span></span>
<span id="cb59-43"><a href="miscellania.html#cb59-43" tabindex="-1"></a>plt.show()</span></code></pre></div>
<p>Then, we can also scatter the data, using the length of our proteins.</p>
<div class="sourceCode" id="cb60"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb60-1"><a href="miscellania.html#cb60-1" tabindex="-1"></a>plt.figure(figsize<span class="op">=</span>(<span class="dv">8</span>, <span class="dv">6</span>))</span>
<span id="cb60-2"><a href="miscellania.html#cb60-2" tabindex="-1"></a>markers <span class="op">=</span> {<span class="st">&quot;Movement protein&quot;</span>: <span class="st">&quot;o&quot;</span>, <span class="st">&quot;p19&quot;</span>: <span class="st">&quot;X&quot;</span>}</span>
<span id="cb60-3"><a href="miscellania.html#cb60-3" tabindex="-1"></a>sns.scatterplot(data<span class="op">=</span>df_disorder, x<span class="op">=</span><span class="st">&#39;Length&#39;</span>, y<span class="op">=</span><span class="st">&#39;Disordered residues&#39;</span>, hue<span class="op">=</span><span class="st">&#39;Taxonomy origin&#39;</span>, style<span class="op">=</span><span class="st">&quot;Movement protein&quot;</span>, markers<span class="op">=</span>markers, palette<span class="op">=</span>safe_colorblind_palette, s<span class="op">=</span><span class="dv">40</span>)</span>
<span id="cb60-4"><a href="miscellania.html#cb60-4" tabindex="-1"></a>plt.legend([],[], frameon<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb60-5"><a href="miscellania.html#cb60-5" tabindex="-1"></a></span>
<span id="cb60-6"><a href="miscellania.html#cb60-6" tabindex="-1"></a>plt.xlabel(<span class="st">&quot;Length&quot;</span>)</span>
<span id="cb60-7"><a href="miscellania.html#cb60-7" tabindex="-1"></a>plt.ylabel(<span class="st">&quot;Predicted disordered residues&quot;</span>)</span>
<span id="cb60-8"><a href="miscellania.html#cb60-8" tabindex="-1"></a></span>
<span id="cb60-9"><a href="miscellania.html#cb60-9" tabindex="-1"></a>plt.savefig(<span class="st">&quot;disorder_scatter_taxonomy_legend.pdf&quot;</span>, <span class="bu">format</span><span class="op">=</span><span class="st">&quot;pdf&quot;</span>)</span>
<span id="cb60-10"><a href="miscellania.html#cb60-10" tabindex="-1"></a></span>
<span id="cb60-11"><a href="miscellania.html#cb60-11" tabindex="-1"></a>plt.show()</span></code></pre></div>
<p>Last, but not least, the previous plot can be only colored by the fact of if they were overprinted or not.</p>
<div class="sourceCode" id="cb61"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb61-1"><a href="miscellania.html#cb61-1" tabindex="-1"></a>plt.figure(figsize<span class="op">=</span>(<span class="dv">8</span>, <span class="dv">6</span>))</span>
<span id="cb61-2"><a href="miscellania.html#cb61-2" tabindex="-1"></a>ax <span class="op">=</span> sns.scatterplot(data<span class="op">=</span>df_disorder, x<span class="op">=</span><span class="st">&#39;Length&#39;</span>, y<span class="op">=</span><span class="st">&#39;Disordered residues&#39;</span>, hue<span class="op">=</span><span class="st">&#39;Overlap&#39;</span>,style<span class="op">=</span><span class="st">&quot;Movement protein&quot;</span>, palette<span class="op">=</span>safe_colorblind_palette_2, s<span class="op">=</span><span class="dv">40</span>)</span>
<span id="cb61-3"><a href="miscellania.html#cb61-3" tabindex="-1"></a></span>
<span id="cb61-4"><a href="miscellania.html#cb61-4" tabindex="-1"></a></span>
<span id="cb61-5"><a href="miscellania.html#cb61-5" tabindex="-1"></a>plt.xlabel(<span class="st">&quot;Length&quot;</span>)</span>
<span id="cb61-6"><a href="miscellania.html#cb61-6" tabindex="-1"></a>plt.ylabel(<span class="st">&quot;Predicted disordered residues&quot;</span>)</span>
<span id="cb61-7"><a href="miscellania.html#cb61-7" tabindex="-1"></a>plt.legend([],[], frameon<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb61-8"><a href="miscellania.html#cb61-8" tabindex="-1"></a></span>
<span id="cb61-9"><a href="miscellania.html#cb61-9" tabindex="-1"></a></span>
<span id="cb61-10"><a href="miscellania.html#cb61-10" tabindex="-1"></a>plt.savefig(<span class="st">&quot;disorder_scatter_overlap_legend.pdf&quot;</span>, <span class="bu">format</span><span class="op">=</span><span class="st">&quot;pdf&quot;</span>)</span>
<span id="cb61-11"><a href="miscellania.html#cb61-11" tabindex="-1"></a></span>
<span id="cb61-12"><a href="miscellania.html#cb61-12" tabindex="-1"></a>plt.show()</span></code></pre></div>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="comparative-analyses-of-selection.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="references.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/USERNAME/REPO/edit/BRANCH/06-micela.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
