<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 5 Compositional analyses of biological sequences | Nested genes code</title>
  <meta name="description" content="This is the description of the code employed in the research." />
  <meta name="generator" content="bookdown 0.34 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 5 Compositional analyses of biological sequences | Nested genes code" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="This is the description of the code employed in the research." />
  <meta name="github-repo" content="https://github.com/llobarts/nestedgenes" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 5 Compositional analyses of biological sequences | Nested genes code" />
  
  <meta name="twitter:description" content="This is the description of the code employed in the research." />
  

<meta name="author" content="Esmeralda G. Legarda" />


<meta name="date" content="2023-12-30" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="distances-with-chimerax.html"/>
<link rel="next" href="comparative-analyses-of-selection.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Code for 'Nested genes'</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> About</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#abstract"><i class="fa fa-check"></i><b>1.1</b> Abstract</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#additional-files"><i class="fa fa-check"></i><b>1.2</b> Additional files</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="blast-and-e-utilities.html"><a href="blast-and-e-utilities.html"><i class="fa fa-check"></i><b>2</b> Blast and E-utilities</a>
<ul>
<li class="chapter" data-level="2.1" data-path="blast-and-e-utilities.html"><a href="blast-and-e-utilities.html#psi-blast"><i class="fa fa-check"></i><b>2.1</b> PSI-BLAST</a></li>
<li class="chapter" data-level="2.2" data-path="blast-and-e-utilities.html"><a href="blast-and-e-utilities.html#protein-id-to-nucleotides"><i class="fa fa-check"></i><b>2.2</b> Protein ID to nucleotides</a></li>
<li class="chapter" data-level="2.3" data-path="blast-and-e-utilities.html"><a href="blast-and-e-utilities.html#species-to-family"><i class="fa fa-check"></i><b>2.3</b> Species to family</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html"><i class="fa fa-check"></i><b>3</b> Distances with clustering in CLANS</a>
<ul>
<li class="chapter" data-level="3.1" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html#read-the-clans-file"><i class="fa fa-check"></i><b>3.1</b> Read the CLANS file</a></li>
<li class="chapter" data-level="3.2" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html#create-a-dataframe"><i class="fa fa-check"></i><b>3.2</b> Create a DataFrame</a></li>
<li class="chapter" data-level="3.3" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html#calculate-centroids-and-distances"><i class="fa fa-check"></i><b>3.3</b> Calculate centroids and distances</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html#plot-the-clustermap"><i class="fa fa-check"></i><b>3.3.1</b> Plot the clustermap</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html#run"><i class="fa fa-check"></i><b>3.4</b> Run</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html"><i class="fa fa-check"></i><b>4</b> Distances with ChimeraX</a>
<ul>
<li class="chapter" data-level="4.1" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#obtaining-rmsd-with-chimerax"><i class="fa fa-check"></i><b>4.1</b> Obtaining RMSD with ChimeraX</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#import-libraries-and-data"><i class="fa fa-check"></i><b>4.1.1</b> Import libraries and data</a></li>
<li class="chapter" data-level="4.1.2" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#matchmaker"><i class="fa fa-check"></i><b>4.1.2</b> MatchMaker</a></li>
<li class="chapter" data-level="4.1.3" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#save-csv-file"><i class="fa fa-check"></i><b>4.1.3</b> Save csv file</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#hiarachical-clustering"><i class="fa fa-check"></i><b>4.2</b> Hiarachical clustering</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#libraries-and-read-csv"><i class="fa fa-check"></i><b>4.2.1</b> Libraries and read csv</a></li>
<li class="chapter" data-level="4.2.2" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#create-matrix"><i class="fa fa-check"></i><b>4.2.2</b> Create matrix</a></li>
<li class="chapter" data-level="4.2.3" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#plot-the-clustermap-1"><i class="fa fa-check"></i><b>4.2.3</b> Plot the clustermap</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="compositional-analyses-of-biological-sequences.html"><a href="compositional-analyses-of-biological-sequences.html"><i class="fa fa-check"></i><b>5</b> Compositional analyses of biological sequences</a>
<ul>
<li class="chapter" data-level="5.1" data-path="compositional-analyses-of-biological-sequences.html"><a href="compositional-analyses-of-biological-sequences.html#amino-acids"><i class="fa fa-check"></i><b>5.1</b> Amino acids</a></li>
<li class="chapter" data-level="5.2" data-path="compositional-analyses-of-biological-sequences.html"><a href="compositional-analyses-of-biological-sequences.html#nucleotides"><i class="fa fa-check"></i><b>5.2</b> Nucleotides</a></li>
<li class="chapter" data-level="5.3" data-path="compositional-analyses-of-biological-sequences.html"><a href="compositional-analyses-of-biological-sequences.html#dinuclotides"><i class="fa fa-check"></i><b>5.3</b> Dinuclotides</a></li>
<li class="chapter" data-level="5.4" data-path="compositional-analyses-of-biological-sequences.html"><a href="compositional-analyses-of-biological-sequences.html#codons"><i class="fa fa-check"></i><b>5.4</b> Codons</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html"><i class="fa fa-check"></i><b>6</b> Comparative analyses of selection</a>
<ul>
<li class="chapter" data-level="6.1" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#evolution-speed-with-absrel"><i class="fa fa-check"></i><b>6.1</b> Evolution speed with aBSREL</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#libraries-and-json-files"><i class="fa fa-check"></i><b>6.1.1</b> Libraries and json files</a></li>
<li class="chapter" data-level="6.1.2" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#omega-and-significance"><i class="fa fa-check"></i><b>6.1.2</b> Omega and significance</a></li>
<li class="chapter" data-level="6.1.3" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#relative-asinh-transformation"><i class="fa fa-check"></i><b>6.1.3</b> Relative asinh transformation</a></li>
<li class="chapter" data-level="6.1.4" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#utilities"><i class="fa fa-check"></i><b>6.1.4</b> Utilities</a></li>
<li class="chapter" data-level="6.1.5" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#boxplots-with-significance"><i class="fa fa-check"></i><b>6.1.5</b> Boxplots with significance</a></li>
<li class="chapter" data-level="6.1.6" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#test-selection"><i class="fa fa-check"></i><b>6.1.6</b> Test selection</a></li>
<li class="chapter" data-level="6.1.7" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#run-1"><i class="fa fa-check"></i><b>6.1.7</b> Run</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#fel-for-correlations"><i class="fa fa-check"></i><b>6.2</b> FEL for correlations</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#significance-and-difference"><i class="fa fa-check"></i><b>6.2.1</b> Significance and difference</a></li>
<li class="chapter" data-level="6.2.2" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#plot-correlation"><i class="fa fa-check"></i><b>6.2.2</b> Plot correlation</a></li>
<li class="chapter" data-level="6.2.3" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#test-correlation-of-selection"><i class="fa fa-check"></i><b>6.2.3</b> Test correlation of selection</a></li>
<li class="chapter" data-level="6.2.4" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#correlation-2d"><i class="fa fa-check"></i><b>6.2.4</b> Correlation 2D</a></li>
<li class="chapter" data-level="6.2.5" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#selection-in-2d"><i class="fa fa-check"></i><b>6.2.5</b> Selection in 2D</a></li>
<li class="chapter" data-level="6.2.6" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#run-2"><i class="fa fa-check"></i><b>6.2.6</b> Run</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="miscellania.html"><a href="miscellania.html"><i class="fa fa-check"></i><b>7</b> Miscellania</a>
<ul>
<li class="chapter" data-level="7.1" data-path="miscellania.html"><a href="miscellania.html#representation-of-predicted-disordered-amino-acid-data"><i class="fa fa-check"></i><b>7.1</b> Representation of predicted disordered amino acid data</a></li>
<li class="chapter" data-level="7.2" data-path="miscellania.html"><a href="miscellania.html#shannons-entropy"><i class="fa fa-check"></i><b>7.2</b> Shannon’s entropy</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/llobarts/nestedgenes" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Nested genes code</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="compositional-analyses-of-biological-sequences" class="section level1 hasAnchor" number="5">
<h1><span class="header-section-number">Chapter 5</span> Compositional analyses of biological sequences<a href="compositional-analyses-of-biological-sequences.html#compositional-analyses-of-biological-sequences" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>Compositional data can be valuable for comparing two or more groups, potentially revealing overrepresented characteristics. Here, we present code for comparing movement proteins and their corresponding encoding genes, both with and without overlapping frames.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb1-1"><a href="compositional-analyses-of-biological-sequences.html#cb1-1" tabindex="-1"></a><span class="im">from</span> Bio <span class="im">import</span> SeqIO</span>
<span id="cb1-2"><a href="compositional-analyses-of-biological-sequences.html#cb1-2" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb1-3"><a href="compositional-analyses-of-biological-sequences.html#cb1-3" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb1-4"><a href="compositional-analyses-of-biological-sequences.html#cb1-4" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb1-5"><a href="compositional-analyses-of-biological-sequences.html#cb1-5" tabindex="-1"></a><span class="im">import</span> scipy.stats</span>
<span id="cb1-6"><a href="compositional-analyses-of-biological-sequences.html#cb1-6" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb1-7"><a href="compositional-analyses-of-biological-sequences.html#cb1-7" tabindex="-1"></a><span class="im">from</span> statannotations.Annotator <span class="im">import</span> Annotator</span></code></pre></div>
<p>We define a table of codons and amino acids, accompanied by a classification of amino acids based on their physicochemical properties. It’s worth noting that the table of codons is ordered by the degeneracy of the genetic code.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb2-1"><a href="compositional-analyses-of-biological-sequences.html#cb2-1" tabindex="-1"></a>table_DNA_to_codon <span class="op">=</span> {<span class="st">&quot;M&quot;</span>: [<span class="st">&quot;ATG&quot;</span>],</span>
<span id="cb2-2"><a href="compositional-analyses-of-biological-sequences.html#cb2-2" tabindex="-1"></a>                      <span class="st">&quot;W&quot;</span>: [<span class="st">&quot;TGG&quot;</span>],</span>
<span id="cb2-3"><a href="compositional-analyses-of-biological-sequences.html#cb2-3" tabindex="-1"></a>                      <span class="st">&quot;C&quot;</span>: [<span class="st">&quot;TGT&quot;</span>, <span class="st">&quot;TGC&quot;</span>],</span>
<span id="cb2-4"><a href="compositional-analyses-of-biological-sequences.html#cb2-4" tabindex="-1"></a>                      <span class="st">&quot;D&quot;</span>: [<span class="st">&quot;GAT&quot;</span>, <span class="st">&quot;GAC&quot;</span>],</span>
<span id="cb2-5"><a href="compositional-analyses-of-biological-sequences.html#cb2-5" tabindex="-1"></a>                      <span class="st">&quot;E&quot;</span>: [<span class="st">&quot;GAA&quot;</span>, <span class="st">&quot;GAG&quot;</span>],</span>
<span id="cb2-6"><a href="compositional-analyses-of-biological-sequences.html#cb2-6" tabindex="-1"></a>                      <span class="st">&quot;F&quot;</span>: [<span class="st">&quot;TTT&quot;</span>, <span class="st">&quot;TTC&quot;</span>],</span>
<span id="cb2-7"><a href="compositional-analyses-of-biological-sequences.html#cb2-7" tabindex="-1"></a>                      <span class="st">&quot;H&quot;</span>: [<span class="st">&quot;CAT&quot;</span>, <span class="st">&quot;CAC&quot;</span>],</span>
<span id="cb2-8"><a href="compositional-analyses-of-biological-sequences.html#cb2-8" tabindex="-1"></a>                      <span class="st">&quot;K&quot;</span>: [<span class="st">&quot;AAA&quot;</span>, <span class="st">&quot;AAG&quot;</span>],</span>
<span id="cb2-9"><a href="compositional-analyses-of-biological-sequences.html#cb2-9" tabindex="-1"></a>                      <span class="st">&quot;N&quot;</span>: [<span class="st">&quot;AAT&quot;</span>, <span class="st">&quot;AAC&quot;</span>], </span>
<span id="cb2-10"><a href="compositional-analyses-of-biological-sequences.html#cb2-10" tabindex="-1"></a>                      <span class="st">&quot;Q&quot;</span>: [<span class="st">&quot;CAA&quot;</span>, <span class="st">&quot;CAG&quot;</span>],</span>
<span id="cb2-11"><a href="compositional-analyses-of-biological-sequences.html#cb2-11" tabindex="-1"></a>                      <span class="st">&quot;Y&quot;</span>: [<span class="st">&quot;TAT&quot;</span>, <span class="st">&quot;TAC&quot;</span>],</span>
<span id="cb2-12"><a href="compositional-analyses-of-biological-sequences.html#cb2-12" tabindex="-1"></a>                      <span class="st">&quot;I&quot;</span>: [<span class="st">&quot;ATT&quot;</span>, <span class="st">&quot;ATC&quot;</span>, <span class="st">&quot;ATA&quot;</span>],</span>
<span id="cb2-13"><a href="compositional-analyses-of-biological-sequences.html#cb2-13" tabindex="-1"></a>                      <span class="st">&quot;A&quot;</span>: [<span class="st">&quot;GCT&quot;</span>, <span class="st">&quot;GCC&quot;</span>, <span class="st">&quot;GCA&quot;</span>, <span class="st">&quot;GCG&quot;</span>], </span>
<span id="cb2-14"><a href="compositional-analyses-of-biological-sequences.html#cb2-14" tabindex="-1"></a>                      <span class="st">&quot;G&quot;</span>: [<span class="st">&quot;GGT&quot;</span>, <span class="st">&quot;GGC&quot;</span>, <span class="st">&quot;GGA&quot;</span>, <span class="st">&quot;GGG&quot;</span>],</span>
<span id="cb2-15"><a href="compositional-analyses-of-biological-sequences.html#cb2-15" tabindex="-1"></a>                      <span class="st">&quot;P&quot;</span>: [<span class="st">&quot;CCT&quot;</span>, <span class="st">&quot;CCC&quot;</span>, <span class="st">&quot;CCA&quot;</span>, <span class="st">&quot;CCG&quot;</span>],</span>
<span id="cb2-16"><a href="compositional-analyses-of-biological-sequences.html#cb2-16" tabindex="-1"></a>                      <span class="st">&quot;T&quot;</span>: [<span class="st">&quot;ACT&quot;</span>, <span class="st">&quot;ACC&quot;</span>, <span class="st">&quot;ACA&quot;</span>, <span class="st">&quot;ACG&quot;</span>],</span>
<span id="cb2-17"><a href="compositional-analyses-of-biological-sequences.html#cb2-17" tabindex="-1"></a>                      <span class="st">&quot;V&quot;</span>: [<span class="st">&quot;GTT&quot;</span>, <span class="st">&quot;GTC&quot;</span>, <span class="st">&quot;GTA&quot;</span>, <span class="st">&quot;GTG&quot;</span>],</span>
<span id="cb2-18"><a href="compositional-analyses-of-biological-sequences.html#cb2-18" tabindex="-1"></a>                      <span class="st">&quot;L&quot;</span>: [<span class="st">&quot;CTT&quot;</span>, <span class="st">&quot;CTC&quot;</span>, <span class="st">&quot;CTA&quot;</span>, <span class="st">&quot;CTG&quot;</span>, <span class="st">&quot;TTA&quot;</span>, <span class="st">&quot;TTG&quot;</span>],</span>
<span id="cb2-19"><a href="compositional-analyses-of-biological-sequences.html#cb2-19" tabindex="-1"></a>                      <span class="st">&quot;R&quot;</span>: [<span class="st">&quot;CGT&quot;</span>, <span class="st">&quot;CGC&quot;</span>, <span class="st">&quot;CGA&quot;</span>, <span class="st">&quot;CGG&quot;</span>, <span class="st">&quot;AGA&quot;</span>, <span class="st">&quot;AGG&quot;</span>],</span>
<span id="cb2-20"><a href="compositional-analyses-of-biological-sequences.html#cb2-20" tabindex="-1"></a>                      <span class="st">&quot;S&quot;</span>: [<span class="st">&quot;TCT&quot;</span>, <span class="st">&quot;TCC&quot;</span>, <span class="st">&quot;TCA&quot;</span>, <span class="st">&quot;TCG&quot;</span>, <span class="st">&quot;AGT&quot;</span>, <span class="st">&quot;AGC&quot;</span>],</span>
<span id="cb2-21"><a href="compositional-analyses-of-biological-sequences.html#cb2-21" tabindex="-1"></a>                      <span class="st">&quot;STOP&quot;</span>: [<span class="st">&quot;TAA&quot;</span>, <span class="st">&quot;TGA&quot;</span>, <span class="st">&quot;TAG&quot;</span>]</span>
<span id="cb2-22"><a href="compositional-analyses-of-biological-sequences.html#cb2-22" tabindex="-1"></a>                      }</span>
<span id="cb2-23"><a href="compositional-analyses-of-biological-sequences.html#cb2-23" tabindex="-1"></a></span>
<span id="cb2-24"><a href="compositional-analyses-of-biological-sequences.html#cb2-24" tabindex="-1"></a>features_aa <span class="op">=</span> {<span class="st">&quot;nonpolarbig&quot;</span>: [<span class="st">&quot;F&quot;</span>, <span class="st">&quot;W&quot;</span>, <span class="st">&quot;L&quot;</span>, <span class="st">&quot;I&quot;</span>, <span class="st">&quot;M&quot;</span>],</span>
<span id="cb2-25"><a href="compositional-analyses-of-biological-sequences.html#cb2-25" tabindex="-1"></a>               <span class="st">&quot;nonpolar&quot;</span>: [<span class="st">&quot;V&quot;</span>, <span class="st">&quot;P&quot;</span>, <span class="st">&quot;A&quot;</span>, <span class="st">&quot;G&quot;</span>],</span>
<span id="cb2-26"><a href="compositional-analyses-of-biological-sequences.html#cb2-26" tabindex="-1"></a>               <span class="st">&quot;polar&quot;</span>: [<span class="st">&quot;S&quot;</span>, <span class="st">&quot;C&quot;</span>, <span class="st">&quot;N&quot;</span>, <span class="st">&quot;Q&quot;</span>, <span class="st">&quot;T&quot;</span>, <span class="st">&quot;Y&quot;</span>],</span>
<span id="cb2-27"><a href="compositional-analyses-of-biological-sequences.html#cb2-27" tabindex="-1"></a>               <span class="st">&quot;basic&quot;</span>: [<span class="st">&quot;K&quot;</span>, <span class="st">&quot;R&quot;</span>, <span class="st">&quot;H&quot;</span>],</span>
<span id="cb2-28"><a href="compositional-analyses-of-biological-sequences.html#cb2-28" tabindex="-1"></a>               <span class="st">&quot;acidic&quot;</span>: [<span class="st">&quot;D&quot;</span>, <span class="st">&quot;E&quot;</span>]}</span>
<span id="cb2-29"><a href="compositional-analyses-of-biological-sequences.html#cb2-29" tabindex="-1"></a></span>
<span id="cb2-30"><a href="compositional-analyses-of-biological-sequences.html#cb2-30" tabindex="-1"></a><span class="co"># Bind the codon correspondence of amino acids and amino acid features</span></span>
<span id="cb2-31"><a href="compositional-analyses-of-biological-sequences.html#cb2-31" tabindex="-1"></a>codons_byfeatures <span class="op">=</span> {}</span>
<span id="cb2-32"><a href="compositional-analyses-of-biological-sequences.html#cb2-32" tabindex="-1"></a><span class="cf">for</span> feat <span class="kw">in</span> features_aa:</span>
<span id="cb2-33"><a href="compositional-analyses-of-biological-sequences.html#cb2-33" tabindex="-1"></a>    codons_byfeatures[feat] <span class="op">=</span> []</span>
<span id="cb2-34"><a href="compositional-analyses-of-biological-sequences.html#cb2-34" tabindex="-1"></a>    <span class="cf">for</span> aminoacid <span class="kw">in</span> features_aa[feat]:</span>
<span id="cb2-35"><a href="compositional-analyses-of-biological-sequences.html#cb2-35" tabindex="-1"></a>        <span class="cf">for</span> codon <span class="kw">in</span> table_DNA_to_codon[aminoacid]:</span>
<span id="cb2-36"><a href="compositional-analyses-of-biological-sequences.html#cb2-36" tabindex="-1"></a>            codons_byfeatures[feat].append(codon)</span>
<span id="cb2-37"><a href="compositional-analyses-of-biological-sequences.html#cb2-37" tabindex="-1"></a></span>
<span id="cb2-38"><a href="compositional-analyses-of-biological-sequences.html#cb2-38" tabindex="-1"></a><span class="co"># Add STOP codons</span></span>
<span id="cb2-39"><a href="compositional-analyses-of-biological-sequences.html#cb2-39" tabindex="-1"></a>codons_byfeatures[<span class="st">&quot;STOP&quot;</span>] <span class="op">=</span> table_DNA_to_codon[<span class="st">&quot;STOP&quot;</span>]</span></code></pre></div>
<div id="amino-acids" class="section level2 hasAnchor" number="5.1">
<h2><span class="header-section-number">5.1</span> Amino acids<a href="compositional-analyses-of-biological-sequences.html#amino-acids" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>We counted the number of amino acids within each group of sequences, encompassing the 20 essential amino acids.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb3-1"><a href="compositional-analyses-of-biological-sequences.html#cb3-1" tabindex="-1"></a></span>
<span id="cb3-2"><a href="compositional-analyses-of-biological-sequences.html#cb3-2" tabindex="-1"></a><span class="kw">def</span> getAA(seq, letter):</span>
<span id="cb3-3"><a href="compositional-analyses-of-biological-sequences.html#cb3-3" tabindex="-1"></a>    <span class="cf">return</span>(<span class="bu">round</span>((<span class="bu">sum</span>([<span class="fl">1.0</span> <span class="cf">for</span> nucl <span class="kw">in</span> seq <span class="cf">if</span> nucl <span class="kw">in</span> [letter]]) <span class="op">/</span> (<span class="bu">len</span>(seq)<span class="op">-</span>seq.count(<span class="st">&quot;-&quot;</span>))), ndigits<span class="op">=</span><span class="dv">2</span>))</span>
<span id="cb3-4"><a href="compositional-analyses-of-biological-sequences.html#cb3-4" tabindex="-1"></a></span>
<span id="cb3-5"><a href="compositional-analyses-of-biological-sequences.html#cb3-5" tabindex="-1"></a>paths <span class="op">=</span> [<span class="st">&quot;p19_aligned.fa&quot;</span>, <span class="st">&quot;p22_aligned.fa&quot;</span>, <span class="st">&quot;p22 core_aligned.fa&quot;</span>, <span class="st">&quot;other cores_allMP.fa&quot;</span>, <span class="st">&quot;RdRp_aligned.fa&quot;</span>, <span class="st">&quot;CP_aligned.fa&quot;</span>] </span>
<span id="cb3-6"><a href="compositional-analyses-of-biological-sequences.html#cb3-6" tabindex="-1"></a>letters <span class="op">=</span> [<span class="st">&quot;M&quot;</span>, <span class="st">&quot;W&quot;</span>, <span class="st">&quot;C&quot;</span>, <span class="st">&quot;D&quot;</span>, <span class="st">&quot;E&quot;</span>, <span class="st">&quot;F&quot;</span>, <span class="st">&quot;H&quot;</span>, <span class="st">&quot;K&quot;</span>, <span class="st">&quot;N&quot;</span>, <span class="st">&quot;Q&quot;</span>, </span>
<span id="cb3-7"><a href="compositional-analyses-of-biological-sequences.html#cb3-7" tabindex="-1"></a>            <span class="st">&quot;Y&quot;</span>, <span class="st">&quot;I&quot;</span>, <span class="st">&quot;A&quot;</span>, <span class="st">&quot;G&quot;</span>, <span class="st">&quot;P&quot;</span>, <span class="st">&quot;T&quot;</span>, <span class="st">&quot;V&quot;</span>, <span class="st">&quot;L&quot;</span>, <span class="st">&quot;R&quot;</span>, <span class="st">&quot;S&quot;</span>]</span>
<span id="cb3-8"><a href="compositional-analyses-of-biological-sequences.html#cb3-8" tabindex="-1"></a>df_aa <span class="op">=</span> []</span>
<span id="cb3-9"><a href="compositional-analyses-of-biological-sequences.html#cb3-9" tabindex="-1"></a>aa_means <span class="op">=</span> []</span>
<span id="cb3-10"><a href="compositional-analyses-of-biological-sequences.html#cb3-10" tabindex="-1"></a>aa_medians <span class="op">=</span> []</span>
<span id="cb3-11"><a href="compositional-analyses-of-biological-sequences.html#cb3-11" tabindex="-1"></a>aa_std <span class="op">=</span> []</span>
<span id="cb3-12"><a href="compositional-analyses-of-biological-sequences.html#cb3-12" tabindex="-1"></a></span>
<span id="cb3-13"><a href="compositional-analyses-of-biological-sequences.html#cb3-13" tabindex="-1"></a><span class="cf">for</span> i, aminoacid <span class="kw">in</span> <span class="bu">enumerate</span>(letters): </span>
<span id="cb3-14"><a href="compositional-analyses-of-biological-sequences.html#cb3-14" tabindex="-1"></a>    AA_content <span class="op">=</span> []</span>
<span id="cb3-15"><a href="compositional-analyses-of-biological-sequences.html#cb3-15" tabindex="-1"></a>    <span class="cf">for</span> alignment <span class="kw">in</span> paths:</span>
<span id="cb3-16"><a href="compositional-analyses-of-biological-sequences.html#cb3-16" tabindex="-1"></a>        name <span class="op">=</span> alignment[<span class="dv">0</span>:alignment.find(<span class="st">&quot;_&quot;</span>)]</span>
<span id="cb3-17"><a href="compositional-analyses-of-biological-sequences.html#cb3-17" tabindex="-1"></a>        <span class="cf">for</span> record <span class="kw">in</span> SeqIO.parse(alignment, <span class="st">&quot;fasta&quot;</span>):</span>
<span id="cb3-18"><a href="compositional-analyses-of-biological-sequences.html#cb3-18" tabindex="-1"></a>            count <span class="op">=</span> getAA(<span class="bu">str</span>(record.seq), aminoacid)</span>
<span id="cb3-19"><a href="compositional-analyses-of-biological-sequences.html#cb3-19" tabindex="-1"></a>            AA_content.append([count, name])</span>
<span id="cb3-20"><a href="compositional-analyses-of-biological-sequences.html#cb3-20" tabindex="-1"></a>    name <span class="op">=</span> aminoacid</span>
<span id="cb3-21"><a href="compositional-analyses-of-biological-sequences.html#cb3-21" tabindex="-1"></a>    df_aa.append(pd.DataFrame(AA_content, columns<span class="op">=</span>[name, <span class="st">&quot;Sequence&quot;</span>]))</span>
<span id="cb3-22"><a href="compositional-analyses-of-biological-sequences.html#cb3-22" tabindex="-1"></a></span>
<span id="cb3-23"><a href="compositional-analyses-of-biological-sequences.html#cb3-23" tabindex="-1"></a>    aa_means.append(df_aa[i].groupby(<span class="st">&quot;Sequence&quot;</span>).mean())</span>
<span id="cb3-24"><a href="compositional-analyses-of-biological-sequences.html#cb3-24" tabindex="-1"></a>    aa_medians.append(df_aa[i].groupby(<span class="st">&quot;Sequence&quot;</span>).median())</span>
<span id="cb3-25"><a href="compositional-analyses-of-biological-sequences.html#cb3-25" tabindex="-1"></a>    aa_std.append(df_aa[i].groupby(<span class="st">&quot;Sequence&quot;</span>).std())</span></code></pre></div>
<p>In this case, we chose to plot and compare the medians of amino acid proportions (out of the total number of amino acids in the sequence).</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb4-1"><a href="compositional-analyses-of-biological-sequences.html#cb4-1" tabindex="-1"></a>hm_medians <span class="op">=</span> pd.concat(aa_medians, axis <span class="op">=</span> <span class="dv">1</span>)</span>
<span id="cb4-2"><a href="compositional-analyses-of-biological-sequences.html#cb4-2" tabindex="-1"></a>hm_medians <span class="op">=</span> hm_medians.reindex(index <span class="op">=</span> [<span class="st">&#39;RdRp&#39;</span>,<span class="st">&#39;CP&#39;</span>,<span class="st">&#39;p19&#39;</span>, <span class="st">&#39;p22&#39;</span>, <span class="st">&#39;p22 core&#39;</span>, <span class="st">&#39;other cores&#39;</span>])</span>
<span id="cb4-3"><a href="compositional-analyses-of-biological-sequences.html#cb4-3" tabindex="-1"></a></span>
<span id="cb4-4"><a href="compositional-analyses-of-biological-sequences.html#cb4-4" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots()</span>
<span id="cb4-5"><a href="compositional-analyses-of-biological-sequences.html#cb4-5" tabindex="-1"></a>plt.rcParams[<span class="st">&quot;figure.figsize&quot;</span>] <span class="op">=</span> [<span class="dv">12</span>,<span class="fl">1.5</span>]</span>
<span id="cb4-6"><a href="compositional-analyses-of-biological-sequences.html#cb4-6" tabindex="-1"></a>sns.heatmap(hm_medians, ax<span class="op">=</span>ax, cmap<span class="op">=</span><span class="st">&quot;viridis&quot;</span>)</span>
<span id="cb4-7"><a href="compositional-analyses-of-biological-sequences.html#cb4-7" tabindex="-1"></a>plt.xlabel(<span class="st">&quot;Amino acid&quot;</span>)</span>
<span id="cb4-8"><a href="compositional-analyses-of-biological-sequences.html#cb4-8" tabindex="-1"></a></span>
<span id="cb4-9"><a href="compositional-analyses-of-biological-sequences.html#cb4-9" tabindex="-1"></a>plt.tight_layout()</span></code></pre></div>
</div>
<div id="nucleotides" class="section level2 hasAnchor" number="5.2">
<h2><span class="header-section-number">5.2</span> Nucleotides<a href="compositional-analyses-of-biological-sequences.html#nucleotides" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Our initial approach involved quantifying the proportion of nucleotides in various genes, with special attention to the overlapping region of the p19 and p22 genes.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb5-1"><a href="compositional-analyses-of-biological-sequences.html#cb5-1" tabindex="-1"></a><span class="kw">def</span> getNT(seq, letter):</span>
<span id="cb5-2"><a href="compositional-analyses-of-biological-sequences.html#cb5-2" tabindex="-1"></a>    <span class="cf">return</span>(<span class="bu">round</span>((<span class="bu">sum</span>([<span class="fl">1.0</span> <span class="cf">for</span> nucl <span class="kw">in</span> seq <span class="cf">if</span> nucl <span class="kw">in</span> [letter]]) <span class="op">/</span> (<span class="bu">len</span>(seq)<span class="op">-</span> seq.count(<span class="st">&quot;-&quot;</span>))), ndigits<span class="op">=</span><span class="dv">2</span>))</span>
<span id="cb5-3"><a href="compositional-analyses-of-biological-sequences.html#cb5-3" tabindex="-1"></a></span>
<span id="cb5-4"><a href="compositional-analyses-of-biological-sequences.html#cb5-4" tabindex="-1"></a>paths <span class="op">=</span> [<span class="st">&quot;RdRp_nt_aligned.fa&quot;</span>, <span class="st">&quot;CP_nt_aligned.fa&quot;</span>, <span class="st">&quot;p19_nt_aligned.fa&quot;</span>, <span class="st">&quot;p22_nt_aligned.fa&quot;</span>, <span class="st">&quot;/Users/esmeralda/Documents/TFM/RevTrans/final/nt/alignments/concatenate/p22 core_nt_alignment.fa&quot;</span>, <span class="st">&quot;other cores_nt.fa&quot;</span>] </span>
<span id="cb5-5"><a href="compositional-analyses-of-biological-sequences.html#cb5-5" tabindex="-1"></a>letters <span class="op">=</span> [<span class="st">&quot;A&quot;</span>, <span class="st">&quot;T&quot;</span>, <span class="st">&quot;C&quot;</span>, <span class="st">&quot;G&quot;</span>]</span>
<span id="cb5-6"><a href="compositional-analyses-of-biological-sequences.html#cb5-6" tabindex="-1"></a>df_nt <span class="op">=</span> []</span>
<span id="cb5-7"><a href="compositional-analyses-of-biological-sequences.html#cb5-7" tabindex="-1"></a>nt_only_means <span class="op">=</span> []</span>
<span id="cb5-8"><a href="compositional-analyses-of-biological-sequences.html#cb5-8" tabindex="-1"></a>nt_only_medians <span class="op">=</span> []</span>
<span id="cb5-9"><a href="compositional-analyses-of-biological-sequences.html#cb5-9" tabindex="-1"></a>nt_only_std <span class="op">=</span> []</span>
<span id="cb5-10"><a href="compositional-analyses-of-biological-sequences.html#cb5-10" tabindex="-1"></a></span>
<span id="cb5-11"><a href="compositional-analyses-of-biological-sequences.html#cb5-11" tabindex="-1"></a><span class="cf">for</span> i, nucleot <span class="kw">in</span> <span class="bu">enumerate</span>(letters):</span>
<span id="cb5-12"><a href="compositional-analyses-of-biological-sequences.html#cb5-12" tabindex="-1"></a>    NT_content <span class="op">=</span> []</span>
<span id="cb5-13"><a href="compositional-analyses-of-biological-sequences.html#cb5-13" tabindex="-1"></a>    <span class="cf">for</span> alignment <span class="kw">in</span> paths:</span>
<span id="cb5-14"><a href="compositional-analyses-of-biological-sequences.html#cb5-14" tabindex="-1"></a>        name <span class="op">=</span> alignment[<span class="dv">0</span>:alignment.find(<span class="st">&quot;_&quot;</span>)]</span>
<span id="cb5-15"><a href="compositional-analyses-of-biological-sequences.html#cb5-15" tabindex="-1"></a>        <span class="cf">if</span> <span class="bu">len</span>(name) <span class="op">&gt;</span> <span class="dv">50</span>:</span>
<span id="cb5-16"><a href="compositional-analyses-of-biological-sequences.html#cb5-16" tabindex="-1"></a>             name <span class="op">=</span> <span class="st">&quot;p22 core&quot;</span></span>
<span id="cb5-17"><a href="compositional-analyses-of-biological-sequences.html#cb5-17" tabindex="-1"></a>        <span class="cf">for</span> record <span class="kw">in</span> SeqIO.parse(alignment, <span class="st">&quot;fasta&quot;</span>):</span>
<span id="cb5-18"><a href="compositional-analyses-of-biological-sequences.html#cb5-18" tabindex="-1"></a>            count <span class="op">=</span> getNT(<span class="bu">str</span>(record.seq), nucleot)</span>
<span id="cb5-19"><a href="compositional-analyses-of-biological-sequences.html#cb5-19" tabindex="-1"></a>            NT_content.append([count, name])</span>
<span id="cb5-20"><a href="compositional-analyses-of-biological-sequences.html#cb5-20" tabindex="-1"></a>    <span class="cf">if</span> nucleot <span class="op">==</span> <span class="st">&quot;T&quot;</span>:</span>
<span id="cb5-21"><a href="compositional-analyses-of-biological-sequences.html#cb5-21" tabindex="-1"></a>        nucleot <span class="op">=</span> <span class="st">&quot;U&quot;</span></span>
<span id="cb5-22"><a href="compositional-analyses-of-biological-sequences.html#cb5-22" tabindex="-1"></a>    </span>
<span id="cb5-23"><a href="compositional-analyses-of-biological-sequences.html#cb5-23" tabindex="-1"></a>    name <span class="op">=</span> nucleot</span>
<span id="cb5-24"><a href="compositional-analyses-of-biological-sequences.html#cb5-24" tabindex="-1"></a>    df_nt.append(pd.DataFrame(NT_content, columns<span class="op">=</span>[name, <span class="st">&quot;Sequence&quot;</span>]))</span>
<span id="cb5-25"><a href="compositional-analyses-of-biological-sequences.html#cb5-25" tabindex="-1"></a></span>
<span id="cb5-26"><a href="compositional-analyses-of-biological-sequences.html#cb5-26" tabindex="-1"></a>    nt_only_means.append(df_nt[i].groupby(<span class="st">&quot;Sequence&quot;</span>).mean())</span>
<span id="cb5-27"><a href="compositional-analyses-of-biological-sequences.html#cb5-27" tabindex="-1"></a>    nt_only_medians.append(df_nt[i].groupby(<span class="st">&quot;Sequence&quot;</span>).median())</span>
<span id="cb5-28"><a href="compositional-analyses-of-biological-sequences.html#cb5-28" tabindex="-1"></a>    nt_only_std.append(df_nt[i].groupby(<span class="st">&quot;Sequence&quot;</span>).std())</span></code></pre></div>
<p>Additionally, we were interested in the positions of codons (1st, 2nd, and 3rd nucleotides) and their nucleotide abundances.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb6-1"><a href="compositional-analyses-of-biological-sequences.html#cb6-1" tabindex="-1"></a><span class="kw">def</span> getPosCodon(seq, letter, start):</span>
<span id="cb6-2"><a href="compositional-analyses-of-biological-sequences.html#cb6-2" tabindex="-1"></a>    count <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb6-3"><a href="compositional-analyses-of-biological-sequences.html#cb6-3" tabindex="-1"></a>    <span class="cf">for</span> nucl <span class="kw">in</span> <span class="bu">range</span>(start, <span class="bu">len</span>(seq), <span class="dv">3</span>):</span>
<span id="cb6-4"><a href="compositional-analyses-of-biological-sequences.html#cb6-4" tabindex="-1"></a>        <span class="cf">if</span> seq[nucl] <span class="op">==</span> letter:</span>
<span id="cb6-5"><a href="compositional-analyses-of-biological-sequences.html#cb6-5" tabindex="-1"></a>            count <span class="op">+=</span> <span class="fl">1.0</span></span>
<span id="cb6-6"><a href="compositional-analyses-of-biological-sequences.html#cb6-6" tabindex="-1"></a>    </span>
<span id="cb6-7"><a href="compositional-analyses-of-biological-sequences.html#cb6-7" tabindex="-1"></a>    <span class="cf">return</span> <span class="bu">round</span>(count<span class="op">/</span>((<span class="bu">len</span>(seq)<span class="op">-</span> seq.count(<span class="st">&quot;-&quot;</span>))<span class="op">/</span><span class="dv">3</span>), ndigits<span class="op">=</span><span class="dv">2</span>)</span>
<span id="cb6-8"><a href="compositional-analyses-of-biological-sequences.html#cb6-8" tabindex="-1"></a></span>
<span id="cb6-9"><a href="compositional-analyses-of-biological-sequences.html#cb6-9" tabindex="-1"></a>general <span class="op">=</span> []</span>
<span id="cb6-10"><a href="compositional-analyses-of-biological-sequences.html#cb6-10" tabindex="-1"></a></span>
<span id="cb6-11"><a href="compositional-analyses-of-biological-sequences.html#cb6-11" tabindex="-1"></a><span class="cf">for</span> start <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">3</span>):</span>
<span id="cb6-12"><a href="compositional-analyses-of-biological-sequences.html#cb6-12" tabindex="-1"></a>    df_nt <span class="op">=</span> []</span>
<span id="cb6-13"><a href="compositional-analyses-of-biological-sequences.html#cb6-13" tabindex="-1"></a>    nt_pos_means <span class="op">=</span> []</span>
<span id="cb6-14"><a href="compositional-analyses-of-biological-sequences.html#cb6-14" tabindex="-1"></a>    nt_pos_medians <span class="op">=</span> []</span>
<span id="cb6-15"><a href="compositional-analyses-of-biological-sequences.html#cb6-15" tabindex="-1"></a>    nt_pos_std <span class="op">=</span> []</span>
<span id="cb6-16"><a href="compositional-analyses-of-biological-sequences.html#cb6-16" tabindex="-1"></a></span>
<span id="cb6-17"><a href="compositional-analyses-of-biological-sequences.html#cb6-17" tabindex="-1"></a>    <span class="cf">for</span> i, nucleot <span class="kw">in</span> <span class="bu">enumerate</span>(letters):</span>
<span id="cb6-18"><a href="compositional-analyses-of-biological-sequences.html#cb6-18" tabindex="-1"></a>        NT_content <span class="op">=</span> []</span>
<span id="cb6-19"><a href="compositional-analyses-of-biological-sequences.html#cb6-19" tabindex="-1"></a>        <span class="cf">for</span> alignment <span class="kw">in</span> paths:</span>
<span id="cb6-20"><a href="compositional-analyses-of-biological-sequences.html#cb6-20" tabindex="-1"></a>            name <span class="op">=</span> alignment[<span class="dv">0</span>:alignment.find(<span class="st">&quot;_&quot;</span>)]</span>
<span id="cb6-21"><a href="compositional-analyses-of-biological-sequences.html#cb6-21" tabindex="-1"></a>            <span class="cf">if</span> <span class="bu">len</span>(name) <span class="op">&gt;</span> <span class="dv">50</span>:</span>
<span id="cb6-22"><a href="compositional-analyses-of-biological-sequences.html#cb6-22" tabindex="-1"></a>                name <span class="op">=</span> <span class="st">&quot;p22 core&quot;</span></span>
<span id="cb6-23"><a href="compositional-analyses-of-biological-sequences.html#cb6-23" tabindex="-1"></a>            <span class="cf">for</span> record <span class="kw">in</span> SeqIO.parse(alignment, <span class="st">&quot;fasta&quot;</span>):</span>
<span id="cb6-24"><a href="compositional-analyses-of-biological-sequences.html#cb6-24" tabindex="-1"></a>                count <span class="op">=</span> getPosCodon(<span class="bu">str</span>(record.seq), nucleot, start)</span>
<span id="cb6-25"><a href="compositional-analyses-of-biological-sequences.html#cb6-25" tabindex="-1"></a>                NT_content.append([count, name])</span>
<span id="cb6-26"><a href="compositional-analyses-of-biological-sequences.html#cb6-26" tabindex="-1"></a>        </span>
<span id="cb6-27"><a href="compositional-analyses-of-biological-sequences.html#cb6-27" tabindex="-1"></a>        <span class="cf">if</span> nucleot <span class="op">==</span> <span class="st">&quot;T&quot;</span>:</span>
<span id="cb6-28"><a href="compositional-analyses-of-biological-sequences.html#cb6-28" tabindex="-1"></a>            name <span class="op">=</span> <span class="st">&quot;U&quot;</span> <span class="op">+</span> <span class="bu">str</span>(start <span class="op">+</span> <span class="dv">1</span>)</span>
<span id="cb6-29"><a href="compositional-analyses-of-biological-sequences.html#cb6-29" tabindex="-1"></a>        <span class="cf">else</span>:</span>
<span id="cb6-30"><a href="compositional-analyses-of-biological-sequences.html#cb6-30" tabindex="-1"></a>            name <span class="op">=</span> nucleot <span class="op">+</span> <span class="bu">str</span>(start <span class="op">+</span> <span class="dv">1</span>)</span>
<span id="cb6-31"><a href="compositional-analyses-of-biological-sequences.html#cb6-31" tabindex="-1"></a>            </span>
<span id="cb6-32"><a href="compositional-analyses-of-biological-sequences.html#cb6-32" tabindex="-1"></a>        df_nt.append(pd.DataFrame(NT_content, columns<span class="op">=</span>[name, <span class="st">&quot;Sequence&quot;</span>]))</span>
<span id="cb6-33"><a href="compositional-analyses-of-biological-sequences.html#cb6-33" tabindex="-1"></a></span>
<span id="cb6-34"><a href="compositional-analyses-of-biological-sequences.html#cb6-34" tabindex="-1"></a>        nt_pos_means.append(df_nt[i].groupby(<span class="st">&quot;Sequence&quot;</span>).mean())</span>
<span id="cb6-35"><a href="compositional-analyses-of-biological-sequences.html#cb6-35" tabindex="-1"></a>        nt_pos_medians.append(df_nt[i].groupby(<span class="st">&quot;Sequence&quot;</span>).median())</span>
<span id="cb6-36"><a href="compositional-analyses-of-biological-sequences.html#cb6-36" tabindex="-1"></a>        nt_pos_std.append(df_nt[i].groupby(<span class="st">&quot;Sequence&quot;</span>).std())</span>
<span id="cb6-37"><a href="compositional-analyses-of-biological-sequences.html#cb6-37" tabindex="-1"></a></span>
<span id="cb6-38"><a href="compositional-analyses-of-biological-sequences.html#cb6-38" tabindex="-1"></a>    general.append(nt_pos_medians)</span></code></pre></div>
<p>We now combine all nucleotide data and generate a heatmap.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb7-1"><a href="compositional-analyses-of-biological-sequences.html#cb7-1" tabindex="-1"></a>nucleotides_matrix <span class="op">=</span> pd.concat([nt_only_medians[<span class="dv">0</span>],nt_only_medians[<span class="dv">1</span>],nt_only_medians[<span class="dv">2</span>],nt_only_medians[<span class="dv">3</span>],</span>
<span id="cb7-2"><a href="compositional-analyses-of-biological-sequences.html#cb7-2" tabindex="-1"></a>                                                general[<span class="dv">0</span>][<span class="dv">0</span>], general[<span class="dv">0</span>][<span class="dv">1</span>], general[<span class="dv">0</span>][<span class="dv">2</span>], general[<span class="dv">0</span>][<span class="dv">3</span>],</span>
<span id="cb7-3"><a href="compositional-analyses-of-biological-sequences.html#cb7-3" tabindex="-1"></a>                                                general[<span class="dv">1</span>][<span class="dv">0</span>], general[<span class="dv">1</span>][<span class="dv">1</span>], general[<span class="dv">1</span>][<span class="dv">2</span>], general[<span class="dv">1</span>][<span class="dv">3</span>],</span>
<span id="cb7-4"><a href="compositional-analyses-of-biological-sequences.html#cb7-4" tabindex="-1"></a>                                                general[<span class="dv">2</span>][<span class="dv">0</span>],general[<span class="dv">2</span>][<span class="dv">1</span>], general[<span class="dv">2</span>][<span class="dv">2</span>], general[<span class="dv">2</span>][<span class="dv">3</span>]],</span>
<span id="cb7-5"><a href="compositional-analyses-of-biological-sequences.html#cb7-5" tabindex="-1"></a>                                                axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb7-6"><a href="compositional-analyses-of-biological-sequences.html#cb7-6" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots()</span>
<span id="cb7-7"><a href="compositional-analyses-of-biological-sequences.html#cb7-7" tabindex="-1"></a>plt.rcParams[<span class="st">&quot;figure.figsize&quot;</span>] <span class="op">=</span> [<span class="dv">20</span>,<span class="fl">1.5</span>]</span>
<span id="cb7-8"><a href="compositional-analyses-of-biological-sequences.html#cb7-8" tabindex="-1"></a>nucleotides_matrix <span class="op">=</span> nucleotides_matrix.reindex(index <span class="op">=</span> [<span class="st">&#39;RdRp&#39;</span>,<span class="st">&#39;CP&#39;</span>,<span class="st">&#39;p19&#39;</span>, <span class="st">&#39;p22&#39;</span>, <span class="st">&#39;p22 core&#39;</span>, <span class="st">&#39;other cores&#39;</span>])</span>
<span id="cb7-9"><a href="compositional-analyses-of-biological-sequences.html#cb7-9" tabindex="-1"></a>sns.heatmap(nucleotides_matrix, ax<span class="op">=</span>ax, cmap<span class="op">=</span><span class="st">&quot;viridis&quot;</span>)</span>
<span id="cb7-10"><a href="compositional-analyses-of-biological-sequences.html#cb7-10" tabindex="-1"></a>plt.xlabel(<span class="st">&quot;Nucleotide&quot;</span>)</span>
<span id="cb7-11"><a href="compositional-analyses-of-biological-sequences.html#cb7-11" tabindex="-1"></a></span>
<span id="cb7-12"><a href="compositional-analyses-of-biological-sequences.html#cb7-12" tabindex="-1"></a>plt.tight_layout()</span></code></pre></div>
</div>
<div id="dinuclotides" class="section level2 hasAnchor" number="5.3">
<h2><span class="header-section-number">5.3</span> Dinuclotides<a href="compositional-analyses-of-biological-sequences.html#dinuclotides" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>We read dinucleotides, calculate their relativized values, and plot the medians.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb8-1"><a href="compositional-analyses-of-biological-sequences.html#cb8-1" tabindex="-1"></a></span>
<span id="cb8-2"><a href="compositional-analyses-of-biological-sequences.html#cb8-2" tabindex="-1"></a><span class="kw">def</span> read_diNT(path):</span>
<span id="cb8-3"><a href="compositional-analyses-of-biological-sequences.html#cb8-3" tabindex="-1"></a>    chains_dint <span class="op">=</span> SeqIO.parse(path, <span class="st">&quot;fasta&quot;</span>)</span>
<span id="cb8-4"><a href="compositional-analyses-of-biological-sequences.html#cb8-4" tabindex="-1"></a></span>
<span id="cb8-5"><a href="compositional-analyses-of-biological-sequences.html#cb8-5" tabindex="-1"></a>    chain_dint_p22_core <span class="op">=</span> []</span>
<span id="cb8-6"><a href="compositional-analyses-of-biological-sequences.html#cb8-6" tabindex="-1"></a>    <span class="cf">for</span> chain <span class="kw">in</span> chains_dint:</span>
<span id="cb8-7"><a href="compositional-analyses-of-biological-sequences.html#cb8-7" tabindex="-1"></a>        dint <span class="op">=</span> {}</span>
<span id="cb8-8"><a href="compositional-analyses-of-biological-sequences.html#cb8-8" tabindex="-1"></a>        seq <span class="op">=</span> <span class="bu">str</span>(chain.seq)</span>
<span id="cb8-9"><a href="compositional-analyses-of-biological-sequences.html#cb8-9" tabindex="-1"></a>        <span class="cf">for</span> ch <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(seq)<span class="op">-</span><span class="dv">1</span>):</span>
<span id="cb8-10"><a href="compositional-analyses-of-biological-sequences.html#cb8-10" tabindex="-1"></a>            dinuc <span class="op">=</span> seq[ch:ch<span class="op">+</span><span class="dv">2</span>].replace(<span class="st">&quot;T&quot;</span>, <span class="st">&quot;U&quot;</span>)</span>
<span id="cb8-11"><a href="compositional-analyses-of-biological-sequences.html#cb8-11" tabindex="-1"></a>            <span class="cf">if</span> dinuc <span class="kw">in</span> dint:</span>
<span id="cb8-12"><a href="compositional-analyses-of-biological-sequences.html#cb8-12" tabindex="-1"></a>                dint[dinuc] <span class="op">+=</span> <span class="dv">1</span></span>
<span id="cb8-13"><a href="compositional-analyses-of-biological-sequences.html#cb8-13" tabindex="-1"></a>            <span class="cf">else</span>:</span>
<span id="cb8-14"><a href="compositional-analyses-of-biological-sequences.html#cb8-14" tabindex="-1"></a>                <span class="cf">if</span> dinuc.find(<span class="st">&quot;-&quot;</span>)<span class="op">&lt;</span><span class="dv">0</span> <span class="kw">and</span> dinuc.find(<span class="st">&quot;N&quot;</span>)<span class="op">&lt;</span><span class="dv">0</span>:</span>
<span id="cb8-15"><a href="compositional-analyses-of-biological-sequences.html#cb8-15" tabindex="-1"></a>                    dint[dinuc] <span class="op">=</span> <span class="dv">1</span></span>
<span id="cb8-16"><a href="compositional-analyses-of-biological-sequences.html#cb8-16" tabindex="-1"></a>        chain_dint_p22_core.append(dint)</span>
<span id="cb8-17"><a href="compositional-analyses-of-biological-sequences.html#cb8-17" tabindex="-1"></a></span>
<span id="cb8-18"><a href="compositional-analyses-of-biological-sequences.html#cb8-18" tabindex="-1"></a>    <span class="cf">return</span> pd.DataFrame.from_dict(chain_dint_p22_core)</span>
<span id="cb8-19"><a href="compositional-analyses-of-biological-sequences.html#cb8-19" tabindex="-1"></a></span>
<span id="cb8-20"><a href="compositional-analyses-of-biological-sequences.html#cb8-20" tabindex="-1"></a><span class="kw">def</span> rel_diNT(df_dint):</span>
<span id="cb8-21"><a href="compositional-analyses-of-biological-sequences.html#cb8-21" tabindex="-1"></a>    <span class="cf">return</span> df_dint.divide(df_dint.<span class="bu">sum</span>(axis<span class="op">=</span><span class="dv">1</span>), axis<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb8-22"><a href="compositional-analyses-of-biological-sequences.html#cb8-22" tabindex="-1"></a></span>
<span id="cb8-23"><a href="compositional-analyses-of-biological-sequences.html#cb8-23" tabindex="-1"></a><span class="kw">def</span> stats_composition(df_composition, stat<span class="op">=</span><span class="st">&quot;median&quot;</span>):</span>
<span id="cb8-24"><a href="compositional-analyses-of-biological-sequences.html#cb8-24" tabindex="-1"></a>    <span class="cf">if</span> stat <span class="op">==</span> <span class="st">&quot;median&quot;</span>:</span>
<span id="cb8-25"><a href="compositional-analyses-of-biological-sequences.html#cb8-25" tabindex="-1"></a>        <span class="cf">return</span> pd.DataFrame.from_dict(df_composition.mean())</span>
<span id="cb8-26"><a href="compositional-analyses-of-biological-sequences.html#cb8-26" tabindex="-1"></a>    <span class="cf">elif</span> stat <span class="op">==</span> <span class="st">&quot;mean&quot;</span>:</span>
<span id="cb8-27"><a href="compositional-analyses-of-biological-sequences.html#cb8-27" tabindex="-1"></a>        <span class="cf">return</span> pd.DataFrame.from_dict(df_composition.median())</span>
<span id="cb8-28"><a href="compositional-analyses-of-biological-sequences.html#cb8-28" tabindex="-1"></a>    <span class="cf">elif</span> stat <span class="op">==</span> <span class="st">&quot;std&quot;</span>:</span>
<span id="cb8-29"><a href="compositional-analyses-of-biological-sequences.html#cb8-29" tabindex="-1"></a>        <span class="cf">return</span> pd.DataFrame.from_dict(df_composition.std())</span>
<span id="cb8-30"><a href="compositional-analyses-of-biological-sequences.html#cb8-30" tabindex="-1"></a>    </span>
<span id="cb8-31"><a href="compositional-analyses-of-biological-sequences.html#cb8-31" tabindex="-1"></a><span class="kw">def</span> join_medians(median_df):</span>
<span id="cb8-32"><a href="compositional-analyses-of-biological-sequences.html#cb8-32" tabindex="-1"></a>    dint_comparison <span class="op">=</span> pd.concat([median_df[<span class="dv">0</span>],</span>
<span id="cb8-33"><a href="compositional-analyses-of-biological-sequences.html#cb8-33" tabindex="-1"></a>                                median_df[<span class="dv">1</span>],</span>
<span id="cb8-34"><a href="compositional-analyses-of-biological-sequences.html#cb8-34" tabindex="-1"></a>                                median_df[<span class="dv">2</span>],</span>
<span id="cb8-35"><a href="compositional-analyses-of-biological-sequences.html#cb8-35" tabindex="-1"></a>                                median_df[<span class="dv">3</span>],</span>
<span id="cb8-36"><a href="compositional-analyses-of-biological-sequences.html#cb8-36" tabindex="-1"></a>                                median_df[<span class="dv">4</span>],</span>
<span id="cb8-37"><a href="compositional-analyses-of-biological-sequences.html#cb8-37" tabindex="-1"></a>                                median_df[<span class="dv">5</span>][<span class="dv">0</span>][<span class="bu">list</span>(median_df[<span class="dv">0</span>].index)]], axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb8-38"><a href="compositional-analyses-of-biological-sequences.html#cb8-38" tabindex="-1"></a></span>
<span id="cb8-39"><a href="compositional-analyses-of-biological-sequences.html#cb8-39" tabindex="-1"></a>    dint_comparison.columns <span class="op">=</span> [<span class="st">&quot;RdRp&quot;</span>, <span class="st">&quot;CP&quot;</span>, <span class="st">&quot;p19&quot;</span>, <span class="st">&quot;p22&quot;</span>, <span class="st">&quot;p22_core&quot;</span>, <span class="st">&quot;other cores&quot;</span>]</span>
<span id="cb8-40"><a href="compositional-analyses-of-biological-sequences.html#cb8-40" tabindex="-1"></a></span>
<span id="cb8-41"><a href="compositional-analyses-of-biological-sequences.html#cb8-41" tabindex="-1"></a>    <span class="cf">return</span> dint_comparison.T</span></code></pre></div>
<p>Then, we plot a heatmap with the shares of both nucleotides and nucleotides per codon position.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb9-1"><a href="compositional-analyses-of-biological-sequences.html#cb9-1" tabindex="-1"></a><span class="co"># Paths of fasta files with nucleotide sequences</span></span>
<span id="cb9-2"><a href="compositional-analyses-of-biological-sequences.html#cb9-2" tabindex="-1"></a>paths <span class="op">=</span> [<span class="st">&quot;RdRp_nt_aligned.fa&quot;</span>,</span>
<span id="cb9-3"><a href="compositional-analyses-of-biological-sequences.html#cb9-3" tabindex="-1"></a>         <span class="st">&quot;cp_nt_aligned.fa&quot;</span>,</span>
<span id="cb9-4"><a href="compositional-analyses-of-biological-sequences.html#cb9-4" tabindex="-1"></a>         <span class="st">&quot;p19_nt_aligned.fa&quot;</span>,</span>
<span id="cb9-5"><a href="compositional-analyses-of-biological-sequences.html#cb9-5" tabindex="-1"></a>         <span class="st">&quot;p22_nt_aligned.fa&quot;</span>,</span>
<span id="cb9-6"><a href="compositional-analyses-of-biological-sequences.html#cb9-6" tabindex="-1"></a>         <span class="st">&quot;/Users/esmeralda/Documents/TFM/RevTrans/final/nt/alignments/concatenate/p22 core_nt_alignment.fa&quot;</span>,</span>
<span id="cb9-7"><a href="compositional-analyses-of-biological-sequences.html#cb9-7" tabindex="-1"></a>         <span class="st">&quot;other cores_nt.fa&quot;</span>]</span>
<span id="cb9-8"><a href="compositional-analyses-of-biological-sequences.html#cb9-8" tabindex="-1"></a></span>
<span id="cb9-9"><a href="compositional-analyses-of-biological-sequences.html#cb9-9" tabindex="-1"></a>order_names <span class="op">=</span> []</span>
<span id="cb9-10"><a href="compositional-analyses-of-biological-sequences.html#cb9-10" tabindex="-1"></a>median_df <span class="op">=</span> []</span>
<span id="cb9-11"><a href="compositional-analyses-of-biological-sequences.html#cb9-11" tabindex="-1"></a></span>
<span id="cb9-12"><a href="compositional-analyses-of-biological-sequences.html#cb9-12" tabindex="-1"></a><span class="cf">for</span> path <span class="kw">in</span> paths:</span>
<span id="cb9-13"><a href="compositional-analyses-of-biological-sequences.html#cb9-13" tabindex="-1"></a>    df_dint <span class="op">=</span> read_diNT(path)</span>
<span id="cb9-14"><a href="compositional-analyses-of-biological-sequences.html#cb9-14" tabindex="-1"></a>    df_dint_rel <span class="op">=</span> rel_diNT(df_dint)</span>
<span id="cb9-15"><a href="compositional-analyses-of-biological-sequences.html#cb9-15" tabindex="-1"></a>    median_df.append(stats_composition(df_dint_rel))</span>
<span id="cb9-16"><a href="compositional-analyses-of-biological-sequences.html#cb9-16" tabindex="-1"></a></span>
<span id="cb9-17"><a href="compositional-analyses-of-biological-sequences.html#cb9-17" tabindex="-1"></a>summary_diNT <span class="op">=</span> join_medians(median_df)</span>
<span id="cb9-18"><a href="compositional-analyses-of-biological-sequences.html#cb9-18" tabindex="-1"></a></span>
<span id="cb9-19"><a href="compositional-analyses-of-biological-sequences.html#cb9-19" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots()</span>
<span id="cb9-20"><a href="compositional-analyses-of-biological-sequences.html#cb9-20" tabindex="-1"></a>plt.rcParams[<span class="st">&quot;figure.figsize&quot;</span>] <span class="op">=</span> [<span class="dv">12</span>,<span class="fl">1.5</span>]</span>
<span id="cb9-21"><a href="compositional-analyses-of-biological-sequences.html#cb9-21" tabindex="-1"></a>sns.heatmap(summary_diNT,ax<span class="op">=</span>ax, cmap<span class="op">=</span><span class="st">&quot;viridis&quot;</span>)</span>
<span id="cb9-22"><a href="compositional-analyses-of-biological-sequences.html#cb9-22" tabindex="-1"></a>plt.xlabel(<span class="st">&quot;Dinucleotide&quot;</span>)</span>
<span id="cb9-23"><a href="compositional-analyses-of-biological-sequences.html#cb9-23" tabindex="-1"></a></span>
<span id="cb9-24"><a href="compositional-analyses-of-biological-sequences.html#cb9-24" tabindex="-1"></a>plt.tight_layout()</span></code></pre></div>
</div>
<div id="codons" class="section level2 hasAnchor" number="5.4">
<h2><span class="header-section-number">5.4</span> Codons<a href="compositional-analyses-of-biological-sequences.html#codons" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>We define the paths of the FASTA files.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb10-1"><a href="compositional-analyses-of-biological-sequences.html#cb10-1" tabindex="-1"></a>concatenate_dir <span class="op">=</span> <span class="st">&quot;/Users/esmeralda/Documents/TFM/RevTrans/final/nt/alignments/concatenate/&quot;</span></span>
<span id="cb10-2"><a href="compositional-analyses-of-biological-sequences.html#cb10-2" tabindex="-1"></a>paths <span class="op">=</span> [ concatenate_dir <span class="op">+</span> <span class="st">&quot;RdRp_nt_aligned.fasta&quot;</span>,</span>
<span id="cb10-3"><a href="compositional-analyses-of-biological-sequences.html#cb10-3" tabindex="-1"></a>         concatenate_dir <span class="op">+</span> <span class="st">&quot;cp_nt_aligned_filtered.fasta&quot;</span>,</span>
<span id="cb10-4"><a href="compositional-analyses-of-biological-sequences.html#cb10-4" tabindex="-1"></a>         concatenate_dir <span class="op">+</span> <span class="st">&quot;p19_nt_alignment.fasta&quot;</span>,</span>
<span id="cb10-5"><a href="compositional-analyses-of-biological-sequences.html#cb10-5" tabindex="-1"></a>         concatenate_dir <span class="op">+</span> <span class="st">&quot;p22_nt_alignment.fasta&quot;</span>,</span>
<span id="cb10-6"><a href="compositional-analyses-of-biological-sequences.html#cb10-6" tabindex="-1"></a>         concatenate_dir <span class="op">+</span> <span class="st">&quot;p22 core_nt_alignment.fa&quot;</span>,</span>
<span id="cb10-7"><a href="compositional-analyses-of-biological-sequences.html#cb10-7" tabindex="-1"></a>         <span class="co"># concatenate_dir + &quot;union/sequence_data.fas&quot;,</span></span>
<span id="cb10-8"><a href="compositional-analyses-of-biological-sequences.html#cb10-8" tabindex="-1"></a>         <span class="st">&quot;/Users/esmeralda/Documents/TFM/overlap_comparisons/other cores_nt.fa&quot;</span>]</span>
<span id="cb10-9"><a href="compositional-analyses-of-biological-sequences.html#cb10-9" tabindex="-1"></a></span>
<span id="cb10-10"><a href="compositional-analyses-of-biological-sequences.html#cb10-10" tabindex="-1"></a>names <span class="op">=</span> [<span class="st">&quot;RdRp&quot;</span>, <span class="st">&quot;CP&quot;</span>, <span class="st">&quot;p19&quot;</span>, <span class="st">&quot;p22&quot;</span>, <span class="st">&quot;p22 core&quot;</span>, <span class="st">&quot;other cores&quot;</span>]</span></code></pre></div>
<p>The subsequent steps involve reading and counting codons. To underscore the impact of codon usage, we calculated the percentage of each codon within its group of codons coding for the same amino acid. For example, if codons TGT and TGC are used 3 and 12 times, respectively, both encoding the amino acid C, the relative abundance is calculated as 0.2 for the former and 0.8 for the latter. Therefore, the comparison is made between different groups but within the same amino acid.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb11-1"><a href="compositional-analyses-of-biological-sequences.html#cb11-1" tabindex="-1"></a><span class="kw">def</span> read_codons(path):</span>
<span id="cb11-2"><a href="compositional-analyses-of-biological-sequences.html#cb11-2" tabindex="-1"></a>    n <span class="op">=</span> <span class="dv">3</span></span>
<span id="cb11-3"><a href="compositional-analyses-of-biological-sequences.html#cb11-3" tabindex="-1"></a>    record_dict <span class="op">=</span> SeqIO.to_dict(SeqIO.parse(path, <span class="st">&quot;fasta&quot;</span>))</span>
<span id="cb11-4"><a href="compositional-analyses-of-biological-sequences.html#cb11-4" tabindex="-1"></a>    codon_dict <span class="op">=</span> {}</span>
<span id="cb11-5"><a href="compositional-analyses-of-biological-sequences.html#cb11-5" tabindex="-1"></a>    <span class="cf">for</span> seq_name <span class="kw">in</span> record_dict.keys():</span>
<span id="cb11-6"><a href="compositional-analyses-of-biological-sequences.html#cb11-6" tabindex="-1"></a>        seq <span class="op">=</span> <span class="bu">str</span>(record_dict[seq_name].seq)</span>
<span id="cb11-7"><a href="compositional-analyses-of-biological-sequences.html#cb11-7" tabindex="-1"></a>        codon_dict[seq_name] <span class="op">=</span> [seq[i:i<span class="op">+</span>n] <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">0</span>, <span class="bu">len</span>(seq), n)]</span>
<span id="cb11-8"><a href="compositional-analyses-of-biological-sequences.html#cb11-8" tabindex="-1"></a>    </span>
<span id="cb11-9"><a href="compositional-analyses-of-biological-sequences.html#cb11-9" tabindex="-1"></a>    <span class="cf">return</span> pd.DataFrame.from_dict(codon_dict)</span>
<span id="cb11-10"><a href="compositional-analyses-of-biological-sequences.html#cb11-10" tabindex="-1"></a></span>
<span id="cb11-11"><a href="compositional-analyses-of-biological-sequences.html#cb11-11" tabindex="-1"></a><span class="kw">def</span> count_codons(df_codons):</span>
<span id="cb11-12"><a href="compositional-analyses-of-biological-sequences.html#cb11-12" tabindex="-1"></a>    counts <span class="op">=</span> []</span>
<span id="cb11-13"><a href="compositional-analyses-of-biological-sequences.html#cb11-13" tabindex="-1"></a></span>
<span id="cb11-14"><a href="compositional-analyses-of-biological-sequences.html#cb11-14" tabindex="-1"></a>    <span class="cf">for</span> column <span class="kw">in</span> df_codons.columns:</span>
<span id="cb11-15"><a href="compositional-analyses-of-biological-sequences.html#cb11-15" tabindex="-1"></a>        counts_num <span class="op">=</span> df_codons[column].value_counts()</span>
<span id="cb11-16"><a href="compositional-analyses-of-biological-sequences.html#cb11-16" tabindex="-1"></a>        counts.append(counts_num)</span>
<span id="cb11-17"><a href="compositional-analyses-of-biological-sequences.html#cb11-17" tabindex="-1"></a></span>
<span id="cb11-18"><a href="compositional-analyses-of-biological-sequences.html#cb11-18" tabindex="-1"></a>    codon_abs <span class="op">=</span> pd.DataFrame.from_dict(counts)</span>
<span id="cb11-19"><a href="compositional-analyses-of-biological-sequences.html#cb11-19" tabindex="-1"></a>    codon_abs.fillna(<span class="dv">0</span>, inplace<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb11-20"><a href="compositional-analyses-of-biological-sequences.html#cb11-20" tabindex="-1"></a></span>
<span id="cb11-21"><a href="compositional-analyses-of-biological-sequences.html#cb11-21" tabindex="-1"></a>    <span class="cf">return</span> codon_abs</span>
<span id="cb11-22"><a href="compositional-analyses-of-biological-sequences.html#cb11-22" tabindex="-1"></a></span>
<span id="cb11-23"><a href="compositional-analyses-of-biological-sequences.html#cb11-23" tabindex="-1"></a><span class="kw">def</span> relative_codons(codon_matrix, table_DNA_to_codon):</span>
<span id="cb11-24"><a href="compositional-analyses-of-biological-sequences.html#cb11-24" tabindex="-1"></a>    codon_rel <span class="op">=</span> pd.DataFrame()</span>
<span id="cb11-25"><a href="compositional-analyses-of-biological-sequences.html#cb11-25" tabindex="-1"></a>    </span>
<span id="cb11-26"><a href="compositional-analyses-of-biological-sequences.html#cb11-26" tabindex="-1"></a>    <span class="cf">for</span> amino <span class="kw">in</span> table_DNA_to_codon.keys():</span>
<span id="cb11-27"><a href="compositional-analyses-of-biological-sequences.html#cb11-27" tabindex="-1"></a>        <span class="cf">if</span> <span class="bu">len</span>(table_DNA_to_codon[amino]) <span class="op">&gt;</span> <span class="dv">1</span> <span class="kw">and</span> amino <span class="op">!=</span> <span class="st">&quot;STOP&quot;</span>:</span>
<span id="cb11-28"><a href="compositional-analyses-of-biological-sequences.html#cb11-28" tabindex="-1"></a>            total <span class="op">=</span> codon_matrix[table_DNA_to_codon[amino]].<span class="bu">sum</span>(axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb11-29"><a href="compositional-analyses-of-biological-sequences.html#cb11-29" tabindex="-1"></a>            res <span class="op">=</span> codon_matrix[table_DNA_to_codon[amino]].div(total, axis<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb11-30"><a href="compositional-analyses-of-biological-sequences.html#cb11-30" tabindex="-1"></a>            codon_rel <span class="op">=</span> pd.concat([codon_rel, res], axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb11-31"><a href="compositional-analyses-of-biological-sequences.html#cb11-31" tabindex="-1"></a>    </span>
<span id="cb11-32"><a href="compositional-analyses-of-biological-sequences.html#cb11-32" tabindex="-1"></a>    <span class="cf">return</span> codon_rel</span>
<span id="cb11-33"><a href="compositional-analyses-of-biological-sequences.html#cb11-33" tabindex="-1"></a></span>
<span id="cb11-34"><a href="compositional-analyses-of-biological-sequences.html#cb11-34" tabindex="-1"></a><span class="kw">def</span> join_codons(list_dfs, names):</span>
<span id="cb11-35"><a href="compositional-analyses-of-biological-sequences.html#cb11-35" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(names)):</span>
<span id="cb11-36"><a href="compositional-analyses-of-biological-sequences.html#cb11-36" tabindex="-1"></a>        list_dfs[i][<span class="st">&quot;Sequence&quot;</span>] <span class="op">=</span> names[i]</span>
<span id="cb11-37"><a href="compositional-analyses-of-biological-sequences.html#cb11-37" tabindex="-1"></a></span>
<span id="cb11-38"><a href="compositional-analyses-of-biological-sequences.html#cb11-38" tabindex="-1"></a>    <span class="cf">return</span> pd.concat([list_dfs[<span class="dv">0</span>],</span>
<span id="cb11-39"><a href="compositional-analyses-of-biological-sequences.html#cb11-39" tabindex="-1"></a>                     list_dfs[<span class="dv">1</span>], </span>
<span id="cb11-40"><a href="compositional-analyses-of-biological-sequences.html#cb11-40" tabindex="-1"></a>                     list_dfs[<span class="dv">2</span>], </span>
<span id="cb11-41"><a href="compositional-analyses-of-biological-sequences.html#cb11-41" tabindex="-1"></a>                     list_dfs[<span class="dv">3</span>], </span>
<span id="cb11-42"><a href="compositional-analyses-of-biological-sequences.html#cb11-42" tabindex="-1"></a>                     list_dfs[<span class="dv">4</span>],</span>
<span id="cb11-43"><a href="compositional-analyses-of-biological-sequences.html#cb11-43" tabindex="-1"></a>                     list_dfs[<span class="dv">5</span>]], axis<span class="op">=</span><span class="dv">0</span>).T.fillna(<span class="dv">0</span>, inplace<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb11-44"><a href="compositional-analyses-of-biological-sequences.html#cb11-44" tabindex="-1"></a></span>
<span id="cb11-45"><a href="compositional-analyses-of-biological-sequences.html#cb11-45" tabindex="-1"></a><span class="kw">def</span> join_codons_medians(list_dfs):</span>
<span id="cb11-46"><a href="compositional-analyses-of-biological-sequences.html#cb11-46" tabindex="-1"></a>    <span class="cf">return</span> pd.concat([list_dfs[<span class="dv">0</span>].median(axis<span class="op">=</span><span class="dv">0</span>), </span>
<span id="cb11-47"><a href="compositional-analyses-of-biological-sequences.html#cb11-47" tabindex="-1"></a>                          list_dfs[<span class="dv">1</span>].median(axis<span class="op">=</span><span class="dv">0</span>), </span>
<span id="cb11-48"><a href="compositional-analyses-of-biological-sequences.html#cb11-48" tabindex="-1"></a>                          list_dfs[<span class="dv">2</span>].median(axis<span class="op">=</span><span class="dv">0</span>), </span>
<span id="cb11-49"><a href="compositional-analyses-of-biological-sequences.html#cb11-49" tabindex="-1"></a>                          list_dfs[<span class="dv">3</span>].median(axis<span class="op">=</span><span class="dv">0</span>), </span>
<span id="cb11-50"><a href="compositional-analyses-of-biological-sequences.html#cb11-50" tabindex="-1"></a>                          list_dfs[<span class="dv">4</span>].median(axis<span class="op">=</span><span class="dv">0</span>),</span>
<span id="cb11-51"><a href="compositional-analyses-of-biological-sequences.html#cb11-51" tabindex="-1"></a>                          list_dfs[<span class="dv">5</span>].median(axis<span class="op">=</span><span class="dv">0</span>)], axis <span class="op">=</span> <span class="dv">1</span>).T</span></code></pre></div>
<p>Now we run the functions.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb12-1"><a href="compositional-analyses-of-biological-sequences.html#cb12-1" tabindex="-1"></a>dfs_codons <span class="op">=</span> []</span>
<span id="cb12-2"><a href="compositional-analyses-of-biological-sequences.html#cb12-2" tabindex="-1"></a></span>
<span id="cb12-3"><a href="compositional-analyses-of-biological-sequences.html#cb12-3" tabindex="-1"></a><span class="cf">for</span> path <span class="kw">in</span> paths:</span>
<span id="cb12-4"><a href="compositional-analyses-of-biological-sequences.html#cb12-4" tabindex="-1"></a>    codons <span class="op">=</span> read_codons(path)</span>
<span id="cb12-5"><a href="compositional-analyses-of-biological-sequences.html#cb12-5" tabindex="-1"></a>    codons_abs <span class="op">=</span> count_codons(codons)</span>
<span id="cb12-6"><a href="compositional-analyses-of-biological-sequences.html#cb12-6" tabindex="-1"></a>    <span class="cf">if</span> path.find(<span class="st">&quot;p22 core&quot;</span>) <span class="op">&gt;</span> <span class="dv">0</span>:</span>
<span id="cb12-7"><a href="compositional-analyses-of-biological-sequences.html#cb12-7" tabindex="-1"></a>        codons_abs[<span class="st">&quot;CGT&quot;</span>] <span class="op">=</span> <span class="fl">0.0</span></span>
<span id="cb12-8"><a href="compositional-analyses-of-biological-sequences.html#cb12-8" tabindex="-1"></a>    dfs_codons.append(relative_codons(codons_abs, table_DNA_to_codon))</span>
<span id="cb12-9"><a href="compositional-analyses-of-biological-sequences.html#cb12-9" tabindex="-1"></a></span>
<span id="cb12-10"><a href="compositional-analyses-of-biological-sequences.html#cb12-10" tabindex="-1"></a>df_codon_all <span class="op">=</span> join_codons(dfs_codons, names<span class="op">=</span>names)</span>
<span id="cb12-11"><a href="compositional-analyses-of-biological-sequences.html#cb12-11" tabindex="-1"></a>df_codons_medians <span class="op">=</span> join_codons_medians(dfs_codons)</span></code></pre></div>
<p>Finally, we generate a heatmap using a seaborn function.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb13-1"><a href="compositional-analyses-of-biological-sequences.html#cb13-1" tabindex="-1"></a>plt.rcParams[<span class="st">&quot;figure.figsize&quot;</span>] <span class="op">=</span> [<span class="dv">16</span>, <span class="dv">2</span>]</span>
<span id="cb13-2"><a href="compositional-analyses-of-biological-sequences.html#cb13-2" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots(nrows<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb13-3"><a href="compositional-analyses-of-biological-sequences.html#cb13-3" tabindex="-1"></a></span>
<span id="cb13-4"><a href="compositional-analyses-of-biological-sequences.html#cb13-4" tabindex="-1"></a>new_columns <span class="op">=</span> []</span>
<span id="cb13-5"><a href="compositional-analyses-of-biological-sequences.html#cb13-5" tabindex="-1"></a><span class="cf">for</span> codon_name <span class="kw">in</span> df_codons_medians.columns:</span>
<span id="cb13-6"><a href="compositional-analyses-of-biological-sequences.html#cb13-6" tabindex="-1"></a>    new <span class="op">=</span> codon_name.replace(<span class="st">&quot;T&quot;</span>, <span class="st">&quot;U&quot;</span>)</span>
<span id="cb13-7"><a href="compositional-analyses-of-biological-sequences.html#cb13-7" tabindex="-1"></a>    new_columns.append(new)</span>
<span id="cb13-8"><a href="compositional-analyses-of-biological-sequences.html#cb13-8" tabindex="-1"></a></span>
<span id="cb13-9"><a href="compositional-analyses-of-biological-sequences.html#cb13-9" tabindex="-1"></a>df_codons_medians.columns <span class="op">=</span> new_columns</span>
<span id="cb13-10"><a href="compositional-analyses-of-biological-sequences.html#cb13-10" tabindex="-1"></a>df_codons_medians.index <span class="op">=</span> [<span class="st">&quot;RdRp&quot;</span>, <span class="st">&quot;CP&quot;</span>, <span class="st">&quot;p19&quot;</span>, <span class="st">&quot;p22&quot;</span>, <span class="st">&quot;p22 core&quot;</span>, <span class="st">&quot;other cores&quot;</span>]</span>
<span id="cb13-11"><a href="compositional-analyses-of-biological-sequences.html#cb13-11" tabindex="-1"></a><span class="bu">print</span>(<span class="bu">len</span>(new_columns))</span>
<span id="cb13-12"><a href="compositional-analyses-of-biological-sequences.html#cb13-12" tabindex="-1"></a></span>
<span id="cb13-13"><a href="compositional-analyses-of-biological-sequences.html#cb13-13" tabindex="-1"></a>sns.heatmap(df_codons_medians, cmap<span class="op">=</span><span class="st">&quot;viridis&quot;</span>)</span>
<span id="cb13-14"><a href="compositional-analyses-of-biological-sequences.html#cb13-14" tabindex="-1"></a>plt.xlabel(<span class="st">&quot;Codon&quot;</span>)</span>
<span id="cb13-15"><a href="compositional-analyses-of-biological-sequences.html#cb13-15" tabindex="-1"></a>plt.xticks(rotation<span class="op">=</span><span class="dv">45</span>)</span>
<span id="cb13-16"><a href="compositional-analyses-of-biological-sequences.html#cb13-16" tabindex="-1"></a>plt.yticks(rotation<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb13-17"><a href="compositional-analyses-of-biological-sequences.html#cb13-17" tabindex="-1"></a>ax.yaxis.set_ticks_position(<span class="st">&#39;none&#39;</span>) </span>
<span id="cb13-18"><a href="compositional-analyses-of-biological-sequences.html#cb13-18" tabindex="-1"></a></span>
<span id="cb13-19"><a href="compositional-analyses-of-biological-sequences.html#cb13-19" tabindex="-1"></a>plt.tight_layout()</span></code></pre></div>
<p>The following code is useful for assessing the significance between two groups. We employ the Mann-Whitney test to determine if there is statistical significance in the difference between p22 cores and non-overlapping cores.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb14-1"><a href="compositional-analyses-of-biological-sequences.html#cb14-1" tabindex="-1"></a>df_codon_all.fillna(<span class="dv">0</span>, inplace<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb14-2"><a href="compositional-analyses-of-biological-sequences.html#cb14-2" tabindex="-1"></a>codon <span class="op">=</span> <span class="st">&quot;CAT&quot;</span></span>
<span id="cb14-3"><a href="compositional-analyses-of-biological-sequences.html#cb14-3" tabindex="-1"></a>df_plot <span class="op">=</span> df_codon_all.T[[codon, <span class="st">&quot;Sequence&quot;</span>]]</span>
<span id="cb14-4"><a href="compositional-analyses-of-biological-sequences.html#cb14-4" tabindex="-1"></a>p22_core <span class="op">=</span> <span class="bu">list</span>(df_plot[df_plot[<span class="st">&quot;Sequence&quot;</span>] <span class="op">==</span> <span class="st">&quot;p22 core&quot;</span>][codon].values)</span>
<span id="cb14-5"><a href="compositional-analyses-of-biological-sequences.html#cb14-5" tabindex="-1"></a>other_cores <span class="op">=</span> <span class="bu">list</span>(df_plot[df_plot[<span class="st">&quot;Sequence&quot;</span>] <span class="op">==</span> <span class="st">&quot;other cores&quot;</span>][codon].values)</span>
<span id="cb14-6"><a href="compositional-analyses-of-biological-sequences.html#cb14-6" tabindex="-1"></a></span>
<span id="cb14-7"><a href="compositional-analyses-of-biological-sequences.html#cb14-7" tabindex="-1"></a>pvalues <span class="op">=</span> [</span>
<span id="cb14-8"><a href="compositional-analyses-of-biological-sequences.html#cb14-8" tabindex="-1"></a>    scipy.stats.mannwhitneyu(p22_core, other_cores, alternative<span class="op">=</span><span class="st">&quot;two-sided&quot;</span>).pvalue</span>
<span id="cb14-9"><a href="compositional-analyses-of-biological-sequences.html#cb14-9" tabindex="-1"></a>]</span>
<span id="cb14-10"><a href="compositional-analyses-of-biological-sequences.html#cb14-10" tabindex="-1"></a>formatted_pvalues <span class="op">=</span> [<span class="ss">f&#39;p=</span><span class="sc">{</span>pvalue<span class="sc">:.2e}</span><span class="ss">&#39;</span> <span class="cf">for</span> pvalue <span class="kw">in</span> pvalues]</span></code></pre></div>
<p>Subsequently, we plot the data and indicate the significance between the two groups.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb15-1"><a href="compositional-analyses-of-biological-sequences.html#cb15-1" tabindex="-1"></a>plt.rcParams[<span class="st">&quot;figure.figsize&quot;</span>] <span class="op">=</span> [<span class="dv">16</span>,<span class="dv">9</span>]</span>
<span id="cb15-2"><a href="compositional-analyses-of-biological-sequences.html#cb15-2" tabindex="-1"></a>sns.set_style(<span class="st">&quot;white&quot;</span>)</span>
<span id="cb15-3"><a href="compositional-analyses-of-biological-sequences.html#cb15-3" tabindex="-1"></a>colors <span class="op">=</span> sns.color_palette(<span class="st">&quot;colorblind&quot;</span>, <span class="dv">5</span>)</span>
<span id="cb15-4"><a href="compositional-analyses-of-biological-sequences.html#cb15-4" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots()</span>
<span id="cb15-5"><a href="compositional-analyses-of-biological-sequences.html#cb15-5" tabindex="-1"></a></span>
<span id="cb15-6"><a href="compositional-analyses-of-biological-sequences.html#cb15-6" tabindex="-1"></a>pairs <span class="op">=</span> [(<span class="st">&#39;p22 core&#39;</span>, <span class="st">&#39;other cores&#39;</span>)]</span>
<span id="cb15-7"><a href="compositional-analyses-of-biological-sequences.html#cb15-7" tabindex="-1"></a></span>
<span id="cb15-8"><a href="compositional-analyses-of-biological-sequences.html#cb15-8" tabindex="-1"></a><span class="cf">with</span> sns.plotting_context(<span class="st">&#39;notebook&#39;</span>, font_scale <span class="op">=</span> <span class="fl">1.2</span>):</span>
<span id="cb15-9"><a href="compositional-analyses-of-biological-sequences.html#cb15-9" tabindex="-1"></a>        sns.swarmplot(data<span class="op">=</span>df_plot, x<span class="op">=</span><span class="st">&quot;Sequence&quot;</span>, y<span class="op">=</span>codon, ax<span class="op">=</span>ax, palette<span class="op">=</span>colors)</span>
<span id="cb15-10"><a href="compositional-analyses-of-biological-sequences.html#cb15-10" tabindex="-1"></a>        sns.boxplot(data<span class="op">=</span>df_plot, x<span class="op">=</span><span class="st">&quot;Sequence&quot;</span>, y<span class="op">=</span>codon, zorder<span class="op">=</span><span class="dv">0</span>, dodge<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>ax, boxprops<span class="op">=</span>{<span class="st">&#39;facecolor&#39;</span>:<span class="st">&#39;None&#39;</span>})</span>
<span id="cb15-11"><a href="compositional-analyses-of-biological-sequences.html#cb15-11" tabindex="-1"></a></span>
<span id="cb15-12"><a href="compositional-analyses-of-biological-sequences.html#cb15-12" tabindex="-1"></a>        <span class="co"># Add annotations</span></span>
<span id="cb15-13"><a href="compositional-analyses-of-biological-sequences.html#cb15-13" tabindex="-1"></a>        annotator <span class="op">=</span> Annotator(ax<span class="op">=</span>ax, pairs <span class="op">=</span> pairs, data<span class="op">=</span>df_plot, x<span class="op">=</span><span class="st">&quot;Sequence&quot;</span>, y<span class="op">=</span>codon)</span>
<span id="cb15-14"><a href="compositional-analyses-of-biological-sequences.html#cb15-14" tabindex="-1"></a>        annotator.set_custom_annotations(formatted_pvalues)</span>
<span id="cb15-15"><a href="compositional-analyses-of-biological-sequences.html#cb15-15" tabindex="-1"></a>        annotator.configure(test_short_name<span class="op">=</span><span class="st">&quot;MWW&quot;</span>)  <span class="co"># text_format is still simple</span></span>
<span id="cb15-16"><a href="compositional-analyses-of-biological-sequences.html#cb15-16" tabindex="-1"></a>        annotator.set_pvalues_and_annotate(pvalues)</span></code></pre></div>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="distances-with-chimerax.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="comparative-analyses-of-selection.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/llobarts/nestedgenes/edit/main/04-composition.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
