<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 2 Blast and E-utilities | Nested genes code</title>
  <meta name="description" content="This is the description of the code employed in the research." />
  <meta name="generator" content="bookdown 0.34 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 2 Blast and E-utilities | Nested genes code" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="This is the description of the code employed in the research." />
  <meta name="github-repo" content="https://github.com/llobarts/nestedgenes" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 2 Blast and E-utilities | Nested genes code" />
  
  <meta name="twitter:description" content="This is the description of the code employed in the research." />
  

<meta name="author" content="Esmeralda G. Legarda" />


<meta name="date" content="2023-12-30" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="index.html"/>
<link rel="next" href="distances-with-clustering-in-clans.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Code for 'Nested genes'</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> About</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#abstract"><i class="fa fa-check"></i><b>1.1</b> Abstract</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#additional-files"><i class="fa fa-check"></i><b>1.2</b> Additional files</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="blast-and-e-utilities.html"><a href="blast-and-e-utilities.html"><i class="fa fa-check"></i><b>2</b> Blast and E-utilities</a>
<ul>
<li class="chapter" data-level="2.1" data-path="blast-and-e-utilities.html"><a href="blast-and-e-utilities.html#psi-blast"><i class="fa fa-check"></i><b>2.1</b> PSI-BLAST</a></li>
<li class="chapter" data-level="2.2" data-path="blast-and-e-utilities.html"><a href="blast-and-e-utilities.html#protein-id-to-nucleotides"><i class="fa fa-check"></i><b>2.2</b> Protein ID to nucleotides</a></li>
<li class="chapter" data-level="2.3" data-path="blast-and-e-utilities.html"><a href="blast-and-e-utilities.html#species-to-family"><i class="fa fa-check"></i><b>2.3</b> Species to family</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html"><i class="fa fa-check"></i><b>3</b> Distances with clustering in CLANS</a>
<ul>
<li class="chapter" data-level="3.1" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html#read-the-clans-file"><i class="fa fa-check"></i><b>3.1</b> Read the CLANS file</a></li>
<li class="chapter" data-level="3.2" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html#create-a-dataframe"><i class="fa fa-check"></i><b>3.2</b> Create a DataFrame</a></li>
<li class="chapter" data-level="3.3" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html#calculate-centroids-and-distances"><i class="fa fa-check"></i><b>3.3</b> Calculate centroids and distances</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html#plot-the-clustermap"><i class="fa fa-check"></i><b>3.3.1</b> Plot the clustermap</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html#run"><i class="fa fa-check"></i><b>3.4</b> Run</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html"><i class="fa fa-check"></i><b>4</b> Distances with ChimeraX</a>
<ul>
<li class="chapter" data-level="4.1" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#obtaining-rmsd-with-chimerax"><i class="fa fa-check"></i><b>4.1</b> Obtaining RMSD with ChimeraX</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#import-libraries-and-data"><i class="fa fa-check"></i><b>4.1.1</b> Import libraries and data</a></li>
<li class="chapter" data-level="4.1.2" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#matchmaker"><i class="fa fa-check"></i><b>4.1.2</b> MatchMaker</a></li>
<li class="chapter" data-level="4.1.3" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#save-csv-file"><i class="fa fa-check"></i><b>4.1.3</b> Save csv file</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#hiarachical-clustering"><i class="fa fa-check"></i><b>4.2</b> Hiarachical clustering</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#libraries-and-read-csv"><i class="fa fa-check"></i><b>4.2.1</b> Libraries and read csv</a></li>
<li class="chapter" data-level="4.2.2" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#create-matrix"><i class="fa fa-check"></i><b>4.2.2</b> Create matrix</a></li>
<li class="chapter" data-level="4.2.3" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#plot-the-clustermap-1"><i class="fa fa-check"></i><b>4.2.3</b> Plot the clustermap</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="compositional-analyses-of-biological-sequences.html"><a href="compositional-analyses-of-biological-sequences.html"><i class="fa fa-check"></i><b>5</b> Compositional analyses of biological sequences</a>
<ul>
<li class="chapter" data-level="5.1" data-path="compositional-analyses-of-biological-sequences.html"><a href="compositional-analyses-of-biological-sequences.html#amino-acids"><i class="fa fa-check"></i><b>5.1</b> Amino acids</a></li>
<li class="chapter" data-level="5.2" data-path="compositional-analyses-of-biological-sequences.html"><a href="compositional-analyses-of-biological-sequences.html#nucleotides"><i class="fa fa-check"></i><b>5.2</b> Nucleotides</a></li>
<li class="chapter" data-level="5.3" data-path="compositional-analyses-of-biological-sequences.html"><a href="compositional-analyses-of-biological-sequences.html#dinuclotides"><i class="fa fa-check"></i><b>5.3</b> Dinuclotides</a></li>
<li class="chapter" data-level="5.4" data-path="compositional-analyses-of-biological-sequences.html"><a href="compositional-analyses-of-biological-sequences.html#codons"><i class="fa fa-check"></i><b>5.4</b> Codons</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html"><i class="fa fa-check"></i><b>6</b> Comparative analyses of selection</a>
<ul>
<li class="chapter" data-level="6.1" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#evolution-speed-with-absrel"><i class="fa fa-check"></i><b>6.1</b> Evolution speed with aBSREL</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#libraries-and-json-files"><i class="fa fa-check"></i><b>6.1.1</b> Libraries and json files</a></li>
<li class="chapter" data-level="6.1.2" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#omega-and-significance"><i class="fa fa-check"></i><b>6.1.2</b> Omega and significance</a></li>
<li class="chapter" data-level="6.1.3" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#relative-asinh-transformation"><i class="fa fa-check"></i><b>6.1.3</b> Relative asinh transformation</a></li>
<li class="chapter" data-level="6.1.4" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#utilities"><i class="fa fa-check"></i><b>6.1.4</b> Utilities</a></li>
<li class="chapter" data-level="6.1.5" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#boxplots-with-significance"><i class="fa fa-check"></i><b>6.1.5</b> Boxplots with significance</a></li>
<li class="chapter" data-level="6.1.6" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#test-selection"><i class="fa fa-check"></i><b>6.1.6</b> Test selection</a></li>
<li class="chapter" data-level="6.1.7" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#run-1"><i class="fa fa-check"></i><b>6.1.7</b> Run</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#fel-for-correlations"><i class="fa fa-check"></i><b>6.2</b> FEL for correlations</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#significance-and-difference"><i class="fa fa-check"></i><b>6.2.1</b> Significance and difference</a></li>
<li class="chapter" data-level="6.2.2" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#plot-correlation"><i class="fa fa-check"></i><b>6.2.2</b> Plot correlation</a></li>
<li class="chapter" data-level="6.2.3" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#test-correlation-of-selection"><i class="fa fa-check"></i><b>6.2.3</b> Test correlation of selection</a></li>
<li class="chapter" data-level="6.2.4" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#correlation-2d"><i class="fa fa-check"></i><b>6.2.4</b> Correlation 2D</a></li>
<li class="chapter" data-level="6.2.5" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#selection-in-2d"><i class="fa fa-check"></i><b>6.2.5</b> Selection in 2D</a></li>
<li class="chapter" data-level="6.2.6" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#run-2"><i class="fa fa-check"></i><b>6.2.6</b> Run</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="miscellania.html"><a href="miscellania.html"><i class="fa fa-check"></i><b>7</b> Miscellania</a>
<ul>
<li class="chapter" data-level="7.1" data-path="miscellania.html"><a href="miscellania.html#representation-of-predicted-disordered-amino-acid-data"><i class="fa fa-check"></i><b>7.1</b> Representation of predicted disordered amino acid data</a></li>
<li class="chapter" data-level="7.2" data-path="miscellania.html"><a href="miscellania.html#shannons-entropy"><i class="fa fa-check"></i><b>7.2</b> Shannon’s entropy</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/llobarts/nestedgenes" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Nested genes code</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="blast-and-e-utilities" class="section level1 hasAnchor" number="2">
<h1><span class="header-section-number">Chapter 2</span> Blast and E-utilities<a href="blast-and-e-utilities.html#blast-and-e-utilities" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<div id="psi-blast" class="section level2 hasAnchor" number="2.1">
<h2><span class="header-section-number">2.1</span> PSI-BLAST<a href="blast-and-e-utilities.html#psi-blast" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>This section shows a Bash file that enables the quick execution of PSI-BLAST with a simple command for various inputs, all saved in the same directory.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="blast-and-e-utilities.html#cb1-1" tabindex="-1"></a><span class="co">#!/bin/bash</span></span>
<span id="cb1-2"><a href="blast-and-e-utilities.html#cb1-2" tabindex="-1"></a><span class="co"># This program generates results from running PSI-BLAST software</span></span>
<span id="cb1-3"><a href="blast-and-e-utilities.html#cb1-3" tabindex="-1"></a><span class="co"># Reading FLAGS</span></span>
<span id="cb1-4"><a href="blast-and-e-utilities.html#cb1-4" tabindex="-1"></a><span class="cf">while</span> <span class="bu">getopts</span> <span class="op">:</span>p<span class="op">:</span>e<span class="op">:</span>i<span class="op">:</span>d<span class="op">:</span>D<span class="op">:</span>t<span class="op">:</span> <span class="va">flag</span></span>
<span id="cb1-5"><a href="blast-and-e-utilities.html#cb1-5" tabindex="-1"></a><span class="cf">do</span></span>
<span id="cb1-6"><a href="blast-and-e-utilities.html#cb1-6" tabindex="-1"></a>    <span class="cf">case</span> <span class="va">${flag}</span> <span class="kw">in</span></span>
<span id="cb1-7"><a href="blast-and-e-utilities.html#cb1-7" tabindex="-1"></a>        <span class="ss">p</span><span class="kw">)</span> <span class="va">psiblast</span><span class="op">=</span><span class="va">${OPTARG}</span><span class="cf">;;</span></span>
<span id="cb1-8"><a href="blast-and-e-utilities.html#cb1-8" tabindex="-1"></a>        <span class="ss">e</span><span class="kw">)</span> <span class="va">evalue</span><span class="op">=</span><span class="va">${OPTARG}</span><span class="cf">;;</span></span>
<span id="cb1-9"><a href="blast-and-e-utilities.html#cb1-9" tabindex="-1"></a>        <span class="ss">i</span><span class="kw">)</span> <span class="va">iter</span><span class="op">=</span><span class="va">${OPTARG}</span><span class="cf">;;</span></span>
<span id="cb1-10"><a href="blast-and-e-utilities.html#cb1-10" tabindex="-1"></a>        <span class="ss">d</span><span class="kw">)</span> <span class="va">dbname</span><span class="op">=</span><span class="va">${OPTARG}</span><span class="cf">;;</span></span>
<span id="cb1-11"><a href="blast-and-e-utilities.html#cb1-11" tabindex="-1"></a>        <span class="ss">D</span><span class="kw">)</span> <span class="va">dbpath</span><span class="op">=</span><span class="va">${OPTARG}</span><span class="cf">;;</span></span>
<span id="cb1-12"><a href="blast-and-e-utilities.html#cb1-12" tabindex="-1"></a>        <span class="ss">t</span><span class="kw">)</span> <span class="va">threads</span><span class="op">=</span><span class="va">${OPTARG}</span><span class="cf">;;</span></span>
<span id="cb1-13"><a href="blast-and-e-utilities.html#cb1-13" tabindex="-1"></a>    <span class="cf">esac</span></span>
<span id="cb1-14"><a href="blast-and-e-utilities.html#cb1-14" tabindex="-1"></a><span class="cf">done</span></span>
<span id="cb1-15"><a href="blast-and-e-utilities.html#cb1-15" tabindex="-1"></a></span>
<span id="cb1-16"><a href="blast-and-e-utilities.html#cb1-16" tabindex="-1"></a><span class="co"># Initial settings</span></span>
<span id="cb1-17"><a href="blast-and-e-utilities.html#cb1-17" tabindex="-1"></a><span class="va">format</span><span class="op">=</span><span class="st">&#39;7 qaccver saccver pident length mismatch gapopen qstart qend sstart send stitle ssciname sskingdom&#39;</span></span>
<span id="cb1-18"><a href="blast-and-e-utilities.html#cb1-18" tabindex="-1"></a><span class="va">currpath</span><span class="op">=</span><span class="va">$(</span><span class="bu">pwd</span><span class="va">)</span></span>
<span id="cb1-19"><a href="blast-and-e-utilities.html#cb1-19" tabindex="-1"></a><span class="va">evalue_float</span><span class="op">=</span><span class="va">$(</span><span class="fu">expr</span> <span class="va">$evalue)</span></span>
<span id="cb1-20"><a href="blast-and-e-utilities.html#cb1-20" tabindex="-1"></a><span class="va">num_itera</span><span class="op">=</span><span class="va">$(</span><span class="fu">expr</span> <span class="va">$iter)</span></span>
<span id="cb1-21"><a href="blast-and-e-utilities.html#cb1-21" tabindex="-1"></a></span>
<span id="cb1-22"><a href="blast-and-e-utilities.html#cb1-22" tabindex="-1"></a><span class="co"># Create directory</span></span>
<span id="cb1-23"><a href="blast-and-e-utilities.html#cb1-23" tabindex="-1"></a><span class="va">OUTPUT_DIR</span><span class="op">=</span><span class="va">${currpath}</span>/Results/temp/psiblast_<span class="va">${evalue}</span>_<span class="va">${iter}</span></span>
<span id="cb1-24"><a href="blast-and-e-utilities.html#cb1-24" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$dbpath</span></span>
<span id="cb1-25"><a href="blast-and-e-utilities.html#cb1-25" tabindex="-1"></a></span>
<span id="cb1-26"><a href="blast-and-e-utilities.html#cb1-26" tabindex="-1"></a><span class="cf">for</span> INPUT <span class="kw">in</span> <span class="va">$(</span><span class="fu">ls</span> <span class="va">${currpath}</span>/Results/temp/Fastafiles<span class="va">)</span></span>
<span id="cb1-27"><a href="blast-and-e-utilities.html#cb1-27" tabindex="-1"></a><span class="cf">do</span></span>
<span id="cb1-28"><a href="blast-and-e-utilities.html#cb1-28" tabindex="-1"></a><span class="co"># Run PSI-BLAST</span></span>
<span id="cb1-29"><a href="blast-and-e-utilities.html#cb1-29" tabindex="-1"></a><span class="va">OUTPUT</span><span class="op">=</span><span class="va">${OUTPUT_DIR}</span>/<span class="va">${INPUT</span><span class="op">%</span>.<span class="pp">*</span><span class="va">}</span>_<span class="va">${evalue}</span>_<span class="va">${iter}</span></span>
<span id="cb1-30"><a href="blast-and-e-utilities.html#cb1-30" tabindex="-1"></a><span class="va">code</span><span class="op">=</span><span class="va">$($psiblast</span> <span class="at">-query</span> <span class="va">${currpath}</span>/Results/temp/Fastafiles/<span class="va">$INPUT</span> <span class="at">-db</span> <span class="va">$dbname</span> <span class="at">-evalue</span> <span class="va">$evalue_float</span> <span class="at">-num_iterations</span> <span class="va">$num_itera</span> <span class="at">-outfmt</span> <span class="st">&quot;</span><span class="va">$format</span><span class="st">&quot;</span> <span class="at">-out</span> <span class="va">$OUTPUT</span>.blast <span class="at">-out_pssm</span> <span class="va">$OUTPUT</span>.matrix.smp <span class="at">-seg</span> yes <span class="at">-save_pssm_after_last_round</span> <span class="at">-num_threads</span> <span class="va">$threads)</span></span>
<span id="cb1-31"><a href="blast-and-e-utilities.html#cb1-31" tabindex="-1"></a><span class="bu">eval</span> <span class="va">$code</span></span>
<span id="cb1-32"><a href="blast-and-e-utilities.html#cb1-32" tabindex="-1"></a><span class="cf">done</span></span></code></pre></div>
<p>Run the following code in the terminal:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="blast-and-e-utilities.html#cb2-1" tabindex="-1"></a><span class="ex">/bin/sh</span> Models/dopsiblast.sh <span class="at">-p</span> path_psiblast <span class="at">-e</span> e_value <span class="at">-i</span> max_iter <span class="at">-d</span> db_name <span class="at">-D</span> db_path <span class="at">-t</span> num_threads</span></code></pre></div>
</div>
<div id="protein-id-to-nucleotides" class="section level2 hasAnchor" number="2.2">
<h2><span class="header-section-number">2.2</span> Protein ID to nucleotides<a href="blast-and-e-utilities.html#protein-id-to-nucleotides" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Here is a command that proves highly useful when changing to nucleotide analysis. The output comprises three columns: the accession number of the protein, the accession number of the nucleotide, and the nucleotide sequence.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="blast-and-e-utilities.html#cb3-1" tabindex="-1"></a><span class="ex">elink</span> <span class="at">-db</span> protein <span class="at">-id</span> CAG38136.1 <span class="at">-target</span> nuccore <span class="kw">|</span> <span class="ex">efetch</span> <span class="at">-format</span> gbc <span class="kw">|</span> <span class="ex">xtract</span> <span class="at">-insd</span> protein_id CDS sub_sequence</span></code></pre></div>
<p>It is also possible to obtain information in a GFF-like format for a nucleotide accession.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="blast-and-e-utilities.html#cb4-1" tabindex="-1"></a><span class="ex">esearch</span> <span class="at">-db</span> nuccore <span class="at">-query</span> OP477335.1 <span class="kw">|</span> <span class="ex">efetch</span> <span class="at">-format</span> ft</span></code></pre></div>
<p>Or in a genbank-like format.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb5-1"><a href="blast-and-e-utilities.html#cb5-1" tabindex="-1"></a><span class="ex">esearch</span> <span class="at">-db</span> nuccore <span class="at">-query</span> OP477335.1 <span class="kw">|</span> <span class="ex">efetch</span> <span class="at">-format</span> gb</span></code></pre></div>
</div>
<div id="species-to-family" class="section level2 hasAnchor" number="2.3">
<h2><span class="header-section-number">2.3</span> Species to family<a href="blast-and-e-utilities.html#species-to-family" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The following code allows us to generate a list of unique families (or other ranks) by providing a file containing species as input. The file should have one species per line.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb6-1"><a href="blast-and-e-utilities.html#cb6-1" tabindex="-1"></a><span class="im">import</span> subprocess</span>
<span id="cb6-2"><a href="blast-and-e-utilities.html#cb6-2" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb6-3"><a href="blast-and-e-utilities.html#cb6-3" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb6-4"><a href="blast-and-e-utilities.html#cb6-4" tabindex="-1"></a></span>
<span id="cb6-5"><a href="blast-and-e-utilities.html#cb6-5" tabindex="-1"></a><span class="co"># Read a list of species from a file</span></span>
<span id="cb6-6"><a href="blast-and-e-utilities.html#cb6-6" tabindex="-1"></a><span class="cf">with</span> <span class="bu">open</span>(<span class="st">&#39;virus_species.txt&#39;</span>, <span class="st">&#39;r&#39;</span>) <span class="im">as</span> f:</span>
<span id="cb6-7"><a href="blast-and-e-utilities.html#cb6-7" tabindex="-1"></a>    virus_list <span class="op">=</span> [line.strip() <span class="cf">for</span> line <span class="kw">in</span> f]</span>
<span id="cb6-8"><a href="blast-and-e-utilities.html#cb6-8" tabindex="-1"></a></span>
<span id="cb6-9"><a href="blast-and-e-utilities.html#cb6-9" tabindex="-1"></a><span class="co"># Initialize an empty dictionary to store frequencies of different families.</span></span>
<span id="cb6-10"><a href="blast-and-e-utilities.html#cb6-10" tabindex="-1"></a>family_counts <span class="op">=</span> {}</span>
<span id="cb6-11"><a href="blast-and-e-utilities.html#cb6-11" tabindex="-1"></a></span>
<span id="cb6-12"><a href="blast-and-e-utilities.html#cb6-12" tabindex="-1"></a><span class="co"># Create a loop to iterate through the virus list and retrieve lineage information</span></span>
<span id="cb6-13"><a href="blast-and-e-utilities.html#cb6-13" tabindex="-1"></a><span class="cf">for</span> virus <span class="kw">in</span> virus_list:</span>
<span id="cb6-14"><a href="blast-and-e-utilities.html#cb6-14" tabindex="-1"></a>    command <span class="op">=</span> <span class="ss">f&#39;esearch -db taxonomy -query &quot;</span><span class="sc">{</span>virus<span class="sc">}</span><span class="ss">[Organism]&quot; | efetch -format docsum | xtract -pattern DocumentSummary -element Lineage&#39;</span></span>
<span id="cb6-15"><a href="blast-and-e-utilities.html#cb6-15" tabindex="-1"></a>    output <span class="op">=</span> subprocess.check_output(command, shell<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb6-16"><a href="blast-and-e-utilities.html#cb6-16" tabindex="-1"></a>    lineage <span class="op">=</span> output.decode().strip()</span>
<span id="cb6-17"><a href="blast-and-e-utilities.html#cb6-17" tabindex="-1"></a>    </span>
<span id="cb6-18"><a href="blast-and-e-utilities.html#cb6-18" tabindex="-1"></a>    <span class="co"># Extract the family name from the lineage and increment the frequency count.</span></span>
<span id="cb6-19"><a href="blast-and-e-utilities.html#cb6-19" tabindex="-1"></a>    <span class="cf">if</span> <span class="bu">len</span>(lineage) <span class="op">!=</span> <span class="dv">0</span>:</span>
<span id="cb6-20"><a href="blast-and-e-utilities.html#cb6-20" tabindex="-1"></a>        family <span class="op">=</span> lineage.split(<span class="st">&quot;;&quot;</span>)[<span class="dv">2</span>].strip()</span>
<span id="cb6-21"><a href="blast-and-e-utilities.html#cb6-21" tabindex="-1"></a>        family_counts[family] <span class="op">=</span> family_counts.get(family, <span class="dv">0</span>) <span class="op">+</span> <span class="dv">1</span></span>
<span id="cb6-22"><a href="blast-and-e-utilities.html#cb6-22" tabindex="-1"></a></span>
<span id="cb6-23"><a href="blast-and-e-utilities.html#cb6-23" tabindex="-1"></a><span class="bu">print</span>(<span class="bu">len</span>(family_counts.keys()))</span>
<span id="cb6-24"><a href="blast-and-e-utilities.html#cb6-24" tabindex="-1"></a><span class="bu">print</span>(<span class="bu">len</span>(family_counts.keys()))</span>
<span id="cb6-25"><a href="blast-and-e-utilities.html#cb6-25" tabindex="-1"></a></span>
<span id="cb6-26"><a href="blast-and-e-utilities.html#cb6-26" tabindex="-1"></a><span class="co"># Convert the family counts into a Pandas Data Frame and visualize them as a bar chart.</span></span>
<span id="cb6-27"><a href="blast-and-e-utilities.html#cb6-27" tabindex="-1"></a>df <span class="op">=</span> pd.DataFrame.from_dict(family_counts, orient<span class="op">=</span><span class="st">&#39;index&#39;</span>, columns<span class="op">=</span>[<span class="st">&#39;Frequency&#39;</span>])</span>
<span id="cb6-28"><a href="blast-and-e-utilities.html#cb6-28" tabindex="-1"></a>ax <span class="op">=</span> df.plot(kind<span class="op">=</span><span class="st">&#39;bar&#39;</span>, rot<span class="op">=</span><span class="dv">0</span>, legend<span class="op">=</span><span class="va">None</span>)</span>
<span id="cb6-29"><a href="blast-and-e-utilities.html#cb6-29" tabindex="-1"></a>ax.set_xlabel(<span class="st">&#39;Virus Family&#39;</span>)</span>
<span id="cb6-30"><a href="blast-and-e-utilities.html#cb6-30" tabindex="-1"></a>ax.set_ylabel(<span class="st">&#39;Frequency&#39;</span>)</span>
<span id="cb6-31"><a href="blast-and-e-utilities.html#cb6-31" tabindex="-1"></a>plt.show()</span></code></pre></div>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="index.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="distances-with-clustering-in-clans.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/llobarts/nestedgenes/edit/main/01-eutilities.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
