<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 6 Comparative analyses of selection | References</title>
  <meta name="description" content="This is the description of the code employed in the research." />
  <meta name="generator" content="bookdown 0.34 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 6 Comparative analyses of selection | References" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="This is the description of the code employed in the research." />
  <meta name="github-repo" content="https://github.com/llobarts/nestedgenes" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 6 Comparative analyses of selection | References" />
  
  <meta name="twitter:description" content="This is the description of the code employed in the research." />
  

<meta name="author" content="Esmeralda G. Legarda" />


<meta name="date" content="2024-01-08" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="compositional-analyses-of-biological-sequences.html"/>
<link rel="next" href="miscellania.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Code for 'Nested genes'</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> About</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#abstract"><i class="fa fa-check"></i><b>1.1</b> Abstract</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#additional-files"><i class="fa fa-check"></i><b>1.2</b> Additional files</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="blast-and-e-utilities.html"><a href="blast-and-e-utilities.html"><i class="fa fa-check"></i><b>2</b> Blast and E-utilities</a>
<ul>
<li class="chapter" data-level="2.1" data-path="blast-and-e-utilities.html"><a href="blast-and-e-utilities.html#psi-blast"><i class="fa fa-check"></i><b>2.1</b> PSI-BLAST</a></li>
<li class="chapter" data-level="2.2" data-path="blast-and-e-utilities.html"><a href="blast-and-e-utilities.html#protein-id-to-nucleotides"><i class="fa fa-check"></i><b>2.2</b> Protein ID to nucleotides</a></li>
<li class="chapter" data-level="2.3" data-path="blast-and-e-utilities.html"><a href="blast-and-e-utilities.html#species-to-family"><i class="fa fa-check"></i><b>2.3</b> Species to family</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html"><i class="fa fa-check"></i><b>3</b> Distances with clustering in CLANS</a>
<ul>
<li class="chapter" data-level="3.1" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html#read-the-clans-file"><i class="fa fa-check"></i><b>3.1</b> Read the CLANS file</a></li>
<li class="chapter" data-level="3.2" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html#create-a-dataframe"><i class="fa fa-check"></i><b>3.2</b> Create a DataFrame</a></li>
<li class="chapter" data-level="3.3" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html#calculate-centroids-and-distances"><i class="fa fa-check"></i><b>3.3</b> Calculate centroids and distances</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html#plot-the-clustermap"><i class="fa fa-check"></i><b>3.3.1</b> Plot the clustermap</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html#run"><i class="fa fa-check"></i><b>3.4</b> Run</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html"><i class="fa fa-check"></i><b>4</b> Distances with ChimeraX</a>
<ul>
<li class="chapter" data-level="4.1" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#obtaining-rmsd-with-chimerax"><i class="fa fa-check"></i><b>4.1</b> Obtaining RMSD with ChimeraX</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#import-libraries-and-data"><i class="fa fa-check"></i><b>4.1.1</b> Import libraries and data</a></li>
<li class="chapter" data-level="4.1.2" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#matchmaker"><i class="fa fa-check"></i><b>4.1.2</b> MatchMaker</a></li>
<li class="chapter" data-level="4.1.3" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#save-csv-file"><i class="fa fa-check"></i><b>4.1.3</b> Save csv file</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#hiarachical-clustering"><i class="fa fa-check"></i><b>4.2</b> Hiarachical clustering</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#libraries-and-read-csv"><i class="fa fa-check"></i><b>4.2.1</b> Libraries and read csv</a></li>
<li class="chapter" data-level="4.2.2" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#create-matrix"><i class="fa fa-check"></i><b>4.2.2</b> Create matrix</a></li>
<li class="chapter" data-level="4.2.3" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#plot-the-clustermap-1"><i class="fa fa-check"></i><b>4.2.3</b> Plot the clustermap</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="compositional-analyses-of-biological-sequences.html"><a href="compositional-analyses-of-biological-sequences.html"><i class="fa fa-check"></i><b>5</b> Compositional analyses of biological sequences</a>
<ul>
<li class="chapter" data-level="5.1" data-path="compositional-analyses-of-biological-sequences.html"><a href="compositional-analyses-of-biological-sequences.html#amino-acids"><i class="fa fa-check"></i><b>5.1</b> Amino acids</a></li>
<li class="chapter" data-level="5.2" data-path="compositional-analyses-of-biological-sequences.html"><a href="compositional-analyses-of-biological-sequences.html#nucleotides"><i class="fa fa-check"></i><b>5.2</b> Nucleotides</a></li>
<li class="chapter" data-level="5.3" data-path="compositional-analyses-of-biological-sequences.html"><a href="compositional-analyses-of-biological-sequences.html#dinuclotides"><i class="fa fa-check"></i><b>5.3</b> Dinuclotides</a></li>
<li class="chapter" data-level="5.4" data-path="compositional-analyses-of-biological-sequences.html"><a href="compositional-analyses-of-biological-sequences.html#codons"><i class="fa fa-check"></i><b>5.4</b> Codons</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html"><i class="fa fa-check"></i><b>6</b> Comparative analyses of selection</a>
<ul>
<li class="chapter" data-level="6.1" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#evolution-speed-with-absrel"><i class="fa fa-check"></i><b>6.1</b> Evolution speed with aBSREL</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#libraries-and-json-files"><i class="fa fa-check"></i><b>6.1.1</b> Libraries and json files</a></li>
<li class="chapter" data-level="6.1.2" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#omega-and-significance"><i class="fa fa-check"></i><b>6.1.2</b> Omega and significance</a></li>
<li class="chapter" data-level="6.1.3" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#relative-asinh-transformation"><i class="fa fa-check"></i><b>6.1.3</b> Relative asinh transformation</a></li>
<li class="chapter" data-level="6.1.4" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#utilities"><i class="fa fa-check"></i><b>6.1.4</b> Utilities</a></li>
<li class="chapter" data-level="6.1.5" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#boxplots-with-significance"><i class="fa fa-check"></i><b>6.1.5</b> Boxplots with significance</a></li>
<li class="chapter" data-level="6.1.6" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#test-selection"><i class="fa fa-check"></i><b>6.1.6</b> Test selection</a></li>
<li class="chapter" data-level="6.1.7" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#run-1"><i class="fa fa-check"></i><b>6.1.7</b> Run</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#fel-for-correlations"><i class="fa fa-check"></i><b>6.2</b> FEL for correlations</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#significance-and-difference"><i class="fa fa-check"></i><b>6.2.1</b> Significance and difference</a></li>
<li class="chapter" data-level="6.2.2" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#plot-correlation"><i class="fa fa-check"></i><b>6.2.2</b> Plot correlation</a></li>
<li class="chapter" data-level="6.2.3" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#test-correlation-of-selection"><i class="fa fa-check"></i><b>6.2.3</b> Test correlation of selection</a></li>
<li class="chapter" data-level="6.2.4" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#correlation-2d"><i class="fa fa-check"></i><b>6.2.4</b> Correlation 2D</a></li>
<li class="chapter" data-level="6.2.5" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#selection-in-2d"><i class="fa fa-check"></i><b>6.2.5</b> Selection in 2D</a></li>
<li class="chapter" data-level="6.2.6" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#run-2"><i class="fa fa-check"></i><b>6.2.6</b> Run</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="miscellania.html"><a href="miscellania.html"><i class="fa fa-check"></i><b>7</b> Miscellania</a>
<ul>
<li class="chapter" data-level="7.1" data-path="miscellania.html"><a href="miscellania.html#representation-of-predicted-disordered-amino-acid-data"><i class="fa fa-check"></i><b>7.1</b> Representation of predicted disordered amino acid data</a></li>
<li class="chapter" data-level="7.2" data-path="miscellania.html"><a href="miscellania.html#shannons-entropy"><i class="fa fa-check"></i><b>7.2</b> Shannon’s entropy</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/llobarts/nestedgenes" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">References</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="comparative-analyses-of-selection" class="section level1 hasAnchor" number="6">
<h1><span class="header-section-number">Chapter 6</span> Comparative analyses of selection<a href="comparative-analyses-of-selection.html#comparative-analyses-of-selection" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<div id="evolution-speed-with-absrel" class="section level2 hasAnchor" number="6.1">
<h2><span class="header-section-number">6.1</span> Evolution speed with aBSREL<a href="comparative-analyses-of-selection.html#evolution-speed-with-absrel" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Here, we leverage the complete sequences of tombusvirus genomes available in NCBI. Using the adaptive branch-site random effects likelihood (aBSREL) method <span class="citation">(<a href="#ref-msv022">Smith et al. 2015</a>)</span>, we estimate the omega (dN/dS rates ratio) per branch for p19 and p22, with RdRp as a reference. For each branch, the estimated omega values for p19 and p22 are normalized by the corresponding omega values for RdRp. This normalization brings all branches to a common relative scale, and the values are then asinh-transformed: ω_x^=ω_x⁄ω_RdRp with x  {p19, p22}.</p>
<div id="libraries-and-json-files" class="section level3 hasAnchor" number="6.1.1">
<h3><span class="header-section-number">6.1.1</span> Libraries and json files<a href="comparative-analyses-of-selection.html#libraries-and-json-files" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb1"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb1-1"><a href="comparative-analyses-of-selection.html#cb1-1" tabindex="-1"></a><span class="im">import</span> json</span>
<span id="cb1-2"><a href="comparative-analyses-of-selection.html#cb1-2" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb1-3"><a href="comparative-analyses-of-selection.html#cb1-3" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb1-4"><a href="comparative-analyses-of-selection.html#cb1-4" tabindex="-1"></a><span class="im">import</span> scipy.stats</span>
<span id="cb1-5"><a href="comparative-analyses-of-selection.html#cb1-5" tabindex="-1"></a><span class="im">from</span> statannotations.Annotator <span class="im">import</span> Annotator</span></code></pre></div>
<p>First, we need to read the JSON files containing the results of aBSREL for different genes from the same genomes.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb2-1"><a href="comparative-analyses-of-selection.html#cb2-1" tabindex="-1"></a><span class="co"># Read json files of each gene</span></span>
<span id="cb2-2"><a href="comparative-analyses-of-selection.html#cb2-2" tabindex="-1"></a>paths <span class="op">=</span> [path_p22, path_p19, path_RdRp, path_cp]</span>
<span id="cb2-3"><a href="comparative-analyses-of-selection.html#cb2-3" tabindex="-1"></a><span class="cf">for</span> path <span class="kw">in</span> paths:</span>
<span id="cb2-4"><a href="comparative-analyses-of-selection.html#cb2-4" tabindex="-1"></a>  df <span class="op">=</span> read_branches(path)</span>
<span id="cb2-5"><a href="comparative-analyses-of-selection.html#cb2-5" tabindex="-1"></a>  dfs.append(add_info(df, significance))</span></code></pre></div>
</div>
<div id="omega-and-significance" class="section level3 hasAnchor" number="6.1.2">
<h3><span class="header-section-number">6.1.2</span> Omega and significance<a href="comparative-analyses-of-selection.html#omega-and-significance" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Afterwards, we add significance information, and calculate the overall omega. Additionally, we identify nodes that are missing in the output of aBSREL. These nodes are then defined as null and non-significant.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb3-1"><a href="comparative-analyses-of-selection.html#cb3-1" tabindex="-1"></a><span class="kw">def</span> add_info(data, significance):</span>
<span id="cb3-2"><a href="comparative-analyses-of-selection.html#cb3-2" tabindex="-1"></a>    </span>
<span id="cb3-3"><a href="comparative-analyses-of-selection.html#cb3-3" tabindex="-1"></a>    data[<span class="st">&#39;omega_avg&#39;</span>] <span class="op">=</span> (data[<span class="st">&#39;omega1&#39;</span>]<span class="op">*</span>data[<span class="st">&#39;prop1&#39;</span>]) <span class="op">+</span> (data[<span class="st">&#39;omega2&#39;</span>]<span class="op">*</span>data[<span class="st">&#39;prop2&#39;</span>])</span>
<span id="cb3-4"><a href="comparative-analyses-of-selection.html#cb3-4" tabindex="-1"></a>    data[<span class="st">&quot;signif&quot;</span>] <span class="op">=</span> np.where(data[<span class="st">&#39;pvalue&#39;</span>] <span class="op">&lt;</span> significance, <span class="va">True</span>, <span class="va">False</span>)</span>
<span id="cb3-5"><a href="comparative-analyses-of-selection.html#cb3-5" tabindex="-1"></a></span>
<span id="cb3-6"><a href="comparative-analyses-of-selection.html#cb3-6" tabindex="-1"></a>    <span class="cf">return</span> data</span>
<span id="cb3-7"><a href="comparative-analyses-of-selection.html#cb3-7" tabindex="-1"></a></span>
<span id="cb3-8"><a href="comparative-analyses-of-selection.html#cb3-8" tabindex="-1"></a><span class="kw">def</span> fill_missing_nodes(data, missing_nodes):</span>
<span id="cb3-9"><a href="comparative-analyses-of-selection.html#cb3-9" tabindex="-1"></a>    </span>
<span id="cb3-10"><a href="comparative-analyses-of-selection.html#cb3-10" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> missing_nodes:</span>
<span id="cb3-11"><a href="comparative-analyses-of-selection.html#cb3-11" tabindex="-1"></a>        new_row <span class="op">=</span> {<span class="st">&quot;name&quot;</span>: i,</span>
<span id="cb3-12"><a href="comparative-analyses-of-selection.html#cb3-12" tabindex="-1"></a>                   <span class="st">&quot;type&quot;</span> : <span class="st">&quot;Node&quot;</span>,</span>
<span id="cb3-13"><a href="comparative-analyses-of-selection.html#cb3-13" tabindex="-1"></a>                   <span class="st">&quot;omega1&quot;</span>: <span class="fl">0.0</span>,</span>
<span id="cb3-14"><a href="comparative-analyses-of-selection.html#cb3-14" tabindex="-1"></a>                   <span class="st">&quot;prop1&quot;</span> : <span class="fl">0.0</span>,</span>
<span id="cb3-15"><a href="comparative-analyses-of-selection.html#cb3-15" tabindex="-1"></a>                   <span class="st">&quot;omega2&quot;</span>: <span class="fl">0.0</span>,</span>
<span id="cb3-16"><a href="comparative-analyses-of-selection.html#cb3-16" tabindex="-1"></a>                   <span class="st">&quot;prop2&quot;</span> : <span class="fl">0.0</span>,</span>
<span id="cb3-17"><a href="comparative-analyses-of-selection.html#cb3-17" tabindex="-1"></a>                   <span class="st">&quot;pvalue&quot;</span> : <span class="fl">1.0</span>,</span>
<span id="cb3-18"><a href="comparative-analyses-of-selection.html#cb3-18" tabindex="-1"></a>                   <span class="st">&quot;omega_avg&quot;</span> : <span class="fl">0.0</span>,</span>
<span id="cb3-19"><a href="comparative-analyses-of-selection.html#cb3-19" tabindex="-1"></a>                   <span class="st">&quot;Signif.&quot;</span> : <span class="va">False</span>}</span>
<span id="cb3-20"><a href="comparative-analyses-of-selection.html#cb3-20" tabindex="-1"></a>    </span>
<span id="cb3-21"><a href="comparative-analyses-of-selection.html#cb3-21" tabindex="-1"></a>        data <span class="op">=</span> data.append(new_row, ignore_index<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb3-22"><a href="comparative-analyses-of-selection.html#cb3-22" tabindex="-1"></a></span>
<span id="cb3-23"><a href="comparative-analyses-of-selection.html#cb3-23" tabindex="-1"></a>    <span class="cf">return</span> data</span></code></pre></div>
</div>
<div id="relative-asinh-transformation" class="section level3 hasAnchor" number="6.1.3">
<h3><span class="header-section-number">6.1.3</span> Relative asinh transformation<a href="comparative-analyses-of-selection.html#relative-asinh-transformation" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The next step consists of calculating the relative measure for comparing different genes, using an asinh transformation.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb4-1"><a href="comparative-analyses-of-selection.html#cb4-1" tabindex="-1"></a><span class="kw">def</span> relativize(genes, gene_ref):</span>
<span id="cb4-2"><a href="comparative-analyses-of-selection.html#cb4-2" tabindex="-1"></a>    </span>
<span id="cb4-3"><a href="comparative-analyses-of-selection.html#cb4-3" tabindex="-1"></a>    results <span class="op">=</span> []</span>
<span id="cb4-4"><a href="comparative-analyses-of-selection.html#cb4-4" tabindex="-1"></a>    type_tree <span class="op">=</span> []</span>
<span id="cb4-5"><a href="comparative-analyses-of-selection.html#cb4-5" tabindex="-1"></a></span>
<span id="cb4-6"><a href="comparative-analyses-of-selection.html#cb4-6" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">3</span>):</span>
<span id="cb4-7"><a href="comparative-analyses-of-selection.html#cb4-7" tabindex="-1"></a>        results_gene <span class="op">=</span> []</span>
<span id="cb4-8"><a href="comparative-analyses-of-selection.html#cb4-8" tabindex="-1"></a>        <span class="cf">for</span> name <span class="kw">in</span> gene_ref[<span class="st">&quot;name&quot;</span>]:</span>
<span id="cb4-9"><a href="comparative-analyses-of-selection.html#cb4-9" tabindex="-1"></a></span>
<span id="cb4-10"><a href="comparative-analyses-of-selection.html#cb4-10" tabindex="-1"></a>            omega_x <span class="op">=</span> <span class="bu">float</span>(genes[i][genes[i][<span class="st">&quot;name&quot;</span>] <span class="op">==</span> name][<span class="st">&quot;omega_avg&quot;</span>])</span>
<span id="cb4-11"><a href="comparative-analyses-of-selection.html#cb4-11" tabindex="-1"></a>            omega_std <span class="op">=</span> <span class="bu">float</span>(gene_ref[gene_ref[<span class="st">&quot;name&quot;</span>] <span class="op">==</span> name][<span class="st">&quot;omega_avg&quot;</span>])</span>
<span id="cb4-12"><a href="comparative-analyses-of-selection.html#cb4-12" tabindex="-1"></a></span>
<span id="cb4-13"><a href="comparative-analyses-of-selection.html#cb4-13" tabindex="-1"></a>            results_gene.append(np.log(omega_x <span class="op">+</span> (np.sqrt(omega_x<span class="op">**</span><span class="dv">2</span> <span class="op">+</span> <span class="dv">1</span>))<span class="op">/</span>(omega_std <span class="op">+</span> (np.sqrt(omega_std<span class="op">**</span><span class="dv">2</span> <span class="op">+</span> <span class="dv">1</span>)))))</span>
<span id="cb4-14"><a href="comparative-analyses-of-selection.html#cb4-14" tabindex="-1"></a>        </span>
<span id="cb4-15"><a href="comparative-analyses-of-selection.html#cb4-15" tabindex="-1"></a>        results.append(results_gene)</span>
<span id="cb4-16"><a href="comparative-analyses-of-selection.html#cb4-16" tabindex="-1"></a></span>
<span id="cb4-17"><a href="comparative-analyses-of-selection.html#cb4-17" tabindex="-1"></a>    <span class="cf">return</span> results</span></code></pre></div>
</div>
<div id="utilities" class="section level3 hasAnchor" number="6.1.4">
<h3><span class="header-section-number">6.1.4</span> Utilities<a href="comparative-analyses-of-selection.html#utilities" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Other functions are employed to restructure the data for visualization, while additional functions are used to calculate p-values using the Wilcoxon-Mann-Whitney test.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb5-1"><a href="comparative-analyses-of-selection.html#cb5-1" tabindex="-1"></a><span class="kw">def</span> divide_branches(results, gene_ref):</span>
<span id="cb5-2"><a href="comparative-analyses-of-selection.html#cb5-2" tabindex="-1"></a>    pos_tips <span class="op">=</span> gene_ref.index[gene_ref[<span class="st">&#39;type&#39;</span>] <span class="op">==</span> <span class="st">&quot;Tip&quot;</span>].tolist()</span>
<span id="cb5-3"><a href="comparative-analyses-of-selection.html#cb5-3" tabindex="-1"></a>    tips <span class="op">=</span> []</span>
<span id="cb5-4"><a href="comparative-analyses-of-selection.html#cb5-4" tabindex="-1"></a>    nodes <span class="op">=</span> []</span>
<span id="cb5-5"><a href="comparative-analyses-of-selection.html#cb5-5" tabindex="-1"></a></span>
<span id="cb5-6"><a href="comparative-analyses-of-selection.html#cb5-6" tabindex="-1"></a>    <span class="cf">for</span> gene <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">3</span>):</span>
<span id="cb5-7"><a href="comparative-analyses-of-selection.html#cb5-7" tabindex="-1"></a>        tip_selection <span class="op">=</span> []</span>
<span id="cb5-8"><a href="comparative-analyses-of-selection.html#cb5-8" tabindex="-1"></a>        node_selection <span class="op">=</span> []</span>
<span id="cb5-9"><a href="comparative-analyses-of-selection.html#cb5-9" tabindex="-1"></a>        <span class="cf">for</span> i, value <span class="kw">in</span> <span class="bu">enumerate</span>(results[gene]):</span>
<span id="cb5-10"><a href="comparative-analyses-of-selection.html#cb5-10" tabindex="-1"></a>            <span class="cf">if</span> i <span class="kw">in</span> pos_tips:</span>
<span id="cb5-11"><a href="comparative-analyses-of-selection.html#cb5-11" tabindex="-1"></a>                tip_selection.append(value)</span>
<span id="cb5-12"><a href="comparative-analyses-of-selection.html#cb5-12" tabindex="-1"></a>            <span class="cf">else</span>:</span>
<span id="cb5-13"><a href="comparative-analyses-of-selection.html#cb5-13" tabindex="-1"></a>                node_selection.append(value)</span>
<span id="cb5-14"><a href="comparative-analyses-of-selection.html#cb5-14" tabindex="-1"></a>                </span>
<span id="cb5-15"><a href="comparative-analyses-of-selection.html#cb5-15" tabindex="-1"></a>        tips.append(tip_selection)</span>
<span id="cb5-16"><a href="comparative-analyses-of-selection.html#cb5-16" tabindex="-1"></a>        nodes.append(node_selection)</span>
<span id="cb5-17"><a href="comparative-analyses-of-selection.html#cb5-17" tabindex="-1"></a>    </span>
<span id="cb5-18"><a href="comparative-analyses-of-selection.html#cb5-18" tabindex="-1"></a>    <span class="cf">return</span> (tips, nodes)</span>
<span id="cb5-19"><a href="comparative-analyses-of-selection.html#cb5-19" tabindex="-1"></a></span>
<span id="cb5-20"><a href="comparative-analyses-of-selection.html#cb5-20" tabindex="-1"></a><span class="kw">def</span> calc_pvalues(chunk):</span>
<span id="cb5-21"><a href="comparative-analyses-of-selection.html#cb5-21" tabindex="-1"></a>    pvalues_tips <span class="op">=</span> [scipy.stats.wilcoxon(chunk[<span class="dv">0</span>], chunk[<span class="dv">1</span>], alternative<span class="op">=</span><span class="st">&quot;two-sided&quot;</span>).pvalue, <span class="co"># p22 vs p19</span></span>
<span id="cb5-22"><a href="comparative-analyses-of-selection.html#cb5-22" tabindex="-1"></a>    scipy.stats.wilcoxon(chunk[<span class="dv">0</span>], chunk[<span class="dv">2</span>], alternative<span class="op">=</span><span class="st">&quot;two-sided&quot;</span>).pvalue, <span class="co"># p22 vs CP</span></span>
<span id="cb5-23"><a href="comparative-analyses-of-selection.html#cb5-23" tabindex="-1"></a>    scipy.stats.wilcoxon(chunk[<span class="dv">1</span>], chunk[<span class="dv">2</span>], alternative<span class="op">=</span><span class="st">&quot;two-sided&quot;</span>).pvalue]   <span class="co"># p19 vs CP</span></span>
<span id="cb5-24"><a href="comparative-analyses-of-selection.html#cb5-24" tabindex="-1"></a></span>
<span id="cb5-25"><a href="comparative-analyses-of-selection.html#cb5-25" tabindex="-1"></a>    <span class="cf">return</span> pvalues_tips</span>
<span id="cb5-26"><a href="comparative-analyses-of-selection.html#cb5-26" tabindex="-1"></a></span>
<span id="cb5-27"><a href="comparative-analyses-of-selection.html#cb5-27" tabindex="-1"></a><span class="kw">def</span> format_pvalues(pvalues_1to1):</span>
<span id="cb5-28"><a href="comparative-analyses-of-selection.html#cb5-28" tabindex="-1"></a>    <span class="cf">return</span> [[<span class="ss">f&#39;p=</span><span class="sc">{</span>pvalue<span class="sc">:.3e}</span><span class="ss">&#39;</span> <span class="cf">for</span> pvalue <span class="kw">in</span> pvalues_1to1[<span class="dv">0</span>]], </span>
<span id="cb5-29"><a href="comparative-analyses-of-selection.html#cb5-29" tabindex="-1"></a>            [<span class="ss">f&#39;p=</span><span class="sc">{</span>pvalue<span class="sc">:.3e}</span><span class="ss">&#39;</span> <span class="cf">for</span> pvalue <span class="kw">in</span> pvalues_1to1[<span class="dv">1</span>]], </span>
<span id="cb5-30"><a href="comparative-analyses-of-selection.html#cb5-30" tabindex="-1"></a>            [<span class="ss">f&#39;p=</span><span class="sc">{</span>pvalue<span class="sc">:.3e}</span><span class="ss">&#39;</span> <span class="cf">for</span> pvalue <span class="kw">in</span> pvalues_1to1[<span class="dv">2</span>]]]</span>
<span id="cb5-31"><a href="comparative-analyses-of-selection.html#cb5-31" tabindex="-1"></a></span>
<span id="cb5-32"><a href="comparative-analyses-of-selection.html#cb5-32" tabindex="-1"></a><span class="kw">def</span> info_to_dataframe(data):</span>
<span id="cb5-33"><a href="comparative-analyses-of-selection.html#cb5-33" tabindex="-1"></a>    <span class="cf">return</span> pd.DataFrame.from_dict({<span class="st">&quot;omega&quot;</span>: data[<span class="dv">0</span>] <span class="op">+</span> data[<span class="dv">1</span>] <span class="op">+</span> data[<span class="dv">2</span>], </span>
<span id="cb5-34"><a href="comparative-analyses-of-selection.html#cb5-34" tabindex="-1"></a>                            <span class="st">&quot;Gene&quot;</span> : <span class="bu">len</span>(data[<span class="dv">0</span>])<span class="op">*</span>[<span class="st">&#39;p22&#39;</span>] <span class="op">+</span> <span class="bu">len</span>(data[<span class="dv">0</span>])<span class="op">*</span>[<span class="st">&#39;p19&#39;</span>] <span class="op">+</span> <span class="bu">len</span>(data[<span class="dv">0</span>])<span class="op">*</span>[<span class="st">&#39;CP&#39;</span>]})</span></code></pre></div>
</div>
<div id="boxplots-with-significance" class="section level3 hasAnchor" number="6.1.5">
<h3><span class="header-section-number">6.1.5</span> Boxplots with significance<a href="comparative-analyses-of-selection.html#boxplots-with-significance" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Two different plots can be generated using the following code. The first plot includes all elements, nodes (or branches), and tips, while the second plot is focused on the entire tree. Both plots include p-value annotations.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb6-1"><a href="comparative-analyses-of-selection.html#cb6-1" tabindex="-1"></a><span class="kw">def</span> plot_diff_types(data, pvalues_1to1, formatted_pvalues):</span>
<span id="cb6-2"><a href="comparative-analyses-of-selection.html#cb6-2" tabindex="-1"></a></span>
<span id="cb6-3"><a href="comparative-analyses-of-selection.html#cb6-3" tabindex="-1"></a>    colors <span class="op">=</span> sns.color_palette(<span class="st">&quot;colorblind&quot;</span>, <span class="dv">5</span>)</span>
<span id="cb6-4"><a href="comparative-analyses-of-selection.html#cb6-4" tabindex="-1"></a>    fig, ax <span class="op">=</span> plt.subplot_mosaic(<span class="st">&quot;ABC&quot;</span>, figsize <span class="op">=</span> (<span class="dv">20</span>,<span class="dv">8</span>))</span>
<span id="cb6-5"><a href="comparative-analyses-of-selection.html#cb6-5" tabindex="-1"></a>    axes <span class="op">=</span> [ax[<span class="st">&quot;A&quot;</span>], ax[<span class="st">&quot;B&quot;</span>], ax[<span class="st">&quot;C&quot;</span>]]</span>
<span id="cb6-6"><a href="comparative-analyses-of-selection.html#cb6-6" tabindex="-1"></a>    annot <span class="op">=</span> []</span>
<span id="cb6-7"><a href="comparative-analyses-of-selection.html#cb6-7" tabindex="-1"></a>    pairs <span class="op">=</span> [(<span class="st">&quot;p22&quot;</span>, <span class="st">&quot;p19&quot;</span>),(<span class="st">&quot;p22&quot;</span>, <span class="st">&quot;CP&quot;</span>), (<span class="st">&quot;p19&quot;</span>, <span class="st">&quot;CP&quot;</span>)]</span>
<span id="cb6-8"><a href="comparative-analyses-of-selection.html#cb6-8" tabindex="-1"></a>    titles <span class="op">=</span> [<span class="st">&quot;TIPS&quot;</span>, <span class="st">&quot;BRANCHES&quot;</span>, <span class="st">&quot;TREE&quot;</span>]</span>
<span id="cb6-9"><a href="comparative-analyses-of-selection.html#cb6-9" tabindex="-1"></a></span>
<span id="cb6-10"><a href="comparative-analyses-of-selection.html#cb6-10" tabindex="-1"></a>    <span class="cf">for</span> part <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(data)):</span>
<span id="cb6-11"><a href="comparative-analyses-of-selection.html#cb6-11" tabindex="-1"></a>        <span class="cf">with</span> sns.plotting_context(<span class="st">&#39;notebook&#39;</span>, font_scale <span class="op">=</span> <span class="fl">1.2</span>):</span>
<span id="cb6-12"><a href="comparative-analyses-of-selection.html#cb6-12" tabindex="-1"></a>                        sns.swarmplot(data<span class="op">=</span>data[part], x<span class="op">=</span><span class="st">&quot;Gene&quot;</span>, y<span class="op">=</span><span class="st">&quot;omega&quot;</span>, ax<span class="op">=</span>axes[part],palette<span class="op">=</span>colors, zorder<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb6-13"><a href="comparative-analyses-of-selection.html#cb6-13" tabindex="-1"></a>                        sns.boxplot(data<span class="op">=</span>data[part], x<span class="op">=</span><span class="st">&quot;Gene&quot;</span>, y<span class="op">=</span><span class="st">&quot;omega&quot;</span>, dodge<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axes[part], boxprops<span class="op">=</span>{<span class="st">&#39;facecolor&#39;</span>:<span class="st">&#39;None&#39;</span>})</span>
<span id="cb6-14"><a href="comparative-analyses-of-selection.html#cb6-14" tabindex="-1"></a>                        axes[part].set_title(titles[part])</span>
<span id="cb6-15"><a href="comparative-analyses-of-selection.html#cb6-15" tabindex="-1"></a>                        plt.axhline(<span class="dv">1</span>)</span>
<span id="cb6-16"><a href="comparative-analyses-of-selection.html#cb6-16" tabindex="-1"></a></span>
<span id="cb6-17"><a href="comparative-analyses-of-selection.html#cb6-17" tabindex="-1"></a>                        <span class="co"># Add annotations</span></span>
<span id="cb6-18"><a href="comparative-analyses-of-selection.html#cb6-18" tabindex="-1"></a>                        annot.append(Annotator(ax<span class="op">=</span>axes[part], pairs <span class="op">=</span> pairs, data<span class="op">=</span>data[part], x<span class="op">=</span><span class="st">&quot;Gene&quot;</span>, y<span class="op">=</span><span class="st">&quot;omega&quot;</span>)) </span>
<span id="cb6-19"><a href="comparative-analyses-of-selection.html#cb6-19" tabindex="-1"></a>                        annot[part].set_custom_annotations(formatted_pvalues[part])</span>
<span id="cb6-20"><a href="comparative-analyses-of-selection.html#cb6-20" tabindex="-1"></a>                        annot[part].configure(test_short_name<span class="op">=</span><span class="st">&quot;MWW&quot;</span>)  <span class="co"># text_format is still simple</span></span>
<span id="cb6-21"><a href="comparative-analyses-of-selection.html#cb6-21" tabindex="-1"></a>                        annot[part].set_pvalues_and_annotate(pvalues_1to1[part])</span>
<span id="cb6-22"><a href="comparative-analyses-of-selection.html#cb6-22" tabindex="-1"></a>    </span>
<span id="cb6-23"><a href="comparative-analyses-of-selection.html#cb6-23" tabindex="-1"></a>    plt.tight_layout(pad<span class="op">=</span><span class="fl">1.5</span>)</span>
<span id="cb6-24"><a href="comparative-analyses-of-selection.html#cb6-24" tabindex="-1"></a>    plt.show()</span>
<span id="cb6-25"><a href="comparative-analyses-of-selection.html#cb6-25" tabindex="-1"></a></span>
<span id="cb6-26"><a href="comparative-analyses-of-selection.html#cb6-26" tabindex="-1"></a><span class="kw">def</span> plot_one_type(data, pvalues_1to1, formatted_pvalues, one):</span>
<span id="cb6-27"><a href="comparative-analyses-of-selection.html#cb6-27" tabindex="-1"></a>    colors <span class="op">=</span> [<span class="st">&quot;#9867F4&quot;</span>, <span class="st">&quot;#FFD500&quot;</span>]</span>
<span id="cb6-28"><a href="comparative-analyses-of-selection.html#cb6-28" tabindex="-1"></a>    plt.style.use(<span class="st">&quot;default&quot;</span>)</span>
<span id="cb6-29"><a href="comparative-analyses-of-selection.html#cb6-29" tabindex="-1"></a>    fig, ax <span class="op">=</span> plt.subplots(figsize <span class="op">=</span> (<span class="dv">6</span>,<span class="dv">12</span>))</span>
<span id="cb6-30"><a href="comparative-analyses-of-selection.html#cb6-30" tabindex="-1"></a>    part <span class="op">=</span> <span class="dv">2</span> <span class="co"># 2: tree</span></span>
<span id="cb6-31"><a href="comparative-analyses-of-selection.html#cb6-31" tabindex="-1"></a>    annot <span class="op">=</span> []</span>
<span id="cb6-32"><a href="comparative-analyses-of-selection.html#cb6-32" tabindex="-1"></a>    pairs <span class="op">=</span> [(<span class="st">&quot;p22&quot;</span>, <span class="st">&quot;p19&quot;</span>)]</span>
<span id="cb6-33"><a href="comparative-analyses-of-selection.html#cb6-33" tabindex="-1"></a>    size_labels <span class="op">=</span> <span class="dv">24</span></span>
<span id="cb6-34"><a href="comparative-analyses-of-selection.html#cb6-34" tabindex="-1"></a></span>
<span id="cb6-35"><a href="comparative-analyses-of-selection.html#cb6-35" tabindex="-1"></a>    <span class="cf">with</span> sns.plotting_context(<span class="st">&#39;notebook&#39;</span>, font_scale <span class="op">=</span> <span class="fl">1.2</span>):</span>
<span id="cb6-36"><a href="comparative-analyses-of-selection.html#cb6-36" tabindex="-1"></a>        sns.<span class="bu">set</span>(font_scale<span class="op">=</span><span class="dv">4</span>)</span>
<span id="cb6-37"><a href="comparative-analyses-of-selection.html#cb6-37" tabindex="-1"></a></span>
<span id="cb6-38"><a href="comparative-analyses-of-selection.html#cb6-38" tabindex="-1"></a>        sns.swarmplot(data<span class="op">=</span>data[part][data[part][<span class="st">&quot;Gene&quot;</span>] <span class="op">!=</span> <span class="st">&quot;CP&quot;</span>], x<span class="op">=</span><span class="st">&quot;Gene&quot;</span>, y<span class="op">=</span><span class="st">&quot;omega&quot;</span>, palette<span class="op">=</span>colors, zorder<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb6-39"><a href="comparative-analyses-of-selection.html#cb6-39" tabindex="-1"></a>        sns.boxplot(data<span class="op">=</span>data[part][data[part][<span class="st">&quot;Gene&quot;</span>] <span class="op">!=</span> <span class="st">&quot;CP&quot;</span>], x<span class="op">=</span><span class="st">&quot;Gene&quot;</span>, y<span class="op">=</span><span class="st">&quot;omega&quot;</span>, dodge<span class="op">=</span><span class="va">True</span>, boxprops<span class="op">=</span>{<span class="st">&#39;facecolor&#39;</span>:<span class="st">&#39;None&#39;</span>})</span>
<span id="cb6-40"><a href="comparative-analyses-of-selection.html#cb6-40" tabindex="-1"></a>        plt.axhline(<span class="dv">0</span>, color<span class="op">=</span><span class="st">&#39;grey&#39;</span>, linestyle<span class="op">=</span><span class="st">&#39;dotted&#39;</span>)</span>
<span id="cb6-41"><a href="comparative-analyses-of-selection.html#cb6-41" tabindex="-1"></a>        ax.set_xlabel(<span class="st">&quot;Gene sequence&quot;</span>, fontsize <span class="op">=</span> size_labels)</span>
<span id="cb6-42"><a href="comparative-analyses-of-selection.html#cb6-42" tabindex="-1"></a>        ax.set_ylabel(<span class="st">&quot;Transformed omega ratio&quot;</span>, fontsize <span class="op">=</span> size_labels)</span>
<span id="cb6-43"><a href="comparative-analyses-of-selection.html#cb6-43" tabindex="-1"></a>        ax.yaxis.set_tick_params(labelsize <span class="op">=</span> size_labels <span class="op">-</span> <span class="dv">1</span>)</span>
<span id="cb6-44"><a href="comparative-analyses-of-selection.html#cb6-44" tabindex="-1"></a>        ax.xaxis.set_tick_params(labelsize <span class="op">=</span> size_labels <span class="op">-</span> <span class="dv">1</span>)</span>
<span id="cb6-45"><a href="comparative-analyses-of-selection.html#cb6-45" tabindex="-1"></a></span>
<span id="cb6-46"><a href="comparative-analyses-of-selection.html#cb6-46" tabindex="-1"></a>        <span class="co"># Add annotations</span></span>
<span id="cb6-47"><a href="comparative-analyses-of-selection.html#cb6-47" tabindex="-1"></a>        annot <span class="op">=</span> Annotator(ax<span class="op">=</span>ax, pairs <span class="op">=</span> pairs, data<span class="op">=</span>data[part][data[part][<span class="st">&quot;Gene&quot;</span>] <span class="op">!=</span> <span class="st">&quot;CP&quot;</span>], x<span class="op">=</span><span class="st">&quot;Gene&quot;</span>, y<span class="op">=</span><span class="st">&quot;omega&quot;</span>)</span>
<span id="cb6-48"><a href="comparative-analyses-of-selection.html#cb6-48" tabindex="-1"></a>        annot.set_custom_annotations([formatted_pvalues[part][one]])</span>
<span id="cb6-49"><a href="comparative-analyses-of-selection.html#cb6-49" tabindex="-1"></a>        annot.configure(test_short_name<span class="op">=</span><span class="st">&quot;MWW&quot;</span>)  <span class="co"># text_format is still simple</span></span>
<span id="cb6-50"><a href="comparative-analyses-of-selection.html#cb6-50" tabindex="-1"></a>        annot.set_pvalues_and_annotate([pvalues_1to1[part][one]])</span>
<span id="cb6-51"><a href="comparative-analyses-of-selection.html#cb6-51" tabindex="-1"></a></span>
<span id="cb6-52"><a href="comparative-analyses-of-selection.html#cb6-52" tabindex="-1"></a>    plt.tight_layout(pad<span class="op">=</span><span class="fl">1.5</span>)</span>
<span id="cb6-53"><a href="comparative-analyses-of-selection.html#cb6-53" tabindex="-1"></a>    <span class="co"># plt.savefig(&quot;/Users/esmeralda/Documents/TFM/article/omegaratio.pdf&quot;, format= &quot;pdf&quot;)</span></span>
<span id="cb6-54"><a href="comparative-analyses-of-selection.html#cb6-54" tabindex="-1"></a>    plt.show()</span></code></pre></div>
</div>
<div id="test-selection" class="section level3 hasAnchor" number="6.1.6">
<h3><span class="header-section-number">6.1.6</span> Test selection<a href="comparative-analyses-of-selection.html#test-selection" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Finally, we determine whether the omega values significantly deviate from zero, indicating the presence or absence of selection.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb7-1"><a href="comparative-analyses-of-selection.html#cb7-1" tabindex="-1"></a><span class="kw">def</span> test_one(df_values, mean):</span>
<span id="cb7-2"><a href="comparative-analyses-of-selection.html#cb7-2" tabindex="-1"></a>    results_test <span class="op">=</span> {<span class="st">&quot;Tips:&quot;</span>: [scipy.stats.ttest_1samp(df_values[<span class="dv">0</span>][<span class="dv">0</span>], popmean<span class="op">=</span>mean).pvalue,</span>
<span id="cb7-3"><a href="comparative-analyses-of-selection.html#cb7-3" tabindex="-1"></a>                              scipy.stats.ttest_1samp(df_values[<span class="dv">0</span>][<span class="dv">1</span>], popmean<span class="op">=</span>mean).pvalue,</span>
<span id="cb7-4"><a href="comparative-analyses-of-selection.html#cb7-4" tabindex="-1"></a>                              scipy.stats.ttest_1samp(df_values[<span class="dv">0</span>][<span class="dv">2</span>], popmean<span class="op">=</span>mean).pvalue],</span>
<span id="cb7-5"><a href="comparative-analyses-of-selection.html#cb7-5" tabindex="-1"></a>                    <span class="st">&quot;Branches:&quot;</span>: [scipy.stats.ttest_1samp(df_values[<span class="dv">1</span>][<span class="dv">0</span>], popmean<span class="op">=</span>mean).pvalue,</span>
<span id="cb7-6"><a href="comparative-analyses-of-selection.html#cb7-6" tabindex="-1"></a>                                 scipy.stats.ttest_1samp(df_values[<span class="dv">1</span>][<span class="dv">1</span>], popmean<span class="op">=</span>mean).pvalue,</span>
<span id="cb7-7"><a href="comparative-analyses-of-selection.html#cb7-7" tabindex="-1"></a>                                 scipy.stats.ttest_1samp(df_values[<span class="dv">1</span>][<span class="dv">2</span>], popmean<span class="op">=</span>mean).pvalue],</span>
<span id="cb7-8"><a href="comparative-analyses-of-selection.html#cb7-8" tabindex="-1"></a>                    <span class="st">&quot;Tree&quot;</span>: [scipy.stats.ttest_1samp(df_values[<span class="dv">2</span>][<span class="dv">0</span>], popmean<span class="op">=</span>mean).pvalue,</span>
<span id="cb7-9"><a href="comparative-analyses-of-selection.html#cb7-9" tabindex="-1"></a>                             scipy.stats.ttest_1samp(df_values[<span class="dv">2</span>][<span class="dv">1</span>], popmean<span class="op">=</span>mean).pvalue,</span>
<span id="cb7-10"><a href="comparative-analyses-of-selection.html#cb7-10" tabindex="-1"></a>                             scipy.stats.ttest_1samp(df_values[<span class="dv">2</span>][<span class="dv">2</span>], popmean<span class="op">=</span>mean).pvalue]}</span>
<span id="cb7-11"><a href="comparative-analyses-of-selection.html#cb7-11" tabindex="-1"></a>    </span>
<span id="cb7-12"><a href="comparative-analyses-of-selection.html#cb7-12" tabindex="-1"></a>    df <span class="op">=</span> pd.DataFrame.from_dict(results_test)</span>
<span id="cb7-13"><a href="comparative-analyses-of-selection.html#cb7-13" tabindex="-1"></a>    df.index <span class="op">=</span> [<span class="st">&quot;p22&quot;</span>, <span class="st">&quot;p19&quot;</span>, <span class="st">&quot;CP&quot;</span>]</span>
<span id="cb7-14"><a href="comparative-analyses-of-selection.html#cb7-14" tabindex="-1"></a></span>
<span id="cb7-15"><a href="comparative-analyses-of-selection.html#cb7-15" tabindex="-1"></a>    <span class="cf">return</span> df</span></code></pre></div>
</div>
<div id="run-1" class="section level3 hasAnchor" number="6.1.7">
<h3><span class="header-section-number">6.1.7</span> Run<a href="comparative-analyses-of-selection.html#run-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb8"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb8-1"><a href="comparative-analyses-of-selection.html#cb8-1" tabindex="-1"></a><span class="kw">def</span> main():</span>
<span id="cb8-2"><a href="comparative-analyses-of-selection.html#cb8-2" tabindex="-1"></a>    dfs <span class="op">=</span> [] </span>
<span id="cb8-3"><a href="comparative-analyses-of-selection.html#cb8-3" tabindex="-1"></a>    significance <span class="op">=</span> <span class="fl">0.1</span></span>
<span id="cb8-4"><a href="comparative-analyses-of-selection.html#cb8-4" tabindex="-1"></a></span>
<span id="cb8-5"><a href="comparative-analyses-of-selection.html#cb8-5" tabindex="-1"></a>    path_p22 <span class="op">=</span> <span class="st">&#39;p22_nt_alignment.fasta.ABSREL.json&#39;</span></span>
<span id="cb8-6"><a href="comparative-analyses-of-selection.html#cb8-6" tabindex="-1"></a>    path_p19 <span class="op">=</span> <span class="st">&#39;p19_nt_alignment.fasta.ABSREL.json&#39;</span></span>
<span id="cb8-7"><a href="comparative-analyses-of-selection.html#cb8-7" tabindex="-1"></a>    path_RdRp <span class="op">=</span> <span class="st">&#39;RdRp_nt_alignment.fasta.ABSREL.json&#39;</span></span>
<span id="cb8-8"><a href="comparative-analyses-of-selection.html#cb8-8" tabindex="-1"></a>    path_cp <span class="op">=</span> <span class="st">&#39;cp_nt_alignment.fasta.ABSREL.json&#39;</span></span>
<span id="cb8-9"><a href="comparative-analyses-of-selection.html#cb8-9" tabindex="-1"></a></span>
<span id="cb8-10"><a href="comparative-analyses-of-selection.html#cb8-10" tabindex="-1"></a>    <span class="co"># Read json files of each gene</span></span>
<span id="cb8-11"><a href="comparative-analyses-of-selection.html#cb8-11" tabindex="-1"></a>    paths <span class="op">=</span> [path_p22, path_p19, path_RdRp, path_cp]</span>
<span id="cb8-12"><a href="comparative-analyses-of-selection.html#cb8-12" tabindex="-1"></a>    <span class="cf">for</span> path <span class="kw">in</span> paths:</span>
<span id="cb8-13"><a href="comparative-analyses-of-selection.html#cb8-13" tabindex="-1"></a>        df <span class="op">=</span> read_branches(path)</span>
<span id="cb8-14"><a href="comparative-analyses-of-selection.html#cb8-14" tabindex="-1"></a>        dfs.append(add_info(df, significance))</span>
<span id="cb8-15"><a href="comparative-analyses-of-selection.html#cb8-15" tabindex="-1"></a></span>
<span id="cb8-16"><a href="comparative-analyses-of-selection.html#cb8-16" tabindex="-1"></a>    <span class="co"># Filling missing nodes with zero values</span></span>
<span id="cb8-17"><a href="comparative-analyses-of-selection.html#cb8-17" tabindex="-1"></a>    missing_nodes_p22 <span class="op">=</span> [<span class="st">&quot;Node11&quot;</span>, <span class="st">&quot;Node18&quot;</span>, <span class="st">&quot;Node6&quot;</span>]</span>
<span id="cb8-18"><a href="comparative-analyses-of-selection.html#cb8-18" tabindex="-1"></a>    missing_nodes_p19 <span class="op">=</span> [<span class="st">&quot;Node11&quot;</span>, <span class="st">&quot;Node18&quot;</span>, <span class="st">&quot;Node3&quot;</span>, <span class="st">&quot;Node6&quot;</span>]</span>
<span id="cb8-19"><a href="comparative-analyses-of-selection.html#cb8-19" tabindex="-1"></a></span>
<span id="cb8-20"><a href="comparative-analyses-of-selection.html#cb8-20" tabindex="-1"></a>    df_p22 <span class="op">=</span> fill_missing_nodes(dfs[<span class="dv">0</span>], missing_nodes_p22)</span>
<span id="cb8-21"><a href="comparative-analyses-of-selection.html#cb8-21" tabindex="-1"></a>    df_p19 <span class="op">=</span> fill_missing_nodes(dfs[<span class="dv">1</span>], missing_nodes_p19)</span>
<span id="cb8-22"><a href="comparative-analyses-of-selection.html#cb8-22" tabindex="-1"></a>    df_RdRp <span class="op">=</span> dfs[<span class="dv">2</span>]</span>
<span id="cb8-23"><a href="comparative-analyses-of-selection.html#cb8-23" tabindex="-1"></a>    df_cp <span class="op">=</span> dfs[<span class="dv">3</span>]</span>
<span id="cb8-24"><a href="comparative-analyses-of-selection.html#cb8-24" tabindex="-1"></a></span>
<span id="cb8-25"><a href="comparative-analyses-of-selection.html#cb8-25" tabindex="-1"></a>    <span class="co"># Generate relative values for all, tips and nodes</span></span>
<span id="cb8-26"><a href="comparative-analyses-of-selection.html#cb8-26" tabindex="-1"></a>    genes <span class="op">=</span> [df_p22, df_p19, df_cp]</span>
<span id="cb8-27"><a href="comparative-analyses-of-selection.html#cb8-27" tabindex="-1"></a>    results <span class="op">=</span> relativize(genes, df_RdRp)</span>
<span id="cb8-28"><a href="comparative-analyses-of-selection.html#cb8-28" tabindex="-1"></a>    tips, nodes <span class="op">=</span> divide_branches(results, df_RdRp)</span>
<span id="cb8-29"><a href="comparative-analyses-of-selection.html#cb8-29" tabindex="-1"></a></span>
<span id="cb8-30"><a href="comparative-analyses-of-selection.html#cb8-30" tabindex="-1"></a>    <span class="co"># Calculate pvalues in pairs</span></span>
<span id="cb8-31"><a href="comparative-analyses-of-selection.html#cb8-31" tabindex="-1"></a>    pvalues_1to1 <span class="op">=</span> []</span>
<span id="cb8-32"><a href="comparative-analyses-of-selection.html#cb8-32" tabindex="-1"></a>    df_values <span class="op">=</span> []</span>
<span id="cb8-33"><a href="comparative-analyses-of-selection.html#cb8-33" tabindex="-1"></a>    sep_types <span class="op">=</span> [tips, nodes, results]</span>
<span id="cb8-34"><a href="comparative-analyses-of-selection.html#cb8-34" tabindex="-1"></a>    </span>
<span id="cb8-35"><a href="comparative-analyses-of-selection.html#cb8-35" tabindex="-1"></a>    <span class="cf">for</span> chunk <span class="kw">in</span> sep_types:</span>
<span id="cb8-36"><a href="comparative-analyses-of-selection.html#cb8-36" tabindex="-1"></a>        pvalues_1to1.append(calc_pvalues(chunk))</span>
<span id="cb8-37"><a href="comparative-analyses-of-selection.html#cb8-37" tabindex="-1"></a>        df_values.append(info_to_dataframe(chunk))</span>
<span id="cb8-38"><a href="comparative-analyses-of-selection.html#cb8-38" tabindex="-1"></a></span>
<span id="cb8-39"><a href="comparative-analyses-of-selection.html#cb8-39" tabindex="-1"></a>    <span class="co"># Get formatted pvalues</span></span>
<span id="cb8-40"><a href="comparative-analyses-of-selection.html#cb8-40" tabindex="-1"></a>    formatted_pvalues <span class="op">=</span> format_pvalues(pvalues_1to1)</span>
<span id="cb8-41"><a href="comparative-analyses-of-selection.html#cb8-41" tabindex="-1"></a></span>
<span id="cb8-42"><a href="comparative-analyses-of-selection.html#cb8-42" tabindex="-1"></a>    <span class="co"># Boxplots with significance</span></span>
<span id="cb8-43"><a href="comparative-analyses-of-selection.html#cb8-43" tabindex="-1"></a>    plot_diff_types(df_values, pvalues_1to1, formatted_pvalues)</span>
<span id="cb8-44"><a href="comparative-analyses-of-selection.html#cb8-44" tabindex="-1"></a>    plot_one_type(df_values,  pvalues_1to1, formatted_pvalues, <span class="dv">0</span>)</span>
<span id="cb8-45"><a href="comparative-analyses-of-selection.html#cb8-45" tabindex="-1"></a></span>
<span id="cb8-46"><a href="comparative-analyses-of-selection.html#cb8-46" tabindex="-1"></a>    <span class="co"># T test one sample vs mean = 0</span></span>
<span id="cb8-47"><a href="comparative-analyses-of-selection.html#cb8-47" tabindex="-1"></a>    <span class="bu">print</span>(test_one(sep_types, <span class="dv">0</span>))</span></code></pre></div>
</div>
</div>
<div id="fel-for-correlations" class="section level2 hasAnchor" number="6.2">
<h2><span class="header-section-number">6.2</span> FEL for correlations<a href="comparative-analyses-of-selection.html#fel-for-correlations" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In this section, we read JSON files containing the results of p19 and p22 analyses, generated with the FEL method of Hyphy <span class="citation">(<a href="#ref-Kosakovsky2005">Kosakovsky Pond and Frost 2005</a>)</span>. The objective is to establish correlations between selection and infer the extent to which the correlation is influenced by the secondary structure of the proteins they encode.</p>
<p>###Libraries and json files</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb9-1"><a href="comparative-analyses-of-selection.html#cb9-1" tabindex="-1"></a><span class="im">import</span> json</span>
<span id="cb9-2"><a href="comparative-analyses-of-selection.html#cb9-2" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb9-3"><a href="comparative-analyses-of-selection.html#cb9-3" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb9-4"><a href="comparative-analyses-of-selection.html#cb9-4" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb9-5"><a href="comparative-analyses-of-selection.html#cb9-5" tabindex="-1"></a><span class="im">import</span> scipy.stats</span>
<span id="cb9-6"><a href="comparative-analyses-of-selection.html#cb9-6" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span></code></pre></div>
<p>First, we need to read the JSON files containing the results of the FEL analysis for both genes.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb10-1"><a href="comparative-analyses-of-selection.html#cb10-1" tabindex="-1"></a><span class="kw">def</span> read_json(path):</span>
<span id="cb10-2"><a href="comparative-analyses-of-selection.html#cb10-2" tabindex="-1"></a>    <span class="cf">with</span> <span class="bu">open</span>(path) <span class="im">as</span> user_file:</span>
<span id="cb10-3"><a href="comparative-analyses-of-selection.html#cb10-3" tabindex="-1"></a>        file_contents <span class="op">=</span> user_file.read()</span>
<span id="cb10-4"><a href="comparative-analyses-of-selection.html#cb10-4" tabindex="-1"></a>    parsed_json <span class="op">=</span> json.loads(file_contents)</span>
<span id="cb10-5"><a href="comparative-analyses-of-selection.html#cb10-5" tabindex="-1"></a></span>
<span id="cb10-6"><a href="comparative-analyses-of-selection.html#cb10-6" tabindex="-1"></a>    values <span class="op">=</span> {<span class="st">&quot;codon&quot;</span>: <span class="bu">list</span>(<span class="bu">range</span>(<span class="bu">len</span>(parsed_json[<span class="st">&#39;MLE&#39;</span>][<span class="st">&#39;content&#39;</span>][<span class="st">&quot;0&quot;</span>]))),</span>
<span id="cb10-7"><a href="comparative-analyses-of-selection.html#cb10-7" tabindex="-1"></a>               <span class="st">&quot;alpha&quot;</span>: [], </span>
<span id="cb10-8"><a href="comparative-analyses-of-selection.html#cb10-8" tabindex="-1"></a>               <span class="st">&quot;beta&quot;</span>: [],</span>
<span id="cb10-9"><a href="comparative-analyses-of-selection.html#cb10-9" tabindex="-1"></a>               <span class="st">&quot;pvalue&quot;</span>: []}</span>
<span id="cb10-10"><a href="comparative-analyses-of-selection.html#cb10-10" tabindex="-1"></a></span>
<span id="cb10-11"><a href="comparative-analyses-of-selection.html#cb10-11" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(parsed_json[<span class="st">&#39;MLE&#39;</span>][<span class="st">&#39;content&#39;</span>][<span class="st">&quot;0&quot;</span>])):</span>
<span id="cb10-12"><a href="comparative-analyses-of-selection.html#cb10-12" tabindex="-1"></a>        values[<span class="st">&#39;alpha&#39;</span>].append(parsed_json[<span class="st">&#39;MLE&#39;</span>][<span class="st">&#39;content&#39;</span>][<span class="st">&quot;0&quot;</span>][i][<span class="dv">0</span>])</span>
<span id="cb10-13"><a href="comparative-analyses-of-selection.html#cb10-13" tabindex="-1"></a>        values[<span class="st">&#39;beta&#39;</span>].append(parsed_json[<span class="st">&#39;MLE&#39;</span>][<span class="st">&#39;content&#39;</span>][<span class="st">&quot;0&quot;</span>][i][<span class="dv">1</span>])</span>
<span id="cb10-14"><a href="comparative-analyses-of-selection.html#cb10-14" tabindex="-1"></a>        values[<span class="st">&#39;pvalue&#39;</span>].append(parsed_json[<span class="st">&#39;MLE&#39;</span>][<span class="st">&#39;content&#39;</span>][<span class="st">&quot;0&quot;</span>][i][<span class="dv">4</span>])</span>
<span id="cb10-15"><a href="comparative-analyses-of-selection.html#cb10-15" tabindex="-1"></a></span>
<span id="cb10-16"><a href="comparative-analyses-of-selection.html#cb10-16" tabindex="-1"></a>    <span class="cf">return</span> pd.DataFrame.from_dict(values)</span></code></pre></div>
<div id="significance-and-difference" class="section level3 hasAnchor" number="6.2.1">
<h3><span class="header-section-number">6.2.1</span> Significance and difference<a href="comparative-analyses-of-selection.html#significance-and-difference" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Now, information about significance and the difference between the estimated rates of non-synonymous (beta) and synonymous (alpha) substitutions.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb11-1"><a href="comparative-analyses-of-selection.html#cb11-1" tabindex="-1"></a><span class="kw">def</span> add_info(data, significance):</span>
<span id="cb11-2"><a href="comparative-analyses-of-selection.html#cb11-2" tabindex="-1"></a>    data[<span class="st">&#39;beta - alpha&#39;</span>] <span class="op">=</span> data[<span class="st">&#39;beta&#39;</span>] <span class="op">-</span> data[<span class="st">&#39;alpha&#39;</span>]</span>
<span id="cb11-3"><a href="comparative-analyses-of-selection.html#cb11-3" tabindex="-1"></a>    data[<span class="st">&quot;signif&quot;</span>] <span class="op">=</span> np.where(data[<span class="st">&#39;pvalue&#39;</span>] <span class="op">&lt;</span> significance, <span class="va">True</span>, <span class="va">False</span>)</span>
<span id="cb11-4"><a href="comparative-analyses-of-selection.html#cb11-4" tabindex="-1"></a></span>
<span id="cb11-5"><a href="comparative-analyses-of-selection.html#cb11-5" tabindex="-1"></a>    <span class="cf">return</span> data</span></code></pre></div>
<p>We can even print the codon positions that are detected to be under significant selection.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb12-1"><a href="comparative-analyses-of-selection.html#cb12-1" tabindex="-1"></a><span class="kw">def</span> significative_codons(data):</span>
<span id="cb12-2"><a href="comparative-analyses-of-selection.html#cb12-2" tabindex="-1"></a>    <span class="cf">return</span> (<span class="bu">list</span>(data[(data[<span class="st">&quot;signif&quot;</span>] <span class="op">==</span> <span class="va">True</span>) <span class="op">&amp;</span> (data[<span class="st">&quot;beta - alpha&quot;</span>] <span class="op">&gt;</span> <span class="dv">0</span>)].index <span class="op">+</span> <span class="dv">1</span>) <span class="op">+\</span></span>
<span id="cb12-3"><a href="comparative-analyses-of-selection.html#cb12-3" tabindex="-1"></a>            <span class="bu">list</span>(data[(data[<span class="st">&quot;signif&quot;</span>] <span class="op">==</span> <span class="va">True</span>) <span class="op">&amp;</span> (data[<span class="st">&quot;beta - alpha&quot;</span>] <span class="op">&lt;</span> <span class="dv">0</span>)].index <span class="op">+</span> <span class="dv">1</span>))</span></code></pre></div>
</div>
<div id="plot-correlation" class="section level3 hasAnchor" number="6.2.2">
<h3><span class="header-section-number">6.2.2</span> Plot correlation<a href="comparative-analyses-of-selection.html#plot-correlation" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>As we know the frameshift of one gene, we can plot selection values, specifically the difference between beta and alpha, by positions in p19 and p22. Additionally, we can denote significance using colors.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb13-1"><a href="comparative-analyses-of-selection.html#cb13-1" tabindex="-1"></a><span class="kw">def</span> plot_correlation(data1, data2, shift, colors):</span>
<span id="cb13-2"><a href="comparative-analyses-of-selection.html#cb13-2" tabindex="-1"></a>    plt.style.use(<span class="st">&#39;default&#39;</span>)</span>
<span id="cb13-3"><a href="comparative-analyses-of-selection.html#cb13-3" tabindex="-1"></a>    fig <span class="op">=</span> plt.figure(figsize<span class="op">=</span>(<span class="dv">6</span>,<span class="dv">12</span>))</span>
<span id="cb13-4"><a href="comparative-analyses-of-selection.html#cb13-4" tabindex="-1"></a>    ax <span class="op">=</span> fig.add_subplot(<span class="dv">111</span>)</span>
<span id="cb13-5"><a href="comparative-analyses-of-selection.html#cb13-5" tabindex="-1"></a>    fontsize <span class="op">=</span> <span class="dv">24</span></span>
<span id="cb13-6"><a href="comparative-analyses-of-selection.html#cb13-6" tabindex="-1"></a>    points_size <span class="op">=</span> <span class="dv">20</span></span>
<span id="cb13-7"><a href="comparative-analyses-of-selection.html#cb13-7" tabindex="-1"></a></span>
<span id="cb13-8"><a href="comparative-analyses-of-selection.html#cb13-8" tabindex="-1"></a>    plt.ylim(<span class="op">-</span><span class="dv">14</span>,<span class="dv">11</span>)</span>
<span id="cb13-9"><a href="comparative-analyses-of-selection.html#cb13-9" tabindex="-1"></a>    plt.xlim(<span class="op">-</span><span class="dv">16</span>,<span class="dv">7</span>)</span>
<span id="cb13-10"><a href="comparative-analyses-of-selection.html#cb13-10" tabindex="-1"></a></span>
<span id="cb13-11"><a href="comparative-analyses-of-selection.html#cb13-11" tabindex="-1"></a>    plt.scatter(data2[<span class="st">&quot;beta - alpha&quot;</span>][shift:<span class="bu">len</span>(data1[<span class="st">&quot;beta - alpha&quot;</span>]) <span class="op">+</span> shift], data1[<span class="st">&quot;beta - alpha&quot;</span>], s<span class="op">=</span>points_size, c<span class="op">=</span>colors)</span>
<span id="cb13-12"><a href="comparative-analyses-of-selection.html#cb13-12" tabindex="-1"></a></span>
<span id="cb13-13"><a href="comparative-analyses-of-selection.html#cb13-13" tabindex="-1"></a>    plt.xlabel(<span class="st">&quot;$</span><span class="ch">\\</span><span class="st">beta$ - $</span><span class="ch">\\</span><span class="st">alpha$ (p22)&quot;</span>)</span>
<span id="cb13-14"><a href="comparative-analyses-of-selection.html#cb13-14" tabindex="-1"></a>    plt.ylabel(<span class="st">&quot;$</span><span class="ch">\\</span><span class="st">beta$ - $</span><span class="ch">\\</span><span class="st">alpha$ (p19)&quot;</span>)</span>
<span id="cb13-15"><a href="comparative-analyses-of-selection.html#cb13-15" tabindex="-1"></a></span>
<span id="cb13-16"><a href="comparative-analyses-of-selection.html#cb13-16" tabindex="-1"></a>    plt.axhline(<span class="dv">0</span>,color<span class="op">=</span><span class="st">&#39;black&#39;</span>,</span>
<span id="cb13-17"><a href="comparative-analyses-of-selection.html#cb13-17" tabindex="-1"></a>                linewidth<span class="op">=</span><span class="dv">1</span>,</span>
<span id="cb13-18"><a href="comparative-analyses-of-selection.html#cb13-18" tabindex="-1"></a>                ls<span class="op">=</span><span class="st">&quot;--&quot;</span>, </span>
<span id="cb13-19"><a href="comparative-analyses-of-selection.html#cb13-19" tabindex="-1"></a>                alpha<span class="op">=</span><span class="fl">0.5</span>) <span class="co"># x = 0</span></span>
<span id="cb13-20"><a href="comparative-analyses-of-selection.html#cb13-20" tabindex="-1"></a>    plt.axvline(<span class="dv">0</span>,color<span class="op">=</span><span class="st">&#39;black&#39;</span>,</span>
<span id="cb13-21"><a href="comparative-analyses-of-selection.html#cb13-21" tabindex="-1"></a>                linewidth<span class="op">=</span><span class="dv">1</span>,</span>
<span id="cb13-22"><a href="comparative-analyses-of-selection.html#cb13-22" tabindex="-1"></a>                ls<span class="op">=</span><span class="st">&quot;--&quot;</span>,</span>
<span id="cb13-23"><a href="comparative-analyses-of-selection.html#cb13-23" tabindex="-1"></a>                alpha<span class="op">=</span><span class="fl">0.5</span>) <span class="co"># x = 0</span></span>
<span id="cb13-24"><a href="comparative-analyses-of-selection.html#cb13-24" tabindex="-1"></a></span>
<span id="cb13-25"><a href="comparative-analyses-of-selection.html#cb13-25" tabindex="-1"></a>    ax.xaxis.label.set_fontsize(fontsize)</span>
<span id="cb13-26"><a href="comparative-analyses-of-selection.html#cb13-26" tabindex="-1"></a>    ax.yaxis.label.set_fontsize(fontsize)</span>
<span id="cb13-27"><a href="comparative-analyses-of-selection.html#cb13-27" tabindex="-1"></a>    plt.xticks(fontsize<span class="op">=</span>fontsize <span class="op">-</span> <span class="dv">1</span>)</span>
<span id="cb13-28"><a href="comparative-analyses-of-selection.html#cb13-28" tabindex="-1"></a>    plt.yticks(fontsize<span class="op">=</span>fontsize <span class="op">-</span> <span class="dv">1</span>)</span>
<span id="cb13-29"><a href="comparative-analyses-of-selection.html#cb13-29" tabindex="-1"></a></span>
<span id="cb13-30"><a href="comparative-analyses-of-selection.html#cb13-30" tabindex="-1"></a>    plt.tight_layout(pad<span class="op">=</span><span class="fl">1.5</span>)</span>
<span id="cb13-31"><a href="comparative-analyses-of-selection.html#cb13-31" tabindex="-1"></a>    plt.show()</span>
<span id="cb13-32"><a href="comparative-analyses-of-selection.html#cb13-32" tabindex="-1"></a></span>
<span id="cb13-33"><a href="comparative-analyses-of-selection.html#cb13-33" tabindex="-1"></a><span class="kw">def</span> define_colors(data1, data2):</span>
<span id="cb13-34"><a href="comparative-analyses-of-selection.html#cb13-34" tabindex="-1"></a>    color_p19 <span class="op">=</span> <span class="st">&quot;#FFD500&quot;</span></span>
<span id="cb13-35"><a href="comparative-analyses-of-selection.html#cb13-35" tabindex="-1"></a>    color_p22 <span class="op">=</span> <span class="st">&quot;#9867F4&quot;</span></span>
<span id="cb13-36"><a href="comparative-analyses-of-selection.html#cb13-36" tabindex="-1"></a>    color_both <span class="op">=</span> <span class="st">&quot;#242038&quot;</span></span>
<span id="cb13-37"><a href="comparative-analyses-of-selection.html#cb13-37" tabindex="-1"></a>    others <span class="op">=</span> <span class="st">&quot;#A2A2A2&quot;</span></span>
<span id="cb13-38"><a href="comparative-analyses-of-selection.html#cb13-38" tabindex="-1"></a></span>
<span id="cb13-39"><a href="comparative-analyses-of-selection.html#cb13-39" tabindex="-1"></a>    colors <span class="op">=</span> []</span>
<span id="cb13-40"><a href="comparative-analyses-of-selection.html#cb13-40" tabindex="-1"></a></span>
<span id="cb13-41"><a href="comparative-analyses-of-selection.html#cb13-41" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(data1)):</span>
<span id="cb13-42"><a href="comparative-analyses-of-selection.html#cb13-42" tabindex="-1"></a>        <span class="cf">if</span> data1[<span class="st">&quot;signif&quot;</span>][i] <span class="kw">and</span> data2[<span class="st">&quot;signif&quot;</span>][i]:</span>
<span id="cb13-43"><a href="comparative-analyses-of-selection.html#cb13-43" tabindex="-1"></a>            colors.append(color_both)</span>
<span id="cb13-44"><a href="comparative-analyses-of-selection.html#cb13-44" tabindex="-1"></a>        <span class="cf">elif</span> data2[<span class="st">&quot;signif&quot;</span>][i]:</span>
<span id="cb13-45"><a href="comparative-analyses-of-selection.html#cb13-45" tabindex="-1"></a>            colors.append(color_p22)</span>
<span id="cb13-46"><a href="comparative-analyses-of-selection.html#cb13-46" tabindex="-1"></a>        <span class="cf">elif</span> data1[<span class="st">&quot;signif&quot;</span>][i]:</span>
<span id="cb13-47"><a href="comparative-analyses-of-selection.html#cb13-47" tabindex="-1"></a>            colors.append(color_p19)</span>
<span id="cb13-48"><a href="comparative-analyses-of-selection.html#cb13-48" tabindex="-1"></a>        <span class="cf">else</span>:</span>
<span id="cb13-49"><a href="comparative-analyses-of-selection.html#cb13-49" tabindex="-1"></a>            colors.append(others)</span>
<span id="cb13-50"><a href="comparative-analyses-of-selection.html#cb13-50" tabindex="-1"></a></span>
<span id="cb13-51"><a href="comparative-analyses-of-selection.html#cb13-51" tabindex="-1"></a>    <span class="cf">return</span> colors</span></code></pre></div>
</div>
<div id="test-correlation-of-selection" class="section level3 hasAnchor" number="6.2.3">
<h3><span class="header-section-number">6.2.3</span> Test correlation of selection<a href="comparative-analyses-of-selection.html#test-correlation-of-selection" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>After creating a combined dataframe based on the shift between both genes, we proceed to search for correlation, considering all values or only those that were deemed significant (of p19, p22, both, or at least one of them). During this analysis, we observed a potential linear dependence in the upper left quadrant. As a result, we calculated the Pearson correlation coefficient and its corresponding p-value.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb14-1"><a href="comparative-analyses-of-selection.html#cb14-1" tabindex="-1"></a><span class="kw">def</span> combine_genes(data1, data2, shift):</span>
<span id="cb14-2"><a href="comparative-analyses-of-selection.html#cb14-2" tabindex="-1"></a>    combi <span class="op">=</span> pd.DataFrame()</span>
<span id="cb14-3"><a href="comparative-analyses-of-selection.html#cb14-3" tabindex="-1"></a>    combi[<span class="st">&quot;p22 diff&quot;</span>] <span class="op">=</span> <span class="bu">list</span>(data2[<span class="st">&quot;beta - alpha&quot;</span>][shift:<span class="bu">len</span>(data1[<span class="st">&quot;beta - alpha&quot;</span>]) <span class="op">+</span> shift])</span>
<span id="cb14-4"><a href="comparative-analyses-of-selection.html#cb14-4" tabindex="-1"></a>    combi[<span class="st">&quot;p19 diff&quot;</span>] <span class="op">=</span> <span class="bu">list</span>(data1[<span class="st">&quot;beta - alpha&quot;</span>])</span>
<span id="cb14-5"><a href="comparative-analyses-of-selection.html#cb14-5" tabindex="-1"></a></span>
<span id="cb14-6"><a href="comparative-analyses-of-selection.html#cb14-6" tabindex="-1"></a>    combi[<span class="st">&quot;p22_signif&quot;</span>] <span class="op">=</span> <span class="bu">list</span>(data2[<span class="st">&quot;signif&quot;</span>][shift:<span class="bu">len</span>(data1[<span class="st">&quot;beta - alpha&quot;</span>]) <span class="op">+</span> shift])</span>
<span id="cb14-7"><a href="comparative-analyses-of-selection.html#cb14-7" tabindex="-1"></a>    combi[<span class="st">&quot;p19_signif&quot;</span>] <span class="op">=</span> <span class="bu">list</span>(data1[<span class="st">&quot;signif&quot;</span>])</span>
<span id="cb14-8"><a href="comparative-analyses-of-selection.html#cb14-8" tabindex="-1"></a></span>
<span id="cb14-9"><a href="comparative-analyses-of-selection.html#cb14-9" tabindex="-1"></a>    <span class="cf">return</span> combi</span>
<span id="cb14-10"><a href="comparative-analyses-of-selection.html#cb14-10" tabindex="-1"></a></span>
<span id="cb14-11"><a href="comparative-analyses-of-selection.html#cb14-11" tabindex="-1"></a><span class="kw">def</span> search_correlation(combi, num):</span>
<span id="cb14-12"><a href="comparative-analyses-of-selection.html#cb14-12" tabindex="-1"></a>    <span class="cf">if</span> num <span class="op">==</span> <span class="dv">0</span>: <span class="co"># ALL</span></span>
<span id="cb14-13"><a href="comparative-analyses-of-selection.html#cb14-13" tabindex="-1"></a>        corr <span class="op">=</span> correl_total(combi)</span>
<span id="cb14-14"><a href="comparative-analyses-of-selection.html#cb14-14" tabindex="-1"></a>    <span class="cf">elif</span> num <span class="op">==</span> <span class="dv">1</span>:  <span class="co"># p19</span></span>
<span id="cb14-15"><a href="comparative-analyses-of-selection.html#cb14-15" tabindex="-1"></a>        corr <span class="op">=</span> correl_one(combi, num)</span>
<span id="cb14-16"><a href="comparative-analyses-of-selection.html#cb14-16" tabindex="-1"></a>    <span class="cf">elif</span> num <span class="op">==</span> <span class="dv">2</span>:  <span class="co"># p22</span></span>
<span id="cb14-17"><a href="comparative-analyses-of-selection.html#cb14-17" tabindex="-1"></a>        corr <span class="op">=</span> correl_one(combi, num)</span>
<span id="cb14-18"><a href="comparative-analyses-of-selection.html#cb14-18" tabindex="-1"></a>    <span class="cf">elif</span> num <span class="op">==</span> <span class="dv">3</span>: <span class="co"># At least one</span></span>
<span id="cb14-19"><a href="comparative-analyses-of-selection.html#cb14-19" tabindex="-1"></a>        corr <span class="op">=</span> correl_least_one(combi)</span>
<span id="cb14-20"><a href="comparative-analyses-of-selection.html#cb14-20" tabindex="-1"></a>    <span class="cf">else</span>: <span class="co"># Both significative</span></span>
<span id="cb14-21"><a href="comparative-analyses-of-selection.html#cb14-21" tabindex="-1"></a>        corr <span class="op">=</span> correl_both(combi)</span>
<span id="cb14-22"><a href="comparative-analyses-of-selection.html#cb14-22" tabindex="-1"></a>    </span>
<span id="cb14-23"><a href="comparative-analyses-of-selection.html#cb14-23" tabindex="-1"></a>    <span class="cf">return</span> (corr, scipy.stats.fisher_exact(corr))</span>
<span id="cb14-24"><a href="comparative-analyses-of-selection.html#cb14-24" tabindex="-1"></a></span>
<span id="cb14-25"><a href="comparative-analyses-of-selection.html#cb14-25" tabindex="-1"></a><span class="kw">def</span> correl_total(combi):</span>
<span id="cb14-26"><a href="comparative-analyses-of-selection.html#cb14-26" tabindex="-1"></a>    total <span class="op">=</span> [[<span class="bu">len</span>(combi[(combi[<span class="st">&quot;p22 diff&quot;</span>] <span class="op">&gt;</span> <span class="dv">0</span>) <span class="op">&amp;</span> (combi[<span class="st">&quot;p19 diff&quot;</span>] <span class="op">&gt;</span> <span class="dv">0</span>)]),</span>
<span id="cb14-27"><a href="comparative-analyses-of-selection.html#cb14-27" tabindex="-1"></a>    <span class="bu">len</span>(combi[(combi[<span class="st">&quot;p22 diff&quot;</span>] <span class="op">&lt;</span> <span class="dv">0</span>) <span class="op">&amp;</span> (combi[<span class="st">&quot;p19 diff&quot;</span>] <span class="op">&gt;</span> <span class="dv">0</span>)])],</span>
<span id="cb14-28"><a href="comparative-analyses-of-selection.html#cb14-28" tabindex="-1"></a>    [<span class="bu">len</span>(combi[(combi[<span class="st">&quot;p22 diff&quot;</span>] <span class="op">&gt;</span> <span class="dv">0</span>) <span class="op">&amp;</span> (combi[<span class="st">&quot;p19 diff&quot;</span>] <span class="op">&lt;</span> <span class="dv">0</span>)]),</span>
<span id="cb14-29"><a href="comparative-analyses-of-selection.html#cb14-29" tabindex="-1"></a>    <span class="bu">len</span>(combi[(combi[<span class="st">&quot;p22 diff&quot;</span>] <span class="op">&lt;</span> <span class="dv">0</span>) <span class="op">&amp;</span> (combi[<span class="st">&quot;p19 diff&quot;</span>] <span class="op">&lt;</span> <span class="dv">0</span>)])]]</span>
<span id="cb14-30"><a href="comparative-analyses-of-selection.html#cb14-30" tabindex="-1"></a></span>
<span id="cb14-31"><a href="comparative-analyses-of-selection.html#cb14-31" tabindex="-1"></a>    <span class="cf">return</span> total</span>
<span id="cb14-32"><a href="comparative-analyses-of-selection.html#cb14-32" tabindex="-1"></a></span>
<span id="cb14-33"><a href="comparative-analyses-of-selection.html#cb14-33" tabindex="-1"></a><span class="kw">def</span> correl_least_one(combi):</span>
<span id="cb14-34"><a href="comparative-analyses-of-selection.html#cb14-34" tabindex="-1"></a>    one_atleast <span class="op">=</span> [[<span class="bu">len</span>(combi[(((combi[<span class="st">&quot;p22 diff&quot;</span>] <span class="op">&gt;</span> <span class="dv">0</span>) <span class="op">&amp;</span> (combi[<span class="st">&quot;p19 diff&quot;</span>] <span class="op">&gt;</span> <span class="dv">0</span>)) <span class="op">&amp;</span> (combi[<span class="st">&quot;p22_signif&quot;</span>] <span class="op">|</span> combi[<span class="st">&quot;p19_signif&quot;</span>]))]),</span>
<span id="cb14-35"><a href="comparative-analyses-of-selection.html#cb14-35" tabindex="-1"></a>    <span class="bu">len</span>(combi[(((combi[<span class="st">&quot;p22 diff&quot;</span>] <span class="op">&lt;</span> <span class="dv">0</span>) <span class="op">&amp;</span> (combi[<span class="st">&quot;p19 diff&quot;</span>] <span class="op">&gt;</span> <span class="dv">0</span>)) <span class="op">&amp;</span> (combi[<span class="st">&quot;p22_signif&quot;</span>] <span class="op">|</span> combi[<span class="st">&quot;p19_signif&quot;</span>]))])],</span>
<span id="cb14-36"><a href="comparative-analyses-of-selection.html#cb14-36" tabindex="-1"></a>    [<span class="bu">len</span>(combi[(((combi[<span class="st">&quot;p22 diff&quot;</span>] <span class="op">&gt;</span> <span class="dv">0</span>) <span class="op">&amp;</span> (combi[<span class="st">&quot;p19 diff&quot;</span>] <span class="op">&lt;</span> <span class="dv">0</span>)) <span class="op">&amp;</span> (combi[<span class="st">&quot;p22_signif&quot;</span>] <span class="op">|</span> combi[<span class="st">&quot;p19_signif&quot;</span>]))]),</span>
<span id="cb14-37"><a href="comparative-analyses-of-selection.html#cb14-37" tabindex="-1"></a>    <span class="bu">len</span>(combi[(((combi[<span class="st">&quot;p22 diff&quot;</span>] <span class="op">&lt;</span> <span class="dv">0</span>) <span class="op">&amp;</span> (combi[<span class="st">&quot;p19 diff&quot;</span>] <span class="op">&lt;</span> <span class="dv">0</span>)) <span class="op">&amp;</span> (combi[<span class="st">&quot;p22_signif&quot;</span>] <span class="op">|</span> combi[<span class="st">&quot;p19_signif&quot;</span>]))])]]</span>
<span id="cb14-38"><a href="comparative-analyses-of-selection.html#cb14-38" tabindex="-1"></a></span>
<span id="cb14-39"><a href="comparative-analyses-of-selection.html#cb14-39" tabindex="-1"></a>    <span class="cf">return</span> one_atleast</span>
<span id="cb14-40"><a href="comparative-analyses-of-selection.html#cb14-40" tabindex="-1"></a></span>
<span id="cb14-41"><a href="comparative-analyses-of-selection.html#cb14-41" tabindex="-1"></a><span class="kw">def</span> correl_both(combi):</span>
<span id="cb14-42"><a href="comparative-analyses-of-selection.html#cb14-42" tabindex="-1"></a>    both_sig_table <span class="op">=</span> [[<span class="bu">len</span>(combi[(((combi[<span class="st">&quot;p22 diff&quot;</span>] <span class="op">&gt;</span> <span class="dv">0</span>) <span class="op">&amp;</span> (combi[<span class="st">&quot;p19 diff&quot;</span>] <span class="op">&gt;</span> <span class="dv">0</span>)) <span class="op">&amp;</span> (combi[<span class="st">&quot;p22_signif&quot;</span>] <span class="op">&amp;</span> combi[<span class="st">&quot;p19_signif&quot;</span>]))]),</span>
<span id="cb14-43"><a href="comparative-analyses-of-selection.html#cb14-43" tabindex="-1"></a>    <span class="bu">len</span>(combi[(((combi[<span class="st">&quot;p22 diff&quot;</span>] <span class="op">&lt;</span> <span class="dv">0</span>) <span class="op">&amp;</span> (combi[<span class="st">&quot;p19 diff&quot;</span>] <span class="op">&gt;</span> <span class="dv">0</span>)) <span class="op">&amp;</span> (combi[<span class="st">&quot;p22_signif&quot;</span>] <span class="op">|</span> combi[<span class="st">&quot;p19_signif&quot;</span>]))])],</span>
<span id="cb14-44"><a href="comparative-analyses-of-selection.html#cb14-44" tabindex="-1"></a>    [<span class="bu">len</span>(combi[(((combi[<span class="st">&quot;p22 diff&quot;</span>] <span class="op">&gt;</span> <span class="dv">0</span>) <span class="op">&amp;</span> (combi[<span class="st">&quot;p19 diff&quot;</span>] <span class="op">&lt;</span> <span class="dv">0</span>)) <span class="op">&amp;</span> (combi[<span class="st">&quot;p22_signif&quot;</span>] <span class="op">|</span> combi[<span class="st">&quot;p19_signif&quot;</span>]))]),</span>
<span id="cb14-45"><a href="comparative-analyses-of-selection.html#cb14-45" tabindex="-1"></a>    <span class="bu">len</span>(combi[(((combi[<span class="st">&quot;p22 diff&quot;</span>] <span class="op">&lt;</span> <span class="dv">0</span>) <span class="op">&amp;</span> (combi[<span class="st">&quot;p19 diff&quot;</span>] <span class="op">&lt;</span> <span class="dv">0</span>)) <span class="op">&amp;</span> (combi[<span class="st">&quot;p22_signif&quot;</span>] <span class="op">|</span> combi[<span class="st">&quot;p19_signif&quot;</span>]))])]]</span>
<span id="cb14-46"><a href="comparative-analyses-of-selection.html#cb14-46" tabindex="-1"></a></span>
<span id="cb14-47"><a href="comparative-analyses-of-selection.html#cb14-47" tabindex="-1"></a>    <span class="cf">return</span> both_sig_table</span>
<span id="cb14-48"><a href="comparative-analyses-of-selection.html#cb14-48" tabindex="-1"></a></span>
<span id="cb14-49"><a href="comparative-analyses-of-selection.html#cb14-49" tabindex="-1"></a><span class="kw">def</span> correl_one(combi, num):</span>
<span id="cb14-50"><a href="comparative-analyses-of-selection.html#cb14-50" tabindex="-1"></a>    <span class="cf">if</span> num <span class="op">==</span> <span class="dv">1</span>:</span>
<span id="cb14-51"><a href="comparative-analyses-of-selection.html#cb14-51" tabindex="-1"></a>        sig_table <span class="op">=</span> [[<span class="bu">len</span>(combi[(((combi[<span class="st">&quot;p22 diff&quot;</span>] <span class="op">&gt;</span> <span class="dv">0</span>) <span class="op">&amp;</span> (combi[<span class="st">&quot;p19 diff&quot;</span>] <span class="op">&gt;</span> <span class="dv">0</span>)) <span class="op">&amp;</span> (combi[<span class="st">&quot;p19_signif&quot;</span>]))]),</span>
<span id="cb14-52"><a href="comparative-analyses-of-selection.html#cb14-52" tabindex="-1"></a>        <span class="bu">len</span>(combi[(((combi[<span class="st">&quot;p22 diff&quot;</span>] <span class="op">&lt;</span> <span class="dv">0</span>) <span class="op">&amp;</span> (combi[<span class="st">&quot;p19 diff&quot;</span>] <span class="op">&gt;</span> <span class="dv">0</span>)) <span class="op">&amp;</span> (combi[<span class="st">&quot;p19_signif&quot;</span>]))])],</span>
<span id="cb14-53"><a href="comparative-analyses-of-selection.html#cb14-53" tabindex="-1"></a>        [<span class="bu">len</span>(combi[(((combi[<span class="st">&quot;p22 diff&quot;</span>] <span class="op">&gt;</span> <span class="dv">0</span>) <span class="op">&amp;</span> (combi[<span class="st">&quot;p19 diff&quot;</span>] <span class="op">&lt;</span> <span class="dv">0</span>)) <span class="op">&amp;</span> (combi[<span class="st">&quot;p19_signif&quot;</span>]))]),</span>
<span id="cb14-54"><a href="comparative-analyses-of-selection.html#cb14-54" tabindex="-1"></a>        <span class="bu">len</span>(combi[(((combi[<span class="st">&quot;p22 diff&quot;</span>] <span class="op">&lt;</span> <span class="dv">0</span>) <span class="op">&amp;</span> (combi[<span class="st">&quot;p19 diff&quot;</span>] <span class="op">&lt;</span> <span class="dv">0</span>)) <span class="op">&amp;</span> (combi[<span class="st">&quot;p19_signif&quot;</span>]))])]]</span>
<span id="cb14-55"><a href="comparative-analyses-of-selection.html#cb14-55" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb14-56"><a href="comparative-analyses-of-selection.html#cb14-56" tabindex="-1"></a>        sig_table <span class="op">=</span> [[<span class="bu">len</span>(combi[(((combi[<span class="st">&quot;p22 diff&quot;</span>] <span class="op">&gt;</span> <span class="dv">0</span>) <span class="op">&amp;</span> (combi[<span class="st">&quot;p19 diff&quot;</span>] <span class="op">&gt;</span> <span class="dv">0</span>)) <span class="op">&amp;</span> (combi[<span class="st">&quot;p22_signif&quot;</span>]))]),</span>
<span id="cb14-57"><a href="comparative-analyses-of-selection.html#cb14-57" tabindex="-1"></a>        <span class="bu">len</span>(combi[(((combi[<span class="st">&quot;p22 diff&quot;</span>] <span class="op">&lt;</span> <span class="dv">0</span>) <span class="op">&amp;</span> (combi[<span class="st">&quot;p19 diff&quot;</span>] <span class="op">&gt;</span> <span class="dv">0</span>)) <span class="op">&amp;</span> (combi[<span class="st">&quot;p22_signif&quot;</span>]))])],</span>
<span id="cb14-58"><a href="comparative-analyses-of-selection.html#cb14-58" tabindex="-1"></a>        [<span class="bu">len</span>(combi[(((combi[<span class="st">&quot;p22 diff&quot;</span>] <span class="op">&gt;</span> <span class="dv">0</span>) <span class="op">&amp;</span> (combi[<span class="st">&quot;p19 diff&quot;</span>] <span class="op">&lt;</span> <span class="dv">0</span>)) <span class="op">&amp;</span> (combi[<span class="st">&quot;p22_signif&quot;</span>]))]),</span>
<span id="cb14-59"><a href="comparative-analyses-of-selection.html#cb14-59" tabindex="-1"></a>        <span class="bu">len</span>(combi[(((combi[<span class="st">&quot;p22 diff&quot;</span>] <span class="op">&lt;</span> <span class="dv">0</span>) <span class="op">&amp;</span> (combi[<span class="st">&quot;p19 diff&quot;</span>] <span class="op">&lt;</span> <span class="dv">0</span>)) <span class="op">&amp;</span> (combi[<span class="st">&quot;p22_signif&quot;</span>]))])]]</span>
<span id="cb14-60"><a href="comparative-analyses-of-selection.html#cb14-60" tabindex="-1"></a></span>
<span id="cb14-61"><a href="comparative-analyses-of-selection.html#cb14-61" tabindex="-1"></a>    <span class="cf">return</span> sig_table</span>
<span id="cb14-62"><a href="comparative-analyses-of-selection.html#cb14-62" tabindex="-1"></a></span>
<span id="cb14-63"><a href="comparative-analyses-of-selection.html#cb14-63" tabindex="-1"></a><span class="kw">def</span> pearson_upper_left(combi):</span>
<span id="cb14-64"><a href="comparative-analyses-of-selection.html#cb14-64" tabindex="-1"></a>    upper_left <span class="op">=</span> combi[(combi[<span class="st">&quot;p19 diff&quot;</span>] <span class="op">&gt;</span> <span class="dv">0</span>) <span class="op">&amp;</span> (combi[<span class="st">&quot;p22 diff&quot;</span>] <span class="op">&lt;</span> <span class="dv">0</span>)]</span>
<span id="cb14-65"><a href="comparative-analyses-of-selection.html#cb14-65" tabindex="-1"></a>    y <span class="op">=</span> upper_left[<span class="st">&quot;p19 diff&quot;</span>]</span>
<span id="cb14-66"><a href="comparative-analyses-of-selection.html#cb14-66" tabindex="-1"></a>    x <span class="op">=</span> upper_left[<span class="st">&quot;p22 diff&quot;</span>]</span>
<span id="cb14-67"><a href="comparative-analyses-of-selection.html#cb14-67" tabindex="-1"></a>    </span>
<span id="cb14-68"><a href="comparative-analyses-of-selection.html#cb14-68" tabindex="-1"></a>    <span class="cf">return</span> scipy.stats.pearsonr(x, y)</span></code></pre></div>
</div>
<div id="correlation-2d" class="section level3 hasAnchor" number="6.2.4">
<h3><span class="header-section-number">6.2.4</span> Correlation 2D<a href="comparative-analyses-of-selection.html#correlation-2d" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Another point to be investigated is the possible correlation between the secondary structures of both proteins.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb15-1"><a href="comparative-analyses-of-selection.html#cb15-1" tabindex="-1"></a><span class="kw">def</span> corr_struct(struct1, struct2, shift):</span>
<span id="cb15-2"><a href="comparative-analyses-of-selection.html#cb15-2" tabindex="-1"></a>    combinations_3d <span class="op">=</span> pd.DataFrame(<span class="dv">0</span>, columns<span class="op">=</span>[<span class="st">&quot;H&quot;</span>, <span class="st">&quot;S&quot;</span>, <span class="st">&quot;L&quot;</span>],</span>
<span id="cb15-3"><a href="comparative-analyses-of-selection.html#cb15-3" tabindex="-1"></a>                                    index<span class="op">=</span>[<span class="st">&quot;H&quot;</span>, <span class="st">&quot;S&quot;</span>, <span class="st">&quot;L&quot;</span>])</span>
<span id="cb15-4"><a href="comparative-analyses-of-selection.html#cb15-4" tabindex="-1"></a>    <span class="cf">for</span> pos <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(struct1)):</span>
<span id="cb15-5"><a href="comparative-analyses-of-selection.html#cb15-5" tabindex="-1"></a>        combinations_3d[struct2[pos<span class="op">+</span>shift]][struct1[pos]] <span class="op">+=</span> <span class="dv">1</span></span>
<span id="cb15-6"><a href="comparative-analyses-of-selection.html#cb15-6" tabindex="-1"></a>    res <span class="op">=</span> <span class="bu">list</span>(scipy.stats.chi2_contingency(combinations_3d))</span>
<span id="cb15-7"><a href="comparative-analyses-of-selection.html#cb15-7" tabindex="-1"></a>    res.append(combinations_3d)</span>
<span id="cb15-8"><a href="comparative-analyses-of-selection.html#cb15-8" tabindex="-1"></a>    </span>
<span id="cb15-9"><a href="comparative-analyses-of-selection.html#cb15-9" tabindex="-1"></a>    <span class="cf">return</span> res</span></code></pre></div>
<p>The resulting contingency table can be visually represented with colors, indicating the distribution of data. Additionally, this representation can be combined with the expected frequencies, allowing us to create a table displaying the ratio of observed to expected values. This ratio provides informative insights into the data distribution.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb16-1"><a href="comparative-analyses-of-selection.html#cb16-1" tabindex="-1"></a><span class="kw">def</span> plot_dist_3D(combinations_3d, expected<span class="op">=</span><span class="dv">0</span>):</span>
<span id="cb16-2"><a href="comparative-analyses-of-selection.html#cb16-2" tabindex="-1"></a>    <span class="cf">if</span> <span class="bu">isinstance</span>(expected, <span class="bu">int</span>):</span>
<span id="cb16-3"><a href="comparative-analyses-of-selection.html#cb16-3" tabindex="-1"></a>        data <span class="op">=</span> combinations_3d </span>
<span id="cb16-4"><a href="comparative-analyses-of-selection.html#cb16-4" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb16-5"><a href="comparative-analyses-of-selection.html#cb16-5" tabindex="-1"></a>        data <span class="op">=</span> combinations_3d <span class="op">/</span> pd.DataFrame(expected, columns<span class="op">=</span>[<span class="st">&quot;H&quot;</span>, <span class="st">&quot;S&quot;</span>, <span class="st">&quot;L&quot;</span>], index<span class="op">=</span>[<span class="st">&quot;H&quot;</span>, <span class="st">&quot;S&quot;</span>, <span class="st">&quot;L&quot;</span>])</span>
<span id="cb16-6"><a href="comparative-analyses-of-selection.html#cb16-6" tabindex="-1"></a></span>
<span id="cb16-7"><a href="comparative-analyses-of-selection.html#cb16-7" tabindex="-1"></a>    plt.figure(figsize<span class="op">=</span>(<span class="dv">8</span>, <span class="dv">6</span>))</span>
<span id="cb16-8"><a href="comparative-analyses-of-selection.html#cb16-8" tabindex="-1"></a>    sns.heatmap(data, annot<span class="op">=</span><span class="va">True</span>, cmap<span class="op">=</span><span class="st">&#39;YlGnBu&#39;</span>)</span>
<span id="cb16-9"><a href="comparative-analyses-of-selection.html#cb16-9" tabindex="-1"></a>    plt.xlabel(<span class="st">&#39;p22&#39;</span>)</span>
<span id="cb16-10"><a href="comparative-analyses-of-selection.html#cb16-10" tabindex="-1"></a>    plt.ylabel(<span class="st">&#39;p19&#39;</span>)</span>
<span id="cb16-11"><a href="comparative-analyses-of-selection.html#cb16-11" tabindex="-1"></a>    plt.show()</span></code></pre></div>
</div>
<div id="selection-in-2d" class="section level3 hasAnchor" number="6.2.5">
<h3><span class="header-section-number">6.2.5</span> Selection in 2D<a href="comparative-analyses-of-selection.html#selection-in-2d" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>This additional division aims to direct attention to the secondary structures separately. Using the following function, nine contingency tables (3x3) are calculated, considering neutral, positive, and negative selection in both frames. The distinction between these tables lies in the exclusive presence of codons that belong to regions forming helices, strands, or loops (or all).</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb17-1"><a href="comparative-analyses-of-selection.html#cb17-1" tabindex="-1"></a><span class="kw">def</span> selection_structure(p19_struct, p22_struct, df_p19, df_p22, </span>
<span id="cb17-2"><a href="comparative-analyses-of-selection.html#cb17-2" tabindex="-1"></a>                        comparison_letters, shift):</span>
<span id="cb17-3"><a href="comparative-analyses-of-selection.html#cb17-3" tabindex="-1"></a>    macro_results <span class="op">=</span> []</span>
<span id="cb17-4"><a href="comparative-analyses-of-selection.html#cb17-4" tabindex="-1"></a></span>
<span id="cb17-5"><a href="comparative-analyses-of-selection.html#cb17-5" tabindex="-1"></a>    <span class="cf">for</span> letters <span class="kw">in</span> comparison_letters:</span>
<span id="cb17-6"><a href="comparative-analyses-of-selection.html#cb17-6" tabindex="-1"></a>        result <span class="op">=</span> pd.DataFrame(<span class="dv">0</span>, columns <span class="op">=</span> [<span class="st">&quot;Neg.&quot;</span>, <span class="st">&quot;Pos.&quot;</span>, <span class="st">&quot;-&quot;</span>], index <span class="op">=</span> [<span class="st">&quot;Neg.&quot;</span>, <span class="st">&quot;Pos.&quot;</span>, <span class="st">&quot;-&quot;</span>])</span>
<span id="cb17-7"><a href="comparative-analyses-of-selection.html#cb17-7" tabindex="-1"></a>        <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(p19_struct)):</span>
<span id="cb17-8"><a href="comparative-analyses-of-selection.html#cb17-8" tabindex="-1"></a>            <span class="cf">if</span> p19_struct[i] <span class="op">==</span> letters[<span class="dv">0</span>] <span class="kw">and</span> p22_struct[i<span class="op">+</span>shift] <span class="op">==</span> letters[<span class="dv">1</span>]:</span>
<span id="cb17-9"><a href="comparative-analyses-of-selection.html#cb17-9" tabindex="-1"></a>                </span>
<span id="cb17-10"><a href="comparative-analyses-of-selection.html#cb17-10" tabindex="-1"></a>                p19_value <span class="op">=</span> df_p19[<span class="st">&quot;beta - alpha&quot;</span>][i]</span>
<span id="cb17-11"><a href="comparative-analyses-of-selection.html#cb17-11" tabindex="-1"></a>                p22_value <span class="op">=</span> df_p22[<span class="st">&quot;beta - alpha&quot;</span>][i<span class="op">+</span>shift]</span>
<span id="cb17-12"><a href="comparative-analyses-of-selection.html#cb17-12" tabindex="-1"></a></span>
<span id="cb17-13"><a href="comparative-analyses-of-selection.html#cb17-13" tabindex="-1"></a>                <span class="cf">if</span> p19_value <span class="op">==</span> <span class="dv">0</span> <span class="kw">and</span> p22_value <span class="op">==</span> <span class="dv">0</span>:</span>
<span id="cb17-14"><a href="comparative-analyses-of-selection.html#cb17-14" tabindex="-1"></a>                    result[<span class="st">&quot;-&quot;</span>][<span class="st">&quot;-&quot;</span>] <span class="op">+=</span> <span class="dv">1</span></span>
<span id="cb17-15"><a href="comparative-analyses-of-selection.html#cb17-15" tabindex="-1"></a>                <span class="cf">elif</span> p19_value <span class="op">==</span> <span class="dv">0</span> <span class="kw">and</span> p22_value <span class="op">&gt;</span> <span class="dv">0</span>:</span>
<span id="cb17-16"><a href="comparative-analyses-of-selection.html#cb17-16" tabindex="-1"></a>                    result[<span class="st">&quot;Pos.&quot;</span>][<span class="st">&quot;-&quot;</span>] <span class="op">+=</span> <span class="dv">1</span></span>
<span id="cb17-17"><a href="comparative-analyses-of-selection.html#cb17-17" tabindex="-1"></a>                <span class="cf">elif</span> p19_value <span class="op">==</span> <span class="dv">0</span> <span class="kw">and</span> p22_value <span class="op">&lt;</span> <span class="dv">0</span>:</span>
<span id="cb17-18"><a href="comparative-analyses-of-selection.html#cb17-18" tabindex="-1"></a>                    result[<span class="st">&quot;Neg.&quot;</span>][<span class="st">&quot;-&quot;</span>] <span class="op">+=</span> <span class="dv">1</span></span>
<span id="cb17-19"><a href="comparative-analyses-of-selection.html#cb17-19" tabindex="-1"></a>                <span class="cf">elif</span> p19_value <span class="op">&gt;</span> <span class="dv">0</span> <span class="kw">and</span> p22_value <span class="op">==</span> <span class="dv">0</span>:</span>
<span id="cb17-20"><a href="comparative-analyses-of-selection.html#cb17-20" tabindex="-1"></a>                    result[<span class="st">&quot;-&quot;</span>][<span class="st">&quot;Pos.&quot;</span>] <span class="op">+=</span> <span class="dv">1</span></span>
<span id="cb17-21"><a href="comparative-analyses-of-selection.html#cb17-21" tabindex="-1"></a>                <span class="cf">elif</span> p19_value <span class="op">&lt;</span> <span class="dv">0</span> <span class="kw">and</span> p22_value <span class="op">==</span> <span class="dv">0</span>:</span>
<span id="cb17-22"><a href="comparative-analyses-of-selection.html#cb17-22" tabindex="-1"></a>                    result[<span class="st">&quot;-&quot;</span>][<span class="st">&quot;Neg.&quot;</span>] <span class="op">+=</span> <span class="dv">1</span></span>
<span id="cb17-23"><a href="comparative-analyses-of-selection.html#cb17-23" tabindex="-1"></a>                <span class="cf">elif</span> p19_value <span class="op">&gt;</span> <span class="dv">0</span> <span class="kw">and</span> p22_value <span class="op">&gt;</span> <span class="dv">0</span>:</span>
<span id="cb17-24"><a href="comparative-analyses-of-selection.html#cb17-24" tabindex="-1"></a>                    result[<span class="st">&quot;Pos.&quot;</span>][<span class="st">&quot;Pos.&quot;</span>] <span class="op">+=</span> <span class="dv">1</span></span>
<span id="cb17-25"><a href="comparative-analyses-of-selection.html#cb17-25" tabindex="-1"></a>                <span class="cf">elif</span> p19_value <span class="op">&lt;</span> <span class="dv">0</span> <span class="kw">and</span> p22_value <span class="op">&lt;</span> <span class="dv">0</span>:</span>
<span id="cb17-26"><a href="comparative-analyses-of-selection.html#cb17-26" tabindex="-1"></a>                    result[<span class="st">&quot;Neg.&quot;</span>][<span class="st">&quot;Neg.&quot;</span>] <span class="op">+=</span> <span class="dv">1</span></span>
<span id="cb17-27"><a href="comparative-analyses-of-selection.html#cb17-27" tabindex="-1"></a>                <span class="cf">elif</span> p19_value <span class="op">&gt;</span> <span class="dv">0</span> <span class="kw">and</span> p22_value <span class="op">&lt;</span> <span class="dv">0</span>:</span>
<span id="cb17-28"><a href="comparative-analyses-of-selection.html#cb17-28" tabindex="-1"></a>                    result[<span class="st">&quot;Neg.&quot;</span>][<span class="st">&quot;Pos.&quot;</span>] <span class="op">+=</span> <span class="dv">1</span></span>
<span id="cb17-29"><a href="comparative-analyses-of-selection.html#cb17-29" tabindex="-1"></a>                <span class="cf">elif</span> p19_value <span class="op">&lt;</span> <span class="dv">0</span> <span class="kw">and</span> p22_value <span class="op">&gt;</span> <span class="dv">0</span>:</span>
<span id="cb17-30"><a href="comparative-analyses-of-selection.html#cb17-30" tabindex="-1"></a>                    result[<span class="st">&quot;Pos.&quot;</span>][<span class="st">&quot;Neg.&quot;</span>] <span class="op">+=</span> <span class="dv">1</span></span>
<span id="cb17-31"><a href="comparative-analyses-of-selection.html#cb17-31" tabindex="-1"></a>            </span>
<span id="cb17-32"><a href="comparative-analyses-of-selection.html#cb17-32" tabindex="-1"></a>        macro_results.append(result)</span>
<span id="cb17-33"><a href="comparative-analyses-of-selection.html#cb17-33" tabindex="-1"></a></span>
<span id="cb17-34"><a href="comparative-analyses-of-selection.html#cb17-34" tabindex="-1"></a>    <span class="cf">return</span> macro_results</span></code></pre></div>
</div>
<div id="run-2" class="section level3 hasAnchor" number="6.2.6">
<h3><span class="header-section-number">6.2.6</span> Run<a href="comparative-analyses-of-selection.html#run-2" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Notice that, for every position in p19 and p22, we indicate the corresponding secondary structure. H: helix, S:strand and L: loop.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb18-1"><a href="comparative-analyses-of-selection.html#cb18-1" tabindex="-1"></a><span class="kw">def</span> main():</span>
<span id="cb18-2"><a href="comparative-analyses-of-selection.html#cb18-2" tabindex="-1"></a>    path_p19 <span class="op">=</span> <span class="st">&#39;/Users/esmeralda/Documents/TFM/RevTrans/final/nt/alignments/FEL_p19_local/p19_FEL.json&#39;</span></span>
<span id="cb18-3"><a href="comparative-analyses-of-selection.html#cb18-3" tabindex="-1"></a>    path_p22 <span class="op">=</span> <span class="st">&#39;/Users/esmeralda/Documents/TFM/RevTrans/final/nt/alignments/FEL_p22_local/firts.json&#39;</span></span>
<span id="cb18-4"><a href="comparative-analyses-of-selection.html#cb18-4" tabindex="-1"></a>    </span>
<span id="cb18-5"><a href="comparative-analyses-of-selection.html#cb18-5" tabindex="-1"></a>    <span class="co"># Read json files and add information</span></span>
<span id="cb18-6"><a href="comparative-analyses-of-selection.html#cb18-6" tabindex="-1"></a>    significance <span class="op">=</span> <span class="fl">0.1</span></span>
<span id="cb18-7"><a href="comparative-analyses-of-selection.html#cb18-7" tabindex="-1"></a>    df_p19 <span class="op">=</span> add_info(read_json(path_p19), significance)</span>
<span id="cb18-8"><a href="comparative-analyses-of-selection.html#cb18-8" tabindex="-1"></a>    df_p22 <span class="op">=</span> add_info(read_json(path_p22), significance)</span>
<span id="cb18-9"><a href="comparative-analyses-of-selection.html#cb18-9" tabindex="-1"></a></span>
<span id="cb18-10"><a href="comparative-analyses-of-selection.html#cb18-10" tabindex="-1"></a>    <span class="co"># Print significative codons</span></span>
<span id="cb18-11"><a href="comparative-analyses-of-selection.html#cb18-11" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">&quot;Signif. p19:&quot;</span>, significative_codons(df_p19), </span>
<span id="cb18-12"><a href="comparative-analyses-of-selection.html#cb18-12" tabindex="-1"></a>          <span class="st">&quot;Signif. p22:&quot;</span>, significative_codons(df_p22))</span>
<span id="cb18-13"><a href="comparative-analyses-of-selection.html#cb18-13" tabindex="-1"></a></span>
<span id="cb18-14"><a href="comparative-analyses-of-selection.html#cb18-14" tabindex="-1"></a>    <span class="co"># Plot dots +10 and +11 codons</span></span>
<span id="cb18-15"><a href="comparative-analyses-of-selection.html#cb18-15" tabindex="-1"></a>    colors <span class="op">=</span> define_colors(df_p19, df_p22)</span>
<span id="cb18-16"><a href="comparative-analyses-of-selection.html#cb18-16" tabindex="-1"></a>    plot_correlation(df_p19, df_p22, <span class="dv">10</span>, colors)</span>
<span id="cb18-17"><a href="comparative-analyses-of-selection.html#cb18-17" tabindex="-1"></a>    plot_correlation(df_p19, df_p22, <span class="dv">11</span>, colors)</span>
<span id="cb18-18"><a href="comparative-analyses-of-selection.html#cb18-18" tabindex="-1"></a></span>
<span id="cb18-19"><a href="comparative-analyses-of-selection.html#cb18-19" tabindex="-1"></a>    <span class="co"># Combine both sequences with a shift</span></span>
<span id="cb18-20"><a href="comparative-analyses-of-selection.html#cb18-20" tabindex="-1"></a>    p19_p22 <span class="op">=</span> combine_genes(df_p19, df_p22, <span class="dv">10</span>)</span>
<span id="cb18-21"><a href="comparative-analyses-of-selection.html#cb18-21" tabindex="-1"></a></span>
<span id="cb18-22"><a href="comparative-analyses-of-selection.html#cb18-22" tabindex="-1"></a>    <span class="co"># Search correlations</span></span>
<span id="cb18-23"><a href="comparative-analyses-of-selection.html#cb18-23" tabindex="-1"></a>    correlations <span class="op">=</span> [<span class="st">&quot;All&quot;</span>, <span class="st">&quot;Only p19 significant&quot;</span>, <span class="st">&quot;Only p22 significant&quot;</span>, </span>
<span id="cb18-24"><a href="comparative-analyses-of-selection.html#cb18-24" tabindex="-1"></a>                    <span class="st">&quot;At least one significant&quot;</span>, <span class="st">&quot;Both significant&quot;</span>]</span>
<span id="cb18-25"><a href="comparative-analyses-of-selection.html#cb18-25" tabindex="-1"></a>    <span class="cf">for</span> option <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">5</span>):</span>
<span id="cb18-26"><a href="comparative-analyses-of-selection.html#cb18-26" tabindex="-1"></a>        len_hyphen <span class="op">=</span> (<span class="dv">50</span><span class="op">-</span><span class="bu">len</span>(correlations[option]))<span class="op">//</span><span class="dv">2</span></span>
<span id="cb18-27"><a href="comparative-analyses-of-selection.html#cb18-27" tabindex="-1"></a>        <span class="bu">print</span>(<span class="st">&quot;-&quot;</span><span class="op">*</span>len_hyphen, correlations[option], <span class="st">&quot;-&quot;</span><span class="op">*</span>len_hyphen)</span>
<span id="cb18-28"><a href="comparative-analyses-of-selection.html#cb18-28" tabindex="-1"></a>        res <span class="op">=</span> search_correlation(p19_p22, option)</span>
<span id="cb18-29"><a href="comparative-analyses-of-selection.html#cb18-29" tabindex="-1"></a>        <span class="bu">print</span>(<span class="st">&quot;Contingency table:</span><span class="ch">\n</span><span class="st">&quot;</span>, res[<span class="dv">0</span>])</span>
<span id="cb18-30"><a href="comparative-analyses-of-selection.html#cb18-30" tabindex="-1"></a>        <span class="bu">print</span>(<span class="st">&quot;Statistic (Fisher&#39;s exact test):</span><span class="ch">\n</span><span class="st">&quot;</span>, res[<span class="dv">0</span>][<span class="dv">0</span>])    </span>
<span id="cb18-31"><a href="comparative-analyses-of-selection.html#cb18-31" tabindex="-1"></a>        <span class="bu">print</span>(<span class="st">&quot;p value (Fisher&#39;s exact test):</span><span class="ch">\n</span><span class="st">&quot;</span>, res[<span class="dv">0</span>][<span class="dv">1</span>])</span>
<span id="cb18-32"><a href="comparative-analyses-of-selection.html#cb18-32" tabindex="-1"></a></span>
<span id="cb18-33"><a href="comparative-analyses-of-selection.html#cb18-33" tabindex="-1"></a>    <span class="co"># Pearson correlation</span></span>
<span id="cb18-34"><a href="comparative-analyses-of-selection.html#cb18-34" tabindex="-1"></a>    correlation_coef, p_value <span class="op">=</span> pearson_upper_left(p19_p22)</span>
<span id="cb18-35"><a href="comparative-analyses-of-selection.html#cb18-35" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">&quot;Pearson correlation coefficient:&quot;</span>, correlation_coef)</span>
<span id="cb18-36"><a href="comparative-analyses-of-selection.html#cb18-36" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">&quot;p-value:&quot;</span>, p_value)</span>
<span id="cb18-37"><a href="comparative-analyses-of-selection.html#cb18-37" tabindex="-1"></a></span>
<span id="cb18-38"><a href="comparative-analyses-of-selection.html#cb18-38" tabindex="-1"></a>    <span class="co"># Info structure</span></span>
<span id="cb18-39"><a href="comparative-analyses-of-selection.html#cb18-39" tabindex="-1"></a>    order <span class="op">=</span> [<span class="st">&quot;H&quot;</span>, <span class="st">&quot;S&quot;</span>, <span class="st">&quot;L&quot;</span>]</span>
<span id="cb18-40"><a href="comparative-analyses-of-selection.html#cb18-40" tabindex="-1"></a>    p22_struct <span class="op">=</span> [<span class="st">&quot;L&quot;</span>]<span class="op">*</span><span class="dv">16</span> <span class="op">+</span> [<span class="st">&quot;S&quot;</span>]<span class="op">*</span><span class="dv">4</span> <span class="op">+</span> [<span class="st">&quot;L&quot;</span>]<span class="op">*</span><span class="dv">4</span> <span class="op">+</span> [<span class="st">&quot;S&quot;</span>]<span class="op">*</span><span class="dv">4</span> <span class="op">+</span> [<span class="st">&quot;L&quot;</span>]<span class="op">*</span><span class="dv">6</span> <span class="op">+</span> [<span class="st">&quot;S&quot;</span>]<span class="op">*</span><span class="dv">3</span> <span class="op">+\</span></span>
<span id="cb18-41"><a href="comparative-analyses-of-selection.html#cb18-41" tabindex="-1"></a>                 [<span class="st">&quot;L&quot;</span>]<span class="op">*</span><span class="dv">3</span> <span class="op">+</span> [<span class="st">&quot;H&quot;</span>]<span class="op">*</span><span class="dv">10</span> <span class="op">+</span> [<span class="st">&quot;L&quot;</span>]<span class="op">*</span><span class="dv">4</span> <span class="op">+</span> [<span class="st">&quot;S&quot;</span>]<span class="op">*</span><span class="dv">13</span> <span class="op">+</span> [<span class="st">&quot;L&quot;</span>]<span class="op">*</span><span class="dv">10</span> <span class="op">+</span> [<span class="st">&quot;S&quot;</span>]<span class="op">*</span><span class="dv">5</span> <span class="op">+\</span></span>
<span id="cb18-42"><a href="comparative-analyses-of-selection.html#cb18-42" tabindex="-1"></a>                 [<span class="st">&quot;L&quot;</span>]<span class="op">*</span><span class="dv">9</span> <span class="op">+</span> [<span class="st">&quot;S&quot;</span>]<span class="op">*</span><span class="dv">3</span> <span class="op">+</span> [<span class="st">&quot;L&quot;</span>]<span class="op">*</span><span class="dv">9</span> <span class="op">+</span> [<span class="st">&quot;S&quot;</span>]<span class="op">*</span><span class="dv">5</span> <span class="op">+</span> [<span class="st">&quot;L&quot;</span>]<span class="op">*</span><span class="dv">15</span> <span class="op">+</span> [<span class="st">&quot;S&quot;</span>]<span class="op">*</span><span class="dv">5</span> <span class="op">+\</span></span>
<span id="cb18-43"><a href="comparative-analyses-of-selection.html#cb18-43" tabindex="-1"></a>                 [<span class="st">&quot;L&quot;</span>]<span class="op">*</span><span class="dv">10</span> <span class="op">+</span> [<span class="st">&quot;S&quot;</span>]<span class="op">*</span><span class="dv">14</span> <span class="op">+</span> [<span class="st">&quot;L&quot;</span>]<span class="op">*</span><span class="dv">11</span> <span class="op">+</span> [<span class="st">&quot;S&quot;</span>]<span class="op">*</span><span class="dv">4</span> <span class="op">+</span> [<span class="st">&quot;H&quot;</span>]<span class="op">*</span><span class="dv">20</span> <span class="op">+</span> [<span class="st">&quot;L&quot;</span>]<span class="op">*</span><span class="dv">2</span></span>
<span id="cb18-44"><a href="comparative-analyses-of-selection.html#cb18-44" tabindex="-1"></a></span>
<span id="cb18-45"><a href="comparative-analyses-of-selection.html#cb18-45" tabindex="-1"></a>    p19_struct <span class="op">=</span> [<span class="st">&quot;L&quot;</span>]<span class="op">*</span><span class="dv">5</span> <span class="op">+</span> [<span class="st">&quot;H&quot;</span>]<span class="op">*</span><span class="dv">13</span> <span class="op">+</span> [<span class="st">&quot;L&quot;</span>]<span class="op">*</span><span class="dv">19</span> <span class="op">+</span> [<span class="st">&quot;H&quot;</span>]<span class="op">*</span><span class="dv">9</span> <span class="op">+</span> [<span class="st">&quot;L&quot;</span>]<span class="op">*</span><span class="dv">11</span> <span class="op">+</span> [<span class="st">&quot;S&quot;</span>]<span class="op">*</span><span class="dv">8</span> <span class="op">+\</span></span>
<span id="cb18-46"><a href="comparative-analyses-of-selection.html#cb18-46" tabindex="-1"></a>                 [<span class="st">&quot;L&quot;</span>]<span class="op">*</span><span class="dv">2</span> <span class="op">+</span> [<span class="st">&quot;S&quot;</span>]<span class="op">*</span><span class="dv">8</span> <span class="op">+</span> [<span class="st">&quot;L&quot;</span>]<span class="op">*</span><span class="dv">3</span> <span class="op">+</span> [<span class="st">&quot;H&quot;</span>]<span class="op">*</span><span class="dv">10</span> <span class="op">+</span> [<span class="st">&quot;L&quot;</span>]<span class="op">*</span><span class="dv">2</span> <span class="op">+</span> [<span class="st">&quot;H&quot;</span>]<span class="op">*</span><span class="dv">11</span> <span class="op">+\</span></span>
<span id="cb18-47"><a href="comparative-analyses-of-selection.html#cb18-47" tabindex="-1"></a>                 [<span class="st">&quot;L&quot;</span>]<span class="op">*</span><span class="dv">7</span> <span class="op">+</span> [<span class="st">&quot;S&quot;</span>]<span class="op">*</span><span class="dv">8</span> <span class="op">+</span> [<span class="st">&quot;L&quot;</span>]<span class="op">*</span><span class="dv">2</span> <span class="op">+</span> [<span class="st">&quot;S&quot;</span>]<span class="op">*</span><span class="dv">8</span> <span class="op">+</span> [<span class="st">&quot;L&quot;</span>]<span class="op">*</span><span class="dv">2</span> <span class="op">+</span> [<span class="st">&quot;H&quot;</span>]<span class="op">*</span><span class="dv">17</span> <span class="op">+\</span></span>
<span id="cb18-48"><a href="comparative-analyses-of-selection.html#cb18-48" tabindex="-1"></a>                 [<span class="st">&quot;L&quot;</span>]<span class="op">*</span><span class="dv">27</span></span>
<span id="cb18-49"><a href="comparative-analyses-of-selection.html#cb18-49" tabindex="-1"></a>    </span>
<span id="cb18-50"><a href="comparative-analyses-of-selection.html#cb18-50" tabindex="-1"></a>    <span class="co"># (+10) Combinations 3D; Columns: p22 ; Rows: p19 ; PLOT</span></span>
<span id="cb18-51"><a href="comparative-analyses-of-selection.html#cb18-51" tabindex="-1"></a>    chi2_stat, p_value, dof, expected, cont_table<span class="op">=</span> <span class="op">\</span></span>
<span id="cb18-52"><a href="comparative-analyses-of-selection.html#cb18-52" tabindex="-1"></a>                                        corr_struct(p19_struct, p22_struct, <span class="dv">10</span>)</span>
<span id="cb18-53"><a href="comparative-analyses-of-selection.html#cb18-53" tabindex="-1"></a>    expected <span class="op">=</span> pd.DataFrame(expected, index<span class="op">=</span>order, columns<span class="op">=</span>order)</span>
<span id="cb18-54"><a href="comparative-analyses-of-selection.html#cb18-54" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">&quot;+10</span><span class="ch">\n</span><span class="st">Chi-Square Statistic:&quot;</span>, chi2_stat)</span>
<span id="cb18-55"><a href="comparative-analyses-of-selection.html#cb18-55" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">&quot;P-value:&quot;</span>, p_value)</span>
<span id="cb18-56"><a href="comparative-analyses-of-selection.html#cb18-56" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">&quot;Degrees of Freedom:&quot;</span>, dof)</span>
<span id="cb18-57"><a href="comparative-analyses-of-selection.html#cb18-57" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">&quot;Expected Frequencies:</span><span class="ch">\n</span><span class="st">&quot;</span>, expected,</span>
<span id="cb18-58"><a href="comparative-analyses-of-selection.html#cb18-58" tabindex="-1"></a>           <span class="st">&quot;</span><span class="ch">\n</span><span class="st">Counts:</span><span class="ch">\n</span><span class="st">&quot;</span>, cont_table)</span>
<span id="cb18-59"><a href="comparative-analyses-of-selection.html#cb18-59" tabindex="-1"></a>    plot_dist_3D(cont_table)</span>
<span id="cb18-60"><a href="comparative-analyses-of-selection.html#cb18-60" tabindex="-1"></a>    plot_dist_3D(cont_table, expected)</span>
<span id="cb18-61"><a href="comparative-analyses-of-selection.html#cb18-61" tabindex="-1"></a>    </span>
<span id="cb18-62"><a href="comparative-analyses-of-selection.html#cb18-62" tabindex="-1"></a>    <span class="co"># (+11) Combinations 3D; Columns: p22 ; Rows: p19 ; PLOT</span></span>
<span id="cb18-63"><a href="comparative-analyses-of-selection.html#cb18-63" tabindex="-1"></a>    chi2_stat, p_value, dof, expected, cont_table <span class="op">=</span> <span class="op">\</span></span>
<span id="cb18-64"><a href="comparative-analyses-of-selection.html#cb18-64" tabindex="-1"></a>                                        corr_struct(p19_struct, p22_struct, <span class="dv">11</span>)</span>
<span id="cb18-65"><a href="comparative-analyses-of-selection.html#cb18-65" tabindex="-1"></a>    expected <span class="op">=</span> pd.DataFrame(expected, index<span class="op">=</span>order, columns<span class="op">=</span>order)</span>
<span id="cb18-66"><a href="comparative-analyses-of-selection.html#cb18-66" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">&quot;+11</span><span class="ch">\n</span><span class="st">Chi-Square Statistic:&quot;</span>, chi2_stat)</span>
<span id="cb18-67"><a href="comparative-analyses-of-selection.html#cb18-67" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">&quot;P-value:&quot;</span>, p_value)</span>
<span id="cb18-68"><a href="comparative-analyses-of-selection.html#cb18-68" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">&quot;Degrees of Freedom:&quot;</span>, dof)</span>
<span id="cb18-69"><a href="comparative-analyses-of-selection.html#cb18-69" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">&quot;Expected Frequencies:</span><span class="ch">\n</span><span class="st">&quot;</span>, expected,</span>
<span id="cb18-70"><a href="comparative-analyses-of-selection.html#cb18-70" tabindex="-1"></a>          <span class="st">&quot;</span><span class="ch">\n</span><span class="st">Counts:</span><span class="ch">\n</span><span class="st">&quot;</span>, cont_table)</span>
<span id="cb18-71"><a href="comparative-analyses-of-selection.html#cb18-71" tabindex="-1"></a>    plot_dist_3D(cont_table)</span>
<span id="cb18-72"><a href="comparative-analyses-of-selection.html#cb18-72" tabindex="-1"></a>    plot_dist_3D(cont_table, expected)</span>
<span id="cb18-73"><a href="comparative-analyses-of-selection.html#cb18-73" tabindex="-1"></a></span>
<span id="cb18-74"><a href="comparative-analyses-of-selection.html#cb18-74" tabindex="-1"></a>    <span class="co"># Comparison selection structure</span></span>
<span id="cb18-75"><a href="comparative-analyses-of-selection.html#cb18-75" tabindex="-1"></a>    comparison_letters <span class="op">=</span> [(<span class="st">&quot;H&quot;</span>, <span class="st">&quot;H&quot;</span>), </span>
<span id="cb18-76"><a href="comparative-analyses-of-selection.html#cb18-76" tabindex="-1"></a>                      (<span class="st">&quot;H&quot;</span>, <span class="st">&quot;S&quot;</span>),</span>
<span id="cb18-77"><a href="comparative-analyses-of-selection.html#cb18-77" tabindex="-1"></a>                      (<span class="st">&quot;H&quot;</span>, <span class="st">&quot;L&quot;</span>),</span>
<span id="cb18-78"><a href="comparative-analyses-of-selection.html#cb18-78" tabindex="-1"></a>                      (<span class="st">&quot;S&quot;</span>, <span class="st">&quot;S&quot;</span>),</span>
<span id="cb18-79"><a href="comparative-analyses-of-selection.html#cb18-79" tabindex="-1"></a>                      (<span class="st">&quot;S&quot;</span>, <span class="st">&quot;L&quot;</span>),</span>
<span id="cb18-80"><a href="comparative-analyses-of-selection.html#cb18-80" tabindex="-1"></a>                      (<span class="st">&quot;S&quot;</span>, <span class="st">&quot;H&quot;</span>),</span>
<span id="cb18-81"><a href="comparative-analyses-of-selection.html#cb18-81" tabindex="-1"></a>                      (<span class="st">&quot;L&quot;</span>, <span class="st">&quot;L&quot;</span>),</span>
<span id="cb18-82"><a href="comparative-analyses-of-selection.html#cb18-82" tabindex="-1"></a>                      (<span class="st">&quot;L&quot;</span>, <span class="st">&quot;S&quot;</span>),</span>
<span id="cb18-83"><a href="comparative-analyses-of-selection.html#cb18-83" tabindex="-1"></a>                      (<span class="st">&quot;L&quot;</span>, <span class="st">&quot;H&quot;</span>)]</span>
<span id="cb18-84"><a href="comparative-analyses-of-selection.html#cb18-84" tabindex="-1"></a>    </span>
<span id="cb18-85"><a href="comparative-analyses-of-selection.html#cb18-85" tabindex="-1"></a>    list_tables <span class="op">=</span>  selection_structure(p19_struct, p22_struct, df_p19, </span>
<span id="cb18-86"><a href="comparative-analyses-of-selection.html#cb18-86" tabindex="-1"></a>                                                df_p22, comparison_letters, <span class="dv">10</span>)</span>
<span id="cb18-87"><a href="comparative-analyses-of-selection.html#cb18-87" tabindex="-1"></a>    <span class="bu">print</span>(<span class="dv">25</span><span class="op">*</span><span class="st">&quot;-&quot;</span> <span class="op">+</span> <span class="st">&quot;+10&quot;</span> <span class="op">+</span> <span class="dv">25</span><span class="op">*</span><span class="st">&quot;-&quot;</span>)</span>
<span id="cb18-88"><a href="comparative-analyses-of-selection.html#cb18-88" tabindex="-1"></a></span>
<span id="cb18-89"><a href="comparative-analyses-of-selection.html#cb18-89" tabindex="-1"></a>    <span class="cf">for</span> pos, res  <span class="kw">in</span> <span class="bu">enumerate</span>(comparison_letters):</span>
<span id="cb18-90"><a href="comparative-analyses-of-selection.html#cb18-90" tabindex="-1"></a>        <span class="bu">print</span>(<span class="st">&quot;</span><span class="ch">\n</span><span class="st">Comparison:&quot;</span>, res)</span>
<span id="cb18-91"><a href="comparative-analyses-of-selection.html#cb18-91" tabindex="-1"></a>        <span class="bu">print</span>(list_tables[pos])</span>
<span id="cb18-92"><a href="comparative-analyses-of-selection.html#cb18-92" tabindex="-1"></a>    </span>
<span id="cb18-93"><a href="comparative-analyses-of-selection.html#cb18-93" tabindex="-1"></a>    list_tables <span class="op">=</span>  selection_structure(p19_struct, p22_struct, df_p19, </span>
<span id="cb18-94"><a href="comparative-analyses-of-selection.html#cb18-94" tabindex="-1"></a>                                                df_p22, comparison_letters, <span class="dv">11</span>)</span>
<span id="cb18-95"><a href="comparative-analyses-of-selection.html#cb18-95" tabindex="-1"></a>    <span class="bu">print</span>(<span class="dv">25</span><span class="op">*</span><span class="st">&quot;-&quot;</span> <span class="op">+</span> <span class="st">&quot;+11&quot;</span> <span class="op">+</span> <span class="dv">25</span><span class="op">*</span><span class="st">&quot;-&quot;</span>)</span>
<span id="cb18-96"><a href="comparative-analyses-of-selection.html#cb18-96" tabindex="-1"></a>    <span class="cf">for</span> pos, res  <span class="kw">in</span> <span class="bu">enumerate</span>(comparison_letters):</span>
<span id="cb18-97"><a href="comparative-analyses-of-selection.html#cb18-97" tabindex="-1"></a>        <span class="bu">print</span>(<span class="st">&quot;</span><span class="ch">\n</span><span class="st">Comparison:&quot;</span>, res)</span>
<span id="cb18-98"><a href="comparative-analyses-of-selection.html#cb18-98" tabindex="-1"></a>        <span class="bu">print</span>(list_tables[pos])</span>
<span id="cb18-99"><a href="comparative-analyses-of-selection.html#cb18-99" tabindex="-1"></a></span>
<span id="cb18-100"><a href="comparative-analyses-of-selection.html#cb18-100" tabindex="-1"></a><span class="cf">if</span> <span class="va">__name__</span> <span class="op">==</span> <span class="st">&quot;__main__&quot;</span>:</span>
<span id="cb18-101"><a href="comparative-analyses-of-selection.html#cb18-101" tabindex="-1"></a>    main()</span>
<span id="cb18-102"><a href="comparative-analyses-of-selection.html#cb18-102" tabindex="-1"></a>    </span></code></pre></div>

</div>
</div>
</div>
<h3>References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Kosakovsky2005" class="csl-entry">
Kosakovsky Pond, Sergei L., and Simon D. W. Frost. 2005. <span>“Not so Different After All: A Comparison of Methods for Detecting Amino Acid Sites Under Selection.”</span> <em>Molecular Biology and Evolution</em> 22 (5): 1208–22. https://doi.org/<a href="https://doi.org/10.1093/molbev/msi105">https://doi.org/10.1093/molbev/msi105</a>.
</div>
<div id="ref-msv022" class="csl-entry">
Smith, Martin D., Joel O. Wertheim, Steven Weaver, Ben Murrell, Konrad Scheffler, and Sergei L. Kosakovsky Pond. 2015. <span>“<span class="nocase">Less Is More: An Adaptive Branch-Site Random Effects Model for Efficient Detection of Episodic Diversifying Selection</span>.”</span> <em>Molecular Biology and Evolution</em> 32 (5): 1342–53. <a href="https://doi.org/10.1093/molbev/msv022">https://doi.org/10.1093/molbev/msv022</a>.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="compositional-analyses-of-biological-sequences.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="miscellania.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/llobarts/nestedgenes/edit/main/05-selection.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
