<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 6 Comparative analyses of inferred selection | Nested genes code</title>
  <meta name="description" content="This is the description of the code employed in the research." />
  <meta name="generator" content="bookdown 0.34 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 6 Comparative analyses of inferred selection | Nested genes code" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="This is the description of the code employed in the research." />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 6 Comparative analyses of inferred selection | Nested genes code" />
  
  <meta name="twitter:description" content="This is the description of the code employed in the research." />
  

<meta name="author" content="Esmeralda G. Legarda" />


<meta name="date" content="2023-08-07" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="compositional-analyses-of-biological-sequences.html"/>
<link rel="next" href="references.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A Minimal Book Example</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> About</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#abstract"><i class="fa fa-check"></i><b>1.1</b> Abstract</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="e-utilities-on-the-unix-command-line.html"><a href="e-utilities-on-the-unix-command-line.html"><i class="fa fa-check"></i><b>2</b> E-utilities on the Unix command line</a>
<ul>
<li class="chapter" data-level="2.1" data-path="e-utilities-on-the-unix-command-line.html"><a href="e-utilities-on-the-unix-command-line.html#a-section"><i class="fa fa-check"></i><b>2.1</b> A section</a>
<ul>
<li class="chapter" data-level="" data-path="e-utilities-on-the-unix-command-line.html"><a href="e-utilities-on-the-unix-command-line.html#an-unnumbered-section"><i class="fa fa-check"></i>An unnumbered section</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html"><i class="fa fa-check"></i><b>3</b> Distances with clustering in CLANS</a>
<ul>
<li class="chapter" data-level="3.1" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html#read-clans-file"><i class="fa fa-check"></i><b>3.1</b> Read CLANS file</a></li>
<li class="chapter" data-level="3.2" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html#create-a-dataframe"><i class="fa fa-check"></i><b>3.2</b> Create a DataFrame</a></li>
<li class="chapter" data-level="3.3" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html#calculate-centroids-and-distances"><i class="fa fa-check"></i><b>3.3</b> Calculate centroids and distances</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html#plot-the-clustermap"><i class="fa fa-check"></i><b>3.3.1</b> Plot the clustermap</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html#run"><i class="fa fa-check"></i><b>3.4</b> Run</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html"><i class="fa fa-check"></i><b>4</b> Distances with ChimeraX</a>
<ul>
<li class="chapter" data-level="4.1" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#obtaining-rmsd-with-chimerax"><i class="fa fa-check"></i><b>4.1</b> Obtaining RMSD with ChimeraX</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#import-libraries-and-data"><i class="fa fa-check"></i><b>4.1.1</b> Import libraries and data</a></li>
<li class="chapter" data-level="4.1.2" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#matchmaker"><i class="fa fa-check"></i><b>4.1.2</b> MatchMaker</a></li>
<li class="chapter" data-level="4.1.3" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#save-csv-file"><i class="fa fa-check"></i><b>4.1.3</b> Save csv file</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#hiarachical-clustering"><i class="fa fa-check"></i><b>4.2</b> Hiarachical clustering</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#libraries-and-read-csv"><i class="fa fa-check"></i><b>4.2.1</b> Libraries and read csv</a></li>
<li class="chapter" data-level="4.2.2" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#create-matrix"><i class="fa fa-check"></i><b>4.2.2</b> Create matrix</a></li>
<li class="chapter" data-level="4.2.3" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#plot-the-clustermap-1"><i class="fa fa-check"></i><b>4.2.3</b> Plot the clustermap</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="compositional-analyses-of-biological-sequences.html"><a href="compositional-analyses-of-biological-sequences.html"><i class="fa fa-check"></i><b>5</b> Compositional analyses of biological sequences</a>
<ul>
<li class="chapter" data-level="5.1" data-path="compositional-analyses-of-biological-sequences.html"><a href="compositional-analyses-of-biological-sequences.html#footnotes"><i class="fa fa-check"></i><b>5.1</b> Footnotes</a></li>
<li class="chapter" data-level="5.2" data-path="compositional-analyses-of-biological-sequences.html"><a href="compositional-analyses-of-biological-sequences.html#citations"><i class="fa fa-check"></i><b>5.2</b> Citations</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="comparative-analyses-of-inferred-selection.html"><a href="comparative-analyses-of-inferred-selection.html"><i class="fa fa-check"></i><b>6</b> Comparative analyses of inferred selection</a>
<ul>
<li class="chapter" data-level="6.1" data-path="comparative-analyses-of-inferred-selection.html"><a href="comparative-analyses-of-inferred-selection.html#evolution-speed-with-absrel"><i class="fa fa-check"></i><b>6.1</b> Evolution speed with aBSREL</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="comparative-analyses-of-inferred-selection.html"><a href="comparative-analyses-of-inferred-selection.html#libraries-and-json-files"><i class="fa fa-check"></i><b>6.1.1</b> Libraries and json files</a></li>
<li class="chapter" data-level="6.1.2" data-path="comparative-analyses-of-inferred-selection.html"><a href="comparative-analyses-of-inferred-selection.html#omega-and-significance"><i class="fa fa-check"></i><b>6.1.2</b> Omega and significance</a></li>
<li class="chapter" data-level="6.1.3" data-path="comparative-analyses-of-inferred-selection.html"><a href="comparative-analyses-of-inferred-selection.html#relative-asinh-transformation"><i class="fa fa-check"></i><b>6.1.3</b> Relative asinh transformation</a></li>
<li class="chapter" data-level="6.1.4" data-path="comparative-analyses-of-inferred-selection.html"><a href="comparative-analyses-of-inferred-selection.html#utilities"><i class="fa fa-check"></i><b>6.1.4</b> Utilities</a></li>
<li class="chapter" data-level="6.1.5" data-path="comparative-analyses-of-inferred-selection.html"><a href="comparative-analyses-of-inferred-selection.html#boxplots-with-significance"><i class="fa fa-check"></i><b>6.1.5</b> Boxplots with significance</a></li>
<li class="chapter" data-level="6.1.6" data-path="comparative-analyses-of-inferred-selection.html"><a href="comparative-analyses-of-inferred-selection.html#test-selection"><i class="fa fa-check"></i><b>6.1.6</b> Test selection</a></li>
<li class="chapter" data-level="6.1.7" data-path="comparative-analyses-of-inferred-selection.html"><a href="comparative-analyses-of-inferred-selection.html#run-1"><i class="fa fa-check"></i><b>6.1.7</b> Run</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Nested genes code</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="comparative-analyses-of-inferred-selection" class="section level1 hasAnchor" number="6">
<h1><span class="header-section-number">Chapter 6</span> Comparative analyses of inferred selection<a href="comparative-analyses-of-inferred-selection.html#comparative-analyses-of-inferred-selection" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<div id="evolution-speed-with-absrel" class="section level2 hasAnchor" number="6.1">
<h2><span class="header-section-number">6.1</span> Evolution speed with aBSREL<a href="comparative-analyses-of-inferred-selection.html#evolution-speed-with-absrel" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Here we take advantage of the complete sequences of tombusvirus genomes available in NCBI. We use the adaptive branch-site random effects likelihood (aBSREL) method to estimate the omega = dN/dS rates ratio per branch for p19 and p22 as well as for RdRp as a reference. For each branch, estimated omega p19 and omega p22 values are normalized by the corresponding omega RdRp values, so all branches were brought to a common relative scale and asinh-transformed: ω_x^=ω_x⁄ω_RdRp with x  {p19, p22}.</p>
<div id="libraries-and-json-files" class="section level3 hasAnchor" number="6.1.1">
<h3><span class="header-section-number">6.1.1</span> Libraries and json files<a href="comparative-analyses-of-inferred-selection.html#libraries-and-json-files" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb16"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb16-1"><a href="comparative-analyses-of-inferred-selection.html#cb16-1" tabindex="-1"></a><span class="im">import</span> json</span>
<span id="cb16-2"><a href="comparative-analyses-of-inferred-selection.html#cb16-2" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb16-3"><a href="comparative-analyses-of-inferred-selection.html#cb16-3" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb16-4"><a href="comparative-analyses-of-inferred-selection.html#cb16-4" tabindex="-1"></a><span class="im">import</span> scipy.stats</span>
<span id="cb16-5"><a href="comparative-analyses-of-inferred-selection.html#cb16-5" tabindex="-1"></a><span class="im">from</span> statannotations.Annotator <span class="im">import</span> Annotator</span></code></pre></div>
<p>First, we need to read the json files that contain the results of aBSREL with different genes from the same genomes.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb17-1"><a href="comparative-analyses-of-inferred-selection.html#cb17-1" tabindex="-1"></a><span class="co"># Read json files of each gene</span></span>
<span id="cb17-2"><a href="comparative-analyses-of-inferred-selection.html#cb17-2" tabindex="-1"></a>paths <span class="op">=</span> [path_p22, path_p19, path_RdRp, path_cp]</span>
<span id="cb17-3"><a href="comparative-analyses-of-inferred-selection.html#cb17-3" tabindex="-1"></a><span class="cf">for</span> path <span class="kw">in</span> paths:</span>
<span id="cb17-4"><a href="comparative-analyses-of-inferred-selection.html#cb17-4" tabindex="-1"></a>  df <span class="op">=</span> read_branches(path)</span>
<span id="cb17-5"><a href="comparative-analyses-of-inferred-selection.html#cb17-5" tabindex="-1"></a>  dfs.append(add_info(df, significance))</span></code></pre></div>
</div>
<div id="omega-and-significance" class="section level3 hasAnchor" number="6.1.2">
<h3><span class="header-section-number">6.1.2</span> Omega and significance<a href="comparative-analyses-of-inferred-selection.html#omega-and-significance" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Afterwards, information about the significance is added and the overall omega is calculated. Besides, we need to detect what nodes are missing in the output of aBSREL and then, we define them as null and non-significative.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb18-1"><a href="comparative-analyses-of-inferred-selection.html#cb18-1" tabindex="-1"></a><span class="kw">def</span> add_info(data, significance):</span>
<span id="cb18-2"><a href="comparative-analyses-of-inferred-selection.html#cb18-2" tabindex="-1"></a>    </span>
<span id="cb18-3"><a href="comparative-analyses-of-inferred-selection.html#cb18-3" tabindex="-1"></a>    data[<span class="st">&#39;omega_avg&#39;</span>] <span class="op">=</span> (data[<span class="st">&#39;omega1&#39;</span>]<span class="op">*</span>data[<span class="st">&#39;prop1&#39;</span>]) <span class="op">+</span> (data[<span class="st">&#39;omega2&#39;</span>]<span class="op">*</span>data[<span class="st">&#39;prop2&#39;</span>])</span>
<span id="cb18-4"><a href="comparative-analyses-of-inferred-selection.html#cb18-4" tabindex="-1"></a>    data[<span class="st">&quot;signif&quot;</span>] <span class="op">=</span> np.where(data[<span class="st">&#39;pvalue&#39;</span>] <span class="op">&lt;</span> significance, <span class="va">True</span>, <span class="va">False</span>)</span>
<span id="cb18-5"><a href="comparative-analyses-of-inferred-selection.html#cb18-5" tabindex="-1"></a></span>
<span id="cb18-6"><a href="comparative-analyses-of-inferred-selection.html#cb18-6" tabindex="-1"></a>    <span class="cf">return</span> data</span>
<span id="cb18-7"><a href="comparative-analyses-of-inferred-selection.html#cb18-7" tabindex="-1"></a></span>
<span id="cb18-8"><a href="comparative-analyses-of-inferred-selection.html#cb18-8" tabindex="-1"></a><span class="kw">def</span> fill_missing_nodes(data, missing_nodes):</span>
<span id="cb18-9"><a href="comparative-analyses-of-inferred-selection.html#cb18-9" tabindex="-1"></a>    </span>
<span id="cb18-10"><a href="comparative-analyses-of-inferred-selection.html#cb18-10" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> missing_nodes:</span>
<span id="cb18-11"><a href="comparative-analyses-of-inferred-selection.html#cb18-11" tabindex="-1"></a>        new_row <span class="op">=</span> {<span class="st">&quot;name&quot;</span>: i,</span>
<span id="cb18-12"><a href="comparative-analyses-of-inferred-selection.html#cb18-12" tabindex="-1"></a>                   <span class="st">&quot;type&quot;</span> : <span class="st">&quot;Node&quot;</span>,</span>
<span id="cb18-13"><a href="comparative-analyses-of-inferred-selection.html#cb18-13" tabindex="-1"></a>                   <span class="st">&quot;omega1&quot;</span>: <span class="fl">0.0</span>,</span>
<span id="cb18-14"><a href="comparative-analyses-of-inferred-selection.html#cb18-14" tabindex="-1"></a>                   <span class="st">&quot;prop1&quot;</span> : <span class="fl">0.0</span>,</span>
<span id="cb18-15"><a href="comparative-analyses-of-inferred-selection.html#cb18-15" tabindex="-1"></a>                   <span class="st">&quot;omega2&quot;</span>: <span class="fl">0.0</span>,</span>
<span id="cb18-16"><a href="comparative-analyses-of-inferred-selection.html#cb18-16" tabindex="-1"></a>                   <span class="st">&quot;prop2&quot;</span> : <span class="fl">0.0</span>,</span>
<span id="cb18-17"><a href="comparative-analyses-of-inferred-selection.html#cb18-17" tabindex="-1"></a>                   <span class="st">&quot;pvalue&quot;</span> : <span class="fl">1.0</span>,</span>
<span id="cb18-18"><a href="comparative-analyses-of-inferred-selection.html#cb18-18" tabindex="-1"></a>                   <span class="st">&quot;omega_avg&quot;</span> : <span class="fl">0.0</span>,</span>
<span id="cb18-19"><a href="comparative-analyses-of-inferred-selection.html#cb18-19" tabindex="-1"></a>                   <span class="st">&quot;Signif.&quot;</span> : <span class="va">False</span>}</span>
<span id="cb18-20"><a href="comparative-analyses-of-inferred-selection.html#cb18-20" tabindex="-1"></a>    </span>
<span id="cb18-21"><a href="comparative-analyses-of-inferred-selection.html#cb18-21" tabindex="-1"></a>        data <span class="op">=</span> data.append(new_row, ignore_index<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb18-22"><a href="comparative-analyses-of-inferred-selection.html#cb18-22" tabindex="-1"></a></span>
<span id="cb18-23"><a href="comparative-analyses-of-inferred-selection.html#cb18-23" tabindex="-1"></a>    <span class="cf">return</span> data</span></code></pre></div>
</div>
<div id="relative-asinh-transformation" class="section level3 hasAnchor" number="6.1.3">
<h3><span class="header-section-number">6.1.3</span> Relative asinh transformation<a href="comparative-analyses-of-inferred-selection.html#relative-asinh-transformation" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Next step consists of calculating the relative measure for comparing the different genes, based on an asinh transformation.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb19-1"><a href="comparative-analyses-of-inferred-selection.html#cb19-1" tabindex="-1"></a><span class="kw">def</span> relativize(genes, gene_ref):</span>
<span id="cb19-2"><a href="comparative-analyses-of-inferred-selection.html#cb19-2" tabindex="-1"></a>    </span>
<span id="cb19-3"><a href="comparative-analyses-of-inferred-selection.html#cb19-3" tabindex="-1"></a>    results <span class="op">=</span> []</span>
<span id="cb19-4"><a href="comparative-analyses-of-inferred-selection.html#cb19-4" tabindex="-1"></a>    type_tree <span class="op">=</span> []</span>
<span id="cb19-5"><a href="comparative-analyses-of-inferred-selection.html#cb19-5" tabindex="-1"></a></span>
<span id="cb19-6"><a href="comparative-analyses-of-inferred-selection.html#cb19-6" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">3</span>):</span>
<span id="cb19-7"><a href="comparative-analyses-of-inferred-selection.html#cb19-7" tabindex="-1"></a>        results_gene <span class="op">=</span> []</span>
<span id="cb19-8"><a href="comparative-analyses-of-inferred-selection.html#cb19-8" tabindex="-1"></a>        <span class="cf">for</span> name <span class="kw">in</span> gene_ref[<span class="st">&quot;name&quot;</span>]:</span>
<span id="cb19-9"><a href="comparative-analyses-of-inferred-selection.html#cb19-9" tabindex="-1"></a></span>
<span id="cb19-10"><a href="comparative-analyses-of-inferred-selection.html#cb19-10" tabindex="-1"></a>            omega_x <span class="op">=</span> <span class="bu">float</span>(genes[i][genes[i][<span class="st">&quot;name&quot;</span>] <span class="op">==</span> name][<span class="st">&quot;omega_avg&quot;</span>])</span>
<span id="cb19-11"><a href="comparative-analyses-of-inferred-selection.html#cb19-11" tabindex="-1"></a>            omega_std <span class="op">=</span> <span class="bu">float</span>(gene_ref[gene_ref[<span class="st">&quot;name&quot;</span>] <span class="op">==</span> name][<span class="st">&quot;omega_avg&quot;</span>])</span>
<span id="cb19-12"><a href="comparative-analyses-of-inferred-selection.html#cb19-12" tabindex="-1"></a></span>
<span id="cb19-13"><a href="comparative-analyses-of-inferred-selection.html#cb19-13" tabindex="-1"></a>            results_gene.append(np.log(omega_x <span class="op">+</span> (np.sqrt(omega_x<span class="op">**</span><span class="dv">2</span> <span class="op">+</span> <span class="dv">1</span>))<span class="op">/</span>(omega_std <span class="op">+</span> (np.sqrt(omega_std<span class="op">**</span><span class="dv">2</span> <span class="op">+</span> <span class="dv">1</span>)))))</span>
<span id="cb19-14"><a href="comparative-analyses-of-inferred-selection.html#cb19-14" tabindex="-1"></a>        </span>
<span id="cb19-15"><a href="comparative-analyses-of-inferred-selection.html#cb19-15" tabindex="-1"></a>        results.append(results_gene)</span>
<span id="cb19-16"><a href="comparative-analyses-of-inferred-selection.html#cb19-16" tabindex="-1"></a></span>
<span id="cb19-17"><a href="comparative-analyses-of-inferred-selection.html#cb19-17" tabindex="-1"></a>    <span class="cf">return</span> results</span></code></pre></div>
</div>
<div id="utilities" class="section level3 hasAnchor" number="6.1.4">
<h3><span class="header-section-number">6.1.4</span> Utilities<a href="comparative-analyses-of-inferred-selection.html#utilities" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Some other functions are employed in order to change the structure of data for its visualization and other to calculate p values with Wilcoxon-Mann-Whitney test.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb20-1"><a href="comparative-analyses-of-inferred-selection.html#cb20-1" tabindex="-1"></a><span class="kw">def</span> divide_branches(results, gene_ref):</span>
<span id="cb20-2"><a href="comparative-analyses-of-inferred-selection.html#cb20-2" tabindex="-1"></a>    pos_tips <span class="op">=</span> gene_ref.index[gene_ref[<span class="st">&#39;type&#39;</span>] <span class="op">==</span> <span class="st">&quot;Tip&quot;</span>].tolist()</span>
<span id="cb20-3"><a href="comparative-analyses-of-inferred-selection.html#cb20-3" tabindex="-1"></a>    tips <span class="op">=</span> []</span>
<span id="cb20-4"><a href="comparative-analyses-of-inferred-selection.html#cb20-4" tabindex="-1"></a>    nodes <span class="op">=</span> []</span>
<span id="cb20-5"><a href="comparative-analyses-of-inferred-selection.html#cb20-5" tabindex="-1"></a></span>
<span id="cb20-6"><a href="comparative-analyses-of-inferred-selection.html#cb20-6" tabindex="-1"></a>    <span class="cf">for</span> gene <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">3</span>):</span>
<span id="cb20-7"><a href="comparative-analyses-of-inferred-selection.html#cb20-7" tabindex="-1"></a>        tip_selection <span class="op">=</span> []</span>
<span id="cb20-8"><a href="comparative-analyses-of-inferred-selection.html#cb20-8" tabindex="-1"></a>        node_selection <span class="op">=</span> []</span>
<span id="cb20-9"><a href="comparative-analyses-of-inferred-selection.html#cb20-9" tabindex="-1"></a>        <span class="cf">for</span> i, value <span class="kw">in</span> <span class="bu">enumerate</span>(results[gene]):</span>
<span id="cb20-10"><a href="comparative-analyses-of-inferred-selection.html#cb20-10" tabindex="-1"></a>            <span class="cf">if</span> i <span class="kw">in</span> pos_tips:</span>
<span id="cb20-11"><a href="comparative-analyses-of-inferred-selection.html#cb20-11" tabindex="-1"></a>                tip_selection.append(value)</span>
<span id="cb20-12"><a href="comparative-analyses-of-inferred-selection.html#cb20-12" tabindex="-1"></a>            <span class="cf">else</span>:</span>
<span id="cb20-13"><a href="comparative-analyses-of-inferred-selection.html#cb20-13" tabindex="-1"></a>                node_selection.append(value)</span>
<span id="cb20-14"><a href="comparative-analyses-of-inferred-selection.html#cb20-14" tabindex="-1"></a>                </span>
<span id="cb20-15"><a href="comparative-analyses-of-inferred-selection.html#cb20-15" tabindex="-1"></a>        tips.append(tip_selection)</span>
<span id="cb20-16"><a href="comparative-analyses-of-inferred-selection.html#cb20-16" tabindex="-1"></a>        nodes.append(node_selection)</span>
<span id="cb20-17"><a href="comparative-analyses-of-inferred-selection.html#cb20-17" tabindex="-1"></a>    </span>
<span id="cb20-18"><a href="comparative-analyses-of-inferred-selection.html#cb20-18" tabindex="-1"></a>    <span class="cf">return</span> (tips, nodes)</span>
<span id="cb20-19"><a href="comparative-analyses-of-inferred-selection.html#cb20-19" tabindex="-1"></a></span>
<span id="cb20-20"><a href="comparative-analyses-of-inferred-selection.html#cb20-20" tabindex="-1"></a><span class="kw">def</span> calc_pvalues(chunk):</span>
<span id="cb20-21"><a href="comparative-analyses-of-inferred-selection.html#cb20-21" tabindex="-1"></a>    pvalues_tips <span class="op">=</span> [scipy.stats.wilcoxon(chunk[<span class="dv">0</span>], chunk[<span class="dv">1</span>], alternative<span class="op">=</span><span class="st">&quot;two-sided&quot;</span>).pvalue, <span class="co"># p22 vs p19</span></span>
<span id="cb20-22"><a href="comparative-analyses-of-inferred-selection.html#cb20-22" tabindex="-1"></a>    scipy.stats.wilcoxon(chunk[<span class="dv">0</span>], chunk[<span class="dv">2</span>], alternative<span class="op">=</span><span class="st">&quot;two-sided&quot;</span>).pvalue, <span class="co"># p22 vs CP</span></span>
<span id="cb20-23"><a href="comparative-analyses-of-inferred-selection.html#cb20-23" tabindex="-1"></a>    scipy.stats.wilcoxon(chunk[<span class="dv">1</span>], chunk[<span class="dv">2</span>], alternative<span class="op">=</span><span class="st">&quot;two-sided&quot;</span>).pvalue]   <span class="co"># p19 vs CP</span></span>
<span id="cb20-24"><a href="comparative-analyses-of-inferred-selection.html#cb20-24" tabindex="-1"></a></span>
<span id="cb20-25"><a href="comparative-analyses-of-inferred-selection.html#cb20-25" tabindex="-1"></a>    <span class="cf">return</span> pvalues_tips</span>
<span id="cb20-26"><a href="comparative-analyses-of-inferred-selection.html#cb20-26" tabindex="-1"></a></span>
<span id="cb20-27"><a href="comparative-analyses-of-inferred-selection.html#cb20-27" tabindex="-1"></a><span class="kw">def</span> format_pvalues(pvalues_1to1):</span>
<span id="cb20-28"><a href="comparative-analyses-of-inferred-selection.html#cb20-28" tabindex="-1"></a>    <span class="cf">return</span> [[<span class="ss">f&#39;p=</span><span class="sc">{</span>pvalue<span class="sc">:.3e}</span><span class="ss">&#39;</span> <span class="cf">for</span> pvalue <span class="kw">in</span> pvalues_1to1[<span class="dv">0</span>]], </span>
<span id="cb20-29"><a href="comparative-analyses-of-inferred-selection.html#cb20-29" tabindex="-1"></a>            [<span class="ss">f&#39;p=</span><span class="sc">{</span>pvalue<span class="sc">:.3e}</span><span class="ss">&#39;</span> <span class="cf">for</span> pvalue <span class="kw">in</span> pvalues_1to1[<span class="dv">1</span>]], </span>
<span id="cb20-30"><a href="comparative-analyses-of-inferred-selection.html#cb20-30" tabindex="-1"></a>            [<span class="ss">f&#39;p=</span><span class="sc">{</span>pvalue<span class="sc">:.3e}</span><span class="ss">&#39;</span> <span class="cf">for</span> pvalue <span class="kw">in</span> pvalues_1to1[<span class="dv">2</span>]]]</span>
<span id="cb20-31"><a href="comparative-analyses-of-inferred-selection.html#cb20-31" tabindex="-1"></a></span>
<span id="cb20-32"><a href="comparative-analyses-of-inferred-selection.html#cb20-32" tabindex="-1"></a><span class="kw">def</span> info_to_dataframe(data):</span>
<span id="cb20-33"><a href="comparative-analyses-of-inferred-selection.html#cb20-33" tabindex="-1"></a>    <span class="cf">return</span> pd.DataFrame.from_dict({<span class="st">&quot;omega&quot;</span>: data[<span class="dv">0</span>] <span class="op">+</span> data[<span class="dv">1</span>] <span class="op">+</span> data[<span class="dv">2</span>], </span>
<span id="cb20-34"><a href="comparative-analyses-of-inferred-selection.html#cb20-34" tabindex="-1"></a>                            <span class="st">&quot;Gene&quot;</span> : <span class="bu">len</span>(data[<span class="dv">0</span>])<span class="op">*</span>[<span class="st">&#39;p22&#39;</span>] <span class="op">+</span> <span class="bu">len</span>(data[<span class="dv">0</span>])<span class="op">*</span>[<span class="st">&#39;p19&#39;</span>] <span class="op">+</span> <span class="bu">len</span>(data[<span class="dv">0</span>])<span class="op">*</span>[<span class="st">&#39;CP&#39;</span>]})</span></code></pre></div>
</div>
<div id="boxplots-with-significance" class="section level3 hasAnchor" number="6.1.5">
<h3><span class="header-section-number">6.1.5</span> Boxplots with significance<a href="comparative-analyses-of-inferred-selection.html#boxplots-with-significance" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Two different plots can be generated with the following plots. The first one includes all, only nodes (or branches) and only tips, while the second is only focused on the whole tree. Both includes p-value annotation.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb21-1"><a href="comparative-analyses-of-inferred-selection.html#cb21-1" tabindex="-1"></a><span class="kw">def</span> plot_diff_types(data, pvalues_1to1, formatted_pvalues):</span>
<span id="cb21-2"><a href="comparative-analyses-of-inferred-selection.html#cb21-2" tabindex="-1"></a></span>
<span id="cb21-3"><a href="comparative-analyses-of-inferred-selection.html#cb21-3" tabindex="-1"></a>    colors <span class="op">=</span> sns.color_palette(<span class="st">&quot;colorblind&quot;</span>, <span class="dv">5</span>)</span>
<span id="cb21-4"><a href="comparative-analyses-of-inferred-selection.html#cb21-4" tabindex="-1"></a>    fig, ax <span class="op">=</span> plt.subplot_mosaic(<span class="st">&quot;ABC&quot;</span>, figsize <span class="op">=</span> (<span class="dv">20</span>,<span class="dv">8</span>))</span>
<span id="cb21-5"><a href="comparative-analyses-of-inferred-selection.html#cb21-5" tabindex="-1"></a>    axes <span class="op">=</span> [ax[<span class="st">&quot;A&quot;</span>], ax[<span class="st">&quot;B&quot;</span>], ax[<span class="st">&quot;C&quot;</span>]]</span>
<span id="cb21-6"><a href="comparative-analyses-of-inferred-selection.html#cb21-6" tabindex="-1"></a>    annot <span class="op">=</span> []</span>
<span id="cb21-7"><a href="comparative-analyses-of-inferred-selection.html#cb21-7" tabindex="-1"></a>    pairs <span class="op">=</span> [(<span class="st">&quot;p22&quot;</span>, <span class="st">&quot;p19&quot;</span>),(<span class="st">&quot;p22&quot;</span>, <span class="st">&quot;CP&quot;</span>), (<span class="st">&quot;p19&quot;</span>, <span class="st">&quot;CP&quot;</span>)]</span>
<span id="cb21-8"><a href="comparative-analyses-of-inferred-selection.html#cb21-8" tabindex="-1"></a>    titles <span class="op">=</span> [<span class="st">&quot;TIPS&quot;</span>, <span class="st">&quot;BRANCHES&quot;</span>, <span class="st">&quot;TREE&quot;</span>]</span>
<span id="cb21-9"><a href="comparative-analyses-of-inferred-selection.html#cb21-9" tabindex="-1"></a></span>
<span id="cb21-10"><a href="comparative-analyses-of-inferred-selection.html#cb21-10" tabindex="-1"></a>    <span class="cf">for</span> part <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(data)):</span>
<span id="cb21-11"><a href="comparative-analyses-of-inferred-selection.html#cb21-11" tabindex="-1"></a>        <span class="cf">with</span> sns.plotting_context(<span class="st">&#39;notebook&#39;</span>, font_scale <span class="op">=</span> <span class="fl">1.2</span>):</span>
<span id="cb21-12"><a href="comparative-analyses-of-inferred-selection.html#cb21-12" tabindex="-1"></a>                        sns.swarmplot(data<span class="op">=</span>data[part], x<span class="op">=</span><span class="st">&quot;Gene&quot;</span>, y<span class="op">=</span><span class="st">&quot;omega&quot;</span>, ax<span class="op">=</span>axes[part],palette<span class="op">=</span>colors, zorder<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb21-13"><a href="comparative-analyses-of-inferred-selection.html#cb21-13" tabindex="-1"></a>                        sns.boxplot(data<span class="op">=</span>data[part], x<span class="op">=</span><span class="st">&quot;Gene&quot;</span>, y<span class="op">=</span><span class="st">&quot;omega&quot;</span>, dodge<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>axes[part], boxprops<span class="op">=</span>{<span class="st">&#39;facecolor&#39;</span>:<span class="st">&#39;None&#39;</span>})</span>
<span id="cb21-14"><a href="comparative-analyses-of-inferred-selection.html#cb21-14" tabindex="-1"></a>                        axes[part].set_title(titles[part])</span>
<span id="cb21-15"><a href="comparative-analyses-of-inferred-selection.html#cb21-15" tabindex="-1"></a>                        plt.axhline(<span class="dv">1</span>)</span>
<span id="cb21-16"><a href="comparative-analyses-of-inferred-selection.html#cb21-16" tabindex="-1"></a></span>
<span id="cb21-17"><a href="comparative-analyses-of-inferred-selection.html#cb21-17" tabindex="-1"></a>                        <span class="co"># Add annotations</span></span>
<span id="cb21-18"><a href="comparative-analyses-of-inferred-selection.html#cb21-18" tabindex="-1"></a>                        annot.append(Annotator(ax<span class="op">=</span>axes[part], pairs <span class="op">=</span> pairs, data<span class="op">=</span>data[part], x<span class="op">=</span><span class="st">&quot;Gene&quot;</span>, y<span class="op">=</span><span class="st">&quot;omega&quot;</span>)) </span>
<span id="cb21-19"><a href="comparative-analyses-of-inferred-selection.html#cb21-19" tabindex="-1"></a>                        annot[part].set_custom_annotations(formatted_pvalues[part])</span>
<span id="cb21-20"><a href="comparative-analyses-of-inferred-selection.html#cb21-20" tabindex="-1"></a>                        annot[part].configure(test_short_name<span class="op">=</span><span class="st">&quot;MWW&quot;</span>)  <span class="co"># text_format is still simple</span></span>
<span id="cb21-21"><a href="comparative-analyses-of-inferred-selection.html#cb21-21" tabindex="-1"></a>                        annot[part].set_pvalues_and_annotate(pvalues_1to1[part])</span>
<span id="cb21-22"><a href="comparative-analyses-of-inferred-selection.html#cb21-22" tabindex="-1"></a>    </span>
<span id="cb21-23"><a href="comparative-analyses-of-inferred-selection.html#cb21-23" tabindex="-1"></a>    plt.tight_layout(pad<span class="op">=</span><span class="fl">1.5</span>)</span>
<span id="cb21-24"><a href="comparative-analyses-of-inferred-selection.html#cb21-24" tabindex="-1"></a>    plt.show()</span>
<span id="cb21-25"><a href="comparative-analyses-of-inferred-selection.html#cb21-25" tabindex="-1"></a></span>
<span id="cb21-26"><a href="comparative-analyses-of-inferred-selection.html#cb21-26" tabindex="-1"></a><span class="kw">def</span> plot_one_type(data, pvalues_1to1, formatted_pvalues, one):</span>
<span id="cb21-27"><a href="comparative-analyses-of-inferred-selection.html#cb21-27" tabindex="-1"></a>    colors <span class="op">=</span> [<span class="st">&quot;#9867F4&quot;</span>, <span class="st">&quot;#FFD500&quot;</span>]</span>
<span id="cb21-28"><a href="comparative-analyses-of-inferred-selection.html#cb21-28" tabindex="-1"></a>    plt.style.use(<span class="st">&quot;default&quot;</span>)</span>
<span id="cb21-29"><a href="comparative-analyses-of-inferred-selection.html#cb21-29" tabindex="-1"></a>    fig, ax <span class="op">=</span> plt.subplots(figsize <span class="op">=</span> (<span class="dv">6</span>,<span class="dv">12</span>))</span>
<span id="cb21-30"><a href="comparative-analyses-of-inferred-selection.html#cb21-30" tabindex="-1"></a>    part <span class="op">=</span> <span class="dv">2</span> <span class="co"># 2: tree</span></span>
<span id="cb21-31"><a href="comparative-analyses-of-inferred-selection.html#cb21-31" tabindex="-1"></a>    annot <span class="op">=</span> []</span>
<span id="cb21-32"><a href="comparative-analyses-of-inferred-selection.html#cb21-32" tabindex="-1"></a>    pairs <span class="op">=</span> [(<span class="st">&quot;p22&quot;</span>, <span class="st">&quot;p19&quot;</span>)]</span>
<span id="cb21-33"><a href="comparative-analyses-of-inferred-selection.html#cb21-33" tabindex="-1"></a>    size_labels <span class="op">=</span> <span class="dv">24</span></span>
<span id="cb21-34"><a href="comparative-analyses-of-inferred-selection.html#cb21-34" tabindex="-1"></a></span>
<span id="cb21-35"><a href="comparative-analyses-of-inferred-selection.html#cb21-35" tabindex="-1"></a>    <span class="cf">with</span> sns.plotting_context(<span class="st">&#39;notebook&#39;</span>, font_scale <span class="op">=</span> <span class="fl">1.2</span>):</span>
<span id="cb21-36"><a href="comparative-analyses-of-inferred-selection.html#cb21-36" tabindex="-1"></a>        sns.<span class="bu">set</span>(font_scale<span class="op">=</span><span class="dv">4</span>)</span>
<span id="cb21-37"><a href="comparative-analyses-of-inferred-selection.html#cb21-37" tabindex="-1"></a></span>
<span id="cb21-38"><a href="comparative-analyses-of-inferred-selection.html#cb21-38" tabindex="-1"></a>        sns.swarmplot(data<span class="op">=</span>data[part][data[part][<span class="st">&quot;Gene&quot;</span>] <span class="op">!=</span> <span class="st">&quot;CP&quot;</span>], x<span class="op">=</span><span class="st">&quot;Gene&quot;</span>, y<span class="op">=</span><span class="st">&quot;omega&quot;</span>, palette<span class="op">=</span>colors, zorder<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb21-39"><a href="comparative-analyses-of-inferred-selection.html#cb21-39" tabindex="-1"></a>        sns.boxplot(data<span class="op">=</span>data[part][data[part][<span class="st">&quot;Gene&quot;</span>] <span class="op">!=</span> <span class="st">&quot;CP&quot;</span>], x<span class="op">=</span><span class="st">&quot;Gene&quot;</span>, y<span class="op">=</span><span class="st">&quot;omega&quot;</span>, dodge<span class="op">=</span><span class="va">True</span>, boxprops<span class="op">=</span>{<span class="st">&#39;facecolor&#39;</span>:<span class="st">&#39;None&#39;</span>})</span>
<span id="cb21-40"><a href="comparative-analyses-of-inferred-selection.html#cb21-40" tabindex="-1"></a>        plt.axhline(<span class="dv">0</span>, color<span class="op">=</span><span class="st">&#39;grey&#39;</span>, linestyle<span class="op">=</span><span class="st">&#39;dotted&#39;</span>)</span>
<span id="cb21-41"><a href="comparative-analyses-of-inferred-selection.html#cb21-41" tabindex="-1"></a>        ax.set_xlabel(<span class="st">&quot;Gene sequence&quot;</span>, fontsize <span class="op">=</span> size_labels)</span>
<span id="cb21-42"><a href="comparative-analyses-of-inferred-selection.html#cb21-42" tabindex="-1"></a>        ax.set_ylabel(<span class="st">&quot;Transformed omega ratio&quot;</span>, fontsize <span class="op">=</span> size_labels)</span>
<span id="cb21-43"><a href="comparative-analyses-of-inferred-selection.html#cb21-43" tabindex="-1"></a>        ax.yaxis.set_tick_params(labelsize <span class="op">=</span> size_labels <span class="op">-</span> <span class="dv">1</span>)</span>
<span id="cb21-44"><a href="comparative-analyses-of-inferred-selection.html#cb21-44" tabindex="-1"></a>        ax.xaxis.set_tick_params(labelsize <span class="op">=</span> size_labels <span class="op">-</span> <span class="dv">1</span>)</span>
<span id="cb21-45"><a href="comparative-analyses-of-inferred-selection.html#cb21-45" tabindex="-1"></a></span>
<span id="cb21-46"><a href="comparative-analyses-of-inferred-selection.html#cb21-46" tabindex="-1"></a>        <span class="co"># Add annotations</span></span>
<span id="cb21-47"><a href="comparative-analyses-of-inferred-selection.html#cb21-47" tabindex="-1"></a>        annot <span class="op">=</span> Annotator(ax<span class="op">=</span>ax, pairs <span class="op">=</span> pairs, data<span class="op">=</span>data[part][data[part][<span class="st">&quot;Gene&quot;</span>] <span class="op">!=</span> <span class="st">&quot;CP&quot;</span>], x<span class="op">=</span><span class="st">&quot;Gene&quot;</span>, y<span class="op">=</span><span class="st">&quot;omega&quot;</span>)</span>
<span id="cb21-48"><a href="comparative-analyses-of-inferred-selection.html#cb21-48" tabindex="-1"></a>        annot.set_custom_annotations([formatted_pvalues[part][one]])</span>
<span id="cb21-49"><a href="comparative-analyses-of-inferred-selection.html#cb21-49" tabindex="-1"></a>        annot.configure(test_short_name<span class="op">=</span><span class="st">&quot;MWW&quot;</span>)  <span class="co"># text_format is still simple</span></span>
<span id="cb21-50"><a href="comparative-analyses-of-inferred-selection.html#cb21-50" tabindex="-1"></a>        annot.set_pvalues_and_annotate([pvalues_1to1[part][one]])</span>
<span id="cb21-51"><a href="comparative-analyses-of-inferred-selection.html#cb21-51" tabindex="-1"></a></span>
<span id="cb21-52"><a href="comparative-analyses-of-inferred-selection.html#cb21-52" tabindex="-1"></a>    plt.tight_layout(pad<span class="op">=</span><span class="fl">1.5</span>)</span>
<span id="cb21-53"><a href="comparative-analyses-of-inferred-selection.html#cb21-53" tabindex="-1"></a>    <span class="co"># plt.savefig(&quot;/Users/esmeralda/Documents/TFM/article/omegaratio.pdf&quot;, format= &quot;pdf&quot;)</span></span>
<span id="cb21-54"><a href="comparative-analyses-of-inferred-selection.html#cb21-54" tabindex="-1"></a>    plt.show()</span></code></pre></div>
</div>
<div id="test-selection" class="section level3 hasAnchor" number="6.1.6">
<h3><span class="header-section-number">6.1.6</span> Test selection<a href="comparative-analyses-of-inferred-selection.html#test-selection" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Finally, we determine whether the omega values are different from zero (no selection).</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb22-1"><a href="comparative-analyses-of-inferred-selection.html#cb22-1" tabindex="-1"></a><span class="kw">def</span> test_one(df_values, mean):</span>
<span id="cb22-2"><a href="comparative-analyses-of-inferred-selection.html#cb22-2" tabindex="-1"></a>    results_test <span class="op">=</span> {<span class="st">&quot;Tips:&quot;</span>: [scipy.stats.ttest_1samp(df_values[<span class="dv">0</span>][<span class="dv">0</span>], popmean<span class="op">=</span>mean).pvalue,</span>
<span id="cb22-3"><a href="comparative-analyses-of-inferred-selection.html#cb22-3" tabindex="-1"></a>                              scipy.stats.ttest_1samp(df_values[<span class="dv">0</span>][<span class="dv">1</span>], popmean<span class="op">=</span>mean).pvalue,</span>
<span id="cb22-4"><a href="comparative-analyses-of-inferred-selection.html#cb22-4" tabindex="-1"></a>                              scipy.stats.ttest_1samp(df_values[<span class="dv">0</span>][<span class="dv">2</span>], popmean<span class="op">=</span>mean).pvalue],</span>
<span id="cb22-5"><a href="comparative-analyses-of-inferred-selection.html#cb22-5" tabindex="-1"></a>                    <span class="st">&quot;Branches:&quot;</span>: [scipy.stats.ttest_1samp(df_values[<span class="dv">1</span>][<span class="dv">0</span>], popmean<span class="op">=</span>mean).pvalue,</span>
<span id="cb22-6"><a href="comparative-analyses-of-inferred-selection.html#cb22-6" tabindex="-1"></a>                                 scipy.stats.ttest_1samp(df_values[<span class="dv">1</span>][<span class="dv">1</span>], popmean<span class="op">=</span>mean).pvalue,</span>
<span id="cb22-7"><a href="comparative-analyses-of-inferred-selection.html#cb22-7" tabindex="-1"></a>                                 scipy.stats.ttest_1samp(df_values[<span class="dv">1</span>][<span class="dv">2</span>], popmean<span class="op">=</span>mean).pvalue],</span>
<span id="cb22-8"><a href="comparative-analyses-of-inferred-selection.html#cb22-8" tabindex="-1"></a>                    <span class="st">&quot;Tree&quot;</span>: [scipy.stats.ttest_1samp(df_values[<span class="dv">2</span>][<span class="dv">0</span>], popmean<span class="op">=</span>mean).pvalue,</span>
<span id="cb22-9"><a href="comparative-analyses-of-inferred-selection.html#cb22-9" tabindex="-1"></a>                             scipy.stats.ttest_1samp(df_values[<span class="dv">2</span>][<span class="dv">1</span>], popmean<span class="op">=</span>mean).pvalue,</span>
<span id="cb22-10"><a href="comparative-analyses-of-inferred-selection.html#cb22-10" tabindex="-1"></a>                             scipy.stats.ttest_1samp(df_values[<span class="dv">2</span>][<span class="dv">2</span>], popmean<span class="op">=</span>mean).pvalue]}</span>
<span id="cb22-11"><a href="comparative-analyses-of-inferred-selection.html#cb22-11" tabindex="-1"></a>    </span>
<span id="cb22-12"><a href="comparative-analyses-of-inferred-selection.html#cb22-12" tabindex="-1"></a>    df <span class="op">=</span> pd.DataFrame.from_dict(results_test)</span>
<span id="cb22-13"><a href="comparative-analyses-of-inferred-selection.html#cb22-13" tabindex="-1"></a>    df.index <span class="op">=</span> [<span class="st">&quot;p22&quot;</span>, <span class="st">&quot;p19&quot;</span>, <span class="st">&quot;CP&quot;</span>]</span>
<span id="cb22-14"><a href="comparative-analyses-of-inferred-selection.html#cb22-14" tabindex="-1"></a></span>
<span id="cb22-15"><a href="comparative-analyses-of-inferred-selection.html#cb22-15" tabindex="-1"></a>    <span class="cf">return</span> df</span></code></pre></div>
</div>
<div id="run-1" class="section level3 hasAnchor" number="6.1.7">
<h3><span class="header-section-number">6.1.7</span> Run<a href="comparative-analyses-of-inferred-selection.html#run-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb23"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb23-1"><a href="comparative-analyses-of-inferred-selection.html#cb23-1" tabindex="-1"></a><span class="kw">def</span> main():</span>
<span id="cb23-2"><a href="comparative-analyses-of-inferred-selection.html#cb23-2" tabindex="-1"></a>    dfs <span class="op">=</span> [] </span>
<span id="cb23-3"><a href="comparative-analyses-of-inferred-selection.html#cb23-3" tabindex="-1"></a>    significance <span class="op">=</span> <span class="fl">0.1</span></span>
<span id="cb23-4"><a href="comparative-analyses-of-inferred-selection.html#cb23-4" tabindex="-1"></a></span>
<span id="cb23-5"><a href="comparative-analyses-of-inferred-selection.html#cb23-5" tabindex="-1"></a>    path_p22 <span class="op">=</span> <span class="st">&#39;p22_nt_alignment.fasta.ABSREL.json&#39;</span></span>
<span id="cb23-6"><a href="comparative-analyses-of-inferred-selection.html#cb23-6" tabindex="-1"></a>    path_p19 <span class="op">=</span> <span class="st">&#39;p19_nt_alignment.fasta.ABSREL.json&#39;</span></span>
<span id="cb23-7"><a href="comparative-analyses-of-inferred-selection.html#cb23-7" tabindex="-1"></a>    path_RdRp <span class="op">=</span> <span class="st">&#39;RdRp_nt_alignment.fasta.ABSREL.json&#39;</span></span>
<span id="cb23-8"><a href="comparative-analyses-of-inferred-selection.html#cb23-8" tabindex="-1"></a>    path_cp <span class="op">=</span> <span class="st">&#39;cp_nt_alignment.fasta.ABSREL.json&#39;</span></span>
<span id="cb23-9"><a href="comparative-analyses-of-inferred-selection.html#cb23-9" tabindex="-1"></a></span>
<span id="cb23-10"><a href="comparative-analyses-of-inferred-selection.html#cb23-10" tabindex="-1"></a>    <span class="co"># Read json files of each gene</span></span>
<span id="cb23-11"><a href="comparative-analyses-of-inferred-selection.html#cb23-11" tabindex="-1"></a>    paths <span class="op">=</span> [path_p22, path_p19, path_RdRp, path_cp]</span>
<span id="cb23-12"><a href="comparative-analyses-of-inferred-selection.html#cb23-12" tabindex="-1"></a>    <span class="cf">for</span> path <span class="kw">in</span> paths:</span>
<span id="cb23-13"><a href="comparative-analyses-of-inferred-selection.html#cb23-13" tabindex="-1"></a>        df <span class="op">=</span> read_branches(path)</span>
<span id="cb23-14"><a href="comparative-analyses-of-inferred-selection.html#cb23-14" tabindex="-1"></a>        dfs.append(add_info(df, significance))</span>
<span id="cb23-15"><a href="comparative-analyses-of-inferred-selection.html#cb23-15" tabindex="-1"></a></span>
<span id="cb23-16"><a href="comparative-analyses-of-inferred-selection.html#cb23-16" tabindex="-1"></a>    <span class="co"># Filling missing nodes with zero values</span></span>
<span id="cb23-17"><a href="comparative-analyses-of-inferred-selection.html#cb23-17" tabindex="-1"></a>    missing_nodes_p22 <span class="op">=</span> [<span class="st">&quot;Node11&quot;</span>, <span class="st">&quot;Node18&quot;</span>, <span class="st">&quot;Node6&quot;</span>]</span>
<span id="cb23-18"><a href="comparative-analyses-of-inferred-selection.html#cb23-18" tabindex="-1"></a>    missing_nodes_p19 <span class="op">=</span> [<span class="st">&quot;Node11&quot;</span>, <span class="st">&quot;Node18&quot;</span>, <span class="st">&quot;Node3&quot;</span>, <span class="st">&quot;Node6&quot;</span>]</span>
<span id="cb23-19"><a href="comparative-analyses-of-inferred-selection.html#cb23-19" tabindex="-1"></a></span>
<span id="cb23-20"><a href="comparative-analyses-of-inferred-selection.html#cb23-20" tabindex="-1"></a>    df_p22 <span class="op">=</span> fill_missing_nodes(dfs[<span class="dv">0</span>], missing_nodes_p22)</span>
<span id="cb23-21"><a href="comparative-analyses-of-inferred-selection.html#cb23-21" tabindex="-1"></a>    df_p19 <span class="op">=</span> fill_missing_nodes(dfs[<span class="dv">1</span>], missing_nodes_p19)</span>
<span id="cb23-22"><a href="comparative-analyses-of-inferred-selection.html#cb23-22" tabindex="-1"></a>    df_RdRp <span class="op">=</span> dfs[<span class="dv">2</span>]</span>
<span id="cb23-23"><a href="comparative-analyses-of-inferred-selection.html#cb23-23" tabindex="-1"></a>    df_cp <span class="op">=</span> dfs[<span class="dv">3</span>]</span>
<span id="cb23-24"><a href="comparative-analyses-of-inferred-selection.html#cb23-24" tabindex="-1"></a></span>
<span id="cb23-25"><a href="comparative-analyses-of-inferred-selection.html#cb23-25" tabindex="-1"></a>    <span class="co"># Generate relative values for all, tips and nodes</span></span>
<span id="cb23-26"><a href="comparative-analyses-of-inferred-selection.html#cb23-26" tabindex="-1"></a>    genes <span class="op">=</span> [df_p22, df_p19, df_cp]</span>
<span id="cb23-27"><a href="comparative-analyses-of-inferred-selection.html#cb23-27" tabindex="-1"></a>    results <span class="op">=</span> relativize(genes, df_RdRp)</span>
<span id="cb23-28"><a href="comparative-analyses-of-inferred-selection.html#cb23-28" tabindex="-1"></a>    tips, nodes <span class="op">=</span> divide_branches(results, df_RdRp)</span>
<span id="cb23-29"><a href="comparative-analyses-of-inferred-selection.html#cb23-29" tabindex="-1"></a></span>
<span id="cb23-30"><a href="comparative-analyses-of-inferred-selection.html#cb23-30" tabindex="-1"></a>    <span class="co"># Calculate pvalues in pairs</span></span>
<span id="cb23-31"><a href="comparative-analyses-of-inferred-selection.html#cb23-31" tabindex="-1"></a>    pvalues_1to1 <span class="op">=</span> []</span>
<span id="cb23-32"><a href="comparative-analyses-of-inferred-selection.html#cb23-32" tabindex="-1"></a>    df_values <span class="op">=</span> []</span>
<span id="cb23-33"><a href="comparative-analyses-of-inferred-selection.html#cb23-33" tabindex="-1"></a>    sep_types <span class="op">=</span> [tips, nodes, results]</span>
<span id="cb23-34"><a href="comparative-analyses-of-inferred-selection.html#cb23-34" tabindex="-1"></a>    </span>
<span id="cb23-35"><a href="comparative-analyses-of-inferred-selection.html#cb23-35" tabindex="-1"></a>    <span class="cf">for</span> chunk <span class="kw">in</span> sep_types:</span>
<span id="cb23-36"><a href="comparative-analyses-of-inferred-selection.html#cb23-36" tabindex="-1"></a>        pvalues_1to1.append(calc_pvalues(chunk))</span>
<span id="cb23-37"><a href="comparative-analyses-of-inferred-selection.html#cb23-37" tabindex="-1"></a>        df_values.append(info_to_dataframe(chunk))</span>
<span id="cb23-38"><a href="comparative-analyses-of-inferred-selection.html#cb23-38" tabindex="-1"></a></span>
<span id="cb23-39"><a href="comparative-analyses-of-inferred-selection.html#cb23-39" tabindex="-1"></a>    <span class="co"># Get formatted pvalues</span></span>
<span id="cb23-40"><a href="comparative-analyses-of-inferred-selection.html#cb23-40" tabindex="-1"></a>    formatted_pvalues <span class="op">=</span> format_pvalues(pvalues_1to1)</span>
<span id="cb23-41"><a href="comparative-analyses-of-inferred-selection.html#cb23-41" tabindex="-1"></a></span>
<span id="cb23-42"><a href="comparative-analyses-of-inferred-selection.html#cb23-42" tabindex="-1"></a>    <span class="co"># Boxplots with significance</span></span>
<span id="cb23-43"><a href="comparative-analyses-of-inferred-selection.html#cb23-43" tabindex="-1"></a>    plot_diff_types(df_values, pvalues_1to1, formatted_pvalues)</span>
<span id="cb23-44"><a href="comparative-analyses-of-inferred-selection.html#cb23-44" tabindex="-1"></a>    plot_one_type(df_values,  pvalues_1to1, formatted_pvalues, <span class="dv">0</span>)</span>
<span id="cb23-45"><a href="comparative-analyses-of-inferred-selection.html#cb23-45" tabindex="-1"></a></span>
<span id="cb23-46"><a href="comparative-analyses-of-inferred-selection.html#cb23-46" tabindex="-1"></a>    <span class="co"># T test one sample vs mean = 0</span></span>
<span id="cb23-47"><a href="comparative-analyses-of-inferred-selection.html#cb23-47" tabindex="-1"></a>    <span class="bu">print</span>(test_one(sep_types, <span class="dv">0</span>))</span></code></pre></div>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="compositional-analyses-of-biological-sequences.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="references.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/USERNAME/REPO/edit/BRANCH/05-selection.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
