<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 3 Distances with clustering in CLANS | Nested genes code</title>
  <meta name="description" content="This is the description of the code employed in the research." />
  <meta name="generator" content="bookdown 0.34 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 3 Distances with clustering in CLANS | Nested genes code" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="This is the description of the code employed in the research." />
  <meta name="github-repo" content="https://github.com/llobarts/nestedgenes" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 3 Distances with clustering in CLANS | Nested genes code" />
  
  <meta name="twitter:description" content="This is the description of the code employed in the research." />
  

<meta name="author" content="Esmeralda G. Legarda" />


<meta name="date" content="2023-12-30" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="blast-and-e-utilities.html"/>
<link rel="next" href="distances-with-chimerax.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Code for 'Nested genes'</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> About</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#abstract"><i class="fa fa-check"></i><b>1.1</b> Abstract</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#additional-files"><i class="fa fa-check"></i><b>1.2</b> Additional files</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="blast-and-e-utilities.html"><a href="blast-and-e-utilities.html"><i class="fa fa-check"></i><b>2</b> Blast and E-utilities</a>
<ul>
<li class="chapter" data-level="2.1" data-path="blast-and-e-utilities.html"><a href="blast-and-e-utilities.html#psi-blast"><i class="fa fa-check"></i><b>2.1</b> PSI-BLAST</a></li>
<li class="chapter" data-level="2.2" data-path="blast-and-e-utilities.html"><a href="blast-and-e-utilities.html#protein-id-to-nucleotides"><i class="fa fa-check"></i><b>2.2</b> Protein ID to nucleotides</a></li>
<li class="chapter" data-level="2.3" data-path="blast-and-e-utilities.html"><a href="blast-and-e-utilities.html#species-to-family"><i class="fa fa-check"></i><b>2.3</b> Species to family</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html"><i class="fa fa-check"></i><b>3</b> Distances with clustering in CLANS</a>
<ul>
<li class="chapter" data-level="3.1" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html#read-the-clans-file"><i class="fa fa-check"></i><b>3.1</b> Read the CLANS file</a></li>
<li class="chapter" data-level="3.2" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html#create-a-dataframe"><i class="fa fa-check"></i><b>3.2</b> Create a DataFrame</a></li>
<li class="chapter" data-level="3.3" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html#calculate-centroids-and-distances"><i class="fa fa-check"></i><b>3.3</b> Calculate centroids and distances</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html#plot-the-clustermap"><i class="fa fa-check"></i><b>3.3.1</b> Plot the clustermap</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html#run"><i class="fa fa-check"></i><b>3.4</b> Run</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html"><i class="fa fa-check"></i><b>4</b> Distances with ChimeraX</a>
<ul>
<li class="chapter" data-level="4.1" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#obtaining-rmsd-with-chimerax"><i class="fa fa-check"></i><b>4.1</b> Obtaining RMSD with ChimeraX</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#import-libraries-and-data"><i class="fa fa-check"></i><b>4.1.1</b> Import libraries and data</a></li>
<li class="chapter" data-level="4.1.2" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#matchmaker"><i class="fa fa-check"></i><b>4.1.2</b> MatchMaker</a></li>
<li class="chapter" data-level="4.1.3" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#save-csv-file"><i class="fa fa-check"></i><b>4.1.3</b> Save csv file</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#hiarachical-clustering"><i class="fa fa-check"></i><b>4.2</b> Hiarachical clustering</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#libraries-and-read-csv"><i class="fa fa-check"></i><b>4.2.1</b> Libraries and read csv</a></li>
<li class="chapter" data-level="4.2.2" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#create-matrix"><i class="fa fa-check"></i><b>4.2.2</b> Create matrix</a></li>
<li class="chapter" data-level="4.2.3" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#plot-the-clustermap-1"><i class="fa fa-check"></i><b>4.2.3</b> Plot the clustermap</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="compositional-analyses-of-biological-sequences.html"><a href="compositional-analyses-of-biological-sequences.html"><i class="fa fa-check"></i><b>5</b> Compositional analyses of biological sequences</a>
<ul>
<li class="chapter" data-level="5.1" data-path="compositional-analyses-of-biological-sequences.html"><a href="compositional-analyses-of-biological-sequences.html#amino-acids"><i class="fa fa-check"></i><b>5.1</b> Amino acids</a></li>
<li class="chapter" data-level="5.2" data-path="compositional-analyses-of-biological-sequences.html"><a href="compositional-analyses-of-biological-sequences.html#nucleotides"><i class="fa fa-check"></i><b>5.2</b> Nucleotides</a></li>
<li class="chapter" data-level="5.3" data-path="compositional-analyses-of-biological-sequences.html"><a href="compositional-analyses-of-biological-sequences.html#dinuclotides"><i class="fa fa-check"></i><b>5.3</b> Dinuclotides</a></li>
<li class="chapter" data-level="5.4" data-path="compositional-analyses-of-biological-sequences.html"><a href="compositional-analyses-of-biological-sequences.html#codons"><i class="fa fa-check"></i><b>5.4</b> Codons</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html"><i class="fa fa-check"></i><b>6</b> Comparative analyses of selection</a>
<ul>
<li class="chapter" data-level="6.1" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#evolution-speed-with-absrel"><i class="fa fa-check"></i><b>6.1</b> Evolution speed with aBSREL</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#libraries-and-json-files"><i class="fa fa-check"></i><b>6.1.1</b> Libraries and json files</a></li>
<li class="chapter" data-level="6.1.2" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#omega-and-significance"><i class="fa fa-check"></i><b>6.1.2</b> Omega and significance</a></li>
<li class="chapter" data-level="6.1.3" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#relative-asinh-transformation"><i class="fa fa-check"></i><b>6.1.3</b> Relative asinh transformation</a></li>
<li class="chapter" data-level="6.1.4" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#utilities"><i class="fa fa-check"></i><b>6.1.4</b> Utilities</a></li>
<li class="chapter" data-level="6.1.5" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#boxplots-with-significance"><i class="fa fa-check"></i><b>6.1.5</b> Boxplots with significance</a></li>
<li class="chapter" data-level="6.1.6" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#test-selection"><i class="fa fa-check"></i><b>6.1.6</b> Test selection</a></li>
<li class="chapter" data-level="6.1.7" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#run-1"><i class="fa fa-check"></i><b>6.1.7</b> Run</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#fel-for-correlations"><i class="fa fa-check"></i><b>6.2</b> FEL for correlations</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#significance-and-difference"><i class="fa fa-check"></i><b>6.2.1</b> Significance and difference</a></li>
<li class="chapter" data-level="6.2.2" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#plot-correlation"><i class="fa fa-check"></i><b>6.2.2</b> Plot correlation</a></li>
<li class="chapter" data-level="6.2.3" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#test-correlation-of-selection"><i class="fa fa-check"></i><b>6.2.3</b> Test correlation of selection</a></li>
<li class="chapter" data-level="6.2.4" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#correlation-2d"><i class="fa fa-check"></i><b>6.2.4</b> Correlation 2D</a></li>
<li class="chapter" data-level="6.2.5" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#selection-in-2d"><i class="fa fa-check"></i><b>6.2.5</b> Selection in 2D</a></li>
<li class="chapter" data-level="6.2.6" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#run-2"><i class="fa fa-check"></i><b>6.2.6</b> Run</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="miscellania.html"><a href="miscellania.html"><i class="fa fa-check"></i><b>7</b> Miscellania</a>
<ul>
<li class="chapter" data-level="7.1" data-path="miscellania.html"><a href="miscellania.html#representation-of-predicted-disordered-amino-acid-data"><i class="fa fa-check"></i><b>7.1</b> Representation of predicted disordered amino acid data</a></li>
<li class="chapter" data-level="7.2" data-path="miscellania.html"><a href="miscellania.html#shannons-entropy"><i class="fa fa-check"></i><b>7.2</b> Shannon’s entropy</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/llobarts/nestedgenes" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Nested genes code</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="distances-with-clustering-in-clans" class="section level1 hasAnchor" number="3">
<h1><span class="header-section-number">Chapter 3</span> Distances with clustering in CLANS<a href="distances-with-clustering-in-clans.html#distances-with-clustering-in-clans" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>In this chapter, we proceed to parse the output of the clustering performed with CLANS (CLuster ANalysis of Sequences) software. CLANS conducts an all-against-all comparison of sequences and calculates p-values for the High-Scoring Segment Pairs (HSPs)<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>. The output is a .clans file structured with the following sections:</p>
<ul>
<li>Parameters</li>
<li>Coordinates of the last 3D visualization with CLANS software</li>
<li>Sequences</li>
<li>Clusters of sequences</li>
<li>Positions for each sequence in 3D space</li>
<li>HSP values</li>
</ul>
<p>Our aim here is to generate a matrix of distances between protein groups clustered with CLANS for their hierarchical clustering, using Euclidean distances.</p>
<div id="read-the-clans-file" class="section level2 hasAnchor" number="3.1">
<h2><span class="header-section-number">3.1</span> Read the CLANS file<a href="distances-with-clustering-in-clans.html#read-the-clans-file" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb1-1"><a href="distances-with-clustering-in-clans.html#cb1-1" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb1-2"><a href="distances-with-clustering-in-clans.html#cb1-2" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb1-3"><a href="distances-with-clustering-in-clans.html#cb1-3" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb1-4"><a href="distances-with-clustering-in-clans.html#cb1-4" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb1-5"><a href="distances-with-clustering-in-clans.html#cb1-5" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb1-6"><a href="distances-with-clustering-in-clans.html#cb1-6" tabindex="-1"></a><span class="im">from</span> scipy.cluster.hierarchy <span class="im">import</span> linkage</span>
<span id="cb1-7"><a href="distances-with-clustering-in-clans.html#cb1-7" tabindex="-1"></a><span class="im">from</span> scipy.spatial.distance <span class="im">import</span> squareform</span>
<span id="cb1-8"><a href="distances-with-clustering-in-clans.html#cb1-8" tabindex="-1"></a><span class="im">from</span> scipy.cluster <span class="im">import</span> hierarchy</span></code></pre></div>
<p>There are two functions responsible for reading information. The first one, <code>read_clans_file</code>, collects three types of information at the sequence, group, and position per sequence levels, using the second function, <code>get_info_chunk</code>.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb2-1"><a href="distances-with-clustering-in-clans.html#cb2-1" tabindex="-1"></a><span class="kw">def</span> read_clans_file(filename):</span>
<span id="cb2-2"><a href="distances-with-clustering-in-clans.html#cb2-2" tabindex="-1"></a>    listobjects <span class="op">=</span> [[],[],[]]</span>
<span id="cb2-3"><a href="distances-with-clustering-in-clans.html#cb2-3" tabindex="-1"></a>    wordinit_list <span class="op">=</span> [<span class="st">&quot;&lt;seq&gt;&quot;</span>, <span class="st">&quot;&lt;seqgroups&gt;&quot;</span>, <span class="st">&quot;&lt;pos&gt;&quot;</span>]</span>
<span id="cb2-4"><a href="distances-with-clustering-in-clans.html#cb2-4" tabindex="-1"></a>    wordend_list <span class="op">=</span> [<span class="st">&quot;&lt;/seq&gt;&quot;</span>, <span class="st">&quot;&lt;/seqgroups&gt;&quot;</span>, <span class="st">&quot;&lt;/pos&gt;&quot;</span>]</span>
<span id="cb2-5"><a href="distances-with-clustering-in-clans.html#cb2-5" tabindex="-1"></a></span>
<span id="cb2-6"><a href="distances-with-clustering-in-clans.html#cb2-6" tabindex="-1"></a>    <span class="cf">with</span> <span class="bu">open</span>(filename, <span class="st">&#39;r&#39;</span>) <span class="im">as</span> fileopen:</span>
<span id="cb2-7"><a href="distances-with-clustering-in-clans.html#cb2-7" tabindex="-1"></a>        <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(wordinit_list)):</span>
<span id="cb2-8"><a href="distances-with-clustering-in-clans.html#cb2-8" tabindex="-1"></a>            get_info_chunk(wordinit_list[i], wordend_list[i], listobjects[i], </span>
<span id="cb2-9"><a href="distances-with-clustering-in-clans.html#cb2-9" tabindex="-1"></a>                                                    fileopen)</span>
<span id="cb2-10"><a href="distances-with-clustering-in-clans.html#cb2-10" tabindex="-1"></a>    <span class="cf">return</span> listobjects</span>
<span id="cb2-11"><a href="distances-with-clustering-in-clans.html#cb2-11" tabindex="-1"></a>  </span>
<span id="cb2-12"><a href="distances-with-clustering-in-clans.html#cb2-12" tabindex="-1"></a><span class="kw">def</span> get_info_chunk(wordinit, wordend, listobjects, fileopen):</span>
<span id="cb2-13"><a href="distances-with-clustering-in-clans.html#cb2-13" tabindex="-1"></a>    end <span class="op">=</span> <span class="va">True</span></span>
<span id="cb2-14"><a href="distances-with-clustering-in-clans.html#cb2-14" tabindex="-1"></a>    line <span class="op">=</span> fileopen.readline().strip()</span>
<span id="cb2-15"><a href="distances-with-clustering-in-clans.html#cb2-15" tabindex="-1"></a></span>
<span id="cb2-16"><a href="distances-with-clustering-in-clans.html#cb2-16" tabindex="-1"></a>    <span class="cf">while</span> <span class="kw">not</span> line.startswith(wordend):</span>
<span id="cb2-17"><a href="distances-with-clustering-in-clans.html#cb2-17" tabindex="-1"></a>        <span class="cf">if</span> line.startswith(wordinit):</span>
<span id="cb2-18"><a href="distances-with-clustering-in-clans.html#cb2-18" tabindex="-1"></a>            end <span class="op">=</span> <span class="va">False</span></span>
<span id="cb2-19"><a href="distances-with-clustering-in-clans.html#cb2-19" tabindex="-1"></a>        <span class="cf">elif</span> <span class="kw">not</span> end:</span>
<span id="cb2-20"><a href="distances-with-clustering-in-clans.html#cb2-20" tabindex="-1"></a>            listobjects.append(line.strip().split(<span class="st">&quot; &quot;</span>))</span>
<span id="cb2-21"><a href="distances-with-clustering-in-clans.html#cb2-21" tabindex="-1"></a></span>
<span id="cb2-22"><a href="distances-with-clustering-in-clans.html#cb2-22" tabindex="-1"></a>        line <span class="op">=</span> fileopen.readline().strip()</span></code></pre></div>
<p>We retrieve the names of the groups previously assigned by the CLANS software.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb3-1"><a href="distances-with-clustering-in-clans.html#cb3-1" tabindex="-1"></a><span class="kw">def</span> getgroups(groups):</span>
<span id="cb3-2"><a href="distances-with-clustering-in-clans.html#cb3-2" tabindex="-1"></a>    sep_groups <span class="op">=</span> []</span>
<span id="cb3-3"><a href="distances-with-clustering-in-clans.html#cb3-3" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> groups:</span>
<span id="cb3-4"><a href="distances-with-clustering-in-clans.html#cb3-4" tabindex="-1"></a>        <span class="cf">for</span> l <span class="kw">in</span> i:</span>
<span id="cb3-5"><a href="distances-with-clustering-in-clans.html#cb3-5" tabindex="-1"></a>            <span class="cf">if</span> l.startswith(<span class="st">&quot;numbers&quot;</span>):</span>
<span id="cb3-6"><a href="distances-with-clustering-in-clans.html#cb3-6" tabindex="-1"></a>                sep_groups.append(l.replace(<span class="st">&quot;numbers=&quot;</span>, <span class="st">&quot;&quot;</span>).split(<span class="st">&quot;;&quot;</span>)[:<span class="op">-</span><span class="dv">1</span>])</span>
<span id="cb3-7"><a href="distances-with-clustering-in-clans.html#cb3-7" tabindex="-1"></a></span>
<span id="cb3-8"><a href="distances-with-clustering-in-clans.html#cb3-8" tabindex="-1"></a>    <span class="cf">return</span> sep_groups</span>
<span id="cb3-9"><a href="distances-with-clustering-in-clans.html#cb3-9" tabindex="-1"></a></span>
<span id="cb3-10"><a href="distances-with-clustering-in-clans.html#cb3-10" tabindex="-1"></a><span class="kw">def</span> assign_groups(groups, size):</span>
<span id="cb3-11"><a href="distances-with-clustering-in-clans.html#cb3-11" tabindex="-1"></a>    last_group <span class="op">=</span> <span class="bu">len</span>(groups)</span>
<span id="cb3-12"><a href="distances-with-clustering-in-clans.html#cb3-12" tabindex="-1"></a>    numbers_group <span class="op">=</span> [last_group] <span class="op">*</span> size</span>
<span id="cb3-13"><a href="distances-with-clustering-in-clans.html#cb3-13" tabindex="-1"></a>    cont <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb3-14"><a href="distances-with-clustering-in-clans.html#cb3-14" tabindex="-1"></a>    </span>
<span id="cb3-15"><a href="distances-with-clustering-in-clans.html#cb3-15" tabindex="-1"></a>    <span class="cf">for</span> group <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(groups)):</span>
<span id="cb3-16"><a href="distances-with-clustering-in-clans.html#cb3-16" tabindex="-1"></a>        <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(groups[group])):</span>
<span id="cb3-17"><a href="distances-with-clustering-in-clans.html#cb3-17" tabindex="-1"></a>            numbers_group[<span class="bu">int</span>(groups[group][i])] <span class="op">=</span> cont</span>
<span id="cb3-18"><a href="distances-with-clustering-in-clans.html#cb3-18" tabindex="-1"></a>        cont <span class="op">+=</span> <span class="dv">1</span></span>
<span id="cb3-19"><a href="distances-with-clustering-in-clans.html#cb3-19" tabindex="-1"></a></span>
<span id="cb3-20"><a href="distances-with-clustering-in-clans.html#cb3-20" tabindex="-1"></a>    <span class="cf">return</span> numbers_group</span>
<span id="cb3-21"><a href="distances-with-clustering-in-clans.html#cb3-21" tabindex="-1"></a>  </span>
<span id="cb3-22"><a href="distances-with-clustering-in-clans.html#cb3-22" tabindex="-1"></a><span class="kw">def</span> assingnamegroup(groups, numbers_group):</span>
<span id="cb3-23"><a href="distances-with-clustering-in-clans.html#cb3-23" tabindex="-1"></a></span>
<span id="cb3-24"><a href="distances-with-clustering-in-clans.html#cb3-24" tabindex="-1"></a>    names_group <span class="op">=</span> []</span>
<span id="cb3-25"><a href="distances-with-clustering-in-clans.html#cb3-25" tabindex="-1"></a>    names <span class="op">=</span> []</span>
<span id="cb3-26"><a href="distances-with-clustering-in-clans.html#cb3-26" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> groups:</span>
<span id="cb3-27"><a href="distances-with-clustering-in-clans.html#cb3-27" tabindex="-1"></a>        <span class="cf">for</span> l <span class="kw">in</span> i:</span>
<span id="cb3-28"><a href="distances-with-clustering-in-clans.html#cb3-28" tabindex="-1"></a>            <span class="cf">if</span> l.startswith(<span class="st">&quot;name&quot;</span>) <span class="kw">and</span> l.find(<span class="st">&quot;_&quot;</span>) <span class="op">&gt;</span> <span class="dv">0</span>:</span>
<span id="cb3-29"><a href="distances-with-clustering-in-clans.html#cb3-29" tabindex="-1"></a>                names.append(l.replace(<span class="st">&quot;name=&quot;</span>, <span class="st">&quot;&quot;</span>).replace(<span class="st">&quot;_&quot;</span>, <span class="st">&quot;, &quot;</span>))</span>
<span id="cb3-30"><a href="distances-with-clustering-in-clans.html#cb3-30" tabindex="-1"></a>            <span class="cf">elif</span> l.startswith(<span class="st">&quot;name&quot;</span>):</span>
<span id="cb3-31"><a href="distances-with-clustering-in-clans.html#cb3-31" tabindex="-1"></a>                names.append(l.replace(<span class="st">&quot;name=&quot;</span>, <span class="st">&quot;&quot;</span>))</span>
<span id="cb3-32"><a href="distances-with-clustering-in-clans.html#cb3-32" tabindex="-1"></a>    cont <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb3-33"><a href="distances-with-clustering-in-clans.html#cb3-33" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> numbers_group:</span>
<span id="cb3-34"><a href="distances-with-clustering-in-clans.html#cb3-34" tabindex="-1"></a>        <span class="cf">if</span> i <span class="op">&lt;</span> <span class="bu">len</span>(names):</span>
<span id="cb3-35"><a href="distances-with-clustering-in-clans.html#cb3-35" tabindex="-1"></a>            names_group.append(names[i])</span>
<span id="cb3-36"><a href="distances-with-clustering-in-clans.html#cb3-36" tabindex="-1"></a>        <span class="cf">else</span>: </span>
<span id="cb3-37"><a href="distances-with-clustering-in-clans.html#cb3-37" tabindex="-1"></a>            names_group.append(<span class="st">&quot;NaN&quot;</span>)</span>
<span id="cb3-38"><a href="distances-with-clustering-in-clans.html#cb3-38" tabindex="-1"></a></span>
<span id="cb3-39"><a href="distances-with-clustering-in-clans.html#cb3-39" tabindex="-1"></a>    <span class="cf">return</span> names, names_group</span></code></pre></div>
</div>
<div id="create-a-dataframe" class="section level2 hasAnchor" number="3.2">
<h2><span class="header-section-number">3.2</span> Create a DataFrame<a href="distances-with-clustering-in-clans.html#create-a-dataframe" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>We create a DataFrame that includes information about the position and group for each sequence.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb4-1"><a href="distances-with-clustering-in-clans.html#cb4-1" tabindex="-1"></a><span class="kw">def</span> joininfo(info, coord, groups_names):</span>
<span id="cb4-2"><a href="distances-with-clustering-in-clans.html#cb4-2" tabindex="-1"></a></span>
<span id="cb4-3"><a href="distances-with-clustering-in-clans.html#cb4-3" tabindex="-1"></a>    info_df <span class="op">=</span> pd.Series([ x <span class="cf">for</span> x <span class="kw">in</span> info <span class="cf">if</span> info.index(x) <span class="op">%</span> <span class="dv">2</span> <span class="op">==</span> <span class="dv">0</span> ])</span>
<span id="cb4-4"><a href="distances-with-clustering-in-clans.html#cb4-4" tabindex="-1"></a>    info_df_aux <span class="op">=</span> pd.Series([ x <span class="cf">for</span> x <span class="kw">in</span> info <span class="cf">if</span> info.index(x) <span class="op">%</span> <span class="dv">2</span> <span class="op">!=</span> <span class="dv">0</span> ])</span>
<span id="cb4-5"><a href="distances-with-clustering-in-clans.html#cb4-5" tabindex="-1"></a></span>
<span id="cb4-6"><a href="distances-with-clustering-in-clans.html#cb4-6" tabindex="-1"></a>    coords_df <span class="op">=</span> pd.DataFrame(coord, columns <span class="op">=</span> [<span class="st">&quot;number&quot;</span>, <span class="st">&quot;x&quot;</span>, <span class="st">&quot;y&quot;</span>, <span class="st">&quot;z&quot;</span>])</span>
<span id="cb4-7"><a href="distances-with-clustering-in-clans.html#cb4-7" tabindex="-1"></a>    <span class="kw">del</span> coords_df[<span class="st">&quot;number&quot;</span>]</span>
<span id="cb4-8"><a href="distances-with-clustering-in-clans.html#cb4-8" tabindex="-1"></a></span>
<span id="cb4-9"><a href="distances-with-clustering-in-clans.html#cb4-9" tabindex="-1"></a>    group_df <span class="op">=</span> pd.DataFrame(groups_names, columns <span class="op">=</span> [<span class="st">&quot;group&quot;</span>])</span>
<span id="cb4-10"><a href="distances-with-clustering-in-clans.html#cb4-10" tabindex="-1"></a>    info_df <span class="op">=</span> pd.concat([info_df, info_df_aux, coords_df, group_df], axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb4-11"><a href="distances-with-clustering-in-clans.html#cb4-11" tabindex="-1"></a>    </span>
<span id="cb4-12"><a href="distances-with-clustering-in-clans.html#cb4-12" tabindex="-1"></a>    <span class="cf">return</span> info_df</span></code></pre></div>
</div>
<div id="calculate-centroids-and-distances" class="section level2 hasAnchor" number="3.3">
<h2><span class="header-section-number">3.3</span> Calculate centroids and distances<a href="distances-with-clustering-in-clans.html#calculate-centroids-and-distances" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>We will use Euclidean distance as a measure of similarity between two groups. The purpose of the <code>calculate_centroids</code> function is to calculate the centroids of each group. Subsequently, the <code>calculate_distances</code> function returns a DataFrame with pairwise distances.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb5-1"><a href="distances-with-clustering-in-clans.html#cb5-1" tabindex="-1"></a><span class="kw">def</span> calculate_centroids(info_df, names):</span>
<span id="cb5-2"><a href="distances-with-clustering-in-clans.html#cb5-2" tabindex="-1"></a>    centroids <span class="op">=</span> {}</span>
<span id="cb5-3"><a href="distances-with-clustering-in-clans.html#cb5-3" tabindex="-1"></a>    <span class="cf">for</span> element <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(names)):</span>
<span id="cb5-4"><a href="distances-with-clustering-in-clans.html#cb5-4" tabindex="-1"></a>        info_df_group <span class="op">=</span> info_df[info_df[<span class="st">&quot;group&quot;</span>] <span class="op">==</span> names[element]]</span>
<span id="cb5-5"><a href="distances-with-clustering-in-clans.html#cb5-5" tabindex="-1"></a>        centroids[element] <span class="op">=</span> [np.mean([<span class="bu">float</span>(i) <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">list</span>(info_df_group[<span class="st">&quot;x&quot;</span>])]), </span>
<span id="cb5-6"><a href="distances-with-clustering-in-clans.html#cb5-6" tabindex="-1"></a>            np.mean([<span class="bu">float</span>(i) <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">list</span>(info_df_group[<span class="st">&quot;y&quot;</span>])]), </span>
<span id="cb5-7"><a href="distances-with-clustering-in-clans.html#cb5-7" tabindex="-1"></a>            np.mean([<span class="bu">float</span>(i) <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">list</span>(info_df_group[<span class="st">&quot;z&quot;</span>])])]</span>
<span id="cb5-8"><a href="distances-with-clustering-in-clans.html#cb5-8" tabindex="-1"></a></span>
<span id="cb5-9"><a href="distances-with-clustering-in-clans.html#cb5-9" tabindex="-1"></a>    <span class="cf">return</span> centroids</span>
<span id="cb5-10"><a href="distances-with-clustering-in-clans.html#cb5-10" tabindex="-1"></a></span>
<span id="cb5-11"><a href="distances-with-clustering-in-clans.html#cb5-11" tabindex="-1"></a><span class="kw">def</span> calculate_distances(centroids, names):</span>
<span id="cb5-12"><a href="distances-with-clustering-in-clans.html#cb5-12" tabindex="-1"></a>    matrix_distances <span class="op">=</span> np.zeros(shape<span class="op">=</span>(<span class="bu">len</span>(centroids),<span class="bu">len</span>(centroids)))</span>
<span id="cb5-13"><a href="distances-with-clustering-in-clans.html#cb5-13" tabindex="-1"></a>    <span class="cf">for</span> centroid_A <span class="kw">in</span> centroids.keys():</span>
<span id="cb5-14"><a href="distances-with-clustering-in-clans.html#cb5-14" tabindex="-1"></a>        <span class="cf">for</span> centroid_B <span class="kw">in</span> centroids.keys():</span>
<span id="cb5-15"><a href="distances-with-clustering-in-clans.html#cb5-15" tabindex="-1"></a>            matrix_distances[centroid_A][centroid_B] <span class="op">=</span>  <span class="op">\</span></span>
<span id="cb5-16"><a href="distances-with-clustering-in-clans.html#cb5-16" tabindex="-1"></a>            np.sqrt((centroids[centroid_B][<span class="dv">0</span>] <span class="op">-</span> centroids[centroid_A][<span class="dv">0</span>])<span class="op">**</span><span class="dv">2</span> <span class="op">+</span> <span class="op">\</span></span>
<span id="cb5-17"><a href="distances-with-clustering-in-clans.html#cb5-17" tabindex="-1"></a>            (centroids[centroid_B][<span class="dv">1</span>] <span class="op">-</span> centroids[centroid_A][<span class="dv">1</span>])<span class="op">**</span><span class="dv">2</span> <span class="op">+</span> <span class="op">\</span></span>
<span id="cb5-18"><a href="distances-with-clustering-in-clans.html#cb5-18" tabindex="-1"></a>                (centroids[centroid_B][<span class="dv">2</span>] <span class="op">-</span> centroids[centroid_A][<span class="dv">2</span>])<span class="op">**</span><span class="dv">2</span>)</span>
<span id="cb5-19"><a href="distances-with-clustering-in-clans.html#cb5-19" tabindex="-1"></a></span>
<span id="cb5-20"><a href="distances-with-clustering-in-clans.html#cb5-20" tabindex="-1"></a>    <span class="cf">return</span> pd.DataFrame(matrix_distances, columns <span class="op">=</span> names, index <span class="op">=</span> names)</span></code></pre></div>
<div id="plot-the-clustermap" class="section level3 hasAnchor" number="3.3.1">
<h3><span class="header-section-number">3.3.1</span> Plot the clustermap<a href="distances-with-clustering-in-clans.html#plot-the-clustermap" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>This function plots the distances using hierarchical clustering, which can be performed with various methods. The cophenetic correlation coefficient serves as a metric to assess their ability to recreate pairwise distances, with higher values indicating better explanation of the pairwise relationships. Some instances of methods include <code>simple</code>, <code>average</code>, <code>complete</code>, or <code>centroid</code>.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb6-1"><a href="distances-with-clustering-in-clans.html#cb6-1" tabindex="-1"></a><span class="kw">def</span> plot_clans_distances(data, method):</span>
<span id="cb6-2"><a href="distances-with-clustering-in-clans.html#cb6-2" tabindex="-1"></a></span>
<span id="cb6-3"><a href="distances-with-clustering-in-clans.html#cb6-3" tabindex="-1"></a>    cmap <span class="op">=</span> sns.color_palette(<span class="st">&quot;YlGn&quot;</span>, as_cmap<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb6-4"><a href="distances-with-clustering-in-clans.html#cb6-4" tabindex="-1"></a>    cmap <span class="op">=</span> cmap.<span class="bu">reversed</span>()</span>
<span id="cb6-5"><a href="distances-with-clustering-in-clans.html#cb6-5" tabindex="-1"></a></span>
<span id="cb6-6"><a href="distances-with-clustering-in-clans.html#cb6-6" tabindex="-1"></a>    <span class="co"># Compute a hierarchical clustering</span></span>
<span id="cb6-7"><a href="distances-with-clustering-in-clans.html#cb6-7" tabindex="-1"></a>    data_square <span class="op">=</span> squareform(data)</span>
<span id="cb6-8"><a href="distances-with-clustering-in-clans.html#cb6-8" tabindex="-1"></a>    linkage <span class="op">=</span> hierarchy.linkage(data_square, method<span class="op">=</span>method)</span>
<span id="cb6-9"><a href="distances-with-clustering-in-clans.html#cb6-9" tabindex="-1"></a></span>
<span id="cb6-10"><a href="distances-with-clustering-in-clans.html#cb6-10" tabindex="-1"></a>    <span class="co"># Plot a heatmap with clustered rows and columns</span></span>
<span id="cb6-11"><a href="distances-with-clustering-in-clans.html#cb6-11" tabindex="-1"></a>    sns.clustermap(data,</span>
<span id="cb6-12"><a href="distances-with-clustering-in-clans.html#cb6-12" tabindex="-1"></a>                   figsize<span class="op">=</span>(<span class="dv">8</span>,<span class="dv">8</span>),</span>
<span id="cb6-13"><a href="distances-with-clustering-in-clans.html#cb6-13" tabindex="-1"></a>                   method<span class="op">=</span>method,</span>
<span id="cb6-14"><a href="distances-with-clustering-in-clans.html#cb6-14" tabindex="-1"></a>                   row_linkage<span class="op">=</span>linkage,</span>
<span id="cb6-15"><a href="distances-with-clustering-in-clans.html#cb6-15" tabindex="-1"></a>                   col_linkage<span class="op">=</span>linkage,</span>
<span id="cb6-16"><a href="distances-with-clustering-in-clans.html#cb6-16" tabindex="-1"></a>                   cmap<span class="op">=</span>cmap,</span>
<span id="cb6-17"><a href="distances-with-clustering-in-clans.html#cb6-17" tabindex="-1"></a>                   vmin<span class="op">=</span><span class="dv">0</span>, vmax<span class="op">=</span><span class="dv">90</span>,</span>
<span id="cb6-18"><a href="distances-with-clustering-in-clans.html#cb6-18" tabindex="-1"></a>                   cbar_pos<span class="op">=</span>(<span class="fl">0.02</span>, <span class="fl">0.8225</span>, <span class="fl">0.02</span>, <span class="fl">0.18</span>))</span>
<span id="cb6-19"><a href="distances-with-clustering-in-clans.html#cb6-19" tabindex="-1"></a>    </span>
<span id="cb6-20"><a href="distances-with-clustering-in-clans.html#cb6-20" tabindex="-1"></a>    c, coph_dists <span class="op">=</span> hierarchy.cophenet(linkage, data_square)</span>
<span id="cb6-21"><a href="distances-with-clustering-in-clans.html#cb6-21" tabindex="-1"></a></span>
<span id="cb6-22"><a href="distances-with-clustering-in-clans.html#cb6-22" tabindex="-1"></a>    <span class="co"># Print, save pdf and show</span></span>
<span id="cb6-23"><a href="distances-with-clustering-in-clans.html#cb6-23" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">&quot;Cophenetic correlation coefficient:&quot;</span>, c)</span>
<span id="cb6-24"><a href="distances-with-clustering-in-clans.html#cb6-24" tabindex="-1"></a>    plt.savefig(<span class="st">&quot;/Users/esmeralda/Documents/TFM/3_testprograms/familes_core_AF2/analyse/heatmap.pdf&quot;</span>, <span class="bu">format</span><span class="op">=</span><span class="st">&quot;pdf&quot;</span>, bbox_inches<span class="op">=</span><span class="st">&quot;tight&quot;</span>)</span>
<span id="cb6-25"><a href="distances-with-clustering-in-clans.html#cb6-25" tabindex="-1"></a>    plt.show()</span></code></pre></div>
</div>
</div>
<div id="run" class="section level2 hasAnchor" number="3.4">
<h2><span class="header-section-number">3.4</span> Run<a href="distances-with-clustering-in-clans.html#run" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div class="sourceCode" id="cb7"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb7-1"><a href="distances-with-clustering-in-clans.html#cb7-1" tabindex="-1"></a><span class="co"># Read information</span></span>
<span id="cb7-2"><a href="distances-with-clustering-in-clans.html#cb7-2" tabindex="-1"></a>name <span class="op">=</span>  <span class="st">&quot;4406328_3D_comparison.clans&quot;</span></span>
<span id="cb7-3"><a href="distances-with-clustering-in-clans.html#cb7-3" tabindex="-1"></a>info, groups, coord <span class="op">=</span> read_clans_file(name)</span>
<span id="cb7-4"><a href="distances-with-clustering-in-clans.html#cb7-4" tabindex="-1"></a></span>
<span id="cb7-5"><a href="distances-with-clustering-in-clans.html#cb7-5" tabindex="-1"></a><span class="co"># Parse assigned groups and their names</span></span>
<span id="cb7-6"><a href="distances-with-clustering-in-clans.html#cb7-6" tabindex="-1"></a>sep_groups <span class="op">=</span> getgroups(groups)</span>
<span id="cb7-7"><a href="distances-with-clustering-in-clans.html#cb7-7" tabindex="-1"></a>numbers_group <span class="op">=</span> assigngroups(sep_groups, <span class="bu">len</span>(coord))</span>
<span id="cb7-8"><a href="distances-with-clustering-in-clans.html#cb7-8" tabindex="-1"></a>names, groups_names <span class="op">=</span> assingnamegroup(groups, numbers_group)</span>
<span id="cb7-9"><a href="distances-with-clustering-in-clans.html#cb7-9" tabindex="-1"></a></span>
<span id="cb7-10"><a href="distances-with-clustering-in-clans.html#cb7-10" tabindex="-1"></a><span class="co"># Joined into a DataFrame</span></span>
<span id="cb7-11"><a href="distances-with-clustering-in-clans.html#cb7-11" tabindex="-1"></a>info_df <span class="op">=</span> joininfo(info, coord, groups_names)</span>
<span id="cb7-12"><a href="distances-with-clustering-in-clans.html#cb7-12" tabindex="-1"></a></span>
<span id="cb7-13"><a href="distances-with-clustering-in-clans.html#cb7-13" tabindex="-1"></a><span class="co"># Performing calculations for Euclidean distances in 3D space</span></span>
<span id="cb7-14"><a href="distances-with-clustering-in-clans.html#cb7-14" tabindex="-1"></a>centroids <span class="op">=</span> calculate_centroids(info_df, names)</span>
<span id="cb7-15"><a href="distances-with-clustering-in-clans.html#cb7-15" tabindex="-1"></a>matrix_distances <span class="op">=</span> calculate_distances(centroids, names)</span>
<span id="cb7-16"><a href="distances-with-clustering-in-clans.html#cb7-16" tabindex="-1"></a></span>
<span id="cb7-17"><a href="distances-with-clustering-in-clans.html#cb7-17" tabindex="-1"></a><span class="co"># Plot a clustermap</span></span>
<span id="cb7-18"><a href="distances-with-clustering-in-clans.html#cb7-18" tabindex="-1"></a>plot_clans_distances(matrix_distances, method)</span></code></pre></div>

</div>
</div>
<div class="footnotes">
<hr />
<ol start="1">
<li id="fn1"><p>High-Scoring Segment Pairs: regions of local similarity found during sequence comparison with BLAST (Basic Local Alignment Search Tool)<a href="distances-with-clustering-in-clans.html#fnref1" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="blast-and-e-utilities.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="distances-with-chimerax.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/llobarts/nestedgenes/edit/main/02-clans_dist.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
