<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 7 Miscellania | References</title>
  <meta name="description" content="This is the description of the code employed in the research." />
  <meta name="generator" content="bookdown 0.34 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 7 Miscellania | References" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="This is the description of the code employed in the research." />
  <meta name="github-repo" content="https://github.com/llobarts/nestedgenes" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 7 Miscellania | References" />
  
  <meta name="twitter:description" content="This is the description of the code employed in the research." />
  

<meta name="author" content="Esmeralda G. Legarda" />


<meta name="date" content="2024-01-08" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="comparative-analyses-of-selection.html"/>
<link rel="next" href="references.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Code for 'Nested genes'</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> About</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#abstract"><i class="fa fa-check"></i><b>1.1</b> Abstract</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#additional-files"><i class="fa fa-check"></i><b>1.2</b> Additional files</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="blast-and-e-utilities.html"><a href="blast-and-e-utilities.html"><i class="fa fa-check"></i><b>2</b> Blast and E-utilities</a>
<ul>
<li class="chapter" data-level="2.1" data-path="blast-and-e-utilities.html"><a href="blast-and-e-utilities.html#psi-blast"><i class="fa fa-check"></i><b>2.1</b> PSI-BLAST</a></li>
<li class="chapter" data-level="2.2" data-path="blast-and-e-utilities.html"><a href="blast-and-e-utilities.html#protein-id-to-nucleotides"><i class="fa fa-check"></i><b>2.2</b> Protein ID to nucleotides</a></li>
<li class="chapter" data-level="2.3" data-path="blast-and-e-utilities.html"><a href="blast-and-e-utilities.html#species-to-family"><i class="fa fa-check"></i><b>2.3</b> Species to family</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html"><i class="fa fa-check"></i><b>3</b> Distances with clustering in CLANS</a>
<ul>
<li class="chapter" data-level="3.1" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html#read-the-clans-file"><i class="fa fa-check"></i><b>3.1</b> Read the CLANS file</a></li>
<li class="chapter" data-level="3.2" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html#create-a-dataframe"><i class="fa fa-check"></i><b>3.2</b> Create a DataFrame</a></li>
<li class="chapter" data-level="3.3" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html#calculate-centroids-and-distances"><i class="fa fa-check"></i><b>3.3</b> Calculate centroids and distances</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html#plot-the-clustermap"><i class="fa fa-check"></i><b>3.3.1</b> Plot the clustermap</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="distances-with-clustering-in-clans.html"><a href="distances-with-clustering-in-clans.html#run"><i class="fa fa-check"></i><b>3.4</b> Run</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html"><i class="fa fa-check"></i><b>4</b> Distances with ChimeraX</a>
<ul>
<li class="chapter" data-level="4.1" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#obtaining-rmsd-with-chimerax"><i class="fa fa-check"></i><b>4.1</b> Obtaining RMSD with ChimeraX</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#import-libraries-and-data"><i class="fa fa-check"></i><b>4.1.1</b> Import libraries and data</a></li>
<li class="chapter" data-level="4.1.2" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#matchmaker"><i class="fa fa-check"></i><b>4.1.2</b> MatchMaker</a></li>
<li class="chapter" data-level="4.1.3" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#save-csv-file"><i class="fa fa-check"></i><b>4.1.3</b> Save csv file</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#hiarachical-clustering"><i class="fa fa-check"></i><b>4.2</b> Hiarachical clustering</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#libraries-and-read-csv"><i class="fa fa-check"></i><b>4.2.1</b> Libraries and read csv</a></li>
<li class="chapter" data-level="4.2.2" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#create-matrix"><i class="fa fa-check"></i><b>4.2.2</b> Create matrix</a></li>
<li class="chapter" data-level="4.2.3" data-path="distances-with-chimerax.html"><a href="distances-with-chimerax.html#plot-the-clustermap-1"><i class="fa fa-check"></i><b>4.2.3</b> Plot the clustermap</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="compositional-analyses-of-biological-sequences.html"><a href="compositional-analyses-of-biological-sequences.html"><i class="fa fa-check"></i><b>5</b> Compositional analyses of biological sequences</a>
<ul>
<li class="chapter" data-level="5.1" data-path="compositional-analyses-of-biological-sequences.html"><a href="compositional-analyses-of-biological-sequences.html#amino-acids"><i class="fa fa-check"></i><b>5.1</b> Amino acids</a></li>
<li class="chapter" data-level="5.2" data-path="compositional-analyses-of-biological-sequences.html"><a href="compositional-analyses-of-biological-sequences.html#nucleotides"><i class="fa fa-check"></i><b>5.2</b> Nucleotides</a></li>
<li class="chapter" data-level="5.3" data-path="compositional-analyses-of-biological-sequences.html"><a href="compositional-analyses-of-biological-sequences.html#dinuclotides"><i class="fa fa-check"></i><b>5.3</b> Dinuclotides</a></li>
<li class="chapter" data-level="5.4" data-path="compositional-analyses-of-biological-sequences.html"><a href="compositional-analyses-of-biological-sequences.html#codons"><i class="fa fa-check"></i><b>5.4</b> Codons</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html"><i class="fa fa-check"></i><b>6</b> Comparative analyses of selection</a>
<ul>
<li class="chapter" data-level="6.1" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#evolution-speed-with-absrel"><i class="fa fa-check"></i><b>6.1</b> Evolution speed with aBSREL</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#libraries-and-json-files"><i class="fa fa-check"></i><b>6.1.1</b> Libraries and json files</a></li>
<li class="chapter" data-level="6.1.2" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#omega-and-significance"><i class="fa fa-check"></i><b>6.1.2</b> Omega and significance</a></li>
<li class="chapter" data-level="6.1.3" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#relative-asinh-transformation"><i class="fa fa-check"></i><b>6.1.3</b> Relative asinh transformation</a></li>
<li class="chapter" data-level="6.1.4" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#utilities"><i class="fa fa-check"></i><b>6.1.4</b> Utilities</a></li>
<li class="chapter" data-level="6.1.5" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#boxplots-with-significance"><i class="fa fa-check"></i><b>6.1.5</b> Boxplots with significance</a></li>
<li class="chapter" data-level="6.1.6" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#test-selection"><i class="fa fa-check"></i><b>6.1.6</b> Test selection</a></li>
<li class="chapter" data-level="6.1.7" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#run-1"><i class="fa fa-check"></i><b>6.1.7</b> Run</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#fel-for-correlations"><i class="fa fa-check"></i><b>6.2</b> FEL for correlations</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#significance-and-difference"><i class="fa fa-check"></i><b>6.2.1</b> Significance and difference</a></li>
<li class="chapter" data-level="6.2.2" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#plot-correlation"><i class="fa fa-check"></i><b>6.2.2</b> Plot correlation</a></li>
<li class="chapter" data-level="6.2.3" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#test-correlation-of-selection"><i class="fa fa-check"></i><b>6.2.3</b> Test correlation of selection</a></li>
<li class="chapter" data-level="6.2.4" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#correlation-2d"><i class="fa fa-check"></i><b>6.2.4</b> Correlation 2D</a></li>
<li class="chapter" data-level="6.2.5" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#selection-in-2d"><i class="fa fa-check"></i><b>6.2.5</b> Selection in 2D</a></li>
<li class="chapter" data-level="6.2.6" data-path="comparative-analyses-of-selection.html"><a href="comparative-analyses-of-selection.html#run-2"><i class="fa fa-check"></i><b>6.2.6</b> Run</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="miscellania.html"><a href="miscellania.html"><i class="fa fa-check"></i><b>7</b> Miscellania</a>
<ul>
<li class="chapter" data-level="7.1" data-path="miscellania.html"><a href="miscellania.html#representation-of-predicted-disordered-amino-acid-data"><i class="fa fa-check"></i><b>7.1</b> Representation of predicted disordered amino acid data</a></li>
<li class="chapter" data-level="7.2" data-path="miscellania.html"><a href="miscellania.html#shannons-entropy"><i class="fa fa-check"></i><b>7.2</b> Shannon’s entropy</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/llobarts/nestedgenes" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">References</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="miscellania" class="section level1 hasAnchor" number="7">
<h1><span class="header-section-number">Chapter 7</span> Miscellania<a href="miscellania.html#miscellania" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>Here, you can explore how we implement some calculations and their visualizations of various issues using Python.</p>
<div id="representation-of-predicted-disordered-amino-acid-data" class="section level2 hasAnchor" number="7.1">
<h2><span class="header-section-number">7.1</span> Representation of predicted disordered amino acid data<a href="miscellania.html#representation-of-predicted-disordered-amino-acid-data" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>We import some libraries.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb1-1"><a href="miscellania.html#cb1-1" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb1-2"><a href="miscellania.html#cb1-2" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb1-3"><a href="miscellania.html#cb1-3" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span></code></pre></div>
<p>Now, we read a CSV file containing disordered data and a few other characteristics of each protein.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb2-1"><a href="miscellania.html#cb2-1" tabindex="-1"></a>df_disorder <span class="op">=</span> pd.read_csv(<span class="st">&quot;/Users/esmeralda/Documents/TFM/article/disorder_data.csv&quot;</span>, sep<span class="op">=</span><span class="st">&quot;;&quot;</span>)</span>
<span id="cb2-2"><a href="miscellania.html#cb2-2" tabindex="-1"></a>df_disorder[<span class="st">&#39;Length&#39;</span>] <span class="op">=</span> pd.to_numeric(df_disorder[<span class="st">&#39;Length&#39;</span>], errors<span class="op">=</span><span class="st">&#39;coerce&#39;</span>)</span></code></pre></div>
<p>We define for the following plots in Seaborn:</p>
<ul>
<li>The style</li>
<li>The palette</li>
</ul>
<div class="sourceCode" id="cb3"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb3-1"><a href="miscellania.html#cb3-1" tabindex="-1"></a><span class="co"># Set Seaborn style to match a similar style to Plotly</span></span>
<span id="cb3-2"><a href="miscellania.html#cb3-2" tabindex="-1"></a>sns.<span class="bu">set</span>(rc<span class="op">=</span>{<span class="st">&#39;axes.facecolor&#39;</span>: <span class="st">&#39;white&#39;</span>,</span>
<span id="cb3-3"><a href="miscellania.html#cb3-3" tabindex="-1"></a> <span class="st">&#39;axes.edgecolor&#39;</span>: <span class="st">&#39;white&#39;</span>,</span>
<span id="cb3-4"><a href="miscellania.html#cb3-4" tabindex="-1"></a>            <span class="st">&quot;axes.grid&quot;</span>: <span class="va">True</span>, </span>
<span id="cb3-5"><a href="miscellania.html#cb3-5" tabindex="-1"></a>            <span class="st">&quot;grid.linestyle&quot;</span>: <span class="st">&quot;--&quot;</span>, </span>
<span id="cb3-6"><a href="miscellania.html#cb3-6" tabindex="-1"></a>            <span class="st">&quot;font.size&quot;</span>: <span class="dv">15</span>, </span>
<span id="cb3-7"><a href="miscellania.html#cb3-7" tabindex="-1"></a>            <span class="st">&quot;axes.labelsize&quot;</span>: <span class="dv">15</span>, </span>
<span id="cb3-8"><a href="miscellania.html#cb3-8" tabindex="-1"></a>            <span class="st">&quot;xtick.labelsize&quot;</span>: <span class="dv">12</span>, </span>
<span id="cb3-9"><a href="miscellania.html#cb3-9" tabindex="-1"></a>            <span class="st">&quot;ytick.labelsize&quot;</span>: <span class="dv">12</span>,</span>
<span id="cb3-10"><a href="miscellania.html#cb3-10" tabindex="-1"></a>            <span class="st">&#39;xtick.bottom&#39;</span>: <span class="va">True</span>,</span>
<span id="cb3-11"><a href="miscellania.html#cb3-11" tabindex="-1"></a>            <span class="st">&#39;figure.facecolor&#39;</span>: <span class="st">&#39;white&#39;</span>,</span>
<span id="cb3-12"><a href="miscellania.html#cb3-12" tabindex="-1"></a> <span class="st">&#39;xtick.top&#39;</span>: <span class="va">False</span>,</span>
<span id="cb3-13"><a href="miscellania.html#cb3-13" tabindex="-1"></a> <span class="st">&#39;xtick.bottom&#39;</span>: <span class="va">False</span>,</span>
<span id="cb3-14"><a href="miscellania.html#cb3-14" tabindex="-1"></a> <span class="st">&#39;ytick.left&#39;</span>: <span class="va">True</span>,</span>
<span id="cb3-15"><a href="miscellania.html#cb3-15" tabindex="-1"></a> <span class="st">&#39;ytick.right&#39;</span>: <span class="va">False</span>,</span>
<span id="cb3-16"><a href="miscellania.html#cb3-16" tabindex="-1"></a> <span class="st">&quot;grid.color&quot;</span>: <span class="st">&quot;#DFDFDF&quot;</span>,</span>
<span id="cb3-17"><a href="miscellania.html#cb3-17" tabindex="-1"></a> <span class="st">&#39;font.sans-serif&#39;</span>: <span class="st">&#39;Helvetica&#39;</span>})</span>
<span id="cb3-18"><a href="miscellania.html#cb3-18" tabindex="-1"></a></span>
<span id="cb3-19"><a href="miscellania.html#cb3-19" tabindex="-1"></a>safe_colorblind_palette <span class="op">=</span> [<span class="st">&quot;#E69F00&quot;</span>, <span class="st">&quot;#6D0000&quot;</span>, <span class="st">&quot;#56B4E9&quot;</span>, <span class="st">&quot;#97E4CF&quot;</span>, <span class="st">&quot;#F0E442&quot;</span>, <span class="st">&quot;#0F425F&quot;</span>, <span class="st">&quot;#D55E00&quot;</span>, <span class="st">&quot;#000877&quot;</span>, <span class="st">&quot;#CA4E93&quot;</span>, <span class="st">&quot;#ABB6D3&quot;</span>, <span class="st">&quot;#D9F221&quot;</span>]</span>
<span id="cb3-20"><a href="miscellania.html#cb3-20" tabindex="-1"></a></span>
<span id="cb3-21"><a href="miscellania.html#cb3-21" tabindex="-1"></a><span class="co"># Create a figure and axes</span></span>
<span id="cb3-22"><a href="miscellania.html#cb3-22" tabindex="-1"></a>plt.figure(figsize<span class="op">=</span>(<span class="dv">10</span>, <span class="dv">6</span>))</span>
<span id="cb3-23"><a href="miscellania.html#cb3-23" tabindex="-1"></a>ax <span class="op">=</span> sns.boxplot(data<span class="op">=</span>df_disorder[df_disorder[<span class="st">&quot;Movement protein&quot;</span>] <span class="op">==</span> <span class="st">&quot;Movement protein&quot;</span>], </span>
<span id="cb3-24"><a href="miscellania.html#cb3-24" tabindex="-1"></a>                 x<span class="op">=</span><span class="st">&quot;Taxonomy origin&quot;</span>, </span>
<span id="cb3-25"><a href="miscellania.html#cb3-25" tabindex="-1"></a>                 y<span class="op">=</span><span class="st">&quot;Percentage&quot;</span>, </span>
<span id="cb3-26"><a href="miscellania.html#cb3-26" tabindex="-1"></a>                 color<span class="op">=</span><span class="st">&quot;white&quot;</span>,</span>
<span id="cb3-27"><a href="miscellania.html#cb3-27" tabindex="-1"></a>                 boxprops<span class="op">=</span><span class="bu">dict</span>(edgecolor<span class="op">=</span><span class="st">&#39;black&#39;</span>),</span>
<span id="cb3-28"><a href="miscellania.html#cb3-28" tabindex="-1"></a>                 whiskerprops<span class="op">=</span><span class="bu">dict</span>(color<span class="op">=</span><span class="st">&#39;black&#39;</span>),</span>
<span id="cb3-29"><a href="miscellania.html#cb3-29" tabindex="-1"></a>                 medianprops<span class="op">=</span><span class="bu">dict</span>(color<span class="op">=</span><span class="st">&#39;black&#39;</span>),</span>
<span id="cb3-30"><a href="miscellania.html#cb3-30" tabindex="-1"></a>                 capprops<span class="op">=</span><span class="bu">dict</span>(linewidth<span class="op">=</span><span class="dv">2</span>),</span>
<span id="cb3-31"><a href="miscellania.html#cb3-31" tabindex="-1"></a>                 width<span class="op">=</span><span class="fl">0.7</span>, </span>
<span id="cb3-32"><a href="miscellania.html#cb3-32" tabindex="-1"></a>                 linewidth<span class="op">=</span><span class="fl">0.6</span>)</span>
<span id="cb3-33"><a href="miscellania.html#cb3-33" tabindex="-1"></a></span>
<span id="cb3-34"><a href="miscellania.html#cb3-34" tabindex="-1"></a>marker_style <span class="op">=</span> []</span>
<span id="cb3-35"><a href="miscellania.html#cb3-35" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> df_disorder[<span class="st">&quot;Movement protein&quot;</span>]:</span>
<span id="cb3-36"><a href="miscellania.html#cb3-36" tabindex="-1"></a>    <span class="cf">if</span> i <span class="op">==</span> <span class="st">&#39;o&#39;</span>:</span>
<span id="cb3-37"><a href="miscellania.html#cb3-37" tabindex="-1"></a>        marker_style.append(<span class="st">&#39;o&#39;</span>)</span>
<span id="cb3-38"><a href="miscellania.html#cb3-38" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb3-39"><a href="miscellania.html#cb3-39" tabindex="-1"></a>        marker_style.append(<span class="st">&#39;X&#39;</span>)</span></code></pre></div>
<p>We can plot disorder data classified by taxonomic families.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb4-1"><a href="miscellania.html#cb4-1" tabindex="-1"></a>ax <span class="op">=</span> sns.swarmplot(data<span class="op">=</span>df_disorder, </span>
<span id="cb4-2"><a href="miscellania.html#cb4-2" tabindex="-1"></a>                   x<span class="op">=</span><span class="st">&quot;Taxonomy origin&quot;</span>, </span>
<span id="cb4-3"><a href="miscellania.html#cb4-3" tabindex="-1"></a>                   y<span class="op">=</span><span class="st">&quot;Percentage&quot;</span>,</span>
<span id="cb4-4"><a href="miscellania.html#cb4-4" tabindex="-1"></a>                   color<span class="op">=</span><span class="st">&quot;black&quot;</span>, </span>
<span id="cb4-5"><a href="miscellania.html#cb4-5" tabindex="-1"></a>                   palette<span class="op">=</span>safe_colorblind_palette,</span>
<span id="cb4-6"><a href="miscellania.html#cb4-6" tabindex="-1"></a>                   size<span class="op">=</span><span class="dv">7</span>,</span>
<span id="cb4-7"><a href="miscellania.html#cb4-7" tabindex="-1"></a>                   marker<span class="op">=</span><span class="st">&quot;X&quot;</span>)</span>
<span id="cb4-8"><a href="miscellania.html#cb4-8" tabindex="-1"></a></span>
<span id="cb4-9"><a href="miscellania.html#cb4-9" tabindex="-1"></a>ax <span class="op">=</span> sns.swarmplot(data<span class="op">=</span>df_disorder[df_disorder[<span class="st">&quot;Movement protein&quot;</span>] <span class="op">==</span> <span class="st">&quot;Movement protein&quot;</span>], </span>
<span id="cb4-10"><a href="miscellania.html#cb4-10" tabindex="-1"></a>                   x<span class="op">=</span><span class="st">&quot;Taxonomy origin&quot;</span>, </span>
<span id="cb4-11"><a href="miscellania.html#cb4-11" tabindex="-1"></a>                   y<span class="op">=</span><span class="st">&quot;Percentage&quot;</span>,</span>
<span id="cb4-12"><a href="miscellania.html#cb4-12" tabindex="-1"></a>                   color<span class="op">=</span><span class="st">&quot;black&quot;</span>, </span>
<span id="cb4-13"><a href="miscellania.html#cb4-13" tabindex="-1"></a>                   palette<span class="op">=</span>safe_colorblind_palette,</span>
<span id="cb4-14"><a href="miscellania.html#cb4-14" tabindex="-1"></a>                   marker<span class="op">=</span><span class="st">&quot;o&quot;</span>,</span>
<span id="cb4-15"><a href="miscellania.html#cb4-15" tabindex="-1"></a>                   size<span class="op">=</span><span class="dv">7</span>)</span>
<span id="cb4-16"><a href="miscellania.html#cb4-16" tabindex="-1"></a></span>
<span id="cb4-17"><a href="miscellania.html#cb4-17" tabindex="-1"></a><span class="co"># Set plot labels and title</span></span>
<span id="cb4-18"><a href="miscellania.html#cb4-18" tabindex="-1"></a>plt.xlabel(<span class="st">&quot;Genus&quot;</span>)</span>
<span id="cb4-19"><a href="miscellania.html#cb4-19" tabindex="-1"></a>plt.ylabel(<span class="st">&quot;% predicted disodered residues&quot;</span>)</span>
<span id="cb4-20"><a href="miscellania.html#cb4-20" tabindex="-1"></a></span>
<span id="cb4-21"><a href="miscellania.html#cb4-21" tabindex="-1"></a><span class="co"># Adjust plot layout</span></span>
<span id="cb4-22"><a href="miscellania.html#cb4-22" tabindex="-1"></a>plt.xticks(rotation<span class="op">=</span><span class="dv">45</span>)</span>
<span id="cb4-23"><a href="miscellania.html#cb4-23" tabindex="-1"></a>plt.tight_layout()</span>
<span id="cb4-24"><a href="miscellania.html#cb4-24" tabindex="-1"></a></span>
<span id="cb4-25"><a href="miscellania.html#cb4-25" tabindex="-1"></a>plt.savefig(<span class="st">&quot;percentage_taxonomy.pdf&quot;</span>, <span class="bu">format</span><span class="op">=</span><span class="st">&quot;pdf&quot;</span>)</span>
<span id="cb4-26"><a href="miscellania.html#cb4-26" tabindex="-1"></a><span class="co"># Display the plot</span></span>
<span id="cb4-27"><a href="miscellania.html#cb4-27" tabindex="-1"></a>plt.show()</span></code></pre></div>
<p>In this case, we create a plot for all the proteins, categorizing them based on whether their corresponding open reading frames are overprinted.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb5-1"><a href="miscellania.html#cb5-1" tabindex="-1"></a>safe_colorblind_palette_2 <span class="op">=</span> [<span class="st">&quot;#054C6F&quot;</span>, <span class="st">&quot;#44AA99&quot;</span>]</span>
<span id="cb5-2"><a href="miscellania.html#cb5-2" tabindex="-1"></a></span>
<span id="cb5-3"><a href="miscellania.html#cb5-3" tabindex="-1"></a><span class="co"># Create a figure and axes</span></span>
<span id="cb5-4"><a href="miscellania.html#cb5-4" tabindex="-1"></a>plt.figure(figsize<span class="op">=</span>(<span class="dv">5</span>, <span class="dv">6</span>))</span>
<span id="cb5-5"><a href="miscellania.html#cb5-5" tabindex="-1"></a>ax <span class="op">=</span> sns.boxplot(data<span class="op">=</span>df_disorder[df_disorder[<span class="st">&quot;Movement protein&quot;</span>] <span class="op">==</span> <span class="st">&quot;Movement protein&quot;</span>], </span>
<span id="cb5-6"><a href="miscellania.html#cb5-6" tabindex="-1"></a>                 x<span class="op">=</span><span class="st">&quot;Overlap&quot;</span>, </span>
<span id="cb5-7"><a href="miscellania.html#cb5-7" tabindex="-1"></a>                 y<span class="op">=</span><span class="st">&quot;Percentage&quot;</span>, </span>
<span id="cb5-8"><a href="miscellania.html#cb5-8" tabindex="-1"></a>                 color<span class="op">=</span><span class="st">&quot;white&quot;</span>,</span>
<span id="cb5-9"><a href="miscellania.html#cb5-9" tabindex="-1"></a>                 boxprops<span class="op">=</span><span class="bu">dict</span>(edgecolor<span class="op">=</span><span class="st">&#39;black&#39;</span>),</span>
<span id="cb5-10"><a href="miscellania.html#cb5-10" tabindex="-1"></a>                 whiskerprops<span class="op">=</span><span class="bu">dict</span>(color<span class="op">=</span><span class="st">&#39;black&#39;</span>),</span>
<span id="cb5-11"><a href="miscellania.html#cb5-11" tabindex="-1"></a>                 medianprops<span class="op">=</span><span class="bu">dict</span>(color<span class="op">=</span><span class="st">&#39;black&#39;</span>),</span>
<span id="cb5-12"><a href="miscellania.html#cb5-12" tabindex="-1"></a>                 capprops<span class="op">=</span><span class="bu">dict</span>(linewidth<span class="op">=</span><span class="dv">2</span>),</span>
<span id="cb5-13"><a href="miscellania.html#cb5-13" tabindex="-1"></a>                 width<span class="op">=</span><span class="fl">0.7</span>, </span>
<span id="cb5-14"><a href="miscellania.html#cb5-14" tabindex="-1"></a>                 linewidth<span class="op">=</span><span class="fl">0.6</span>)</span>
<span id="cb5-15"><a href="miscellania.html#cb5-15" tabindex="-1"></a></span>
<span id="cb5-16"><a href="miscellania.html#cb5-16" tabindex="-1"></a>ax <span class="op">=</span> sns.swarmplot(data<span class="op">=</span>df_disorder,  </span>
<span id="cb5-17"><a href="miscellania.html#cb5-17" tabindex="-1"></a>                   x<span class="op">=</span><span class="st">&quot;Overlap&quot;</span>, </span>
<span id="cb5-18"><a href="miscellania.html#cb5-18" tabindex="-1"></a>                   y<span class="op">=</span><span class="st">&quot;Percentage&quot;</span>, </span>
<span id="cb5-19"><a href="miscellania.html#cb5-19" tabindex="-1"></a>                   color<span class="op">=</span><span class="st">&quot;black&quot;</span>, </span>
<span id="cb5-20"><a href="miscellania.html#cb5-20" tabindex="-1"></a>                   palette<span class="op">=</span>safe_colorblind_palette_2,</span>
<span id="cb5-21"><a href="miscellania.html#cb5-21" tabindex="-1"></a>                   marker<span class="op">=</span><span class="st">&quot;X&quot;</span>,      </span>
<span id="cb5-22"><a href="miscellania.html#cb5-22" tabindex="-1"></a>                   size<span class="op">=</span><span class="dv">7</span>)</span>
<span id="cb5-23"><a href="miscellania.html#cb5-23" tabindex="-1"></a></span>
<span id="cb5-24"><a href="miscellania.html#cb5-24" tabindex="-1"></a>ax <span class="op">=</span> sns.swarmplot(data<span class="op">=</span>df_disorder[df_disorder[<span class="st">&quot;Movement protein&quot;</span>] <span class="op">==</span> <span class="st">&quot;Movement protein&quot;</span>],  </span>
<span id="cb5-25"><a href="miscellania.html#cb5-25" tabindex="-1"></a>                   x<span class="op">=</span><span class="st">&quot;Overlap&quot;</span>, </span>
<span id="cb5-26"><a href="miscellania.html#cb5-26" tabindex="-1"></a>                   y<span class="op">=</span><span class="st">&quot;Percentage&quot;</span>, </span>
<span id="cb5-27"><a href="miscellania.html#cb5-27" tabindex="-1"></a>                   color<span class="op">=</span><span class="st">&quot;black&quot;</span>, </span>
<span id="cb5-28"><a href="miscellania.html#cb5-28" tabindex="-1"></a>                   palette<span class="op">=</span>safe_colorblind_palette_2, </span>
<span id="cb5-29"><a href="miscellania.html#cb5-29" tabindex="-1"></a>                   marker<span class="op">=</span><span class="st">&quot;o&quot;</span>,   </span>
<span id="cb5-30"><a href="miscellania.html#cb5-30" tabindex="-1"></a>                   size<span class="op">=</span><span class="dv">7</span>)</span>
<span id="cb5-31"><a href="miscellania.html#cb5-31" tabindex="-1"></a></span>
<span id="cb5-32"><a href="miscellania.html#cb5-32" tabindex="-1"></a><span class="co"># Set plot labels and title</span></span>
<span id="cb5-33"><a href="miscellania.html#cb5-33" tabindex="-1"></a>plt.xlabel(<span class="st">&quot;&quot;</span>)</span>
<span id="cb5-34"><a href="miscellania.html#cb5-34" tabindex="-1"></a>plt.ylabel(<span class="st">&quot;% predicted disodered residues&quot;</span>)</span>
<span id="cb5-35"><a href="miscellania.html#cb5-35" tabindex="-1"></a></span>
<span id="cb5-36"><a href="miscellania.html#cb5-36" tabindex="-1"></a><span class="co"># Adjust plot layout</span></span>
<span id="cb5-37"><a href="miscellania.html#cb5-37" tabindex="-1"></a>plt.xticks(rotation<span class="op">=</span><span class="dv">45</span>)</span>
<span id="cb5-38"><a href="miscellania.html#cb5-38" tabindex="-1"></a>plt.tight_layout()</span>
<span id="cb5-39"><a href="miscellania.html#cb5-39" tabindex="-1"></a></span>
<span id="cb5-40"><a href="miscellania.html#cb5-40" tabindex="-1"></a>plt.savefig(<span class="st">&quot;percentage_overprinted.pdf&quot;</span>, <span class="bu">format</span><span class="op">=</span><span class="st">&quot;pdf&quot;</span>)</span>
<span id="cb5-41"><a href="miscellania.html#cb5-41" tabindex="-1"></a></span>
<span id="cb5-42"><a href="miscellania.html#cb5-42" tabindex="-1"></a><span class="co"># Display the plot</span></span>
<span id="cb5-43"><a href="miscellania.html#cb5-43" tabindex="-1"></a>plt.show()</span></code></pre></div>
<p>Then, we can also create a scatter plot, utilizing the length of our proteins as a variable.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb6-1"><a href="miscellania.html#cb6-1" tabindex="-1"></a>plt.figure(figsize<span class="op">=</span>(<span class="dv">8</span>, <span class="dv">6</span>))</span>
<span id="cb6-2"><a href="miscellania.html#cb6-2" tabindex="-1"></a>markers <span class="op">=</span> {<span class="st">&quot;Movement protein&quot;</span>: <span class="st">&quot;o&quot;</span>, <span class="st">&quot;p19&quot;</span>: <span class="st">&quot;X&quot;</span>}</span>
<span id="cb6-3"><a href="miscellania.html#cb6-3" tabindex="-1"></a>sns.scatterplot(data<span class="op">=</span>df_disorder, x<span class="op">=</span><span class="st">&#39;Length&#39;</span>, y<span class="op">=</span><span class="st">&#39;Disordered residues&#39;</span>, hue<span class="op">=</span><span class="st">&#39;Taxonomy origin&#39;</span>, style<span class="op">=</span><span class="st">&quot;Movement protein&quot;</span>, markers<span class="op">=</span>markers, palette<span class="op">=</span>safe_colorblind_palette, s<span class="op">=</span><span class="dv">40</span>)</span>
<span id="cb6-4"><a href="miscellania.html#cb6-4" tabindex="-1"></a>plt.legend([],[], frameon<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb6-5"><a href="miscellania.html#cb6-5" tabindex="-1"></a></span>
<span id="cb6-6"><a href="miscellania.html#cb6-6" tabindex="-1"></a>plt.xlabel(<span class="st">&quot;Length&quot;</span>)</span>
<span id="cb6-7"><a href="miscellania.html#cb6-7" tabindex="-1"></a>plt.ylabel(<span class="st">&quot;Predicted disordered residues&quot;</span>)</span>
<span id="cb6-8"><a href="miscellania.html#cb6-8" tabindex="-1"></a></span>
<span id="cb6-9"><a href="miscellania.html#cb6-9" tabindex="-1"></a>plt.savefig(<span class="st">&quot;disorder_scatter_taxonomy_legend.pdf&quot;</span>, <span class="bu">format</span><span class="op">=</span><span class="st">&quot;pdf&quot;</span>)</span>
<span id="cb6-10"><a href="miscellania.html#cb6-10" tabindex="-1"></a></span>
<span id="cb6-11"><a href="miscellania.html#cb6-11" tabindex="-1"></a>plt.show()</span></code></pre></div>
<p>Last but not least, in the previous plot, we can use color to distinguish proteins based on whether they were overprinted or not.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb7-1"><a href="miscellania.html#cb7-1" tabindex="-1"></a>plt.figure(figsize<span class="op">=</span>(<span class="dv">8</span>, <span class="dv">6</span>))</span>
<span id="cb7-2"><a href="miscellania.html#cb7-2" tabindex="-1"></a>ax <span class="op">=</span> sns.scatterplot(data<span class="op">=</span>df_disorder, x<span class="op">=</span><span class="st">&#39;Length&#39;</span>, y<span class="op">=</span><span class="st">&#39;Disordered residues&#39;</span>, hue<span class="op">=</span><span class="st">&#39;Overlap&#39;</span>,style<span class="op">=</span><span class="st">&quot;Movement protein&quot;</span>, palette<span class="op">=</span>safe_colorblind_palette_2, s<span class="op">=</span><span class="dv">40</span>)</span>
<span id="cb7-3"><a href="miscellania.html#cb7-3" tabindex="-1"></a></span>
<span id="cb7-4"><a href="miscellania.html#cb7-4" tabindex="-1"></a></span>
<span id="cb7-5"><a href="miscellania.html#cb7-5" tabindex="-1"></a>plt.xlabel(<span class="st">&quot;Length&quot;</span>)</span>
<span id="cb7-6"><a href="miscellania.html#cb7-6" tabindex="-1"></a>plt.ylabel(<span class="st">&quot;Predicted disordered residues&quot;</span>)</span>
<span id="cb7-7"><a href="miscellania.html#cb7-7" tabindex="-1"></a>plt.legend([],[], frameon<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb7-8"><a href="miscellania.html#cb7-8" tabindex="-1"></a></span>
<span id="cb7-9"><a href="miscellania.html#cb7-9" tabindex="-1"></a></span>
<span id="cb7-10"><a href="miscellania.html#cb7-10" tabindex="-1"></a>plt.savefig(<span class="st">&quot;disorder_scatter_overlap_legend.pdf&quot;</span>, <span class="bu">format</span><span class="op">=</span><span class="st">&quot;pdf&quot;</span>)</span>
<span id="cb7-11"><a href="miscellania.html#cb7-11" tabindex="-1"></a></span>
<span id="cb7-12"><a href="miscellania.html#cb7-12" tabindex="-1"></a>plt.show()</span></code></pre></div>
</div>
<div id="shannons-entropy" class="section level2 hasAnchor" number="7.2">
<h2><span class="header-section-number">7.2</span> Shannon’s entropy<a href="miscellania.html#shannons-entropy" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In this final section, we demonstrate how to implement Shannon’s entropy in Python and visually represent the data.</p>
<p>First of all, we load the following libraries.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb8-1"><a href="miscellania.html#cb8-1" tabindex="-1"></a><span class="im">from</span> Bio <span class="im">import</span> SeqIO</span>
<span id="cb8-2"><a href="miscellania.html#cb8-2" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb8-3"><a href="miscellania.html#cb8-3" tabindex="-1"></a><span class="im">import</span> scipy.stats</span>
<span id="cb8-4"><a href="miscellania.html#cb8-4" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb8-5"><a href="miscellania.html#cb8-5" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb8-6"><a href="miscellania.html#cb8-6" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb8-7"><a href="miscellania.html#cb8-7" tabindex="-1"></a><span class="im">from</span> statannotations.Annotator <span class="im">import</span> Annotator</span>
<span id="cb8-8"><a href="miscellania.html#cb8-8" tabindex="-1"></a><span class="im">import</span> itertools</span></code></pre></div>
<p>We have created two functions — one to read and another to drop the dashes from the data.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb9-1"><a href="miscellania.html#cb9-1" tabindex="-1"></a><span class="kw">def</span> reading_alignment(name_file):</span>
<span id="cb9-2"><a href="miscellania.html#cb9-2" tabindex="-1"></a>    record_dict <span class="op">=</span> SeqIO.to_dict(SeqIO.parse(name_file, <span class="st">&quot;fasta&quot;</span>))</span>
<span id="cb9-3"><a href="miscellania.html#cb9-3" tabindex="-1"></a>    df_init <span class="op">=</span> pd.DataFrame.from_dict(record_dict)</span>
<span id="cb9-4"><a href="miscellania.html#cb9-4" tabindex="-1"></a>    </span>
<span id="cb9-5"><a href="miscellania.html#cb9-5" tabindex="-1"></a>    <span class="cf">return</span> df_init</span>
<span id="cb9-6"><a href="miscellania.html#cb9-6" tabindex="-1"></a></span>
<span id="cb9-7"><a href="miscellania.html#cb9-7" tabindex="-1"></a><span class="kw">def</span> dropping_dashes(datafr):</span>
<span id="cb9-8"><a href="miscellania.html#cb9-8" tabindex="-1"></a></span>
<span id="cb9-9"><a href="miscellania.html#cb9-9" tabindex="-1"></a>    datafr_res <span class="op">=</span> []</span>
<span id="cb9-10"><a href="miscellania.html#cb9-10" tabindex="-1"></a>    indexes_dash <span class="op">=</span> []</span>
<span id="cb9-11"><a href="miscellania.html#cb9-11" tabindex="-1"></a></span>
<span id="cb9-12"><a href="miscellania.html#cb9-12" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(datafr.index)):</span>
<span id="cb9-13"><a href="miscellania.html#cb9-13" tabindex="-1"></a>        total <span class="op">=</span> datafr.iloc[i].value_counts()</span>
<span id="cb9-14"><a href="miscellania.html#cb9-14" tabindex="-1"></a>        <span class="cf">if</span> <span class="st">&quot;-&quot;</span> <span class="kw">in</span> total.index <span class="kw">and</span> <span class="bu">len</span>(total.index) <span class="op">==</span> <span class="dv">1</span>:</span>
<span id="cb9-15"><a href="miscellania.html#cb9-15" tabindex="-1"></a>            indexes_dash.append(i)</span>
<span id="cb9-16"><a href="miscellania.html#cb9-16" tabindex="-1"></a>    datafr_res.append(datafr.drop(datafr.index[indexes_dash]))</span>
<span id="cb9-17"><a href="miscellania.html#cb9-17" tabindex="-1"></a></span>
<span id="cb9-18"><a href="miscellania.html#cb9-18" tabindex="-1"></a>    <span class="cf">return</span> datafr_res</span></code></pre></div>
<p>The following function is designed to calculate and store the entropy values.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb10-1"><a href="miscellania.html#cb10-1" tabindex="-1"></a><span class="kw">def</span> calc_entropy(df, base<span class="op">=</span><span class="dv">2</span>):</span>
<span id="cb10-2"><a href="miscellania.html#cb10-2" tabindex="-1"></a></span>
<span id="cb10-3"><a href="miscellania.html#cb10-3" tabindex="-1"></a>    entropies <span class="op">=</span> []</span>
<span id="cb10-4"><a href="miscellania.html#cb10-4" tabindex="-1"></a>    <span class="cf">for</span> row <span class="kw">in</span> df.index:</span>
<span id="cb10-5"><a href="miscellania.html#cb10-5" tabindex="-1"></a>        counts <span class="op">=</span> (df.loc[row].value_counts())</span>
<span id="cb10-6"><a href="miscellania.html#cb10-6" tabindex="-1"></a>        entropies.append(scipy.stats.entropy(counts, base<span class="op">=</span>base))</span>
<span id="cb10-7"><a href="miscellania.html#cb10-7" tabindex="-1"></a></span>
<span id="cb10-8"><a href="miscellania.html#cb10-8" tabindex="-1"></a>    <span class="cf">return</span> pd.DataFrame(entropies, columns<span class="op">=</span>[<span class="st">&quot;Entropy&quot;</span>])</span></code></pre></div>
<p>Finally, we used two functions to generate a visualization of the entropy along the sequences of each sequence analyzed and their distributions. For the distribution plots, the package <code>statannotations</code> allows us to add significance indicators calculated with the Mann-Whitney U test. Both plots are then saved as PDF files.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb11-1"><a href="miscellania.html#cb11-1" tabindex="-1"></a><span class="kw">def</span> plot_positions(dataframe, name, <span class="bu">type</span>):</span>
<span id="cb11-2"><a href="miscellania.html#cb11-2" tabindex="-1"></a>    <span class="cf">with</span> sns.plotting_context(<span class="st">&#39;notebook&#39;</span>, font_scale <span class="op">=</span> <span class="fl">1.2</span>):  </span>
<span id="cb11-3"><a href="miscellania.html#cb11-3" tabindex="-1"></a>        plt.figure(figsize<span class="op">=</span>(<span class="dv">12</span>, <span class="dv">5</span>))</span>
<span id="cb11-4"><a href="miscellania.html#cb11-4" tabindex="-1"></a>        plt.bar(x<span class="op">=</span>dataframe.index, height<span class="op">=</span>dataframe[<span class="st">&quot;Entropy&quot;</span>], color<span class="op">=</span>dataframe[<span class="st">&quot;Color&quot;</span>])</span>
<span id="cb11-5"><a href="miscellania.html#cb11-5" tabindex="-1"></a></span>
<span id="cb11-6"><a href="miscellania.html#cb11-6" tabindex="-1"></a>        <span class="co"># plt.title(&#39;Shannon Entropy per Position&#39;)</span></span>
<span id="cb11-7"><a href="miscellania.html#cb11-7" tabindex="-1"></a>        plt.xlabel(<span class="st">&#39;Position (&#39;</span> <span class="op">+</span> <span class="bu">type</span> <span class="op">+</span> <span class="st">&#39;)&#39;</span>)</span>
<span id="cb11-8"><a href="miscellania.html#cb11-8" tabindex="-1"></a>        plt.ylabel(<span class="st">&quot;Shannon&#39;s Entropy&quot;</span>)</span>
<span id="cb11-9"><a href="miscellania.html#cb11-9" tabindex="-1"></a>        plt.xlim(<span class="dv">0</span>,<span class="bu">len</span>(dataframe.index))</span>
<span id="cb11-10"><a href="miscellania.html#cb11-10" tabindex="-1"></a>        plt.savefig(name, <span class="bu">format</span><span class="op">=</span><span class="st">&#39;pdf&#39;</span>)</span>
<span id="cb11-11"><a href="miscellania.html#cb11-11" tabindex="-1"></a>        plt.show()</span>
<span id="cb11-12"><a href="miscellania.html#cb11-12" tabindex="-1"></a></span>
<span id="cb11-13"><a href="miscellania.html#cb11-13" tabindex="-1"></a><span class="kw">def</span> plot_distribution(result_df, color_orf, x_labels, <span class="bu">type</span>, name):</span>
<span id="cb11-14"><a href="miscellania.html#cb11-14" tabindex="-1"></a>    categories <span class="op">=</span> <span class="bu">list</span>(<span class="bu">set</span>(result_df[<span class="st">&quot;Color&quot;</span>]))</span>
<span id="cb11-15"><a href="miscellania.html#cb11-15" tabindex="-1"></a></span>
<span id="cb11-16"><a href="miscellania.html#cb11-16" tabindex="-1"></a>    pairs <span class="op">=</span> []</span>
<span id="cb11-17"><a href="miscellania.html#cb11-17" tabindex="-1"></a>    p_values <span class="op">=</span> []</span>
<span id="cb11-18"><a href="miscellania.html#cb11-18" tabindex="-1"></a>    group_pairs <span class="op">=</span> <span class="bu">list</span>(itertools.combinations(categories, <span class="dv">2</span>))</span>
<span id="cb11-19"><a href="miscellania.html#cb11-19" tabindex="-1"></a></span>
<span id="cb11-20"><a href="miscellania.html#cb11-20" tabindex="-1"></a>    <span class="co"># Perform Mann-Whitney U test for each pair</span></span>
<span id="cb11-21"><a href="miscellania.html#cb11-21" tabindex="-1"></a>    <span class="cf">for</span> pair <span class="kw">in</span> group_pairs:</span>
<span id="cb11-22"><a href="miscellania.html#cb11-22" tabindex="-1"></a>        categ, categ_2 <span class="op">=</span> pair</span>
<span id="cb11-23"><a href="miscellania.html#cb11-23" tabindex="-1"></a></span>
<span id="cb11-24"><a href="miscellania.html#cb11-24" tabindex="-1"></a>        <span class="co">#print(result_df_clean[nan_mask.any(axis=1)])</span></span>
<span id="cb11-25"><a href="miscellania.html#cb11-25" tabindex="-1"></a>        p_values.append(scipy.stats.mannwhitneyu(result_df[result_df[<span class="st">&quot;Color&quot;</span>] <span class="op">==</span> categ][<span class="st">&quot;Entropy&quot;</span>], </span>
<span id="cb11-26"><a href="miscellania.html#cb11-26" tabindex="-1"></a>                                                     result_df[result_df[<span class="st">&quot;Color&quot;</span>] <span class="op">==</span> categ_2][<span class="st">&quot;Entropy&quot;</span>], </span>
<span id="cb11-27"><a href="miscellania.html#cb11-27" tabindex="-1"></a>                                                     alternative<span class="op">=</span><span class="st">&quot;two-sided&quot;</span>).pvalue)</span>
<span id="cb11-28"><a href="miscellania.html#cb11-28" tabindex="-1"></a>        pairs.append((categ, categ_2))</span>
<span id="cb11-29"><a href="miscellania.html#cb11-29" tabindex="-1"></a></span>
<span id="cb11-30"><a href="miscellania.html#cb11-30" tabindex="-1"></a>    formatted_pvalues <span class="op">=</span> [<span class="ss">f&#39;p=</span><span class="sc">{</span>pvalue<span class="sc">:.3e}</span><span class="ss">&#39;</span> <span class="cf">for</span> pvalue <span class="kw">in</span> p_values]</span>
<span id="cb11-31"><a href="miscellania.html#cb11-31" tabindex="-1"></a>    <span class="cf">with</span> sns.plotting_context(<span class="st">&#39;notebook&#39;</span>, font_scale <span class="op">=</span> <span class="fl">1.2</span>):  </span>
<span id="cb11-32"><a href="miscellania.html#cb11-32" tabindex="-1"></a>        fig, ax <span class="op">=</span> plt.subplots(figsize<span class="op">=</span>(<span class="dv">6</span>, <span class="dv">6</span>))</span>
<span id="cb11-33"><a href="miscellania.html#cb11-33" tabindex="-1"></a>        </span>
<span id="cb11-34"><a href="miscellania.html#cb11-34" tabindex="-1"></a>        sns.stripplot(data<span class="op">=</span>result_df, x<span class="op">=</span><span class="st">&quot;Color&quot;</span>, y<span class="op">=</span><span class="st">&quot;Entropy&quot;</span>, ax<span class="op">=</span>ax, palette<span class="op">=</span>color_orf, jitter<span class="op">=</span><span class="va">True</span>, size<span class="op">=</span><span class="dv">2</span>)</span>
<span id="cb11-35"><a href="miscellania.html#cb11-35" tabindex="-1"></a>        sns.boxplot(data<span class="op">=</span>result_df, x<span class="op">=</span><span class="st">&quot;Color&quot;</span>, y<span class="op">=</span><span class="st">&quot;Entropy&quot;</span>, zorder<span class="op">=</span><span class="dv">0</span>, dodge<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>ax, boxprops<span class="op">=</span>{<span class="st">&#39;facecolor&#39;</span>:<span class="st">&#39;None&#39;</span>})</span>
<span id="cb11-36"><a href="miscellania.html#cb11-36" tabindex="-1"></a>        </span>
<span id="cb11-37"><a href="miscellania.html#cb11-37" tabindex="-1"></a>        ax.set_xticks(ax.get_xticks())  </span>
<span id="cb11-38"><a href="miscellania.html#cb11-38" tabindex="-1"></a>        ax.set_xticklabels(x_labels)</span>
<span id="cb11-39"><a href="miscellania.html#cb11-39" tabindex="-1"></a>        ax.set_xlabel(<span class="bu">type</span>)</span>
<span id="cb11-40"><a href="miscellania.html#cb11-40" tabindex="-1"></a>        ax.set_ylabel(<span class="st">&quot;Shannon&#39;s Entropy&quot;</span>)</span>
<span id="cb11-41"><a href="miscellania.html#cb11-41" tabindex="-1"></a></span>
<span id="cb11-42"><a href="miscellania.html#cb11-42" tabindex="-1"></a>        <span class="co"># Add annotations</span></span>
<span id="cb11-43"><a href="miscellania.html#cb11-43" tabindex="-1"></a>        annot <span class="op">=</span> Annotator(ax<span class="op">=</span>ax, pairs <span class="op">=</span> pairs, data<span class="op">=</span>result_df, x<span class="op">=</span><span class="st">&quot;Color&quot;</span>, y<span class="op">=</span><span class="st">&quot;Entropy&quot;</span>)</span>
<span id="cb11-44"><a href="miscellania.html#cb11-44" tabindex="-1"></a>        annot.set_custom_annotations(formatted_pvalues)</span>
<span id="cb11-45"><a href="miscellania.html#cb11-45" tabindex="-1"></a>        annot.configure(test_short_name<span class="op">=</span><span class="st">&quot;MWW&quot;</span>)  <span class="co"># text_format is still simple</span></span>
<span id="cb11-46"><a href="miscellania.html#cb11-46" tabindex="-1"></a>        annot.set_pvalues_and_annotate(p_values)</span>
<span id="cb11-47"><a href="miscellania.html#cb11-47" tabindex="-1"></a>        </span>
<span id="cb11-48"><a href="miscellania.html#cb11-48" tabindex="-1"></a>        plt.savefig(name, <span class="bu">format</span><span class="op">=</span><span class="st">&#39;pdf&#39;</span>)</span>
<span id="cb11-49"><a href="miscellania.html#cb11-49" tabindex="-1"></a>        plt.show()</span></code></pre></div>
<p>Now, we apply the functions and select specific features as colors for display. We repeat these steps twice, once for nucleotides and once for amino acids.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb12-1"><a href="miscellania.html#cb12-1" tabindex="-1"></a><span class="kw">def</span> main():</span>
<span id="cb12-2"><a href="miscellania.html#cb12-2" tabindex="-1"></a>    <span class="co">############################################################################</span></span>
<span id="cb12-3"><a href="miscellania.html#cb12-3" tabindex="-1"></a>    <span class="co">#                                NUCLEOTIDES                               #</span></span>
<span id="cb12-4"><a href="miscellania.html#cb12-4" tabindex="-1"></a>    <span class="co">############################################################################</span></span>
<span id="cb12-5"><a href="miscellania.html#cb12-5" tabindex="-1"></a></span>
<span id="cb12-6"><a href="miscellania.html#cb12-6" tabindex="-1"></a>    RdRp_nt_file <span class="op">=</span> <span class="st">&quot;/Users/esmeralda/Documents/TFM/article/alignments/RdRp_nt_aligned.fasta&quot;</span></span>
<span id="cb12-7"><a href="miscellania.html#cb12-7" tabindex="-1"></a>    cp_nt_file <span class="op">=</span> <span class="st">&quot;/Users/esmeralda/Documents/TFM/article/alignments/cp_nt_aligned_filtered.fasta&quot;</span></span>
<span id="cb12-8"><a href="miscellania.html#cb12-8" tabindex="-1"></a>    p22_nt_file <span class="op">=</span> <span class="st">&quot;/Users/esmeralda/Documents/TFM/article/alignments/p22_nt_alignment.fasta&quot;</span></span>
<span id="cb12-9"><a href="miscellania.html#cb12-9" tabindex="-1"></a></span>
<span id="cb12-10"><a href="miscellania.html#cb12-10" tabindex="-1"></a>    list_df <span class="op">=</span> [RdRp_nt_file, cp_nt_file, p22_nt_file]</span>
<span id="cb12-11"><a href="miscellania.html#cb12-11" tabindex="-1"></a>    all_res <span class="op">=</span> []</span>
<span id="cb12-12"><a href="miscellania.html#cb12-12" tabindex="-1"></a>    <span class="cf">for</span> name_df <span class="kw">in</span> list_df:</span>
<span id="cb12-13"><a href="miscellania.html#cb12-13" tabindex="-1"></a>        df_init <span class="op">=</span> reading_alignment(name_df)</span>
<span id="cb12-14"><a href="miscellania.html#cb12-14" tabindex="-1"></a>        df_filtered <span class="op">=</span> dropping_dashes(df_init)</span>
<span id="cb12-15"><a href="miscellania.html#cb12-15" tabindex="-1"></a>        all_res.append(calc_entropy(df_filtered[<span class="dv">0</span>]))</span>
<span id="cb12-16"><a href="miscellania.html#cb12-16" tabindex="-1"></a>    </span>
<span id="cb12-17"><a href="miscellania.html#cb12-17" tabindex="-1"></a>    color_orf <span class="op">=</span> [<span class="st">&quot;#E07A5F&quot;</span>, <span class="st">&quot;#ECBC6F&quot;</span>, <span class="st">&quot;#81B29A&quot;</span>, <span class="st">&quot;#3D405B&quot;</span>]</span>
<span id="cb12-18"><a href="miscellania.html#cb12-18" tabindex="-1"></a>    len_orf1 <span class="op">=</span> <span class="dv">993</span> <span class="op">+</span> <span class="dv">81</span></span>
<span id="cb12-19"><a href="miscellania.html#cb12-19" tabindex="-1"></a>    all_res[<span class="dv">0</span>][<span class="st">&quot;Color&quot;</span>] <span class="op">=</span> [color_orf[<span class="dv">0</span>]]<span class="op">*</span>len_orf1 <span class="op">+</span> [color_orf[<span class="dv">1</span>]]<span class="op">*</span>(<span class="bu">len</span>(all_res[<span class="dv">0</span>].index)<span class="op">-</span>len_orf1)</span>
<span id="cb12-20"><a href="miscellania.html#cb12-20" tabindex="-1"></a>    all_res[<span class="dv">1</span>][<span class="st">&quot;Color&quot;</span>] <span class="op">=</span> color_orf[<span class="dv">2</span>]</span>
<span id="cb12-21"><a href="miscellania.html#cb12-21" tabindex="-1"></a>    all_res[<span class="dv">2</span>][<span class="st">&quot;Color&quot;</span>] <span class="op">=</span> color_orf[<span class="dv">3</span>]</span>
<span id="cb12-22"><a href="miscellania.html#cb12-22" tabindex="-1"></a></span>
<span id="cb12-23"><a href="miscellania.html#cb12-23" tabindex="-1"></a>    result_df <span class="op">=</span> pd.concat(all_res, ignore_index<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb12-24"><a href="miscellania.html#cb12-24" tabindex="-1"></a>    plot_positions(result_df, </span>
<span id="cb12-25"><a href="miscellania.html#cb12-25" tabindex="-1"></a>                   <span class="st">&#39;/Users/esmeralda/Documents/TFM/Review/entropy_plot.pdf&#39;</span>, </span>
<span id="cb12-26"><a href="miscellania.html#cb12-26" tabindex="-1"></a>                   <span class="bu">type</span><span class="op">=</span><span class="st">&quot;nucleotides&quot;</span>)</span>
<span id="cb12-27"><a href="miscellania.html#cb12-27" tabindex="-1"></a></span>
<span id="cb12-28"><a href="miscellania.html#cb12-28" tabindex="-1"></a>    x_labels <span class="op">=</span> [<span class="st">&quot;ORF1&quot;</span>,<span class="st">&quot;RdRp&quot;</span>, <span class="st">&quot;CP&quot;</span>, <span class="st">&quot;p22/p19&quot;</span>]</span>
<span id="cb12-29"><a href="miscellania.html#cb12-29" tabindex="-1"></a>    plot_distribution(result_df, color_orf, x_labels, <span class="st">&quot;CDS&quot;</span>,</span>
<span id="cb12-30"><a href="miscellania.html#cb12-30" tabindex="-1"></a>                    <span class="st">&#39;/Users/esmeralda/Documents/TFM/Review/entropy_dist_nt.pdf&#39;</span>)</span>
<span id="cb12-31"><a href="miscellania.html#cb12-31" tabindex="-1"></a>    </span>
<span id="cb12-32"><a href="miscellania.html#cb12-32" tabindex="-1"></a>    results_codons <span class="op">=</span> result_df[<span class="st">&quot;Entropy&quot;</span>]</span>
<span id="cb12-33"><a href="miscellania.html#cb12-33" tabindex="-1"></a></span>
<span id="cb12-34"><a href="miscellania.html#cb12-34" tabindex="-1"></a>    <span class="co">############################################################################</span></span>
<span id="cb12-35"><a href="miscellania.html#cb12-35" tabindex="-1"></a>    <span class="co">#                                AMINO ACIDS                               #</span></span>
<span id="cb12-36"><a href="miscellania.html#cb12-36" tabindex="-1"></a>    <span class="co">############################################################################</span></span>
<span id="cb12-37"><a href="miscellania.html#cb12-37" tabindex="-1"></a></span>
<span id="cb12-38"><a href="miscellania.html#cb12-38" tabindex="-1"></a>    RdRp_file <span class="op">=</span> <span class="st">&quot;/Users/esmeralda/Documents/TFM/article/alignments/RdRp_aligned.fa&quot;</span></span>
<span id="cb12-39"><a href="miscellania.html#cb12-39" tabindex="-1"></a>    cp_file <span class="op">=</span> <span class="st">&quot;/Users/esmeralda/Documents/TFM/article/alignments/cp_aligned.fa&quot;</span></span>
<span id="cb12-40"><a href="miscellania.html#cb12-40" tabindex="-1"></a>    p22_file <span class="op">=</span> <span class="st">&quot;/Users/esmeralda/Documents/TFM/article/alignments/p22_alignment_33.fa&quot;</span></span>
<span id="cb12-41"><a href="miscellania.html#cb12-41" tabindex="-1"></a>    p19_file <span class="op">=</span> <span class="st">&quot;/Users/esmeralda/Documents/TFM/article/alignments/p19_aligned.fa&quot;</span></span>
<span id="cb12-42"><a href="miscellania.html#cb12-42" tabindex="-1"></a>    list_df <span class="op">=</span> [RdRp_file, cp_file, p22_file, p19_file]</span>
<span id="cb12-43"><a href="miscellania.html#cb12-43" tabindex="-1"></a>    </span>
<span id="cb12-44"><a href="miscellania.html#cb12-44" tabindex="-1"></a>    all_res <span class="op">=</span> []</span>
<span id="cb12-45"><a href="miscellania.html#cb12-45" tabindex="-1"></a>    <span class="cf">for</span> name_df <span class="kw">in</span> list_df:</span>
<span id="cb12-46"><a href="miscellania.html#cb12-46" tabindex="-1"></a>        df_init <span class="op">=</span> reading_alignment(name_df)</span>
<span id="cb12-47"><a href="miscellania.html#cb12-47" tabindex="-1"></a>        df_filtered <span class="op">=</span> dropping_dashes(df_init)</span>
<span id="cb12-48"><a href="miscellania.html#cb12-48" tabindex="-1"></a>        all_res.append(calc_entropy(df_filtered[<span class="dv">0</span>]))</span>
<span id="cb12-49"><a href="miscellania.html#cb12-49" tabindex="-1"></a></span>
<span id="cb12-50"><a href="miscellania.html#cb12-50" tabindex="-1"></a>    color_orf <span class="op">=</span> [<span class="st">&quot;#E07A5F&quot;</span>, <span class="st">&quot;#ECBC6F&quot;</span>, <span class="st">&quot;#81B29A&quot;</span>, <span class="st">&quot;#3D405B&quot;</span>, <span class="st">&quot;#B5AA9D&quot;</span>]</span>
<span id="cb12-51"><a href="miscellania.html#cb12-51" tabindex="-1"></a>    len_orf1 <span class="op">=</span> (<span class="dv">993</span> <span class="op">+</span> <span class="dv">81</span>)<span class="op">//</span><span class="dv">3</span></span>
<span id="cb12-52"><a href="miscellania.html#cb12-52" tabindex="-1"></a>    all_res[<span class="dv">0</span>][<span class="st">&quot;Color&quot;</span>] <span class="op">=</span> [color_orf[<span class="dv">0</span>]]<span class="op">*</span>len_orf1 <span class="op">+</span> [color_orf[<span class="dv">1</span>]]<span class="op">*</span>(<span class="bu">len</span>(all_res[<span class="dv">0</span>].index)<span class="op">-</span>(len_orf1))</span>
<span id="cb12-53"><a href="miscellania.html#cb12-53" tabindex="-1"></a>    all_res[<span class="dv">1</span>][<span class="st">&quot;Color&quot;</span>] <span class="op">=</span> color_orf[<span class="dv">2</span>]</span>
<span id="cb12-54"><a href="miscellania.html#cb12-54" tabindex="-1"></a>    all_res[<span class="dv">2</span>][<span class="st">&quot;Color&quot;</span>] <span class="op">=</span> color_orf[<span class="dv">3</span>]</span>
<span id="cb12-55"><a href="miscellania.html#cb12-55" tabindex="-1"></a>    all_res[<span class="dv">3</span>][<span class="st">&quot;Color&quot;</span>] <span class="op">=</span> color_orf[<span class="dv">4</span>]</span>
<span id="cb12-56"><a href="miscellania.html#cb12-56" tabindex="-1"></a></span>
<span id="cb12-57"><a href="miscellania.html#cb12-57" tabindex="-1"></a>    result_df <span class="op">=</span> pd.concat(all_res, ignore_index<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb12-58"><a href="miscellania.html#cb12-58" tabindex="-1"></a>    plot_positions(result_df, </span>
<span id="cb12-59"><a href="miscellania.html#cb12-59" tabindex="-1"></a>                   <span class="st">&#39;/Users/esmeralda/Documents/TFM/Review/entropy_plot_aa.pdf&#39;</span>,</span>
<span id="cb12-60"><a href="miscellania.html#cb12-60" tabindex="-1"></a>                   <span class="st">&#39;amino acids&#39;</span>)</span>
<span id="cb12-61"><a href="miscellania.html#cb12-61" tabindex="-1"></a>    </span>
<span id="cb12-62"><a href="miscellania.html#cb12-62" tabindex="-1"></a>    x_labels <span class="op">=</span> [<span class="st">&quot;p33&quot;</span>,<span class="st">&quot;p92&quot;</span>, <span class="st">&quot;CP&quot;</span>, <span class="st">&quot;p22&quot;</span>, <span class="st">&quot;p19&quot;</span>]</span>
<span id="cb12-63"><a href="miscellania.html#cb12-63" tabindex="-1"></a>    plot_distribution(result_df, color_orf, x_labels, <span class="st">&quot;Product&quot;</span>,</span>
<span id="cb12-64"><a href="miscellania.html#cb12-64" tabindex="-1"></a>                    <span class="st">&#39;/Users/esmeralda/Documents/TFM/Review/entropy_dist_aa.pdf&#39;</span>)</span>
<span id="cb12-65"><a href="miscellania.html#cb12-65" tabindex="-1"></a>    </span>
<span id="cb12-66"><a href="miscellania.html#cb12-66" tabindex="-1"></a>    results_aa <span class="op">=</span> result_df[<span class="st">&quot;Entropy&quot;</span>]</span>
<span id="cb12-67"><a href="miscellania.html#cb12-67" tabindex="-1"></a></span>
<span id="cb12-68"><a href="miscellania.html#cb12-68" tabindex="-1"></a>    <span class="co"># Kruskal-Wallis test</span></span>
<span id="cb12-69"><a href="miscellania.html#cb12-69" tabindex="-1"></a>    statistic, p_value <span class="op">=</span> scipy.stats.kruskal(results_codons, results_aa)</span>
<span id="cb12-70"><a href="miscellania.html#cb12-70" tabindex="-1"></a>    <span class="bu">print</span>(statistic, p_value)</span>
<span id="cb12-71"><a href="miscellania.html#cb12-71" tabindex="-1"></a></span>
<span id="cb12-72"><a href="miscellania.html#cb12-72" tabindex="-1"></a><span class="cf">if</span> <span class="va">__name__</span> <span class="op">==</span> <span class="st">&quot;__main__&quot;</span>:</span>
<span id="cb12-73"><a href="miscellania.html#cb12-73" tabindex="-1"></a>    main()</span></code></pre></div>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="comparative-analyses-of-selection.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="references.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/llobarts/nestedgenes/edit/main/06-miscellania.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
